diff --git a/crates/bevy_pbr/src/render/pbr_fragment.wgsl b/crates/bevy_pbr/src/render/pbr_fragment.wgsl
index a78abcbf52075..b09131c7be1eb 100644
--- a/crates/bevy_pbr/src/render/pbr_fragment.wgsl
+++ b/crates/bevy_pbr/src/render/pbr_fragment.wgsl
@@ -86,10 +86,12 @@ fn pbr_input_from_standard_material(
     let base_color = pbr_bindings::material_array[material_indices[slot].material].base_color;
     let deferred_lighting_pass_id =
         pbr_bindings::material_array[material_indices[slot].material].deferred_lighting_pass_id;
+    let alpha_cutoff = pbr_bindings::material_array[material_indices[slot].material].alpha_cutoff;
 #else   // BINDLESS
     let flags = pbr_bindings::material.flags;
     let base_color = pbr_bindings::material.base_color;
     let deferred_lighting_pass_id = pbr_bindings::material.deferred_lighting_pass_id;
+    let alpha_cutoff = pbr_bindings::material.alpha_cutoff;
 #endif
 
     let double_sided = (flags & pbr_types::STANDARD_MATERIAL_FLAGS_DOUBLE_SIDED_BIT) != 0u;
@@ -237,6 +239,7 @@ pbr_input.material.uv_transform = uv_transform;
 #endif // VERTEX_UVS
 
     pbr_input.material.flags = flags;
+    pbr_input.material.alpha_cutoff = alpha_cutoff;
 
     // NOTE: Unlit bit not set means == 0 is true, so the true case is if lit
     if ((flags & pbr_types::STANDARD_MATERIAL_FLAGS_UNLIT_BIT) == 0u) {
@@ -246,13 +249,10 @@ pbr_input.material.uv_transform = uv_transform;
                 pbr_bindings::material_array[material_indices[slot].material].attenuation_color;
         pbr_input.material.attenuation_distance =
                 pbr_bindings::material_array[material_indices[slot].material].attenuation_distance;
-        pbr_input.material.alpha_cutoff =
-                pbr_bindings::material_array[material_indices[slot].material].alpha_cutoff;
 #else   // BINDLESS
         pbr_input.material.ior = pbr_bindings::material.ior;
         pbr_input.material.attenuation_color = pbr_bindings::material.attenuation_color;
         pbr_input.material.attenuation_distance = pbr_bindings::material.attenuation_distance;
-        pbr_input.material.alpha_cutoff = pbr_bindings::material.alpha_cutoff;
 #endif  // BINDLESS
 
         // reflectance
diff --git a/examples/3d/transparency_3d.rs b/examples/3d/transparency_3d.rs
index 2fa4a56fcdc2b..c12ff500a595c 100644
--- a/examples/3d/transparency_3d.rs
+++ b/examples/3d/transparency_3d.rs
@@ -45,7 +45,7 @@ fn setup(
         Mesh3d(meshes.add(Sphere::new(0.5).mesh().ico(3).unwrap())),
         MeshMaterial3d(materials.add(StandardMaterial {
             base_color: Color::srgba(0.2, 0.7, 0.1, 0.0),
-            alpha_mode: AlphaMode::Mask(0.5),
+            alpha_mode: AlphaMode::Mask(0.1),
             unlit: true,
             ..default()
         })),
