diff --git a/crates/bevy_ecs/src/system/commands/command.rs b/crates/bevy_ecs/src/system/commands/command.rs
index 70e917471a77a..2d81bc2a5d8a5 100644
--- a/crates/bevy_ecs/src/system/commands/command.rs
+++ b/crates/bevy_ecs/src/system/commands/command.rs
@@ -157,6 +157,8 @@ where
 
 /// A [`Command`] that runs the given system with the given input value,
 /// caching its [`SystemId`] in a [`CachedSystemId`](crate::system::CachedSystemId) resource.
+///
+/// To use the supplied input, the system should have a [`SystemInput`] as the first parameter.
 pub fn run_system_cached_with<I, M, S>(system: S, input: I::Inner<'static>) -> impl Command<Result>
 where
     I: SystemInput<Inner<'static>: Send> + Send + 'static,
diff --git a/crates/bevy_ecs/src/system/commands/mod.rs b/crates/bevy_ecs/src/system/commands/mod.rs
index d2da57458ad0c..cc2cd12969579 100644
--- a/crates/bevy_ecs/src/system/commands/mod.rs
+++ b/crates/bevy_ecs/src/system/commands/mod.rs
@@ -1125,6 +1125,8 @@ impl<'w, 's> Commands<'w, 's> {
     ///
     /// Unlike [`Commands::run_system_with`], this method does not require manual registration.
     ///
+    /// To use the supplied input, the system should have a [`SystemInput`] as the first parameter.
+    ///
     /// The first time this method is called for a particular system,
     /// it will register the system and store its [`SystemId`] in a
     /// [`CachedSystemId`](crate::system::CachedSystemId) resource for later.
diff --git a/crates/bevy_ecs/src/system/system_registry.rs b/crates/bevy_ecs/src/system/system_registry.rs
index c7efb2e013b96..3a097dd592c42 100644
--- a/crates/bevy_ecs/src/system/system_registry.rs
+++ b/crates/bevy_ecs/src/system/system_registry.rs
@@ -341,6 +341,7 @@ impl World {
     /// Before running a system, it must first be registered.
     /// The method [`World::register_system`] stores a given system and returns a [`SystemId`].
     ///
+    /// To use the supplied input, the system should have a [`SystemInput`] as the first parameter.
     /// Also runs any queued-up commands.
     ///
     /// # Examples
@@ -510,6 +511,7 @@ impl World {
 
     /// Runs a cached system with an input, registering it if necessary.
     ///
+    /// To use the supplied input, the system should have a [`SystemInput`] as the first parameter.
     /// See [`World::register_system_cached`] for more information.
     pub fn run_system_cached_with<I, O, M, S>(
         &mut self,
