diff --git a/crates/bevy_window/src/cursor/mod.rs b/crates/bevy_window/src/cursor/mod.rs
index 257950d7f645f..bf9bd5c30449f 100644
--- a/crates/bevy_window/src/cursor/mod.rs
+++ b/crates/bevy_window/src/cursor/mod.rs
@@ -24,7 +24,6 @@ pub use crate::cursor::{CustomCursor, CustomCursorImage};
     derive(Reflect),
     reflect(Component, Debug, Default, PartialEq, Clone)
 )]
-
 pub enum CursorIcon {
     #[cfg(feature = "custom_cursor")]
     /// Custom cursor image.
