diff --git a/crates/bevy_dev_tools/Cargo.toml b/crates/bevy_dev_tools/Cargo.toml
index 9d8567af0a6eb..dc218f023baef 100644
--- a/crates/bevy_dev_tools/Cargo.toml
+++ b/crates/bevy_dev_tools/Cargo.toml
@@ -10,6 +10,8 @@ keywords = ["bevy"]
 
 [features]
 bevy_ci_testing = ["serde", "ron"]
+webgl = ["bevy_render/webgl"]
+webgpu = ["bevy_render/webgpu"]
 
 [dependencies]
 # bevy
diff --git a/crates/bevy_dev_tools/src/fps_overlay.rs b/crates/bevy_dev_tools/src/fps_overlay.rs
index f180b6ef9e810..f6a3c10269856 100644
--- a/crates/bevy_dev_tools/src/fps_overlay.rs
+++ b/crates/bevy_dev_tools/src/fps_overlay.rs
@@ -188,35 +188,46 @@ fn setup(
             ))
             .with_child((TextSpan::default(), overlay_config.text_config.clone()));
 
-            let font_size = overlay_config.text_config.font_size;
-            p.spawn((
-                Node {
-                    width: Val::Px(font_size * FRAME_TIME_GRAPH_WIDTH_SCALE),
-                    height: Val::Px(font_size * FRAME_TIME_GRAPH_HEIGHT_SCALE),
-                    display: if overlay_config.frame_time_graph_config.enabled {
-                        bevy_ui::Display::DEFAULT
-                    } else {
-                        bevy_ui::Display::None
-                    },
-                    ..Default::default()
-                },
-                Pickable::IGNORE,
-                MaterialNode::from(frame_time_graph_materials.add(FrametimeGraphMaterial {
-                    values: buffers.add(ShaderStorageBuffer {
-                        // Initialize with dummy data because the default (`data: None`) will
-                        // cause a panic in the shader if the frame time graph is constructed
-                        // with `enabled: false`.
-                        data: Some(vec![0, 0, 0, 0]),
+            #[cfg(all(target_arch = "wasm32", not(feature = "webgpu")))]
+            {
+                if overlay_config.frame_time_graph_config.enabled {
+                    use tracing::warn;
+
+                    warn!("Frame time graph is not supported with WebGL. Consider if WebGPU is viable for your usecase.");
+                }
+            }
+            #[cfg(not(all(target_arch = "wasm32", not(feature = "webgpu"))))]
+            {
+                let font_size = overlay_config.text_config.font_size;
+                p.spawn((
+                    Node {
+                        width: Val::Px(font_size * FRAME_TIME_GRAPH_WIDTH_SCALE),
+                        height: Val::Px(font_size * FRAME_TIME_GRAPH_HEIGHT_SCALE),
+                        display: if overlay_config.frame_time_graph_config.enabled {
+                            bevy_ui::Display::DEFAULT
+                        } else {
+                            bevy_ui::Display::None
+                        },
                         ..Default::default()
-                    }),
-                    config: FrameTimeGraphConfigUniform::new(
-                        overlay_config.frame_time_graph_config.target_fps,
-                        overlay_config.frame_time_graph_config.min_fps,
-                        true,
-                    ),
-                })),
-                FrameTimeGraph,
-            ));
+                    },
+                    Pickable::IGNORE,
+                    MaterialNode::from(frame_time_graph_materials.add(FrametimeGraphMaterial {
+                        values: buffers.add(ShaderStorageBuffer {
+                            // Initialize with dummy data because the default (`data: None`) will
+                            // cause a panic in the shader if the frame time graph is constructed
+                            // with `enabled: false`.
+                            data: Some(vec![0, 0, 0, 0]),
+                            ..Default::default()
+                        }),
+                        config: FrameTimeGraphConfigUniform::new(
+                            overlay_config.frame_time_graph_config.target_fps,
+                            overlay_config.frame_time_graph_config.min_fps,
+                            true,
+                        ),
+                    })),
+                    FrameTimeGraph,
+                ));
+            }
         });
 }
 
diff --git a/crates/bevy_internal/Cargo.toml b/crates/bevy_internal/Cargo.toml
index f6872757b04a7..c894e3664405f 100644
--- a/crates/bevy_internal/Cargo.toml
+++ b/crates/bevy_internal/Cargo.toml
@@ -187,6 +187,7 @@ webgl = [
   "bevy_render?/webgl",
   "bevy_gizmos_render?/webgl",
   "bevy_sprite_render?/webgl",
+  "bevy_dev_tools?/webgl",
 ]
 
 webgpu = [
@@ -196,6 +197,7 @@ webgpu = [
   "bevy_render?/webgpu",
   "bevy_gizmos_render?/webgpu",
   "bevy_sprite_render?/webgpu",
+  "bevy_dev_tools?/webgpu",
 ]
 
 # Enable systems that allow for automated testing on CI
