diff --git a/crates/bevy_render/src/lib.rs b/crates/bevy_render/src/lib.rs
index b1e7837b3ec04..2325139a0af98 100644
--- a/crates/bevy_render/src/lib.rs
+++ b/crates/bevy_render/src/lib.rs
@@ -81,7 +81,7 @@ use crate::{
     gpu_readback::GpuReadbackPlugin,
     mesh::{MeshRenderAssetPlugin, RenderMesh},
     render_asset::prepare_assets,
-    render_resource::{init_empty_bind_group_layout, PipelineCache},
+    render_resource::PipelineCache,
     renderer::{render_system, RenderAdapterInfo},
     settings::RenderCreation,
     storage::StoragePlugin,
@@ -388,8 +388,6 @@ impl Plugin for RenderPlugin {
                     Render,
                     reset_render_asset_bytes_per_frame.in_set(RenderSystems::Cleanup),
                 );
-
-            render_app.add_systems(RenderStartup, init_empty_bind_group_layout);
         }
     }
 
diff --git a/crates/bevy_render/src/render_resource/bind_group_layout.rs b/crates/bevy_render/src/render_resource/bind_group_layout.rs
index fb52cedc2490a..3b8cc000802eb 100644
--- a/crates/bevy_render/src/render_resource/bind_group_layout.rs
+++ b/crates/bevy_render/src/render_resource/bind_group_layout.rs
@@ -1,6 +1,4 @@
-use crate::{define_atomic_id, renderer::RenderDevice, renderer::WgpuWrapper};
-use bevy_ecs::system::Res;
-use bevy_platform::sync::OnceLock;
+use crate::{define_atomic_id, renderer::WgpuWrapper};
 use core::ops::Deref;
 
 define_atomic_id!(BindGroupLayoutId);
@@ -63,19 +61,3 @@ impl Deref for BindGroupLayout {
         &self.value
     }
 }
-
-static EMPTY_BIND_GROUP_LAYOUT: OnceLock<BindGroupLayout> = OnceLock::new();
-
-pub(crate) fn init_empty_bind_group_layout(render_device: Res<RenderDevice>) {
-    let layout = render_device.create_bind_group_layout(Some("empty_bind_group_layout"), &[]);
-    EMPTY_BIND_GROUP_LAYOUT
-        .set(layout)
-        .expect("init_empty_bind_group_layout was called more than once");
-}
-
-pub fn empty_bind_group_layout() -> BindGroupLayout {
-    EMPTY_BIND_GROUP_LAYOUT
-        .get()
-        .expect("init_empty_bind_group_layout was not called")
-        .clone()
-}
