diff --git a/crates/bevy_shader/Cargo.toml b/crates/bevy_shader/Cargo.toml
index f7f0ef3ba75c0..3dec1d1eb6a5f 100644
--- a/crates/bevy_shader/Cargo.toml
+++ b/crates/bevy_shader/Cargo.toml
@@ -19,7 +19,7 @@ wgpu-types = { version = "27", default-features = false }
 naga = { version = "27", features = ["wgsl-in"] }
 serde = { version = "1", features = ["derive"] }
 thiserror = { version = "2", default-features = false }
-wesl = { version = "0.2.0", optional = true }
+wesl = { version = "0.3.1", optional = true }
 tracing = { version = "0.1", default-features = false, features = ["std"] }
 
 
