diff --git a/crates/bevy_animation/src/animation_curves.rs b/crates/bevy_animation/src/animation_curves.rs
index ba77ec4b5ec56..113463152ea6a 100644
--- a/crates/bevy_animation/src/animation_curves.rs
+++ b/crates/bevy_animation/src/animation_curves.rs
@@ -416,6 +416,7 @@ impl<A: Animatable> AnimationCurveEvaluator for AnimatableCurveEvaluator<A> {
             .pop()
             .ok_or_else(inconsistent::<AnimatableCurveEvaluator<A>>)?
             .value;
+        self.evaluator.stack.clear();
         Ok(())
     }
 }
