diff --git a/crates/bevy_render/src/view/mod.rs b/crates/bevy_render/src/view/mod.rs
index b1375b9a1975d..1bc5a32e628a8 100644
--- a/crates/bevy_render/src/view/mod.rs
+++ b/crates/bevy_render/src/view/mod.rs
@@ -117,6 +117,9 @@ impl Plugin for ViewPlugin {
                     clear_view_attachments
                         .in_set(RenderSystems::ManageViews)
                         .before(create_surfaces),
+                    cleanup_view_targets_for_resize
+                        .in_set(RenderSystems::ManageViews)
+                        .before(create_surfaces),
                     prepare_view_attachments
                         .in_set(RenderSystems::ManageViews)
                         .before(prepare_view_targets)
@@ -1040,6 +1043,21 @@ pub fn clear_view_attachments(mut view_target_attachments: ResMut<ViewTargetAtta
     view_target_attachments.clear();
 }
 
+pub fn cleanup_view_targets_for_resize(
+    mut commands: Commands,
+    windows: Res<ExtractedWindows>,
+    cameras: Query<(Entity, &ExtractedCamera), With<ViewTarget>>,
+) {
+    for (entity, camera) in &cameras {
+        if let Some(NormalizedRenderTarget::Window(window_ref)) = &camera.target
+            && let Some(window) = windows.get(&window_ref.entity())
+            && (window.size_changed || window.present_mode_changed)
+        {
+            commands.entity(entity).remove::<ViewTarget>();
+        }
+    }
+}
+
 pub fn prepare_view_targets(
     mut commands: Commands,
     clear_color_global: Res<ClearColor>,
diff --git a/crates/bevy_render/src/view/window/mod.rs b/crates/bevy_render/src/view/window/mod.rs
index 708e873ba3519..02cc0f6c7fde5 100644
--- a/crates/bevy_render/src/view/window/mod.rs
+++ b/crates/bevy_render/src/view/window/mod.rs
@@ -413,6 +413,7 @@ pub fn create_surfaces(
             // normally this is dropped on present but we double check here to be safe as failure to
             // drop it will cause validation errors in wgpu
             drop(window.swap_chain_texture.take());
+            drop(window.swap_chain_texture_view.take());
 
             data.configuration.width = window.physical_width;
             data.configuration.height = window.physical_height;
