diff --git a/crates/bevy_text/src/pipeline.rs b/crates/bevy_text/src/pipeline.rs
index 5d3fe166d1f82..956f8100e61cc 100644
--- a/crates/bevy_text/src/pipeline.rs
+++ b/crates/bevy_text/src/pipeline.rs
@@ -253,9 +253,7 @@ impl TextPipeline {
         font_system: &mut CosmicFontSystem,
         swash_cache: &mut SwashCache,
     ) -> Result<(), TextError> {
-        layout_info.glyphs.clear();
-        layout_info.run_geometry.clear();
-        layout_info.size = Default::default();
+        layout_info.clear();
 
         // Clear this here at the focal point of text rendering to ensure the field's lifecycle has strong boundaries.
         computed.needs_rerender = false;
@@ -521,6 +519,16 @@ pub struct TextLayoutInfo {
     pub size: Vec2,
 }
 
+impl TextLayoutInfo {
+    /// Clear the layout, retaining capacity
+    pub fn clear(&mut self) {
+        self.scale_factor = 1.;
+        self.glyphs.clear();
+        self.run_geometry.clear();
+        self.size = Vec2::ZERO;
+    }
+}
+
 /// Geometry of a text run used to render text decorations like background colors, strikethrough, and underline.
 /// A run in `bevy_text` is a contiguous sequence of glyphs on a line that share the same text attributes like font,
 /// font size, and line height.
