diff --git a/crates/bevy_picking/src/events.rs b/crates/bevy_picking/src/events.rs
index 4528001f0e429..3866d44a42318 100644
--- a/crates/bevy_picking/src/events.rs
+++ b/crates/bevy_picking/src/events.rs
@@ -617,7 +617,7 @@ pub fn pointer_events(
             PointerAction::Release(button) => {
                 let state = pointer_state.get_mut(pointer_id, button);
 
-                // Emit Click and Up events on all the previously hovered entities.
+                // Emit Click and Release events on all the previously hovered entities.
                 for (hovered_entity, hit) in previous_hover_map
                     .get(&pointer_id)
                     .iter()
