diff --git a/crates/bevy_ui/src/layout/mod.rs b/crates/bevy_ui/src/layout/mod.rs
index 68b8c69dbd9e9..a3982ea2d5a4b 100644
--- a/crates/bevy_ui/src/layout/mod.rs
+++ b/crates/bevy_ui/src/layout/mod.rs
@@ -37,7 +37,7 @@ impl LayoutContext {
         scale_factor: 1.0,
         physical_size: Vec2::ZERO,
     };
-    /// create new a [`LayoutContext`] from the window's physical size and scale factor
+    /// Create a new [`LayoutContext`] from the window's physical size and scale factor
     #[inline]
     const fn new(scale_factor: f32, physical_size: Vec2) -> Self {
         Self {
