diff --git a/crates/bevy_ui/src/interaction_states.rs b/crates/bevy_ui/src/interaction_states.rs
index a5793ca778a57..41e80114062a9 100644
--- a/crates/bevy_ui/src/interaction_states.rs
+++ b/crates/bevy_ui/src/interaction_states.rs
@@ -37,15 +37,15 @@ pub(crate) fn on_remove_disabled(
 
 /// Component that indicates whether a button or widget is currently in a pressed or "held down"
 /// state.
-#[derive(Component, Default, Debug)]
+#[derive(Component, Debug, Clone, Copy, Default)]
 pub struct Pressed;
 
 /// Component that indicates that a widget can be checked.
-#[derive(Component, Default, Debug)]
+#[derive(Component, Debug, Clone, Copy, Default)]
 pub struct Checkable;
 
 /// Component that indicates whether a checkbox or radio button is in a checked state.
-#[derive(Component, Default, Debug)]
+#[derive(Component, Debug, Clone, Copy, Default)]
 pub struct Checked;
 
 pub(crate) fn on_add_checkable(add: On<Add, Checked>, mut world: DeferredWorld) {
