diff --git a/crates/bevy_text/src/font_atlas_set.rs b/crates/bevy_text/src/font_atlas_set.rs
index 76d7bfaa460b2..f189b40733fd6 100644
--- a/crates/bevy_text/src/font_atlas_set.rs
+++ b/crates/bevy_text/src/font_atlas_set.rs
@@ -8,15 +8,22 @@ use bevy_platform::collections::HashMap;
 ///
 /// Allows an `f32` font size to be used as a key in a `HashMap`, by its binary representation.
 #[derive(Debug, Hash, PartialEq, Eq, Clone, Copy)]
-pub struct FontAtlasKey(pub AssetId<Font>, pub u32, pub FontSmoothing);
+pub struct FontAtlasKey {
+    /// Font asset id
+    pub id: AssetId<Font>,
+    /// Font size via `f32::to_bits`
+    pub font_size_bits: u32,
+    /// Antialiasing method
+    pub font_smoothing: FontSmoothing,
+}
 
 impl From<&TextFont> for FontAtlasKey {
     fn from(font: &TextFont) -> Self {
-        FontAtlasKey(
-            font.font.id(),
-            font.font_size.to_bits(),
-            font.font_smoothing,
-        )
+        Self {
+            id: font.font.id(),
+            font_size_bits: font.font_size.to_bits(),
+            font_smoothing: font.font_smoothing,
+        }
     }
 }
 
@@ -39,7 +46,7 @@ pub fn free_unused_font_atlases_system(
 ) {
     for event in font_events.read() {
         if let AssetEvent::Removed { id } = event {
-            font_atlas_sets.retain(|key, _| key.0 != *id);
+            font_atlas_sets.retain(|key, _| key.id != *id);
         }
     }
 }
diff --git a/crates/bevy_text/src/pipeline.rs b/crates/bevy_text/src/pipeline.rs
index 720436f678367..665c8bd1e14b6 100644
--- a/crates/bevy_text/src/pipeline.rs
+++ b/crates/bevy_text/src/pipeline.rs
@@ -412,11 +412,11 @@ impl TextPipeline {
                     let physical_glyph = layout_glyph.physical((0., 0.), 1.);
 
                     let font_atlases = font_atlas_set
-                        .entry(FontAtlasKey(
-                            font_id,
-                            physical_glyph.cache_key.font_size_bits,
+                        .entry(FontAtlasKey {
+                            id: font_id,
+                            font_size_bits: physical_glyph.cache_key.font_size_bits,
                             font_smoothing,
-                        ))
+                        })
                         .or_default();
 
                     let atlas_info = get_glyph_atlas_info(font_atlases, physical_glyph.cache_key)
diff --git a/examples/stress_tests/many_text2d.rs b/examples/stress_tests/many_text2d.rs
index d3646c8d2c45d..83c091738053a 100644
--- a/examples/stress_tests/many_text2d.rs
+++ b/examples/stress_tests/many_text2d.rs
@@ -181,7 +181,7 @@ fn print_counts(
     let font_id = font.0.id();
     let num_atlases = font_atlas_set
         .iter()
-        .filter(|(key, _)| key.0 == font_id)
+        .filter(|(key, _)| key.id == font_id)
         .map(|(_, atlases)| atlases.len())
         .sum::<usize>();
 
