diff --git a/crates/bevy_ecs/src/query/access.rs b/crates/bevy_ecs/src/query/access.rs
index 01e3713ad624d..a9cefc6996a08 100644
--- a/crates/bevy_ecs/src/query/access.rs
+++ b/crates/bevy_ecs/src/query/access.rs
@@ -952,7 +952,8 @@ impl AccessConflicts {
         }
     }
 
-    pub(crate) fn is_empty(&self) -> bool {
+    /// Returns true if there are no conflicts present
+    pub fn is_empty(&self) -> bool {
         match self {
             Self::All => false,
             Self::Individual(set) => set.is_empty(),
