diff --git a/examples/tools/scene_viewer/main.rs b/examples/tools/scene_viewer/main.rs
index 0266a22ee0a06..f5ab505d91520 100644
--- a/examples/tools/scene_viewer/main.rs
+++ b/examples/tools/scene_viewer/main.rs
@@ -10,6 +10,7 @@
 
 use argh::FromArgs;
 use bevy::{
+    asset::UnapprovedPathMode,
     core_pipeline::prepass::{DeferredPrepass, DepthPrepass},
     pbr::DefaultOpaqueRendererMethod,
     prelude::*,
@@ -74,6 +75,8 @@ fn main() {
             })
             .set(AssetPlugin {
                 file_path: std::env::var("CARGO_MANIFEST_DIR").unwrap_or_else(|_| ".".to_string()),
+                // Allow scenes to be loaded from anywhere on disk
+                unapproved_path_mode: UnapprovedPathMode::Allow,
                 ..default()
             }),
         CameraControllerPlugin,
