diff --git a/crates/bevy_pbr/src/light/mod.rs b/crates/bevy_pbr/src/light/mod.rs
index e1a28e6a6abeb..91ea9cddd3b4c 100644
--- a/crates/bevy_pbr/src/light/mod.rs
+++ b/crates/bevy_pbr/src/light/mod.rs
@@ -103,7 +103,6 @@ pub mod light_consts {
 #[reflect(Resource, Debug, Default, Clone)]
 pub struct PointLightShadowMap {
     /// The width and height of each of the 6 faces of the cubemap.
-
     ///
     /// Defaults to `1024`.
     pub size: usize,
