diff --git a/examples/testbed/ui.rs b/examples/testbed/ui.rs
index 7cc26d6c90304..1e84d7373d34b 100644
--- a/examples/testbed/ui.rs
+++ b/examples/testbed/ui.rs
@@ -89,7 +89,7 @@ mod image {
 }
 
 mod text {
-    use bevy::prelude::*;
+    use bevy::{color::palettes::css::*, prelude::*};
 
     pub fn setup(mut commands: Commands, asset_server: Res<AssetServer>) {
         commands.spawn((Camera2d, DespawnOnExitState(super::Scene::Text)));
@@ -102,6 +102,34 @@ mod text {
             },
             DespawnOnExitState(super::Scene::Text),
         ));
+
+        commands.spawn((
+            Node {
+                left: px(100.),
+                top: px(250.),
+                ..Default::default()
+            },
+            Text::new("white "),
+            TextFont {
+                font: asset_server.load("fonts/FiraSans-Bold.ttf"),
+                ..default()
+            },
+            DespawnOnExitState(super::Scene::Text),
+            children![
+                (TextSpan::new("red "), TextColor(RED.into()),),
+                (TextSpan::new("green "), TextColor(GREEN.into()),),
+                (TextSpan::new("blue "), TextColor(BLUE.into()),),
+                (
+                    TextSpan::new("black"),
+                    TextColor(Color::BLACK),
+                    TextFont {
+                        font: asset_server.load("fonts/FiraSans-Bold.ttf"),
+                        ..default()
+                    },
+                    TextBackgroundColor(Color::WHITE)
+                ),
+            ],
+        ));
     }
 }
 
