diff --git a/crates/bevy_ecs/macros/src/event.rs b/crates/bevy_ecs/macros/src/event.rs
index 2adb910db3ea1..5137dfe6f408b 100644
--- a/crates/bevy_ecs/macros/src/event.rs
+++ b/crates/bevy_ecs/macros/src/event.rs
@@ -15,9 +15,14 @@ pub const TRIGGER: &str = "trigger";
 pub const EVENT_TARGET: &str = "event_target";
 
 pub fn derive_event(input: TokenStream) -> TokenStream {
-    let ast = parse_macro_input!(input as DeriveInput);
+    let mut ast = parse_macro_input!(input as DeriveInput);
     let bevy_ecs_path: Path = crate::bevy_ecs_path();
 
+    ast.generics
+        .make_where_clause()
+        .predicates
+        .push(parse_quote! { Self: Send + Sync + 'static });
+
     let mut processed_attrs = Vec::new();
     let mut trigger: Option<Type> = None;
 
@@ -55,7 +60,13 @@ pub fn derive_event(input: TokenStream) -> TokenStream {
 }
 
 pub fn derive_entity_event(input: TokenStream) -> TokenStream {
-    let ast = parse_macro_input!(input as DeriveInput);
+    let mut ast = parse_macro_input!(input as DeriveInput);
+
+    ast.generics
+        .make_where_clause()
+        .predicates
+        .push(parse_quote! { Self: Send + Sync + 'static });
+
     let mut auto_propagate = false;
     let mut propagate = false;
     let mut traversal: Option<Type> = None;
