diff --git a/examples/no_std/library/Cargo.toml b/examples/no_std/library/Cargo.toml
index 3cd07b55fbced..f7f20b645e7e1 100644
--- a/examples/no_std/library/Cargo.toml
+++ b/examples/no_std/library/Cargo.toml
@@ -2,6 +2,7 @@
 name = "no_std_library"
 version = "0.1.0"
 edition = "2024"
+publish = false
 
 # Normally we'd put all dependencies in [dependencies], but this syntax is easier to document
 [dependencies.bevy]
diff --git a/tools/export-content/Cargo.toml b/tools/export-content/Cargo.toml
index c4d1281bb00f5..3c574cd05a144 100644
--- a/tools/export-content/Cargo.toml
+++ b/tools/export-content/Cargo.toml
@@ -2,6 +2,7 @@
 name = "export-content"
 version = "0.1.0"
 edition = "2024"
+publish = false
 
 [dependencies]
 miette = { version = "7.6.0", features = ["fancy"] }
