diff --git a/.github/workflows/ci.yml b/.github/workflows/ci.yml
index c7e175b8a5727..3d9bc2a99ba55 100644
--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -339,7 +339,7 @@ jobs:
     steps:
       - uses: actions/checkout@v5
       - name: Check for typos
-        uses: crate-ci/typos@v1.35.5
+        uses: crate-ci/typos@v1.36.2
       - name: Typos info
         if: failure()
         run: |
diff --git a/crates/bevy_ecs/src/storage/table/mod.rs b/crates/bevy_ecs/src/storage/table/mod.rs
index ee79cb9e280a0..3eda97ead1010 100644
--- a/crates/bevy_ecs/src/storage/table/mod.rs
+++ b/crates/bevy_ecs/src/storage/table/mod.rs
@@ -561,7 +561,7 @@ impl Table {
     /// - `current_column_capacity` is indeed the capacity of the columns
     ///
     /// The capacity of all columns is determined by that of the `entities` Vec. This means that
-    /// it must be the correct capacity to allocate, reallocate, and deallocate all columnts. This
+    /// it must be the correct capacity to allocate, reallocate, and deallocate all columns. This
     /// means the safety invariant must be enforced even in `TableBuilder`.
     unsafe fn realloc_columns(
         &mut self,
