diff --git a/crates/bevy_light/src/volumetric.rs b/crates/bevy_light/src/volumetric.rs
index 347731985d2a6..31c73db6bffdb 100644
--- a/crates/bevy_light/src/volumetric.rs
+++ b/crates/bevy_light/src/volumetric.rs
@@ -18,6 +18,8 @@ pub struct VolumetricLight;
 /// When placed on a [`bevy_camera::Camera3d`], enables
 /// volumetric fog and volumetric lighting, also known as light shafts or god
 /// rays.
+///
+/// Requires using WebGPU on Wasm builds.
 #[derive(Clone, Copy, Component, Debug, Reflect)]
 #[reflect(Component, Default, Debug, Clone)]
 pub struct VolumetricFog {
diff --git a/crates/bevy_pbr/src/decal/forward.rs b/crates/bevy_pbr/src/decal/forward.rs
index 240ba2a25268f..65ef0a44663ac 100644
--- a/crates/bevy_pbr/src/decal/forward.rs
+++ b/crates/bevy_pbr/src/decal/forward.rs
@@ -60,6 +60,7 @@ impl Plugin for ForwardDecalPlugin {
 /// * Any camera rendering a forward decal must have the [`bevy_core_pipeline::prepass::DepthPrepass`] component.
 /// * Looking at forward decals at a steep angle can cause distortion. This can be mitigated by padding your decal's
 ///   texture with extra transparent pixels on the edges.
+/// * On Wasm, requires using WebGPU and disabling `Msaa` on your camera.
 #[derive(Component, Reflect)]
 #[require(Mesh3d)]
 #[component(on_add=forward_decal_set_mesh)]
diff --git a/examples/3d/decal.rs b/examples/3d/decal.rs
index 9aa0056c1178b..3bafcabd1d2ad 100644
--- a/examples/3d/decal.rs
+++ b/examples/3d/decal.rs
@@ -1,9 +1,11 @@
 //! Decal rendering.
+//! Note: On Wasm, this example only runs on WebGPU
 
 #[path = "../helpers/camera_controller.rs"]
 mod camera_controller;
 
 use bevy::{
+    anti_alias::fxaa::Fxaa,
     core_pipeline::prepass::DepthPrepass,
     pbr::decal::{ForwardDecal, ForwardDecalMaterial, ForwardDecalMaterialExt},
     prelude::*,
@@ -46,7 +48,12 @@ fn setup(
         Name::new("Camera"),
         Camera3d::default(),
         CameraController::default(),
-        DepthPrepass, // Must enable the depth prepass to render forward decals
+        // Must enable the depth prepass to render forward decals
+        DepthPrepass,
+        // Must disable MSAA to use decals on WebGPU
+        Msaa::Off,
+        // FXAA is a fine alternative to MSAA for anti-aliasing
+        Fxaa::default(),
         Transform::from_xyz(2.0, 9.5, 2.5).looking_at(Vec3::ZERO, Vec3::Y),
     ));
 
diff --git a/examples/3d/volumetric_fog.rs b/examples/3d/volumetric_fog.rs
index 8e3b8f65e581d..4e5f94b37e829 100644
--- a/examples/3d/volumetric_fog.rs
+++ b/examples/3d/volumetric_fog.rs
@@ -1,4 +1,5 @@
 //! Demonstrates volumetric fog and lighting (light shafts or god rays).
+//! Note: On Wasm, this example only runs on WebGPU
 
 use bevy::{
     color::palettes::css::RED,
