diff --git a/crates/bevy_ui_render/src/lib.rs b/crates/bevy_ui_render/src/lib.rs
index 99c18ee31ac35..acaa6aacc15d0 100644
--- a/crates/bevy_ui_render/src/lib.rs
+++ b/crates/bevy_ui_render/src/lib.rs
@@ -973,7 +973,7 @@ pub fn extract_text_sections(
                     .get(
                         computed_block
                             .entities()
-                            .get(*span_index)
+                            .get(*span_index + 1)
                             .map(|t| t.entity)
                             .unwrap_or(Entity::PLACEHOLDER),
                     )
