diff --git a/crates/bevy_log/src/lib.rs b/crates/bevy_log/src/lib.rs
index dd590d566a4a3..6a75a179516c9 100644
--- a/crates/bevy_log/src/lib.rs
+++ b/crates/bevy_log/src/lib.rs
@@ -275,7 +275,18 @@ type PreFmtSubscriber =
 pub type BoxedFmtLayer = Box<dyn Layer<PreFmtSubscriber> + Send + Sync + 'static>;
 
 /// The default [`LogPlugin`] [`EnvFilter`].
-pub const DEFAULT_FILTER: &str = "wgpu=error,naga=warn";
+pub const DEFAULT_FILTER: &str = concat!(
+    "wgpu=error,",
+    "naga=warn,",
+    "symphonia_bundle_mp3::demuxer=warn,",
+    "symphonia_format_caf::demuxer=warn,",
+    "symphonia_format_isompf4::demuxer=warn,",
+    "symphonia_format_mkv::demuxer=warn,",
+    "symphonia_format_ogg::demuxer=warn,",
+    "symphonia_format_riff::demuxer=warn,",
+    "symphonia_format_wav::demuxer=warn,",
+    "calloop::loop_logic=error,",
+);
 
 impl Default for LogPlugin {
     fn default() -> Self {
diff --git a/typos.toml b/typos.toml
index 16aab11772821..23d443e81e0c1 100644
--- a/typos.toml
+++ b/typos.toml
@@ -14,6 +14,7 @@ reparametrization = "reparametrization" # Mathematical term in curve context (re
 reparametrize = "reparametrize"
 reparametrized = "reparametrized"
 mis = "mis"                             # mis - multiple importance sampling
+caf = "caf"                             # CAF - Core Audio Format
 
 # Match a Whole Word - Case Sensitive
 [default.extend-identifiers]
