diff --git a/crates/bevy_sprite_render/src/tilemap_chunk/mod.rs b/crates/bevy_sprite_render/src/tilemap_chunk/mod.rs
index 5139ab704b5a6..8b4f062873262 100644
--- a/crates/bevy_sprite_render/src/tilemap_chunk/mod.rs
+++ b/crates/bevy_sprite_render/src/tilemap_chunk/mod.rs
@@ -1,5 +1,3 @@
-use core::ops::Neg;
-
 use crate::{AlphaMode2d, MeshMaterial2d};
 use bevy_app::{App, Plugin, Update};
 use bevy_asset::{Assets, Handle};
@@ -76,12 +74,12 @@ impl TilemapChunk {
             // tile position
             position.y as f32
             // times display size for a tile
-            * (self.tile_display_size.y as f32).neg()
+            * self.tile_display_size.y as f32
             // minus 1/2 the tile_display_size to correct the center
-            - self.tile_display_size.y as f32 / 2.
+            + self.tile_display_size.y as f32 / 2.
             // plus 1/2 the tilechunk size, in terms of the tile_display_size,
-            // to place the 0 at top of tilemapchunk
-            + self.tile_display_size.y as f32 * self.chunk_size.y as f32 / 2.,
+            // to place the 0 at bottom of tilemapchunk
+            - self.tile_display_size.y as f32 * self.chunk_size.y as f32 / 2.,
             0.,
         )
     }
diff --git a/crates/bevy_sprite_render/src/tilemap_chunk/tilemap_chunk_material.wgsl b/crates/bevy_sprite_render/src/tilemap_chunk/tilemap_chunk_material.wgsl
index a1442aa006f9a..aa3bede413394 100644
--- a/crates/bevy_sprite_render/src/tilemap_chunk/tilemap_chunk_material.wgsl
+++ b/crates/bevy_sprite_render/src/tilemap_chunk/tilemap_chunk_material.wgsl
@@ -35,7 +35,8 @@ fn get_tile_data(coord: vec2<u32>) -> TileData {
 fn fragment(in: VertexOutput) -> @location(0) vec4<f32> {
     let chunk_size = textureDimensions(tile_data, 0);
     let tile_uv = in.uv * vec2<f32>(chunk_size);
-    let tile_coord = clamp(vec2<u32>(floor(tile_uv)), vec2<u32>(0), chunk_size - 1);
+    var tile_coord = clamp(vec2<u32>(floor(tile_uv)), vec2<u32>(0), chunk_size - 1);
+    tile_coord.y = chunk_size.y - 1 - tile_coord.y;
 
     let tile = get_tile_data(tile_coord);
 
diff --git a/release-content/migration-guides/tilemap_chunk_layout_change.md b/release-content/migration-guides/tilemap_chunk_layout_change.md
new file mode 100644
index 0000000000000..194c49e4f7ef6
--- /dev/null
+++ b/release-content/migration-guides/tilemap_chunk_layout_change.md
@@ -0,0 +1,13 @@
+---
+title: Tilemap Chunk Layout
+pull_requests: [21684]
+---
+
+`TilemapChunk` and `TilemapChunkTileData`'s default layout has been changed from the origin being in the top left to the origin being in the bottom left.
+
+The previous layout origin didn't align with Bevy's world coordinate system, so when mapping to and from chunk space (to map a world coordinate to a tile) you would have to account for the chunk y coordinate being inverted.
+
+With the origin of the chunk being in the bottom left, you can simply mod world coordinates to get chunk coordinates.
+
+Some other tiling tools have the convention of the origin being at the top left, but it's more important for Bevy's features
+to be internally consistent as it allows for better ease of use.
