diff --git a/crates/bevy_ecs/src/observer/entity_cloning.rs b/crates/bevy_ecs/src/observer/entity_cloning.rs
index 7af2d2129c594..d01d0caa6315d 100644
--- a/crates/bevy_ecs/src/observer/entity_cloning.rs
+++ b/crates/bevy_ecs/src/observer/entity_cloning.rs
@@ -95,7 +95,6 @@ mod tests {
             .spawn_empty()
             .observe(|_: On<E>, mut res: ResMut<Num>| res.0 += 1)
             .id();
-        world.flush();
 
         world.trigger(E(e));
 
