diff --git a/crates/bevy_ecs/src/entity/index_map.rs b/crates/bevy_ecs/src/entity/index_map.rs
index 6f6cd1bb47af3..9067b12ca28b2 100644
--- a/crates/bevy_ecs/src/entity/index_map.rs
+++ b/crates/bevy_ecs/src/entity/index_map.rs
@@ -17,6 +17,7 @@ use core::{
 
 #[cfg(feature = "bevy_reflect")]
 use bevy_reflect::Reflect;
+pub use indexmap::map::Entry;
 use indexmap::map::{self, IndexMap, IntoValues, ValuesMut};
 
 use super::{Entity, EntityEquivalent, EntityHash, EntitySetIterator};
