diff --git a/crates/bevy_state/macros/src/states.rs b/crates/bevy_state/macros/src/states.rs
index 841cb3ee426ad..416ae4a9f80ec 100644
--- a/crates/bevy_state/macros/src/states.rs
+++ b/crates/bevy_state/macros/src/states.rs
@@ -124,7 +124,7 @@ pub fn derive_substates(input: TokenStream) -> TokenStream {
         }
 
         impl #impl_generics #state_trait_path for #struct_name #ty_generics #where_clause {
-            const DEPENDENCY_DEPTH : usize = <Self as #trait_path>::SourceStates::SET_DEPENDENCY_DEPTH + 1;
+            const DEPENDENCY_DEPTH : usize = <<Self as #trait_path>::SourceStates as #state_set_trait_path>::SET_DEPENDENCY_DEPTH + 1;
         }
 
         impl #impl_generics #state_mutation_trait_path for #struct_name #ty_generics #where_clause {
