diff --git a/examples/testbed/ui.rs b/examples/testbed/ui.rs
index 2b7adfe3d05a5..dd8ca1bcbdab8 100644
--- a/examples/testbed/ui.rs
+++ b/examples/testbed/ui.rs
@@ -164,468 +164,507 @@ mod text {
             DespawnOnExit(super::Scene::Text),
         ));
 
-        commands.spawn((
-            Node {
-                left: px(100.),
-                top: px(230.),
-                ..Default::default()
-            },
-            Text::new("white "),
-            TextFont {
-                font: asset_server.load("fonts/FiraSans-Bold.ttf").into(),
-                ..default()
-            },
-            DespawnOnExit(super::Scene::Text),
-            children![
-                (TextSpan::new("red "), TextColor(RED.into()),),
-                (TextSpan::new("green "), TextColor(GREEN.into()),),
-                (TextSpan::new("blue "), TextColor(BLUE.into()),),
-                (
-                    TextSpan::new("black"),
-                    TextColor(Color::BLACK),
-                    TextFont {
-                        font: asset_server.load("fonts/FiraSans-Bold.ttf").into(),
-                        ..default()
-                    },
-                    TextBackgroundColor(Color::WHITE)
-                ),
-            ],
-        ));
+        for hinting in [FontHinting::Enabled, FontHinting::Disabled] {
+            let mut top = 230.;
+            let left = px(5.
+                + if hinting == FontHinting::Disabled {
+                    650.
+                } else {
+                    0.
+                });
 
-        commands.spawn((
-            Node {
-                left: px(100.),
-                top: px(260.),
-                ..Default::default()
-            },
-            Text::new(""),
-            TextFont {
-                font: asset_server.load("fonts/FiraSans-Bold.ttf").into(),
-                ..default()
-            },
-            DespawnOnExit(super::Scene::Text),
-            children![
-                (
-                    TextSpan::new("white "),
-                    TextFont {
-                        font: asset_server.load("fonts/FiraSans-Bold.ttf").into(),
-                        ..default()
-                    }
-                ),
-                (TextSpan::new("red "), TextColor(RED.into()),),
-                (TextSpan::new("green "), TextColor(GREEN.into()),),
-                (TextSpan::new("blue "), TextColor(BLUE.into()),),
-                (
-                    TextSpan::new("black"),
-                    TextColor(Color::BLACK),
-                    TextFont {
-                        font: asset_server.load("fonts/FiraSans-Bold.ttf").into(),
-                        ..default()
-                    },
-                    TextBackgroundColor(Color::WHITE)
-                ),
-            ],
-        ));
+            commands.spawn((
+                Node {
+                    left,
+                    top: px(top),
+                    ..Default::default()
+                },
+                Text::new(format!("FontHinting::{:?}", hinting)),
+                TextFont {
+                    font: asset_server.load("fonts/FiraSans-Bold.ttf").into(),
+                    ..default()
+                },
+                hinting,
+                DespawnOnExit(super::Scene::Text),
+            ));
 
-        let mut top = 300.;
-        commands.spawn((
-            Node {
-                left: px(100.),
-                top: px(top),
-                ..Default::default()
-            },
-            Text::new(""),
-            TextFont {
-                font: asset_server.load("fonts/FiraSans-Bold.ttf").into(),
-                ..default()
-            },
-            DespawnOnExit(super::Scene::Text),
-            children![
-                (TextSpan::new(""), TextColor(YELLOW.into()),),
-                TextSpan::new(""),
-                (
-                    TextSpan::new("white "),
-                    TextFont {
-                        font: asset_server.load("fonts/FiraSans-Bold.ttf").into(),
-                        ..default()
-                    }
-                ),
-                TextSpan::new(""),
-                (TextSpan::new("red "), TextColor(RED.into()),),
-                TextSpan::new(""),
-                TextSpan::new(""),
-                (TextSpan::new("green "), TextColor(GREEN.into()),),
-                (TextSpan::new(""), TextColor(YELLOW.into()),),
-                (TextSpan::new("blue "), TextColor(BLUE.into()),),
-                TextSpan::new(""),
-                (TextSpan::new(""), TextColor(YELLOW.into()),),
-                (
-                    TextSpan::new("black"),
-                    TextColor(Color::BLACK),
-                    TextFont {
-                        font: asset_server.load("fonts/FiraSans-Bold.ttf").into(),
-                        ..default()
-                    },
-                    TextBackgroundColor(Color::WHITE)
-                ),
-                TextSpan::new(""),
-            ],
-        ));
+            top += 30.;
 
-        top += 35.;
-        commands.spawn((
-            Node {
-                left: px(100.),
-                top: px(top),
-                ..Default::default()
-            },
-            Text::new("FiraSans_"),
-            TextFont {
-                font: asset_server.load("fonts/FiraSans-Bold.ttf").into(),
-                font_size: FontSize::Px(25.),
-                ..default()
-            },
-            DespawnOnExit(super::Scene::Text),
-            children![
-                (
-                    TextSpan::new("MonaSans_"),
-                    TextFont {
-                        font: asset_server.load("fonts/MonaSans-VariableFont.ttf").into(),
-                        font_size: FontSize::Px(25.),
-                        ..default()
-                    }
-                ),
-                (
-                    TextSpan::new("EBGaramond_"),
-                    TextFont {
-                        font: asset_server.load("fonts/EBGaramond12-Regular.otf").into(),
-                        font_size: FontSize::Px(25.),
-                        ..default()
-                    },
-                ),
-                (
-                    TextSpan::new("FiraMono"),
-                    TextFont {
-                        font: asset_server.load("fonts/FiraMono-Medium.ttf").into(),
-                        font_size: FontSize::Px(25.),
-                        ..default()
-                    },
-                ),
-            ],
-        ));
-        top += 35.;
-        commands.spawn((
-            Node {
-                left: px(100.),
-                top: px(top),
-                ..Default::default()
-            },
-            Text::new("FiraSans "),
-            TextFont {
-                font: asset_server.load("fonts/FiraSans-Bold.ttf").into(),
-                font_size: FontSize::Px(25.),
-                ..default()
-            },
-            DespawnOnExit(super::Scene::Text),
-            children![
-                (
-                    TextSpan::new("MonaSans "),
-                    TextFont {
-                        font: asset_server.load("fonts/MonaSans-VariableFont.ttf").into(),
-                        font_size: FontSize::Px(25.),
-                        ..default()
-                    }
-                ),
-                (
-                    TextSpan::new("EBGaramond "),
-                    TextFont {
-                        font: asset_server.load("fonts/EBGaramond12-Regular.otf").into(),
-                        font_size: FontSize::Px(25.),
-                        ..default()
-                    },
-                ),
-                (
-                    TextSpan::new("FiraMono"),
-                    TextFont {
-                        font: asset_server.load("fonts/FiraMono-Medium.ttf").into(),
-                        font_size: FontSize::Px(25.),
-                        ..default()
-                    },
-                ),
-            ],
-        ));
+            commands.spawn((
+                Node {
+                    left,
+                    top: px(top),
+                    ..Default::default()
+                },
+                Text::new("white "),
+                TextFont {
+                    font: asset_server.load("fonts/FiraSans-Bold.ttf").into(),
+                    ..default()
+                },
+                hinting,
+                DespawnOnExit(super::Scene::Text),
+                children![
+                    (TextSpan::new("red "), TextColor(RED.into()),),
+                    (TextSpan::new("green "), TextColor(GREEN.into()),),
+                    (TextSpan::new("blue "), TextColor(BLUE.into()),),
+                    (
+                        TextSpan::new("black"),
+                        TextColor(Color::BLACK),
+                        TextFont {
+                            font: asset_server.load("fonts/FiraSans-Bold.ttf").into(),
+                            ..default()
+                        },
+                        TextBackgroundColor(Color::WHITE)
+                    ),
+                ],
+            ));
 
-        top += 35.;
-        commands.spawn((
-            Node {
-                left: px(100.),
-                top: px(top),
-                ..Default::default()
-            },
-            Text::new("FiraSans "),
-            TextFont {
-                font: asset_server.load("fonts/FiraSans-Bold.ttf").into(),
-                font_size: FontSize::Px(25.),
-                ..default()
-            },
-            DespawnOnExit(super::Scene::Text),
-            children![
-                (
-                    TextSpan::new("MonaSans_"),
-                    TextFont {
-                        font: asset_server.load("fonts/MonaSans-VariableFont.ttf").into(),
-                        font_size: FontSize::Px(25.),
-                        ..default()
-                    }
-                ),
-                (
-                    TextSpan::new("EBGaramond "),
-                    TextFont {
-                        font: asset_server.load("fonts/EBGaramond12-Regular.otf").into(),
-                        font_size: FontSize::Px(25.),
-                        ..default()
-                    },
-                ),
-                (
-                    TextSpan::new("FiraMono"),
-                    TextFont {
-                        font: asset_server.load("fonts/FiraMono-Medium.ttf").into(),
-                        font_size: FontSize::Px(25.),
-                        ..default()
-                    },
-                ),
-            ],
-        ));
+            top += 30.;
+            commands.spawn((
+                Node {
+                    left,
+                    top: px(top),
+                    ..Default::default()
+                },
+                Text::new(""),
+                TextFont {
+                    font: asset_server.load("fonts/FiraSans-Bold.ttf").into(),
+                    ..default()
+                },
+                hinting,
+                DespawnOnExit(super::Scene::Text),
+                children![
+                    (
+                        TextSpan::new("white "),
+                        TextFont {
+                            font: asset_server.load("fonts/FiraSans-Bold.ttf").into(),
+                            ..default()
+                        }
+                    ),
+                    (TextSpan::new("red "), TextColor(RED.into()),),
+                    (TextSpan::new("green "), TextColor(GREEN.into()),),
+                    (TextSpan::new("blue "), TextColor(BLUE.into()),),
+                    (
+                        TextSpan::new("black"),
+                        TextColor(Color::BLACK),
+                        TextFont {
+                            font: asset_server.load("fonts/FiraSans-Bold.ttf").into(),
+                            ..default()
+                        },
+                        TextBackgroundColor(Color::WHITE)
+                    ),
+                ],
+            ));
 
-        top += 35.;
-        commands.spawn((
-            Node {
-                left: px(100.),
-                top: px(top),
-                ..Default::default()
-            },
-            Text::new("FiraSans"),
-            TextFont {
-                font: asset_server.load("fonts/FiraSans-Bold.ttf").into(),
-                font_size: FontSize::Px(25.),
-                ..default()
-            },
-            DespawnOnExit(super::Scene::Text),
-            children![
-                TextSpan::new(" "),
-                (
-                    TextSpan::new("MonaSans"),
-                    TextFont {
-                        font: asset_server.load("fonts/MonaSans-VariableFont.ttf").into(),
-                        font_size: FontSize::Px(25.),
-                        ..default()
-                    }
-                ),
-                TextSpan::new(" "),
-                (
-                    TextSpan::new("EBGaramond"),
-                    TextFont {
-                        font: asset_server.load("fonts/EBGaramond12-Regular.otf").into(),
-                        font_size: FontSize::Px(25.),
-                        ..default()
-                    },
-                ),
-                TextSpan::new(" "),
-                (
-                    TextSpan::new("FiraMono"),
-                    TextFont {
-                        font: asset_server.load("fonts/FiraMono-Medium.ttf").into(),
-                        font_size: FontSize::Px(25.),
-                        ..default()
-                    },
-                ),
-            ],
-        ));
+            top += 30.;
+            commands.spawn((
+                Node {
+                    left,
+                    top: px(top),
+                    ..Default::default()
+                },
+                Text::new(""),
+                TextFont {
+                    font: asset_server.load("fonts/FiraSans-Bold.ttf").into(),
+                    ..default()
+                },
+                hinting,
+                DespawnOnExit(super::Scene::Text),
+                children![
+                    (TextSpan::new(""), TextColor(YELLOW.into()),),
+                    TextSpan::new(""),
+                    (
+                        TextSpan::new("white "),
+                        TextFont {
+                            font: asset_server.load("fonts/FiraSans-Bold.ttf").into(),
+                            ..default()
+                        }
+                    ),
+                    TextSpan::new(""),
+                    (TextSpan::new("red "), TextColor(RED.into()),),
+                    TextSpan::new(""),
+                    TextSpan::new(""),
+                    (TextSpan::new("green "), TextColor(GREEN.into()),),
+                    (TextSpan::new(""), TextColor(YELLOW.into()),),
+                    (TextSpan::new("blue "), TextColor(BLUE.into()),),
+                    TextSpan::new(""),
+                    (TextSpan::new(""), TextColor(YELLOW.into()),),
+                    (
+                        TextSpan::new("black"),
+                        TextColor(Color::BLACK),
+                        TextFont {
+                            font: asset_server.load("fonts/FiraSans-Bold.ttf").into(),
+                            ..default()
+                        },
+                        TextBackgroundColor(Color::WHITE)
+                    ),
+                    TextSpan::new(""),
+                ],
+            ));
 
-        top += 35.;
-        commands.spawn((
-            Node {
-                left: px(100.),
-                top: px(top),
-                ..Default::default()
-            },
-            Text::new("Fira Sans_"),
-            TextFont {
-                font: asset_server.load("fonts/FiraSans-Bold.ttf").into(),
-                font_size: FontSize::Px(25.),
-                ..default()
-            },
-            DespawnOnExit(super::Scene::Text),
-            children![
-                (
-                    TextSpan::new("Mona Sans_"),
-                    TextFont {
-                        font: asset_server.load("fonts/MonaSans-VariableFont.ttf").into(),
-                        font_size: FontSize::Px(25.),
-                        ..default()
-                    }
-                ),
-                (
-                    TextSpan::new("EB Garamond_"),
-                    TextFont {
-                        font: asset_server.load("fonts/EBGaramond12-Regular.otf").into(),
-                        font_size: FontSize::Px(25.),
-                        ..default()
-                    },
-                ),
-                (
-                    TextSpan::new("Fira Mono"),
-                    TextFont {
-                        font: asset_server.load("fonts/FiraMono-Medium.ttf").into(),
-                        font_size: FontSize::Px(25.),
-                        ..default()
-                    },
-                ),
-            ],
-        ));
+            top += 35.;
+            commands.spawn((
+                Node {
+                    left,
+                    top: px(top),
+                    ..Default::default()
+                },
+                hinting,
+                Text::new("FiraSans_"),
+                TextFont {
+                    font: asset_server.load("fonts/FiraSans-Bold.ttf").into(),
+                    font_size: FontSize::Px(25.),
+                    ..default()
+                },
+                DespawnOnExit(super::Scene::Text),
+                children![
+                    (
+                        TextSpan::new("MonaSans_"),
+                        TextFont {
+                            font: asset_server.load("fonts/MonaSans-VariableFont.ttf").into(),
+                            font_size: FontSize::Px(25.),
+                            ..default()
+                        }
+                    ),
+                    (
+                        TextSpan::new("EBGaramond_"),
+                        TextFont {
+                            font: asset_server.load("fonts/EBGaramond12-Regular.otf").into(),
+                            font_size: FontSize::Px(25.),
+                            ..default()
+                        },
+                    ),
+                    (
+                        TextSpan::new("FiraMono"),
+                        TextFont {
+                            font: asset_server.load("fonts/FiraMono-Medium.ttf").into(),
+                            font_size: FontSize::Px(25.),
+                            ..default()
+                        },
+                    ),
+                ],
+            ));
+            top += 35.;
+            commands.spawn((
+                Node {
+                    left,
+                    top: px(top),
+                    ..Default::default()
+                },
+                hinting,
+                Text::new("FiraSans "),
+                TextFont {
+                    font: asset_server.load("fonts/FiraSans-Bold.ttf").into(),
+                    font_size: FontSize::Px(25.),
+                    ..default()
+                },
+                DespawnOnExit(super::Scene::Text),
+                children![
+                    (
+                        TextSpan::new("MonaSans "),
+                        TextFont {
+                            font: asset_server.load("fonts/MonaSans-VariableFont.ttf").into(),
+                            font_size: FontSize::Px(25.),
+                            ..default()
+                        }
+                    ),
+                    (
+                        TextSpan::new("EBGaramond "),
+                        TextFont {
+                            font: asset_server.load("fonts/EBGaramond12-Regular.otf").into(),
+                            font_size: FontSize::Px(25.),
+                            ..default()
+                        },
+                    ),
+                    (
+                        TextSpan::new("FiraMono"),
+                        TextFont {
+                            font: asset_server.load("fonts/FiraMono-Medium.ttf").into(),
+                            font_size: FontSize::Px(25.),
+                            ..default()
+                        },
+                    ),
+                ],
+            ));
 
-        top += 35.;
-        commands.spawn((
-            Node {
-                left: px(100.),
-                top: px(top),
-                ..Default::default()
-            },
-            Text::new("FontWeight(100)_"),
-            TextFont {
-                font: asset_server.load("fonts/MonaSans-VariableFont.ttf").into(),
-                font_size: FontSize::Px(25.),
-                weight: FontWeight(100),
-                ..default()
-            },
-            DespawnOnExit(super::Scene::Text),
-            children![
-                (
-                    TextSpan::new("FontWeight(500)_"),
-                    TextFont {
-                        font: asset_server.load("fonts/MonaSans-VariableFont.ttf").into(),
-                        font_size: FontSize::Px(25.),
-                        weight: FontWeight(500),
-                        ..default()
-                    }
-                ),
-                (
-                    TextSpan::new("FontWeight(900)"),
-                    TextFont {
-                        font: asset_server.load("fonts/MonaSans-VariableFont.ttf").into(),
-                        font_size: FontSize::Px(25.),
-                        weight: FontWeight(900),
-                        ..default()
-                    },
-                ),
-            ],
-        ));
+            top += 35.;
+            commands.spawn((
+                Node {
+                    left,
+                    top: px(top),
+                    ..Default::default()
+                },
+                hinting,
+                Text::new("FiraSans "),
+                TextFont {
+                    font: asset_server.load("fonts/FiraSans-Bold.ttf").into(),
+                    font_size: FontSize::Px(25.),
+                    ..default()
+                },
+                DespawnOnExit(super::Scene::Text),
+                children![
+                    (
+                        TextSpan::new("MonaSans_"),
+                        TextFont {
+                            font: asset_server.load("fonts/MonaSans-VariableFont.ttf").into(),
+                            font_size: FontSize::Px(25.),
+                            ..default()
+                        }
+                    ),
+                    (
+                        TextSpan::new("EBGaramond "),
+                        TextFont {
+                            font: asset_server.load("fonts/EBGaramond12-Regular.otf").into(),
+                            font_size: FontSize::Px(25.),
+                            ..default()
+                        },
+                    ),
+                    (
+                        TextSpan::new("FiraMono"),
+                        TextFont {
+                            font: asset_server.load("fonts/FiraMono-Medium.ttf").into(),
+                            font_size: FontSize::Px(25.),
+                            ..default()
+                        },
+                    ),
+                ],
+            ));
 
-        top += 35.;
-        commands.spawn((
-            Node {
-                left: px(100.),
-                top: px(top),
-                ..Default::default()
-            },
-            Text::new("FiraSans_"),
-            TextFont {
-                font: asset_server.load("fonts/FiraSans-Bold.ttf").into(),
-                font_size: FontSize::Px(25.),
-                weight: FontWeight(900),
-                ..default()
-            },
-            DespawnOnExit(super::Scene::Text),
-            children![
-                (
-                    TextSpan::new("MonaSans_"),
-                    TextFont {
-                        font: asset_server.load("fonts/MonaSans-VariableFont.ttf").into(),
-                        font_size: FontSize::Px(25.),
-                        weight: FontWeight(700),
-                        ..default()
-                    }
-                ),
-                (
-                    TextSpan::new("EBGaramond_"),
-                    TextFont {
-                        font: asset_server.load("fonts/EBGaramond12-Regular.otf").into(),
-                        font_size: FontSize::Px(25.),
-                        weight: FontWeight(500),
-                        ..default()
-                    },
-                ),
-                (
-                    TextSpan::new("FiraMono"),
-                    TextFont {
-                        font: asset_server.load("fonts/FiraMono-Medium.ttf").into(),
-                        font_size: FontSize::Px(25.),
-                        weight: FontWeight(300),
-                        ..default()
-                    },
-                ),
-            ],
-        ));
+            top += 35.;
+            commands.spawn((
+                Node {
+                    left,
+                    top: px(top),
+                    ..Default::default()
+                },
+                hinting,
+                Text::new("FiraSans"),
+                TextFont {
+                    font: asset_server.load("fonts/FiraSans-Bold.ttf").into(),
+                    font_size: FontSize::Px(25.),
+                    ..default()
+                },
+                DespawnOnExit(super::Scene::Text),
+                children![
+                    TextSpan::new(" "),
+                    (
+                        TextSpan::new("MonaSans"),
+                        TextFont {
+                            font: asset_server.load("fonts/MonaSans-VariableFont.ttf").into(),
+                            font_size: FontSize::Px(25.),
+                            ..default()
+                        }
+                    ),
+                    TextSpan::new(" "),
+                    (
+                        TextSpan::new("EBGaramond"),
+                        TextFont {
+                            font: asset_server.load("fonts/EBGaramond12-Regular.otf").into(),
+                            font_size: FontSize::Px(25.),
+                            ..default()
+                        },
+                    ),
+                    TextSpan::new(" "),
+                    (
+                        TextSpan::new("FiraMono"),
+                        TextFont {
+                            font: asset_server.load("fonts/FiraMono-Medium.ttf").into(),
+                            font_size: FontSize::Px(25.),
+                            ..default()
+                        },
+                    ),
+                ],
+            ));
 
-        top += 35.;
-        commands.spawn((
-            Node {
-                left: px(100.),
-                top: px(top),
-                ..Default::default()
-            },
-            Text::new("FiraSans\t"),
-            TextFont {
-                font: asset_server.load("fonts/FiraSans-Bold.ttf").into(),
-                font_size: FontSize::Px(25.),
-                ..default()
-            },
-            DespawnOnExit(super::Scene::Text),
-            children![
-                (
-                    TextSpan::new("MonaSans\t"),
-                    TextFont {
-                        font: asset_server.load("fonts/MonaSans-VariableFont.ttf").into(),
-                        font_size: FontSize::Px(25.),
-                        ..default()
-                    }
-                ),
-                (
-                    TextSpan::new("EBGaramond\t"),
-                    TextFont {
-                        font: asset_server.load("fonts/EBGaramond12-Regular.otf").into(),
-                        font_size: FontSize::Px(25.),
-                        ..default()
-                    },
-                ),
-                (
-                    TextSpan::new("FiraMono"),
-                    TextFont {
-                        font: asset_server.load("fonts/FiraMono-Medium.ttf").into(),
-                        font_size: FontSize::Px(25.),
-                        ..default()
-                    },
-                ),
-            ],
-        ));
+            top += 35.;
+            commands.spawn((
+                Node {
+                    left,
+                    top: px(top),
+                    ..Default::default()
+                },
+                hinting,
+                Text::new("Fira Sans_"),
+                TextFont {
+                    font: asset_server.load("fonts/FiraSans-Bold.ttf").into(),
+                    font_size: FontSize::Px(25.),
+                    ..default()
+                },
+                DespawnOnExit(super::Scene::Text),
+                children![
+                    (
+                        TextSpan::new("Mona Sans_"),
+                        TextFont {
+                            font: asset_server.load("fonts/MonaSans-VariableFont.ttf").into(),
+                            font_size: FontSize::Px(25.),
+                            ..default()
+                        }
+                    ),
+                    (
+                        TextSpan::new("EB Garamond_"),
+                        TextFont {
+                            font: asset_server.load("fonts/EBGaramond12-Regular.otf").into(),
+                            font_size: FontSize::Px(25.),
+                            ..default()
+                        },
+                    ),
+                    (
+                        TextSpan::new("Fira Mono"),
+                        TextFont {
+                            font: asset_server.load("fonts/FiraMono-Medium.ttf").into(),
+                            font_size: FontSize::Px(25.),
+                            ..default()
+                        },
+                    ),
+                ],
+            ));
 
-        for font_smoothing in [FontSmoothing::AntiAliased, FontSmoothing::None] {
             top += 35.;
             commands.spawn((
                 Node {
-                    left: px(100.),
+                    left,
                     top: px(top),
                     ..Default::default()
                 },
-                Text::new(format!("FontSmoothing::{:?}", font_smoothing)),
+                hinting,
+                Text::new("FontWeight(100)_"),
                 TextFont {
                     font: asset_server.load("fonts/MonaSans-VariableFont.ttf").into(),
                     font_size: FontSize::Px(25.),
-                    font_smoothing,
+                    weight: FontWeight(100),
+                    ..default()
+                },
+                DespawnOnExit(super::Scene::Text),
+                children![
+                    (
+                        TextSpan::new("FontWeight(500)_"),
+                        TextFont {
+                            font: asset_server.load("fonts/MonaSans-VariableFont.ttf").into(),
+                            font_size: FontSize::Px(25.),
+                            weight: FontWeight(500),
+                            ..default()
+                        }
+                    ),
+                    (
+                        TextSpan::new("FontWeight(900)"),
+                        TextFont {
+                            font: asset_server.load("fonts/MonaSans-VariableFont.ttf").into(),
+                            font_size: FontSize::Px(25.),
+                            weight: FontWeight(900),
+                            ..default()
+                        },
+                    ),
+                ],
+            ));
+
+            top += 35.;
+            commands.spawn((
+                Node {
+                    left,
+                    top: px(top),
+                    ..Default::default()
+                },
+                hinting,
+                Text::new("FiraSans_"),
+                TextFont {
+                    font: asset_server.load("fonts/FiraSans-Bold.ttf").into(),
+                    font_size: FontSize::Px(25.),
+                    weight: FontWeight(900),
+                    ..default()
+                },
+                DespawnOnExit(super::Scene::Text),
+                children![
+                    (
+                        TextSpan::new("MonaSans_"),
+                        TextFont {
+                            font: asset_server.load("fonts/MonaSans-VariableFont.ttf").into(),
+                            font_size: FontSize::Px(25.),
+                            weight: FontWeight(700),
+                            ..default()
+                        }
+                    ),
+                    (
+                        TextSpan::new("EBGaramond_"),
+                        TextFont {
+                            font: asset_server.load("fonts/EBGaramond12-Regular.otf").into(),
+                            font_size: FontSize::Px(25.),
+                            weight: FontWeight(500),
+                            ..default()
+                        },
+                    ),
+                    (
+                        TextSpan::new("FiraMono"),
+                        TextFont {
+                            font: asset_server.load("fonts/FiraMono-Medium.ttf").into(),
+                            font_size: FontSize::Px(25.),
+                            weight: FontWeight(300),
+                            ..default()
+                        },
+                    ),
+                ],
+            ));
+
+            top += 35.;
+            commands.spawn((
+                Node {
+                    left,
+                    top: px(top),
+                    ..Default::default()
+                },
+                hinting,
+                Text::new("FiraSans\t"),
+                TextFont {
+                    font: asset_server.load("fonts/FiraSans-Bold.ttf").into(),
+                    font_size: FontSize::Px(25.),
                     ..default()
                 },
                 DespawnOnExit(super::Scene::Text),
+                children![
+                    (
+                        TextSpan::new("MonaSans\t"),
+                        TextFont {
+                            font: asset_server.load("fonts/MonaSans-VariableFont.ttf").into(),
+                            font_size: FontSize::Px(25.),
+                            ..default()
+                        }
+                    ),
+                    (
+                        TextSpan::new("EBGaramond\t"),
+                        TextFont {
+                            font: asset_server.load("fonts/EBGaramond12-Regular.otf").into(),
+                            font_size: FontSize::Px(25.),
+                            ..default()
+                        },
+                    ),
+                    (
+                        TextSpan::new("FiraMono"),
+                        TextFont {
+                            font: asset_server.load("fonts/FiraMono-Medium.ttf").into(),
+                            font_size: FontSize::Px(25.),
+                            ..default()
+                        },
+                    ),
+                ],
             ));
+
+            for font_smoothing in [FontSmoothing::AntiAliased, FontSmoothing::None] {
+                top += 35.;
+                commands.spawn((
+                    Node {
+                        left,
+                        top: px(top),
+                        ..Default::default()
+                    },
+                    Text::new(format!("FontSmoothing::{:?}", font_smoothing)),
+                    TextFont {
+                        font: asset_server.load("fonts/MonaSans-VariableFont.ttf").into(),
+                        font_size: FontSize::Px(25.),
+                        font_smoothing,
+                        ..default()
+                    },
+                    DespawnOnExit(super::Scene::Text),
+                ));
+            }
         }
     }
 }
