diff --git a/crates/bevy_anti_alias/src/smaa/mod.rs b/crates/bevy_anti_alias/src/smaa/mod.rs
index f04b484a12c0c..bb496e9cabc2a 100644
--- a/crates/bevy_anti_alias/src/smaa/mod.rs
+++ b/crates/bevy_anti_alias/src/smaa/mod.rs
@@ -41,6 +41,8 @@ use bevy_derive::{Deref, DerefMut};
 use bevy_ecs::{
     component::Component,
     entity::Entity,
+    lifecycle::Remove,
+    observer::On,
     query::With,
     reflect::ReflectComponent,
     resource::Resource,
@@ -329,6 +331,17 @@ impl Plugin for SmaaPlugin {
             return;
         };
 
+        // TODO: remove this manual cleanup when ExtractComponent gets support
+        // for cleanup of derived components
+        render_app.add_observer(|event: On<Remove, Smaa>, mut commands: Commands| {
+            commands.entity(event.entity).remove::<(
+                SmaaTextures,
+                SmaaPipelines,
+                SmaaBindGroups,
+                ViewSmaaPipelines,
+            )>();
+        });
+
         render_app
             .insert_resource(smaa_luts)
             .init_resource::<SmaaSpecializedRenderPipelines>()
diff --git a/crates/bevy_app/src/sub_app.rs b/crates/bevy_app/src/sub_app.rs
index b229da2ef59e6..7790fbf9c7683 100644
--- a/crates/bevy_app/src/sub_app.rs
+++ b/crates/bevy_app/src/sub_app.rs
@@ -2,6 +2,7 @@ use crate::{App, AppLabel, InternedAppLabel, Plugin, Plugins, PluginsState};
 use alloc::{boxed::Box, string::String, vec::Vec};
 use bevy_ecs::{
     message::MessageRegistry,
+    observer::IntoObserver,
     prelude::*,
     schedule::{
         InternedScheduleLabel, InternedSystemSet, ScheduleBuildSettings, ScheduleCleanupPolicy,
@@ -359,6 +360,12 @@ impl SubApp {
         self
     }
 
+    /// See [`App::add_observer`].
+    pub fn add_observer<M>(&mut self, observer: impl IntoObserver<M>) -> &mut Self {
+        self.world_mut().add_observer(observer);
+        self
+    }
+
     /// See [`App::add_message`].
     pub fn add_message<T>(&mut self) -> &mut Self
     where
