diff --git a/crates/bevy_pbr/src/render/light.rs b/crates/bevy_pbr/src/render/light.rs
index 65bd3f8ba0c11..c9dace7006420 100644
--- a/crates/bevy_pbr/src/render/light.rs
+++ b/crates/bevy_pbr/src/render/light.rs
@@ -1388,6 +1388,13 @@ pub fn prepare_lights(
                 continue;
             };
 
+            if !light.shadow_maps_enabled {
+                if let Some(entities) = light_view_entities.remove(&entity) {
+                    despawn_entities(&mut commands, entities);
+                }
+                continue;
+            }
+
             let light_index = *global_clusterable_object_meta
                 .entity_to_index
                 .get(&light_entity)
