diff --git a/crates/bevy_winit/src/state.rs b/crates/bevy_winit/src/state.rs
index 5b873d462039a..8319181e69ccd 100644
--- a/crates/bevy_winit/src/state.rs
+++ b/crates/bevy_winit/src/state.rs
@@ -529,7 +529,11 @@ impl<T: BufferedEvent> ApplicationHandler<T> for WinitAppRunnerState<T> {
                     .iter()
                     .all(|(_, w)| !w.is_visible().unwrap_or(false));
                 if !exiting
-                    && (self.startup_forced_updates > 0 || headless || all_invisible || reactive)
+                    && (self.startup_forced_updates > 0
+                        || headless
+                        || all_invisible
+                        || reactive
+                        || self.window_event_received)
                 {
                     self.redraw_requested(event_loop);
                 }
