diff --git a/crates/bevy_camera/src/primitives.rs b/crates/bevy_camera/src/primitives.rs
index 127a2d9c29e1a..32bb557b93774 100644
--- a/crates/bevy_camera/src/primitives.rs
+++ b/crates/bevy_camera/src/primitives.rs
@@ -442,7 +442,7 @@ pub enum CubemapLayout {
     /// layout in a vertical sequence
     /// ```text
     ///   +x
-    ///   -y
+    ///   -x
     ///   +y
     ///   -y
     ///   -z
@@ -451,7 +451,7 @@ pub enum CubemapLayout {
     SequenceVertical = 2,
     /// layout in a horizontal sequence
     /// ```text
-    /// +x -y +y -y -z +z
+    /// +x -x +y -y -z +z
     /// ```
     SequenceHorizontal = 3,
 }
