diff --git a/crates/bevy_ui/src/ui_node.rs b/crates/bevy_ui/src/ui_node.rs
index 1c5ed364b98f9..d81f80f8b9626 100644
--- a/crates/bevy_ui/src/ui_node.rs
+++ b/crates/bevy_ui/src/ui_node.rs
@@ -2875,6 +2875,8 @@ impl ComputedNodeTarget {
 }
 
 /// Adds a shadow behind text
+///
+/// Not supported by `Text2d`
 #[derive(Component, Copy, Clone, Debug, PartialEq, Reflect)]
 #[reflect(Component, Default, Debug, Clone, PartialEq)]
 pub struct TextShadow {
