diff --git a/crates/bevy_diagnostic/src/system_information_diagnostics_plugin.rs b/crates/bevy_diagnostic/src/system_information_diagnostics_plugin.rs
index 768bbb0828844..83d3663895ca5 100644
--- a/crates/bevy_diagnostic/src/system_information_diagnostics_plugin.rs
+++ b/crates/bevy_diagnostic/src/system_information_diagnostics_plugin.rs
@@ -236,7 +236,7 @@ pub mod internal {
                 memory: format!("{:.1} GiB", sys.total_memory() as f64 * BYTES_TO_GIB),
             };
 
-            info!("{:?}", system_info);
+            info!("{system_info:?}");
             system_info
         }
     }
