diff --git a/crates/bevy_render/src/camera/camera.rs b/crates/bevy_render/src/camera/camera.rs
index f19f57fa5f9dd..d97950240d779 100644
--- a/crates/bevy_render/src/camera/camera.rs
+++ b/crates/bevy_render/src/camera/camera.rs
@@ -1060,6 +1060,7 @@ pub fn camera_system(
 #[reflect(opaque)]
 #[reflect(Component, Default, Clone)]
 pub struct CameraMainTextureUsages(pub TextureUsages);
+
 impl Default for CameraMainTextureUsages {
     fn default() -> Self {
         Self(
@@ -1070,6 +1071,13 @@ impl Default for CameraMainTextureUsages {
     }
 }
 
+impl CameraMainTextureUsages {
+    pub fn with(mut self, usages: TextureUsages) -> Self {
+        self.0 |= usages;
+        self
+    }
+}
+
 #[derive(Component, Debug)]
 pub struct ExtractedCamera {
     pub target: Option<NormalizedRenderTarget>,
