diff --git a/tools/compile_fail_utils/Cargo.toml b/tools/compile_fail_utils/Cargo.toml
index 453b7ea7ad9f1..8668fdaa9e94f 100644
--- a/tools/compile_fail_utils/Cargo.toml
+++ b/tools/compile_fail_utils/Cargo.toml
@@ -8,7 +8,7 @@ license = "MIT OR Apache-2.0"
 publish = false
 
 [dependencies]
-ui_test = "0.29.1"
+ui_test = "0.30.1"
 
 [[test]]
 name = "example"
diff --git a/tools/compile_fail_utils/src/lib.rs b/tools/compile_fail_utils/src/lib.rs
index ecd4f34c687d7..28157400a7cdb 100644
--- a/tools/compile_fail_utils/src/lib.rs
+++ b/tools/compile_fail_utils/src/lib.rs
@@ -123,7 +123,8 @@ pub fn test_with_multiple_configs(
     let emitter: Box<dyn StatusEmitter + Send> = if env::var_os("CI").is_some() {
         Box::new((
             Text::verbose(),
-            Gha::<true> {
+            Gha {
+                group: true,
                 name: test_name.into(),
             },
         ))
