diff --git a/examples/3d/pbr.rs b/examples/3d/pbr.rs
index 12922db03b1fa..da654ce1f3114 100644
--- a/examples/3d/pbr.rs
+++ b/examples/3d/pbr.rs
@@ -85,8 +85,8 @@ fn setup(
             right: Val::ZERO,
             ..default()
         },
-        Transform {
-            rotation: Quat::from_rotation_z(std::f32::consts::PI / 2.0),
+        UiTransform {
+            rotation: Rot2::degrees(90.),
             ..default()
         },
     ));
