diff --git a/clippy.toml b/clippy.toml
index 2c98e8ed02c30..372ffbaf0beae 100644
--- a/clippy.toml
+++ b/clippy.toml
@@ -41,7 +41,6 @@ disallowed-methods = [
   { path = "f32::asinh", reason = "use bevy_math::ops::asinh instead for libm determinism" },
   { path = "f32::acosh", reason = "use bevy_math::ops::acosh instead for libm determinism" },
   { path = "f32::atanh", reason = "use bevy_math::ops::atanh instead for libm determinism" },
-  { path = "criterion::black_box", reason = "use core::hint::black_box instead" },
 ]
 
 # Require `bevy_ecs::children!` to use `[]` braces, instead of `()` or `{}`.
diff --git a/crates/bevy_macro_utils/src/bevy_manifest.rs b/crates/bevy_macro_utils/src/bevy_manifest.rs
index 8d327810695a9..b6df0e0e0f89c 100644
--- a/crates/bevy_macro_utils/src/bevy_manifest.rs
+++ b/crates/bevy_macro_utils/src/bevy_manifest.rs
@@ -95,7 +95,7 @@ impl BevyManifest {
                 return None;
             };
 
-            let mut path = Self::parse_str::<syn::Path>(&format!("::{}", package));
+            let mut path = Self::parse_str::<syn::Path>(&format!("::{package}"));
             if let Some(module) = name.strip_prefix("bevy_") {
                 path.segments.push(Self::parse_str(module));
             }
