diff --git a/crates/bevy_winit/src/state.rs b/crates/bevy_winit/src/state.rs
index d2618e06858a0..08c9e36535bfa 100644
--- a/crates/bevy_winit/src/state.rs
+++ b/crates/bevy_winit/src/state.rs
@@ -478,7 +478,7 @@ impl<M: Message> ApplicationHandler<M> for WinitAppRunnerState<M> {
                 })
             }
 
-            if !self.app_exit.is_some()
+            if self.app_exit.is_none()
                 && (self.startup_forced_updates > 0
                     || matches!(self.update_mode, UpdateMode::Reactive { .. })
                     || self.window_event_received
