diff --git a/clippy.toml b/clippy.toml
index 372ffbaf0beae..50b4a46fe3b82 100644
--- a/clippy.toml
+++ b/clippy.toml
@@ -1,13 +1,10 @@
 doc-valid-idents = [
+  "GameActivity",
   "GilRs",
   "glTF",
-  "macOS",
   "NVidia",
   "OpenXR",
-  "sRGB",
   "VSync",
-  "WebGL2",
-  "WebGPU",
   "..",
 ]
 
diff --git a/src/lib.rs b/src/lib.rs
index 775baac22b06a..79fd8c0a70165 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -1,7 +1,7 @@
 #![cfg_attr(docsrs, feature(doc_cfg))]
 #![expect(
     clippy::doc_markdown,
-    reason = "Android GameActivity does not need to be code-formatted."
+    reason = "Clippy lints for un-backticked identifiers within the cargo features list, which we don't want."
 )]
 //! [![Bevy Logo](https://bevy.org/assets/bevy_logo_docs.svg)](https://bevy.org)
 //!
