diff --git a/examples/2d/2d_viewport_to_world.rs b/examples/2d/2d_viewport_to_world.rs
index f6bf9372f939f..9331382572da3 100644
--- a/examples/2d/2d_viewport_to_world.rs
+++ b/examples/2d/2d_viewport_to_world.rs
@@ -77,6 +77,11 @@ fn controls(
     }
 
     if let Some(viewport) = camera.viewport.as_mut() {
+        // Reset viewport size on window resize
+        if viewport.physical_size.x > window_size.x || viewport.physical_size.y > window_size.y {
+            viewport.physical_size = (window_size.as_vec2() * 0.75).as_uvec2();
+        }
+
         // Viewport movement controls
         if input.pressed(KeyCode::KeyW) {
             viewport.physical_position.y = viewport.physical_position.y.saturating_sub(uspeed);
