diff --git a/crates/bevy_input_focus/Cargo.toml b/crates/bevy_input_focus/Cargo.toml
index 1766ee2243b31..45d6acb60539c 100644
--- a/crates/bevy_input_focus/Cargo.toml
+++ b/crates/bevy_input_focus/Cargo.toml
@@ -64,7 +64,7 @@ bevy_app = { path = "../bevy_app", version = "0.18.0-dev", default-features = fa
 bevy_ecs = { path = "../bevy_ecs", version = "0.18.0-dev", default-features = false }
 bevy_input = { path = "../bevy_input", version = "0.18.0-dev", default-features = false }
 bevy_math = { path = "../bevy_math", version = "0.18.0-dev", default-features = false }
-bevy_picking = { path = "../bevy_picking", version = "0.18.0-dev", default-features = false }
+bevy_picking = { path = "../bevy_picking", version = "0.18.0-dev", default-features = false, optional = true }
 bevy_window = { path = "../bevy_window", version = "0.18.0-dev", default-features = false }
 bevy_reflect = { path = "../bevy_reflect", version = "0.18.0-dev", features = [
   "glam",
diff --git a/crates/bevy_input_focus/src/tab_navigation.rs b/crates/bevy_input_focus/src/tab_navigation.rs
index a490d00dda10e..b8ce6a8d95375 100644
--- a/crates/bevy_input_focus/src/tab_navigation.rs
+++ b/crates/bevy_input_focus/src/tab_navigation.rs
@@ -38,7 +38,6 @@ use bevy_input::{
     keyboard::{KeyCode, KeyboardInput},
     ButtonInput, ButtonState,
 };
-use bevy_picking::events::{Pointer, Press};
 use bevy_window::{PrimaryWindow, Window};
 use log::warn;
 use thiserror::Error;
@@ -346,6 +345,7 @@ impl Plugin for TabNavigationPlugin {
     fn build(&self, app: &mut App) {
         app.add_systems(Startup, setup_tab_navigation);
         app.add_observer(acquire_focus);
+        #[cfg(feature = "bevy_picking")]
         app.add_observer(click_to_focus);
     }
 }
@@ -356,8 +356,9 @@ fn setup_tab_navigation(mut commands: Commands, window: Query<Entity, With<Prima
     }
 }
 
+#[cfg(feature = "bevy_picking")]
 fn click_to_focus(
-    press: On<Pointer<Press>>,
+    press: On<bevy_picking::events::Pointer<bevy_picking::events::Press>>,
     mut focus_visible: ResMut<InputFocusVisible>,
     windows: Query<Entity, With<PrimaryWindow>>,
     mut commands: Commands,
diff --git a/crates/bevy_internal/Cargo.toml b/crates/bevy_internal/Cargo.toml
index 32784290611be..f6872757b04a7 100644
--- a/crates/bevy_internal/Cargo.toml
+++ b/crates/bevy_internal/Cargo.toml
@@ -316,7 +316,7 @@ pan_camera = ["bevy_camera_controller/pan_camera"]
 bevy_remote = ["dep:bevy_remote", "serialize"]
 
 # Provides picking functionality
-bevy_picking = ["dep:bevy_picking"]
+bevy_picking = ["dep:bevy_picking", "bevy_input_focus?/bevy_picking"]
 
 # Provides a mesh picking backend
 mesh_picking = ["bevy_picking", "bevy_picking/mesh_picking"]
