diff --git a/crates/bevy_gizmos/src/arcs.rs b/crates/bevy_gizmos/src/arcs.rs
index 41647f9fe21a2..3594a117fde03 100644
--- a/crates/bevy_gizmos/src/arcs.rs
+++ b/crates/bevy_gizmos/src/arcs.rs
@@ -17,8 +17,6 @@ where
 {
     /// Draw an arc, which is a part of the circumference of a circle, in 2D.
     ///
-    /// This should be called for each frame the arc needs to be rendered.
-    ///
     /// # Arguments
     /// - `isometry` defines the translation and rotation of the arc.
     ///   - the translation specifies the center of the arc
@@ -132,8 +130,6 @@ where
     /// - embedded in the XZ plane
     /// - rotates counterclockwise
     ///
-    /// This should be called for each frame the arc needs to be rendered.
-    ///
     /// # Arguments
     /// - `angle`: sets how much of a circle circumference is passed, e.g. PI is half a circle. This
     ///   value should be in the range (-2 * PI..=2 * PI)
diff --git a/crates/bevy_gizmos/src/arrows.rs b/crates/bevy_gizmos/src/arrows.rs
index 74fbdc97262b5..f1092889c9f37 100644
--- a/crates/bevy_gizmos/src/arrows.rs
+++ b/crates/bevy_gizmos/src/arrows.rs
@@ -108,8 +108,6 @@ where
 {
     /// Draw an arrow in 3D, from `start` to `end`. Has four tips for convenient viewing from any direction.
     ///
-    /// This should be called for each frame the arrow needs to be rendered.
-    ///
     /// # Example
     /// ```
     /// # use bevy_gizmos::prelude::*;
@@ -139,8 +137,6 @@ where
 
     /// Draw an arrow in 2D (on the xy plane), from `start` to `end`.
     ///
-    /// This should be called for each frame the arrow needs to be rendered.
-    ///
     /// # Example
     /// ```
     /// # use bevy_gizmos::prelude::*;
@@ -169,8 +165,6 @@ where
     /// Draw a set of axes local to the given transform (`transform`), with length scaled by a factor
     /// of `base_length`.
     ///
-    /// This should be called for each frame the axes need to be rendered.
-    ///
     /// # Example
     /// ```
     /// # use bevy_gizmos::prelude::*;
@@ -202,8 +196,6 @@ where
     /// Draw a set of axes local to the given transform (`transform`), with length scaled by a factor
     /// of `base_length`.
     ///
-    /// This should be called for each frame the axes need to be rendered.
-    ///
     /// # Example
     /// ```
     /// # use bevy_gizmos::prelude::*;
diff --git a/crates/bevy_gizmos/src/circles.rs b/crates/bevy_gizmos/src/circles.rs
index e5a49398083ab..07ac90ea60e43 100644
--- a/crates/bevy_gizmos/src/circles.rs
+++ b/crates/bevy_gizmos/src/circles.rs
@@ -30,8 +30,6 @@ where
     /// - the center is at `Vec3::ZERO`
     /// - the `half_sizes` are aligned with the `Vec3::X` and `Vec3::Y` axes.
     ///
-    /// This should be called for each frame the ellipse needs to be rendered.
-    ///
     /// # Example
     /// ```
     /// # use bevy_gizmos::prelude::*;
@@ -71,8 +69,6 @@ where
     /// - the center is at `Vec2::ZERO`
     /// - the `half_sizes` are aligned with the `Vec2::X` and `Vec2::Y` axes.
     ///
-    /// This should be called for each frame the ellipse needs to be rendered.
-    ///
     /// # Example
     /// ```
     /// # use bevy_gizmos::prelude::*;
@@ -151,8 +147,6 @@ where
     /// - the center is at `Vec2::ZERO`
     /// - the radius is aligned with the `Vec2::X` and `Vec2::Y` axes.
     ///
-    /// This should be called for each frame the circle needs to be rendered.
-    ///
     /// # Example
     /// ```
     /// # use bevy_gizmos::prelude::*;
@@ -193,8 +187,6 @@ where
     /// - the center is at `Vec3::ZERO`
     /// - the 3 circles are in the XY, YZ and XZ planes.
     ///
-    /// This should be called for each frame the sphere needs to be rendered.
-    ///
     /// # Example
     /// ```
     /// # use bevy_gizmos::prelude::*;
diff --git a/crates/bevy_gizmos/src/cross.rs b/crates/bevy_gizmos/src/cross.rs
index 7c3eff77f3960..b10fd8aeef58b 100644
--- a/crates/bevy_gizmos/src/cross.rs
+++ b/crates/bevy_gizmos/src/cross.rs
@@ -19,8 +19,6 @@ where
     /// - the center is at `Vec3::ZERO`
     /// - the `half_size`s are aligned with the `Vec3::X`, `Vec3::Y` and `Vec3::Z` axes.
     ///
-    /// This should be called for each frame the cross needs to be rendered.
-    ///
     /// # Example
     /// ```
     /// # use bevy_gizmos::prelude::*;
@@ -54,8 +52,6 @@ where
     /// - the center is at `Vec3::ZERO`
     /// - the `half_size`s are aligned with the `Vec3::X` and `Vec3::Y` axes.
     ///
-    /// This should be called for each frame the cross needs to be rendered.
-    ///
     /// # Example
     /// ```
     /// # use bevy_gizmos::prelude::*;
diff --git a/crates/bevy_gizmos/src/curves.rs b/crates/bevy_gizmos/src/curves.rs
index 2d7a350ca29dc..b795ed69061dc 100644
--- a/crates/bevy_gizmos/src/curves.rs
+++ b/crates/bevy_gizmos/src/curves.rs
@@ -18,8 +18,6 @@ where
 {
     /// Draw a curve, at the given time points, sampling in 2D.
     ///
-    /// This should be called for each frame the curve needs to be rendered.
-    ///
     /// Samples of time points outside of the curve's domain will be filtered out and won't
     /// contribute to the rendering. If you wish to render the curve outside of its domain you need
     /// to create a new curve with an extended domain.
@@ -52,8 +50,6 @@ where
 
     /// Draw a curve, at the given time points, sampling in 3D.
     ///
-    /// This should be called for each frame the curve needs to be rendered.
-    ///
     /// Samples of time points outside of the curve's domain will be filtered out and won't
     /// contribute to the rendering. If you wish to render the curve outside of its domain you need
     /// to create a new curve with an extended domain.
@@ -89,8 +85,6 @@ where
 
     /// Draw a curve, at the given time points, sampling in 2D, with a color gradient.
     ///
-    /// This should be called for each frame the curve needs to be rendered.
-    ///
     /// Samples of time points outside of the curve's domain will be filtered out and won't
     /// contribute to the rendering. If you wish to render the curve outside of its domain you need
     /// to create a new curve with an extended domain.
@@ -132,8 +126,6 @@ where
 
     /// Draw a curve, at the given time points, sampling in 3D, with a color gradient.
     ///
-    /// This should be called for each frame the curve needs to be rendered.
-    ///
     /// Samples of time points outside of the curve's domain will be filtered out and won't
     /// contribute to the rendering. If you wish to render the curve outside of its domain you need
     /// to create a new curve with an extended domain.
diff --git a/crates/bevy_gizmos/src/gizmos.rs b/crates/bevy_gizmos/src/gizmos.rs
index 88039631dd413..34f9f476dcaf7 100644
--- a/crates/bevy_gizmos/src/gizmos.rs
+++ b/crates/bevy_gizmos/src/gizmos.rs
@@ -376,8 +376,6 @@ where
     }
     /// Draw a line in 3D from `start` to `end`.
     ///
-    /// This should be called for each frame the line needs to be rendered.
-    ///
     /// # Example
     /// ```
     /// # use bevy_gizmos::prelude::*;
@@ -399,8 +397,6 @@ where
 
     /// Draw a line in 3D with a color gradient from `start` to `end`.
     ///
-    /// This should be called for each frame the line needs to be rendered.
-    ///
     /// # Example
     /// ```
     /// # use bevy_gizmos::prelude::*;
@@ -428,8 +424,6 @@ where
 
     /// Draw a line in 3D from `start` to `start + vector`.
     ///
-    /// This should be called for each frame the line needs to be rendered.
-    ///
     /// # Example
     /// ```
     /// # use bevy_gizmos::prelude::*;
@@ -450,8 +444,6 @@ where
 
     /// Draw a line in 3D with a color gradient from `start` to `start + vector`.
     ///
-    /// This should be called for each frame the line needs to be rendered.
-    ///
     /// # Example
     /// ```
     /// # use bevy_gizmos::prelude::*;
@@ -478,8 +470,6 @@ where
 
     /// Draw a line in 3D made of straight segments between the points.
     ///
-    /// This should be called for each frame the line needs to be rendered.
-    ///
     /// # Example
     /// ```
     /// # use bevy_gizmos::prelude::*;
@@ -508,8 +498,6 @@ where
 
     /// Draw a line in 3D made of straight segments between the points, with a color gradient.
     ///
-    /// This should be called for each frame the lines need to be rendered.
-    ///
     /// # Example
     /// ```
     /// # use bevy_gizmos::prelude::*;
@@ -560,8 +548,6 @@ where
     /// - the center is at `Vec3::ZERO`
     /// - the sizes are aligned with the `Vec3::X` and `Vec3::Y` axes.
     ///
-    /// This should be called for each frame the rectangle needs to be rendered.
-    ///
     /// # Example
     /// ```
     /// # use bevy_gizmos::prelude::*;
@@ -584,8 +570,6 @@ where
 
     /// Draw a wireframe cube in 3D.
     ///
-    /// This should be called for each frame the cube needs to be rendered.
-    ///
     /// # Example
     /// ```
     /// # use bevy_gizmos::prelude::*;
@@ -624,8 +608,6 @@ where
 
     /// Draw a wireframe aabb in 3D.
     ///
-    /// This should be called for each frame the aabb needs to be rendered.
-    ///
     /// # Example
     /// ```
     /// # use bevy_gizmos::prelude::*;
@@ -677,8 +659,6 @@ where
 
     /// Draw a line in 2D from `start` to `end`.
     ///
-    /// This should be called for each frame the line needs to be rendered.
-    ///
     /// # Example
     /// ```
     /// # use bevy_gizmos::prelude::*;
@@ -699,8 +679,6 @@ where
 
     /// Draw a line in 2D with a color gradient from `start` to `end`.
     ///
-    /// This should be called for each frame the line needs to be rendered.
-    ///
     /// # Example
     /// ```
     /// # use bevy_gizmos::prelude::*;
@@ -727,8 +705,6 @@ where
 
     /// Draw a line in 2D made of straight segments between the points.
     ///
-    /// This should be called for each frame the line needs to be rendered.
-    ///
     /// # Example
     /// ```
     /// # use bevy_gizmos::prelude::*;
@@ -753,8 +729,6 @@ where
 
     /// Draw a line in 2D made of straight segments between the points, with a color gradient.
     ///
-    /// This should be called for each frame the line needs to be rendered.
-    ///
     /// # Example
     /// ```
     /// # use bevy_gizmos::prelude::*;
@@ -786,8 +760,6 @@ where
 
     /// Draw a line in 2D from `start` to `start + vector`.
     ///
-    /// This should be called for each frame the line needs to be rendered.
-    ///
     /// # Example
     /// ```
     /// # use bevy_gizmos::prelude::*;
@@ -808,8 +780,6 @@ where
 
     /// Draw a line in 2D with a color gradient from `start` to `start + vector`.
     ///
-    /// This should be called for each frame the line needs to be rendered.
-    ///
     /// # Example
     /// ```
     /// # use bevy_gizmos::prelude::*;
@@ -841,8 +811,6 @@ where
     /// - the center is at `Vec2::ZERO`
     /// - the sizes are aligned with the `Vec2::X` and `Vec2::Y` axes.
     ///
-    /// This should be called for each frame the rectangle needs to be rendered.
-    ///
     /// # Example
     /// ```
     /// # use bevy_gizmos::prelude::*;
diff --git a/crates/bevy_gizmos/src/grid.rs b/crates/bevy_gizmos/src/grid.rs
index 453c98845baf1..dcfb25147576e 100644
--- a/crates/bevy_gizmos/src/grid.rs
+++ b/crates/bevy_gizmos/src/grid.rs
@@ -180,8 +180,6 @@ where
 {
     /// Draw a 2D grid in 3D.
     ///
-    /// This should be called for each frame the grid needs to be rendered.
-    ///
     /// The grid's default orientation aligns with the XY-plane.
     ///
     /// # Arguments
@@ -236,8 +234,6 @@ where
 
     /// Draw a 3D grid of voxel-like cells.
     ///
-    /// This should be called for each frame the grid needs to be rendered.
-    ///
     /// # Arguments
     ///
     /// - `isometry` defines the translation and rotation of the grid.
@@ -289,8 +285,6 @@ where
 
     /// Draw a grid in 2D.
     ///
-    /// This should be called for each frame the grid needs to be rendered.
-    ///
     /// # Arguments
     ///
     /// - `isometry` defines the translation and rotation of the grid.
diff --git a/crates/bevy_gizmos/src/rounded_box.rs b/crates/bevy_gizmos/src/rounded_box.rs
index 0a40d148cf0ca..31a9b4ed876df 100644
--- a/crates/bevy_gizmos/src/rounded_box.rs
+++ b/crates/bevy_gizmos/src/rounded_box.rs
@@ -233,8 +233,6 @@ where
 {
     /// Draw a wireframe rectangle with rounded corners in 3D.
     ///
-    /// This should be called for each frame the rectangle needs to be rendered.
-    ///
     /// # Arguments
     ///
     /// - `isometry` defines the translation and rotation of the rectangle.
@@ -287,8 +285,6 @@ where
 
     /// Draw a wireframe rectangle with rounded corners in 2D.
     ///
-    /// This should be called for each frame the rectangle needs to be rendered.
-    ///
     /// # Arguments
     ///
     /// - `isometry` defines the translation and rotation of the rectangle.
@@ -344,8 +340,6 @@ where
 
     /// Draw a wireframe cuboid with rounded corners in 3D.
     ///
-    /// This should be called for each frame the cuboid needs to be rendered.
-    ///
     /// # Arguments
     ///
     /// - `isometry` defines the translation and rotation of the cuboid.
