diff --git a/crates/bevy_winit/src/lib.rs b/crates/bevy_winit/src/lib.rs
index 6c3521b1e2708..20ec4c198680c 100644
--- a/crates/bevy_winit/src/lib.rs
+++ b/crates/bevy_winit/src/lib.rs
@@ -35,12 +35,12 @@ pub use winit::{
     window::{CustomCursor as WinitCustomCursor, CustomCursorSource},
 };
 pub use winit_config::*;
+pub use winit_monitors::*;
 pub use winit_windows::*;
 
 use crate::{
     accessibility::{AccessKitPlugin, WinitActionRequestHandlers},
     state::winit_runner,
-    winit_monitors::WinitMonitors,
 };
 
 pub mod accessibility;
diff --git a/crates/bevy_winit/src/winit_monitors.rs b/crates/bevy_winit/src/winit_monitors.rs
index 46c62038e2cd4..518c3dc394ba9 100644
--- a/crates/bevy_winit/src/winit_monitors.rs
+++ b/crates/bevy_winit/src/winit_monitors.rs
@@ -21,10 +21,12 @@ pub struct WinitMonitors {
 }
 
 impl WinitMonitors {
+    /// Gets the [`MonitorHandle`] at index `n`.
     pub fn nth(&self, n: usize) -> Option<MonitorHandle> {
         self.monitors.get(n).map(|(monitor, _)| monitor.clone())
     }
 
+    /// Gets the [`MonitorHandle`] associated with a `Monitor` entity.
     pub fn find_entity(&self, entity: Entity) -> Option<MonitorHandle> {
         self.monitors
             .iter()
