diff --git a/crates/bevy_shader/Cargo.toml b/crates/bevy_shader/Cargo.toml
index 3ff64b04dd14f..9a14fc0184258 100644
--- a/crates/bevy_shader/Cargo.toml
+++ b/crates/bevy_shader/Cargo.toml
@@ -19,7 +19,7 @@ wgpu-types = { version = "26", default-features = false }
 naga = { version = "26", features = ["wgsl-in"] }
 serde = { version = "1", features = ["derive"] }
 thiserror = { version = "2", default-features = false }
-wesl = { version = "0.1.2", optional = true }
+wesl = { version = "0.2.0", optional = true }
 tracing = { version = "0.1", default-features = false, features = ["std"] }
 
 
diff --git a/crates/bevy_shader/src/shader_cache.rs b/crates/bevy_shader/src/shader_cache.rs
index c9f8c068bfb26..110573b46b795 100644
--- a/crates/bevy_shader/src/shader_cache.rs
+++ b/crates/bevy_shader/src/shader_cache.rs
@@ -221,7 +221,7 @@ impl<ShaderModule, RenderDevice> ShaderCache<ShaderModule, RenderDevice> {
                             for shader_def in shader_defs {
                                 match shader_def {
                                     ShaderDefVal::Bool(key, value) => {
-                                        compiler_options.features.insert(key.clone(), *value);
+                                        compiler_options.features.flags.insert(key.clone(), (*value).into());
                                     }
                                     _ => debug!(
                                         "ShaderDefVal::Int and ShaderDefVal::UInt are not supported in wesl",
