diff --git a/crates/bevy_core_pipeline/src/core_3d/mod.rs b/crates/bevy_core_pipeline/src/core_3d/mod.rs
index 5a090b5da610a..feeef1701487e 100644
--- a/crates/bevy_core_pipeline/src/core_3d/mod.rs
+++ b/crates/bevy_core_pipeline/src/core_3d/mod.rs
@@ -77,8 +77,8 @@ use bevy_diagnostic::FrameCount;
 use bevy_render::{
     batching::gpu_preprocessing::{GpuPreprocessingMode, GpuPreprocessingSupport},
     camera::CameraRenderGraph,
-    experimental::occlusion_culling::OcclusionCulling,
     mesh::allocator::SlabId,
+    occlusion_culling::OcclusionCulling,
     render_phase::PhaseItemBatchSetKey,
     texture::CachedTexture,
     view::{prepare_view_targets, NoIndirectDrawing, RetainedViewEntity},
diff --git a/crates/bevy_core_pipeline/src/deferred/node.rs b/crates/bevy_core_pipeline/src/deferred/node.rs
index 49bf3dc5adf19..6095cd1f6648e 100644
--- a/crates/bevy_core_pipeline/src/deferred/node.rs
+++ b/crates/bevy_core_pipeline/src/deferred/node.rs
@@ -1,6 +1,6 @@
 use bevy_camera::{MainPassResolutionOverride, Viewport};
 use bevy_ecs::{prelude::*, query::QueryItem};
-use bevy_render::experimental::occlusion_culling::OcclusionCulling;
+use bevy_render::occlusion_culling::OcclusionCulling;
 use bevy_render::render_graph::ViewNode;
 
 use bevy_render::view::{ExtractedView, NoIndirectDrawing};
diff --git a/crates/bevy_core_pipeline/src/mip_generation/experimental/depth.rs b/crates/bevy_core_pipeline/src/mip_generation/experimental/depth.rs
index b17335f9310a4..37d6c22673fdd 100644
--- a/crates/bevy_core_pipeline/src/mip_generation/experimental/depth.rs
+++ b/crates/bevy_core_pipeline/src/mip_generation/experimental/depth.rs
@@ -21,7 +21,7 @@ use bevy_ecs::{
 use bevy_math::{uvec2, UVec2, Vec4Swizzles as _};
 use bevy_render::{
     batching::gpu_preprocessing::GpuPreprocessingSupport,
-    experimental::occlusion_culling::{
+    occlusion_culling::{
         OcclusionCulling, OcclusionCullingSubview, OcclusionCullingSubviewEntities,
     },
     render_graph::{Node, NodeRunError, RenderGraphContext},
diff --git a/crates/bevy_core_pipeline/src/prepass/node.rs b/crates/bevy_core_pipeline/src/prepass/node.rs
index d429fc9289d76..feeba242d97b4 100644
--- a/crates/bevy_core_pipeline/src/prepass/node.rs
+++ b/crates/bevy_core_pipeline/src/prepass/node.rs
@@ -3,7 +3,7 @@ use bevy_ecs::{prelude::*, query::QueryItem};
 use bevy_render::{
     camera::ExtractedCamera,
     diagnostic::RecordDiagnostics,
-    experimental::occlusion_culling::OcclusionCulling,
+    occlusion_culling::OcclusionCulling,
     render_graph::{NodeRunError, RenderGraphContext, ViewNode},
     render_phase::{TrackedRenderPass, ViewBinnedRenderPhases},
     render_resource::{CommandEncoderDescriptor, PipelineCache, RenderPassDescriptor, StoreOp},
diff --git a/crates/bevy_dev_tools/src/render_debug.rs b/crates/bevy_dev_tools/src/render_debug.rs
index e5cedbf507248..ce061af084074 100644
--- a/crates/bevy_dev_tools/src/render_debug.rs
+++ b/crates/bevy_dev_tools/src/render_debug.rs
@@ -122,7 +122,7 @@ pub fn update_overlay(
             Has<bevy_core_pipeline::prepass::NormalPrepass>,
             Has<bevy_core_pipeline::prepass::MotionVectorPrepass>,
             Has<bevy_core_pipeline::prepass::DeferredPrepass>,
-            Has<bevy_render::experimental::occlusion_culling::OcclusionCulling>,
+            Has<bevy_render::occlusion_culling::OcclusionCulling>,
             Has<bevy_pbr::ScreenSpaceReflections>,
         ),
         bevy_ecs::query::With<bevy_camera::Camera>,
diff --git a/crates/bevy_pbr/src/render/gpu_preprocess.rs b/crates/bevy_pbr/src/render/gpu_preprocess.rs
index 88b3bb3a08788..5134441c85fb6 100644
--- a/crates/bevy_pbr/src/render/gpu_preprocess.rs
+++ b/crates/bevy_pbr/src/render/gpu_preprocess.rs
@@ -37,7 +37,7 @@ use bevy_render::{
         UntypedPhaseIndirectParametersBuffers,
     },
     diagnostic::RecordDiagnostics,
-    experimental::occlusion_culling::OcclusionCulling,
+    occlusion_culling::OcclusionCulling,
     render_graph::{Node, NodeRunError, RenderGraphContext, RenderGraphExt},
     render_resource::{
         binding_types::{storage_buffer, storage_buffer_read_only, texture_2d, uniform_buffer},
diff --git a/crates/bevy_pbr/src/render/light.rs b/crates/bevy_pbr/src/render/light.rs
index 25a9e2e4c188c..745815d6317ad 100644
--- a/crates/bevy_pbr/src/render/light.rs
+++ b/crates/bevy_pbr/src/render/light.rs
@@ -38,7 +38,7 @@ use bevy_mesh::MeshVertexBufferLayoutRef;
 use bevy_platform::collections::{HashMap, HashSet};
 use bevy_platform::hash::FixedHasher;
 use bevy_render::erased_render_asset::ErasedRenderAssets;
-use bevy_render::experimental::occlusion_culling::{
+use bevy_render::occlusion_culling::{
     OcclusionCulling, OcclusionCullingSubview, OcclusionCullingSubviewEntities,
 };
 use bevy_render::sync_world::MainEntityHashMap;
diff --git a/crates/bevy_render/src/batching/gpu_preprocessing.rs b/crates/bevy_render/src/batching/gpu_preprocessing.rs
index 71ba6a82872d8..44611b75a803b 100644
--- a/crates/bevy_render/src/batching/gpu_preprocessing.rs
+++ b/crates/bevy_render/src/batching/gpu_preprocessing.rs
@@ -25,7 +25,7 @@ use tracing::{error, info};
 use wgpu::{BindingResource, BufferUsages, DownlevelFlags, Features};
 
 use crate::{
-    experimental::occlusion_culling::OcclusionCulling,
+    occlusion_culling::OcclusionCulling,
     render_phase::{
         BinnedPhaseItem, BinnedRenderPhaseBatch, BinnedRenderPhaseBatchSet,
         BinnedRenderPhaseBatchSets, CachedRenderPipelinePhaseItem, PhaseItem,
diff --git a/crates/bevy_render/src/experimental/mod.rs b/crates/bevy_render/src/experimental/mod.rs
deleted file mode 100644
index 40bb6cf1dcc4d..0000000000000
--- a/crates/bevy_render/src/experimental/mod.rs
+++ /dev/null
@@ -1,6 +0,0 @@
-//! Experimental rendering features.
-//!
-//! Experimental features are features with known problems, but are included
-//! nonetheless for testing purposes.
-
-pub mod occlusion_culling;
diff --git a/crates/bevy_render/src/lib.rs b/crates/bevy_render/src/lib.rs
index 9b8725cd439eb..36160828f8b3f 100644
--- a/crates/bevy_render/src/lib.rs
+++ b/crates/bevy_render/src/lib.rs
@@ -39,7 +39,6 @@ pub mod batching;
 pub mod camera;
 pub mod diagnostic;
 pub mod erased_render_asset;
-pub mod experimental;
 pub mod extract_component;
 pub mod extract_instances;
 mod extract_param;
@@ -48,6 +47,7 @@ pub mod globals;
 pub mod gpu_component_array_buffer;
 pub mod gpu_readback;
 pub mod mesh;
+pub mod occlusion_culling;
 #[cfg(not(target_arch = "wasm32"))]
 pub mod pipelined_rendering;
 pub mod render_asset;
@@ -101,8 +101,8 @@ use bevy_utils::prelude::default;
 use bevy_window::{PrimaryWindow, RawHandleWrapperHolder};
 use bitflags::bitflags;
 use core::ops::{Deref, DerefMut};
-use experimental::occlusion_culling::OcclusionCullingPlugin;
 use globals::GlobalsPlugin;
+use occlusion_culling::OcclusionCullingPlugin;
 use render_asset::{
     extract_render_asset_bytes_per_frame, reset_render_asset_bytes_per_frame,
     RenderAssetBytesPerFrame, RenderAssetBytesPerFrameLimiter,
diff --git a/crates/bevy_render/src/experimental/occlusion_culling/mesh_preprocess_types.wgsl b/crates/bevy_render/src/occlusion_culling/mesh_preprocess_types.wgsl
similarity index 100%
rename from crates/bevy_render/src/experimental/occlusion_culling/mesh_preprocess_types.wgsl
rename to crates/bevy_render/src/occlusion_culling/mesh_preprocess_types.wgsl
diff --git a/crates/bevy_render/src/experimental/occlusion_culling/mod.rs b/crates/bevy_render/src/occlusion_culling/mod.rs
similarity index 92%
rename from crates/bevy_render/src/experimental/occlusion_culling/mod.rs
rename to crates/bevy_render/src/occlusion_culling/mod.rs
index ff76b332a404f..531a93930e78b 100644
--- a/crates/bevy_render/src/experimental/occlusion_culling/mod.rs
+++ b/crates/bevy_render/src/occlusion_culling/mod.rs
@@ -22,13 +22,7 @@ impl Plugin for OcclusionCullingPlugin {
     }
 }
 
-/// Add this component to a view in order to enable experimental GPU occlusion
-/// culling.
-///
-/// *Bevy's occlusion culling is currently marked as experimental.* There are
-/// known issues whereby, in rare circumstances, occlusion culling can result in
-/// meshes being culled that shouldn't be (i.e. meshes that turn invisible).
-/// Please try it out and report issues.
+/// Add this component to a view in order to enable GPU occlusion culling.
 ///
 /// *Occlusion culling* allows Bevy to avoid rendering objects that are fully
 /// behind other opaque or alpha tested objects. This is different from, and
diff --git a/crates/bevy_render/src/view/mod.rs b/crates/bevy_render/src/view/mod.rs
index 1bc5a32e628a8..7a14f86e49b9d 100644
--- a/crates/bevy_render/src/view/mod.rs
+++ b/crates/bevy_render/src/view/mod.rs
@@ -11,8 +11,8 @@ pub use window::*;
 
 use crate::{
     camera::{ExtractedCamera, MipBias, NormalizedRenderTargetExt as _, TemporalJitter},
-    experimental::occlusion_culling::OcclusionCulling,
     extract_component::ExtractComponentPlugin,
+    occlusion_culling::OcclusionCulling,
     render_asset::RenderAssets,
     render_phase::ViewRangefinder3d,
     render_resource::{DynamicUniformBuffer, ShaderType, Texture, TextureView},
diff --git a/examples/3d/occlusion_culling.rs b/examples/3d/occlusion_culling.rs
index d61356b9bca8c..bae0ceb512e1d 100644
--- a/examples/3d/occlusion_culling.rs
+++ b/examples/3d/occlusion_culling.rs
@@ -28,7 +28,7 @@ use bevy::{
         batching::gpu_preprocessing::{
             GpuPreprocessingSupport, IndirectParametersBuffers, IndirectParametersIndexed,
         },
-        experimental::occlusion_culling::OcclusionCulling,
+        occlusion_culling::OcclusionCulling,
         render_graph::{self, NodeRunError, RenderGraphContext, RenderGraphExt, RenderLabel},
         render_resource::{Buffer, BufferDescriptor, BufferUsages, MapMode},
         renderer::{RenderContext, RenderDevice},
diff --git a/examples/large_scenes/bistro/src/main.rs b/examples/large_scenes/bistro/src/main.rs
index 54c87d8f4f583..f27fbe14d2a62 100644
--- a/examples/large_scenes/bistro/src/main.rs
+++ b/examples/large_scenes/bistro/src/main.rs
@@ -20,8 +20,8 @@ use bevy::{
     pbr::{DefaultOpaqueRendererMethod, ScreenSpaceAmbientOcclusion},
     post_process::bloom::Bloom,
     render::{
-        batching::NoAutomaticBatching, experimental::occlusion_culling::OcclusionCulling,
-        render_resource::Face, view::NoIndirectDrawing,
+        batching::NoAutomaticBatching, occlusion_culling::OcclusionCulling, render_resource::Face,
+        view::NoIndirectDrawing,
     },
     scene::SceneInstanceReady,
 };
diff --git a/examples/large_scenes/caldera_hotel/src/main.rs b/examples/large_scenes/caldera_hotel/src/main.rs
index 8cb599096ad83..c28d8c3b0c13f 100644
--- a/examples/large_scenes/caldera_hotel/src/main.rs
+++ b/examples/large_scenes/caldera_hotel/src/main.rs
@@ -19,7 +19,7 @@ use bevy::{
     prelude::*,
     render::{
         batching::NoAutomaticBatching,
-        experimental::occlusion_culling::OcclusionCulling,
+        occlusion_culling::OcclusionCulling,
         render_resource::{
             Extent3d, TextureDescriptor, TextureDimension, TextureFormat, TextureUsages,
         },
diff --git a/examples/tools/scene_viewer/main.rs b/examples/tools/scene_viewer/main.rs
index 1a87c8072e234..015798bc20640 100644
--- a/examples/tools/scene_viewer/main.rs
+++ b/examples/tools/scene_viewer/main.rs
@@ -17,7 +17,7 @@ use bevy::{
     gltf::{convert_coordinates::GltfConvertCoordinates, GltfPlugin},
     pbr::DefaultOpaqueRendererMethod,
     prelude::*,
-    render::experimental::occlusion_culling::OcclusionCulling,
+    render::occlusion_culling::OcclusionCulling,
 };
 
 #[cfg(feature = "gltf_animation")]
diff --git a/release-content/migration-guides/occlusion-culling-no-longer-experimental.md b/release-content/migration-guides/occlusion-culling-no-longer-experimental.md
new file mode 100644
index 0000000000000..b2f2b76749ac2
--- /dev/null
+++ b/release-content/migration-guides/occlusion-culling-no-longer-experimental.md
@@ -0,0 +1,10 @@
+---
+title: "Occlusion culling is no longer experimental"
+pull_requests: [22631]
+---
+
+Occlusion culling is no longer experimental, as the known issues that could
+cause Bevy to cull meshes incorrectly are now fixed.
+ Consequently, the
+`bevy::render::experimental::occlusion_culling` module has been renamed to
+simply `bevy::render::occlusion_culling`.
