diff --git a/Cargo.toml b/Cargo.toml
index 93d8e428a0702..ee71862dbe282 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -752,7 +752,7 @@ argh = "0.1.12"
 thiserror = "2.0"
 event-listener = "5.3.0"
 anyhow = "1"
-accesskit = "0.22"
+accesskit = "0.23"
 nonmax = "0.5"
 gltf = "1.4"
 
diff --git a/crates/bevy_a11y/Cargo.toml b/crates/bevy_a11y/Cargo.toml
index 77ecde3992e35..36b0ee00741e5 100644
--- a/crates/bevy_a11y/Cargo.toml
+++ b/crates/bevy_a11y/Cargo.toml
@@ -46,7 +46,7 @@ bevy_ecs = { path = "../bevy_ecs", version = "0.19.0-dev", default-features = fa
 bevy_reflect = { path = "../bevy_reflect", version = "0.19.0-dev", default-features = false, optional = true }
 
 # other
-accesskit = { version = "0.22", default-features = false }
+accesskit = { version = "0.23", default-features = false }
 serde = { version = "1", default-features = false, features = [
   "alloc",
 ], optional = true }
diff --git a/crates/bevy_feathers/Cargo.toml b/crates/bevy_feathers/Cargo.toml
index 5ff420eb75a0f..5d2ee7860aeda 100644
--- a/crates/bevy_feathers/Cargo.toml
+++ b/crates/bevy_feathers/Cargo.toml
@@ -35,7 +35,7 @@ bevy_derive = { path = "../bevy_derive", version = "0.19.0-dev" }
 smol_str = { version = "0.2", default-features = false }
 
 # other
-accesskit = "0.22"
+accesskit = "0.23"
 
 [features]
 default = []
diff --git a/crates/bevy_ui/Cargo.toml b/crates/bevy_ui/Cargo.toml
index 9cd8cacdf14f4..8518861ed9143 100644
--- a/crates/bevy_ui/Cargo.toml
+++ b/crates/bevy_ui/Cargo.toml
@@ -49,7 +49,7 @@ uuid = { version = "1.1", features = ["v4"], optional = true }
 thiserror = { version = "2", default-features = false }
 derive_more = { version = "2", default-features = false, features = ["from"] }
 smallvec = { version = "1", default-features = false }
-accesskit = "0.22"
+accesskit = "0.23"
 tracing = { version = "0.1", default-features = false, features = ["std"] }
 
 [dev-dependencies]
diff --git a/crates/bevy_ui_widgets/Cargo.toml b/crates/bevy_ui_widgets/Cargo.toml
index 9099e2c261270..0e5093b3561c0 100644
--- a/crates/bevy_ui_widgets/Cargo.toml
+++ b/crates/bevy_ui_widgets/Cargo.toml
@@ -23,7 +23,7 @@ bevy_reflect = { path = "../bevy_reflect", version = "0.19.0-dev" }
 bevy_ui = { path = "../bevy_ui", version = "0.19.0-dev" }
 
 # other
-accesskit = "0.22"
+accesskit = "0.23"
 
 [features]
 default = []
diff --git a/crates/bevy_winit/Cargo.toml b/crates/bevy_winit/Cargo.toml
index 5892bee550055..adc21313d6b9f 100644
--- a/crates/bevy_winit/Cargo.toml
+++ b/crates/bevy_winit/Cargo.toml
@@ -64,12 +64,12 @@ bytemuck = { version = "1.5", optional = true }
 # other
 # feature rwh_06 refers to window_raw_handle@v0.6
 winit = { version = "0.30", default-features = false, features = ["rwh_06"] }
-accesskit_winit = { version = "0.30", default-features = false, features = [
+accesskit_winit = { version = "0.31", default-features = false, features = [
   "rwh_06",
 ] }
 approx = { version = "0.5", default-features = false }
 cfg-if = "1.0"
-accesskit = "0.22"
+accesskit = "0.23"
 tracing = { version = "0.1", default-features = false, features = ["std"] }
 
 [target.'cfg(target_os = "android")'.dependencies]
diff --git a/crates/bevy_winit/src/accessibility.rs b/crates/bevy_winit/src/accessibility.rs
index 244491729f251..f75a6b4e981c2 100644
--- a/crates/bevy_winit/src/accessibility.rs
+++ b/crates/bevy_winit/src/accessibility.rs
@@ -8,7 +8,7 @@ use winit::event_loop::ActiveEventLoop;
 
 use accesskit::{
     ActionHandler, ActionRequest, ActivationHandler, DeactivationHandler, Node, NodeId, Role, Tree,
-    TreeUpdate,
+    TreeId, TreeUpdate,
 };
 use accesskit_winit::Adapter;
 use bevy_a11y::{
@@ -87,6 +87,7 @@ impl AccessKitState {
         TreeUpdate {
             nodes: vec![(accesskit_window_id, root)],
             tree: Some(tree),
+            tree_id: TreeId::ROOT,
             focus: accesskit_window_id,
         }
     }
@@ -266,6 +267,7 @@ fn update_adapter(
     TreeUpdate {
         nodes: to_update,
         tree: None,
+        tree_id: TreeId::ROOT,
         focus: NodeId(focus.0.unwrap_or(primary_window_id).to_bits()),
     }
 }
