diff --git a/crates/bevy_asset/src/io/file/file_asset.rs b/crates/bevy_asset/src/io/file/file_asset.rs
index 0964755f023e6..6366753807550 100644
--- a/crates/bevy_asset/src/io/file/file_asset.rs
+++ b/crates/bevy_asset/src/io/file/file_asset.rs
@@ -91,7 +91,7 @@ impl AssetReader for FileAssetReader {
                 #[cfg(not(target_os = "windows"))]
                 _guard,
                 #[cfg(target_os = "windows")]
-                _lifetime: PhantomData::default(),
+                _lifetime: PhantomData,
             })
     }
 
@@ -115,7 +115,7 @@ impl AssetReader for FileAssetReader {
                 #[cfg(not(target_os = "windows"))]
                 _guard,
                 #[cfg(target_os = "windows")]
-                _lifetime: PhantomData::default(),
+                _lifetime: PhantomData,
             })
     }
 
