diff --git a/examples/gizmos/2d_gizmos.rs b/examples/gizmos/2d_gizmos.rs
index 9f060e6d20c48..50bd46d3b010d 100644
--- a/examples/gizmos/2d_gizmos.rs
+++ b/examples/gizmos/2d_gizmos.rs
@@ -210,10 +210,6 @@ fn update_config(
         };
     }
     if keyboard.just_pressed(KeyCode::Space) {
-        if virtual_time.is_paused() {
-            virtual_time.unpause();
-        } else {
-            virtual_time.pause();
-        }
+        virtual_time.toggle();
     }
 }
diff --git a/examples/gizmos/3d_gizmos.rs b/examples/gizmos/3d_gizmos.rs
index 56478e4fb1c77..aac66225779b8 100644
--- a/examples/gizmos/3d_gizmos.rs
+++ b/examples/gizmos/3d_gizmos.rs
@@ -292,10 +292,6 @@ fn update_config(
         config_store.config_mut::<AabbGizmoConfigGroup>().1.draw_all ^= true;
     }
     if keyboard.just_pressed(KeyCode::Space) {
-        if virtual_time.is_paused() {
-            virtual_time.unpause();
-        } else {
-            virtual_time.pause();
-        }
+        virtual_time.toggle();
     }
 }
