diff --git a/crates/bevy_render/src/camera.rs b/crates/bevy_render/src/camera.rs
index 3890274f41b76..7b20bab14df03 100644
--- a/crates/bevy_render/src/camera.rs
+++ b/crates/bevy_render/src/camera.rs
@@ -41,6 +41,7 @@ use bevy_ecs::{
 };
 use bevy_image::Image;
 use bevy_log::warn;
+use bevy_log::warn_once;
 use bevy_math::{uvec2, vec2, Mat4, URect, UVec2, UVec4, Vec2};
 use bevy_platform::collections::{HashMap, HashSet};
 use bevy_reflect::prelude::*;
@@ -652,7 +653,7 @@ pub fn sort_cameras(
     }
 
     if !ambiguities.is_empty() {
-        warn!(
+        warn_once!(
             "Camera order ambiguities detected for active cameras with the following priorities: {:?}. \
             To fix this, ensure there is exactly one Camera entity spawned with a given order for a given RenderTarget. \
             Ambiguities should be resolved because either (1) multiple active cameras were spawned accidentally, which will \
diff --git a/examples/ui/ui_target_camera.rs b/examples/ui/ui_target_camera.rs
index c2a374bf699a6..5e6260f16606e 100644
--- a/examples/ui/ui_target_camera.rs
+++ b/examples/ui/ui_target_camera.rs
@@ -4,17 +4,11 @@ use bevy::color::palettes::css::BLUE;
 use bevy::color::palettes::css::GREEN;
 use bevy::color::palettes::css::RED;
 use bevy::color::palettes::css::YELLOW;
-use bevy::log::LogPlugin;
-use bevy::log::DEFAULT_FILTER;
 use bevy::prelude::*;
 
 fn main() {
     App::new()
-        .add_plugins(DefaultPlugins.set(LogPlugin {
-            // Disable camera order ambiguity warnings
-            filter: format!("{DEFAULT_FILTER},bevy_render::camera=off"),
-            ..Default::default()
-        }))
+        .add_plugins(DefaultPlugins)
         .add_systems(Startup, setup)
         .run();
 }
