diff --git a/crates/bevy_render/src/batching/gpu_preprocessing.rs b/crates/bevy_render/src/batching/gpu_preprocessing.rs
index 1c31e894523a3..895c74f01a2db 100644
--- a/crates/bevy_render/src/batching/gpu_preprocessing.rs
+++ b/crates/bevy_render/src/batching/gpu_preprocessing.rs
@@ -860,11 +860,6 @@ impl UntypedPhaseIndirectParametersBuffers {
     pub fn new(
         allow_copies_from_indirect_parameter_buffers: bool,
     ) -> UntypedPhaseIndirectParametersBuffers {
-        let mut indirect_parameter_buffer_usages = BufferUsages::STORAGE | BufferUsages::INDIRECT;
-        if allow_copies_from_indirect_parameter_buffers {
-            indirect_parameter_buffer_usages |= BufferUsages::COPY_SRC;
-        }
-
         UntypedPhaseIndirectParametersBuffers {
             non_indexed: MeshClassIndirectParametersBuffers::new(
                 allow_copies_from_indirect_parameter_buffers,
