diff --git a/crates/bevy_text/src/text.rs b/crates/bevy_text/src/text.rs
index 0ec97b0a00d79..12d4bbf143dec 100644
--- a/crates/bevy_text/src/text.rs
+++ b/crates/bevy_text/src/text.rs
@@ -246,7 +246,7 @@ impl From<Justify> for cosmic_text::Align {
     }
 }
 
-#[derive(Component, Clone, Debug, Reflect, PartialEq)]
+#[derive(Clone, Debug, Reflect, PartialEq)]
 /// Specifies how the font face for a text span is sourced.
 ///
 /// A `FontSource` can either reference a font asset or identify a font by family name to be
