diff --git a/crates/bevy_gizmos_render/src/lines.wgsl b/crates/bevy_gizmos_render/src/lines.wgsl
index 9e4c234851161..293f772c788cf 100644
--- a/crates/bevy_gizmos_render/src/lines.wgsl
+++ b/crates/bevy_gizmos_render/src/lines.wgsl
@@ -141,7 +141,10 @@ fn clip_near_plane(a: vec4<f32>, b: vec4<f32>) -> vec4<f32> {
         // Interpolate a towards b until it's at the near plane.
         let distance_a = a.z - a.w;
         let distance_b = b.z - b.w;
-        let t = distance_a / (distance_a - distance_b);
+        // Add an epsilon to the interpolator to ensure that the point is
+        // not just behind the clip plane due to floating-point imprecision,
+        // which can lead to lines showing up where they should not.
+        let t = distance_a / (distance_a - distance_b) + EPSILON;
         return mix(a, b, t);
     }
     return a;
