diff --git a/examples/shader_advanced/fullscreen_material.rs b/examples/shader_advanced/fullscreen_material.rs
index 4e8f24c7724df..225186bf22539 100644
--- a/examples/shader_advanced/fullscreen_material.rs
+++ b/examples/shader_advanced/fullscreen_material.rs
@@ -1,4 +1,9 @@
-//! Demonstrates how to write a custom fullscreen shader.
+//! Demonstrates how to write a custom fullscreen shader
+//!
+//! This example demonstrates working in 3d. To make the example work in 2d,
+//! replace 3d components with their 2d counterparts, and schedule the work
+//! to run in the `Core2d` schedule as described in the `FullscreenMaterial`
+//! comment in this file.
 
 use bevy::{
     core_pipeline::fullscreen_material::{FullscreenMaterial, FullscreenMaterialPlugin},
@@ -50,5 +55,14 @@ impl FullscreenMaterial for FullscreenEffect {
         "shaders/fullscreen_effect.wgsl".into()
     }
 
-    // Uses default run_in (Core3dSystems::PostProcess)
+    // The `FullscreenMaterial` uses 3d schedules by default.
+    // To make this work in 2d, you would need to schedule to
+    // run in `Core2d` and in a `Core2dSystems` set.
+    //
+    // fn schedule() -> impl bevy::ecs::schedule::ScheduleLabel + Clone {
+    //     bevy::core_pipeline::Core2d
+    // }
+    // fn run_in() -> impl SystemSet {
+    //     bevy::core_pipeline::Core2dSystems::PostProcess
+    // }
 }
