diff --git a/crates/bevy_camera_controller/src/pan_camera.rs b/crates/bevy_camera_controller/src/pan_camera.rs
index 283bd10562a50..f2319bba93b91 100644
--- a/crates/bevy_camera_controller/src/pan_camera.rs
+++ b/crates/bevy_camera_controller/src/pan_camera.rs
@@ -81,7 +81,7 @@ pub struct PanCamera {
 /// - Pan speed: 500.0
 /// - Move up/down: W/S
 /// - Move left/right: A/D
-/// - Rotation speed: PI (radiradians per second)
+/// - Rotation speed: PI (radians per second)
 /// - Rotation ccw/cw: Q/E
 impl Default for PanCamera {
     /// Provides the default values for the `PanCamera` controller.
