diff --git a/crates/bevy_diagnostic/src/system_information_diagnostics_plugin.rs b/crates/bevy_diagnostic/src/system_information_diagnostics_plugin.rs
index 376a109ae3bbf..768bbb0828844 100644
--- a/crates/bevy_diagnostic/src/system_information_diagnostics_plugin.rs
+++ b/crates/bevy_diagnostic/src/system_information_diagnostics_plugin.rs
@@ -46,10 +46,15 @@ impl SystemInformationDiagnosticsPlugin {
 /// [`SystemInformationDiagnosticsPlugin`] for more information.
 #[derive(Debug, Resource)]
 pub struct SystemInfo {
+    /// OS name and version.
     pub os: String,
+    /// System kernel version.
     pub kernel: String,
+    /// CPU model name.
     pub cpu: String,
+    /// Physical core count.
     pub core_count: String,
+    /// System RAM.
     pub memory: String,
 }
 
