diff --git a/crates/bevy_diagnostic/src/diagnostic.rs b/crates/bevy_diagnostic/src/diagnostic.rs
index 00a758416be97..675f205a64f93 100644
--- a/crates/bevy_diagnostic/src/diagnostic.rs
+++ b/crates/bevy_diagnostic/src/diagnostic.rs
@@ -293,6 +293,8 @@ impl Diagnostic {
     /// Clear the history of this diagnostic.
     pub fn clear_history(&mut self) {
         self.history.clear();
+        self.sum = 0.0;
+        self.ema = 0.0;
     }
 }
 
@@ -420,3 +422,31 @@ impl RegisterDiagnostic for App {
         self
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_clear_history() {
+        const MEASUREMENT: f64 = 20.0;
+
+        let mut diagnostic =
+            Diagnostic::new(DiagnosticPath::new("test")).with_max_history_length(5);
+        let mut now = Instant::now();
+
+        for _ in 0..3 {
+            for _ in 0..5 {
+                diagnostic.add_measurement(DiagnosticMeasurement {
+                    time: now,
+                    value: MEASUREMENT,
+                });
+                // Increase time to test smoothed average.
+                now += Duration::from_secs(1);
+            }
+            assert!((diagnostic.average().unwrap() - MEASUREMENT).abs() < 0.1);
+            assert!((diagnostic.smoothed().unwrap() - MEASUREMENT).abs() < 0.1);
+            diagnostic.clear_history();
+        }
+    }
+}
