diff --git a/crates/bevy_diagnostic/src/log_diagnostics_plugin.rs b/crates/bevy_diagnostic/src/log_diagnostics_plugin.rs
index 8b01d10402fb2..d91c4235cc4cd 100644
--- a/crates/bevy_diagnostic/src/log_diagnostics_plugin.rs
+++ b/crates/bevy_diagnostic/src/log_diagnostics_plugin.rs
@@ -15,8 +15,14 @@ use log::{debug, info};
 ///
 /// When no diagnostics are provided, this plugin does nothing.
 pub struct LogDiagnosticsPlugin {
+    /// If `true` then the `Debug` representation of each `Diagnostic` is logged.
+    /// If `false` then a (smoothed) current value and historical average are logged.
+    ///
+    /// Defaults to `false`.
     pub debug: bool,
+    /// Time to wait between logging diagnostics and logging them again.
     pub wait_duration: Duration,
+    /// If `Some` then only these diagnostics are logged.
     pub filter: Option<HashSet<DiagnosticPath>>,
 }
 
@@ -107,6 +113,7 @@ impl Plugin for LogDiagnosticsPlugin {
 }
 
 impl LogDiagnosticsPlugin {
+    /// Filter logging to only the paths in `filter`.
     pub fn filtered(filter: HashSet<DiagnosticPath>) -> Self {
         LogDiagnosticsPlugin {
             filter: Some(filter),
