diff --git a/crates/bevy_ecs/src/system/schedule_system.rs b/crates/bevy_ecs/src/system/schedule_system.rs
index 75fad2b7e9af6..b117f2c38731f 100644
--- a/crates/bevy_ecs/src/system/schedule_system.rs
+++ b/crates/bevy_ecs/src/system/schedule_system.rs
@@ -30,6 +30,10 @@ impl<S: System<In = ()>> System for InfallibleSystemWrapper<S> {
         self.0.name()
     }
 
+    fn type_id(&self) -> core::any::TypeId {
+        self.0.type_id()
+    }
+
     #[inline]
     fn component_access(&self) -> &Access<ComponentId> {
         self.0.component_access()
