diff --git a/examples/3d/fog.rs b/examples/3d/fog.rs
index 9793ae0ad356a..cbb8c0fa3b5bc 100644
--- a/examples/3d/fog.rs
+++ b/examples/3d/fog.rs
@@ -180,7 +180,7 @@ fn update_system(
         } else if let FogFalloff::ExponentialSquared { .. } = fog.falloff {
             // No change
         } else {
-            fog.falloff = FogFalloff::Exponential { density: 0.07 };
+            fog.falloff = FogFalloff::ExponentialSquared { density: 0.07 };
         };
     }
 
