diff --git a/crates/bevy_picking/src/pointer.rs b/crates/bevy_picking/src/pointer.rs
index e180a9c1bed78..faba90cbb991a 100644
--- a/crates/bevy_picking/src/pointer.rs
+++ b/crates/bevy_picking/src/pointer.rs
@@ -205,8 +205,8 @@ impl PointerLocation {
 /// - a pointer is not associated with a [`Camera`] because multiple cameras can target the same
 ///   render target. It is up to picking backends to associate a Pointer's `Location` with a
 ///   specific `Camera`, if any.
-#[derive(Debug, Clone, Component, Reflect, PartialEq)]
-#[reflect(Component, Debug, PartialEq, Clone)]
+#[derive(Debug, Clone, Reflect, PartialEq)]
+#[reflect(Debug, PartialEq, Clone)]
 pub struct Location {
     /// The [`NormalizedRenderTarget`] associated with the pointer, usually a window.
     pub target: NormalizedRenderTarget,
diff --git a/release-content/migration-guides/picking_location_not_component.md b/release-content/migration-guides/picking_location_not_component.md
new file mode 100644
index 0000000000000..e70d7058aa8eb
--- /dev/null
+++ b/release-content/migration-guides/picking_location_not_component.md
@@ -0,0 +1,6 @@
+---
+title: Location is not a Component anymore
+pull_requests: [19306]
+---
+
+`bevy_picking::Location` was erroneously made a `Component`. It is no longer one, `bevy_picking::PointerLocation` wraps a `Location` and is the intended usage pattern.
