diff --git a/Cargo.toml b/Cargo.toml
index cdbabd0c1338b..786227de6cab1 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1592,6 +1592,7 @@ wasm = true
 name = "headless"
 path = "examples/app/headless.rs"
 doc-scrape-examples = true
+required-features = ["bevy_log"]
 
 [package.metadata.example.headless]
 name = "Headless"
diff --git a/examples/app/headless.rs b/examples/app/headless.rs
index 536da8c26e0a9..94f01eaddcb8c 100644
--- a/examples/app/headless.rs
+++ b/examples/app/headless.rs
@@ -19,7 +19,7 @@ fn main() {
         println!("Disable the default features and rerun the example to run headless.");
         println!("To do so, run:");
         println!();
-        println!("    cargo run --example headless --no-default-features");
+        println!("    cargo run --example headless --no-default-features --features bevy_log");
         return;
     }
 
