diff --git a/crates/bevy_internal/Cargo.toml b/crates/bevy_internal/Cargo.toml
index e659bdc77d70c..38baf6d09e834 100644
--- a/crates/bevy_internal/Cargo.toml
+++ b/crates/bevy_internal/Cargo.toml
@@ -97,11 +97,11 @@ serialize = [
   "bevy_color?/serialize",
   "bevy_ecs/serialize",
   "bevy_image?/serialize",
-  "bevy_input/serialize",
-  "bevy_math/serialize",
+  "bevy_input?/serialize",
+  "bevy_math?/serialize",
   "bevy_scene?/serialize",
   "bevy_time/serialize",
-  "bevy_transform/serialize",
+  "bevy_transform?/serialize",
   "bevy_ui?/serialize",
   "bevy_window?/serialize",
   "bevy_winit?/serialize",
@@ -317,7 +317,7 @@ critical-section = [
   "bevy_app/critical-section",
   "bevy_diagnostic/critical-section",
   "bevy_ecs/critical-section",
-  "bevy_input/critical-section",
+  "bevy_input?/critical-section",
   "bevy_input_focus?/critical-section",
   "bevy_platform_support/critical-section",
   "bevy_reflect/critical-section",
@@ -343,7 +343,7 @@ async_executor = [
   "std",
   "bevy_tasks/async_executor",
   "bevy_ecs/async_executor",
-  "bevy_transform/async_executor",
+  "bevy_transform?/async_executor",
 ]
 
 # Enables use of browser APIs.
