diff --git a/crates/bevy_camera/src/primitives.rs b/crates/bevy_camera/src/primitives.rs
index eb6246d25990c..52149613385b5 100644
--- a/crates/bevy_camera/src/primitives.rs
+++ b/crates/bevy_camera/src/primitives.rs
@@ -294,6 +294,7 @@ impl Frustum {
                 row3 - row
             });
         }
+        half_spaces[5] = HalfSpace::new(Vec4::new(0.0, 0.0, 0.0, f32::MAX));
         Self { half_spaces }
     }
 
