diff --git a/crates/bevy_pbr/src/render/view_transformations.wgsl b/crates/bevy_pbr/src/render/view_transformations.wgsl
index d21a1d534c3c3..4fe44d48dbd21 100644
--- a/crates/bevy_pbr/src/render/view_transformations.wgsl
+++ b/crates/bevy_pbr/src/render/view_transformations.wgsl
@@ -22,6 +22,10 @@
 ///    Perspective projection: 0.0 is inf far away
 ///    Orthographic projection: 0.0 is far clipping plane
 
+/// Clip space:
+/// This is NDC before the perspective divide, still in homogenous coordinate space.
+/// Dividing a clip space point by its w component yields a point in NDC space.
+
 /// UV space:
 /// 0.0, 0.0 is the top left
 /// 1.0, 1.0 is the bottom right
diff --git a/crates/bevy_render/src/view/view.wgsl b/crates/bevy_render/src/view/view.wgsl
index 0d0efb75b556e..23ded53f40f91 100644
--- a/crates/bevy_render/src/view/view.wgsl
+++ b/crates/bevy_render/src/view/view.wgsl
@@ -84,6 +84,10 @@ struct View {
 ///    Perspective projection: 0.0 is inf far away
 ///    Orthographic projection: 0.0 is far clipping plane
 
+/// Clip space:
+/// This is NDC before the perspective divide, still in homogenous coordinate space.
+/// Dividing a clip space point by its w component yields a point in NDC space.
+
 /// UV space:
 /// 0.0, 0.0 is the top left
 /// 1.0, 1.0 is the bottom right
