diff --git a/crates/bevy_camera/src/projection.rs b/crates/bevy_camera/src/projection.rs
index f95959c43986c..9a1d61f4ee981 100644
--- a/crates/bevy_camera/src/projection.rs
+++ b/crates/bevy_camera/src/projection.rs
@@ -68,7 +68,7 @@ pub trait CameraProjection {
     /// This code is called by [`update_frusta`](crate::visibility::update_frusta) system
     /// for each camera to update its frustum.
     fn compute_frustum(&self, camera_transform: &GlobalTransform) -> Frustum {
-        let clip_from_world = self.get_clip_from_view() * camera_transform.to_matrix().inverse();
+        let clip_from_world = self.get_clip_from_view() * camera_transform.affine().inverse();
         Frustum::from_clip_from_world_custom_far(
             &clip_from_world,
             &camera_transform.translation(),
