diff --git a/Cargo.toml b/Cargo.toml
index 42a880ece5504..4dc83dcc41e12 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1036,7 +1036,7 @@ wasm = true
 name = "tilemap_chunk_orientation"
 path = "examples/2d/tilemap_chunk_orientation.rs"
 # Causes an ICE on docs.rs
-doc-scrape-examples = false
+doc-scrape-examples = true
 
 [package.metadata.example.tilemap_chunk_orientation]
 name = "Tilemap Chunk with Orientations"
