diff --git a/crates/bevy_platform/Cargo.toml b/crates/bevy_platform/Cargo.toml
index bd6402b36a96b..3dc04396b2c5f 100644
--- a/crates/bevy_platform/Cargo.toml
+++ b/crates/bevy_platform/Cargo.toml
@@ -48,7 +48,7 @@ web = ["dep:web-time", "dep:getrandom"]
 [dependencies]
 cfg-if = "1.0.0"
 critical-section = { version = "1.2.0", default-features = false, optional = true }
-spin = { version = "0.9.8", default-features = false, features = [
+spin = { version = "0.10.0", default-features = false, features = [
   "mutex",
   "spin_mutex",
   "rwlock",
@@ -74,8 +74,8 @@ getrandom = { version = "0.2.0", default-features = false, optional = true, feat
 portable-atomic = { version = "1", default-features = false, features = [
   "fallback",
 ] }
-spin = { version = "0.9.8", default-features = false, features = [
-  "portable_atomic",
+spin = { version = "0.10.0", default-features = false, features = [
+  "portable-atomic",
 ] }
 
 [target.'cfg(not(target_has_atomic = "ptr"))'.dependencies]
