diff --git a/crates/bevy_ui/src/ui_node.rs b/crates/bevy_ui/src/ui_node.rs
index 2486296bac3e2..55ce3eb6a2274 100644
--- a/crates/bevy_ui/src/ui_node.rs
+++ b/crates/bevy_ui/src/ui_node.rs
@@ -1178,7 +1178,7 @@ pub struct OverflowClipMargin {
 
 impl OverflowClipMargin {
     pub const DEFAULT: Self = Self {
-        visual_box: OverflowClipBox::ContentBox,
+        visual_box: OverflowClipBox::PaddingBox,
         margin: 0.,
     };
 
@@ -1224,9 +1224,9 @@ impl OverflowClipMargin {
 )]
 pub enum OverflowClipBox {
     /// Clip any content that overflows outside the content box
-    #[default]
     ContentBox,
     /// Clip any content that overflows outside the padding box
+    #[default]
     PaddingBox,
     /// Clip any content that overflows outside the border box
     BorderBox,
diff --git a/release-content/migration-guides/overflowclipbox_default_is_now_paddingbox.md b/release-content/migration-guides/overflowclipbox_default_is_now_paddingbox.md
new file mode 100644
index 0000000000000..a733f7f775a30
--- /dev/null
+++ b/release-content/migration-guides/overflowclipbox_default_is_now_paddingbox.md
@@ -0,0 +1,7 @@
+---
+title: OverflowClipBox's default is now Paddingbox
+pull_requests: [18935]
+---
+
+The default variant for `OverflowClipBox` is now `PaddingBox`.
+The default value for `OverflowClipMargin::visual_box` is now `OverflowClipBox::PaddingBox`.
