diff --git a/crates/bevy_ecs/src/system/commands/mod.rs b/crates/bevy_ecs/src/system/commands/mod.rs
index 8404f1db28172..4cb6d61bc0e9a 100644
--- a/crates/bevy_ecs/src/system/commands/mod.rs
+++ b/crates/bevy_ecs/src/system/commands/mod.rs
@@ -1576,6 +1576,9 @@ impl<'a> EntityCommands<'a> {
 
     /// Removes a [`Bundle`] of components from the entity.
     ///
+    /// This will remove all components that intersect with the provided bundle;
+    /// the entity does not need to have all the components in the bundle.
+    ///
     /// This will emit a warning if the entity does not exist.
     ///
     /// # Example
@@ -1617,6 +1620,9 @@ impl<'a> EntityCommands<'a> {
 
     /// Removes a [`Bundle`] of components from the entity.
     ///
+    /// This will remove all components that intersect with the provided bundle;
+    /// the entity does not need to have all the components in the bundle.
+    ///
     /// Unlike [`Self::remove`],
     /// this will not emit a warning if the entity does not exist.
     ///
@@ -1659,6 +1665,9 @@ impl<'a> EntityCommands<'a> {
     /// Removes a [`Bundle`] of components from the entity,
     /// and also removes any components required by the components in the bundle.
     ///
+    /// This will remove all components that intersect with the provided bundle;
+    /// the entity does not need to have all the components in the bundle.
+    ///
     /// # Example
     ///
     /// ```
