diff --git a/crates/bevy_input/Cargo.toml b/crates/bevy_input/Cargo.toml
index 602d1be8eac76..f1c9497a5d8d8 100644
--- a/crates/bevy_input/Cargo.toml
+++ b/crates/bevy_input/Cargo.toml
@@ -24,7 +24,7 @@ bevy_reflect = [
 ## Adds serialization support through `serde`.
 serialize = [
   "serde",
-  "smol_str/serde",
+  "smol_str?/serde",
   "bevy_ecs/serialize",
   "bevy_math/serialize",
   "bevy_platform_support/serialize",
