diff --git a/crates/bevy_reflect/src/impls/std.rs b/crates/bevy_reflect/src/impls/std.rs
index 54d8393e7afdf..4be1521b7ec6d 100644
--- a/crates/bevy_reflect/src/impls/std.rs
+++ b/crates/bevy_reflect/src/impls/std.rs
@@ -1208,6 +1208,14 @@ macro_rules! impl_reflect_for_hashset {
 
 impl_type_path!(::bevy_platform_support::hash::NoOpHash);
 impl_type_path!(::bevy_platform_support::hash::FixedHasher);
+impl_reflect_opaque!(::core::net::SocketAddr(
+    Clone,
+    Debug,
+    Hash,
+    PartialEq,
+    Serialize,
+    Deserialize
+));
 
 #[cfg(feature = "std")]
 impl_reflect_for_hashset!(::std::collections::HashSet<V,S>);
