diff --git a/crates/bevy_camera/src/lib.rs b/crates/bevy_camera/src/lib.rs
index 1ae5c8107860e..e698e635dab7c 100644
--- a/crates/bevy_camera/src/lib.rs
+++ b/crates/bevy_camera/src/lib.rs
@@ -25,3 +25,15 @@ impl Plugin for CameraPlugin {
         ));
     }
 }
+
+/// The camera prelude.
+///
+/// This includes the most common types in this crate, re-exported for your convenience.
+pub mod prelude {
+    #[doc(hidden)]
+    pub use crate::{
+        visibility::{InheritedVisibility, ViewVisibility, Visibility},
+        Camera, Camera2d, Camera3d, ClearColor, ClearColorConfig, OrthographicProjection,
+        PerspectiveProjection, Projection,
+    };
+}
diff --git a/crates/bevy_internal/src/prelude.rs b/crates/bevy_internal/src/prelude.rs
index 25f6caad2e877..203ded7ccecec 100644
--- a/crates/bevy_internal/src/prelude.rs
+++ b/crates/bevy_internal/src/prelude.rs
@@ -18,9 +18,13 @@ pub use crate::window::prelude::*;
 pub use crate::image::prelude::*;
 
 #[doc(hidden)]
-#[cfg(feature = "bevy_image")]
+#[cfg(feature = "bevy_mesh")]
 pub use crate::mesh::prelude::*;
 
+#[doc(hidden)]
+#[cfg(feature = "bevy_camera")]
+pub use crate::camera::prelude::*;
+
 pub use bevy_derive::{bevy_main, Deref, DerefMut};
 
 #[doc(hidden)]
@@ -39,10 +43,6 @@ pub use crate::animation::prelude::*;
 #[cfg(feature = "bevy_color")]
 pub use crate::color::prelude::*;
 
-#[doc(hidden)]
-#[cfg(feature = "bevy_core_pipeline")]
-pub use crate::core_pipeline::prelude::*;
-
 #[doc(hidden)]
 #[cfg(feature = "bevy_pbr")]
 pub use crate::pbr::prelude::*;
