diff --git a/crates/bevy_render/src/lib.rs b/crates/bevy_render/src/lib.rs
index bd7b350d93da4..81a7641caa7ed 100644
--- a/crates/bevy_render/src/lib.rs
+++ b/crates/bevy_render/src/lib.rs
@@ -1,3 +1,16 @@
+//! # Useful Environment Variables
+//!
+//! Both `bevy_render` and `wgpu` have a number of environment variable options for changing the runtime behavior
+//! of both crates. Many of these may be useful in development or release environments.
+//!
+//! - `WGPU_DEBUG=1` enables debug labels, which can be useful in release builds.
+//! - `WGPU_VALIDATION=0` disables validation layers. This can help with particularly spammy errors.
+//! - `WGPU_FORCE_FALLBACK_ADAPTER=1` attempts to force software rendering. This typically matches what is used in CI.
+//! - `WGPU_ADAPTER_NAME` allows selecting a specific adapter by name.
+//! - `WGPU_SETTINGS_PRIO=webgl2` uses webgl2 limits.
+//! - `WGPU_SETTINGS_PRIO=compatibility` uses webgpu limits.
+//! - `VERBOSE_SHADER_ERROR=1` prints more detailed information about WGSL compilation errors, such as shader defs and shader entrypoint.
+
 #![expect(missing_docs, reason = "Not all docs are written yet, see #3492.")]
 #![expect(unsafe_code, reason = "Unsafe code is used to improve performance.")]
 #![cfg_attr(
