diff --git a/crates/bevy_shader/src/shader.rs b/crates/bevy_shader/src/shader.rs
index 3470e8c617e09..932de7b98ceff 100644
--- a/crates/bevy_shader/src/shader.rs
+++ b/crates/bevy_shader/src/shader.rs
@@ -429,8 +429,10 @@ impl ShaderImport {
 }
 
 /// A reference to a shader asset.
+#[derive(Default)]
 pub enum ShaderRef {
     /// Use the "default" shader for the current context.
+    #[default]
     Default,
     /// A handle to a shader stored in the [`Assets<Shader>`](bevy_asset::Assets) resource
     Handle(Handle<Shader>),
