diff --git a/crates/bevy_color/src/laba.rs b/crates/bevy_color/src/laba.rs
index 010b3df249678..3797c7500d82e 100644
--- a/crates/bevy_color/src/laba.rs
+++ b/crates/bevy_color/src/laba.rs
@@ -275,7 +275,7 @@ impl From<Xyza> for Laba {
         } else {
             (Laba::CIE_KAPPA * yr + 16.0) / 116.0
         };
-        let fz = if yr > Laba::CIE_EPSILON {
+        let fz = if zr > Laba::CIE_EPSILON {
             ops::cbrt(zr)
         } else {
             (Laba::CIE_KAPPA * zr + 16.0) / 116.0
