diff --git a/crates/bevy_solari/src/scene/blas.rs b/crates/bevy_solari/src/scene/blas.rs
index 7e574dd100bb2..58d6b7fd3abbb 100644
--- a/crates/bevy_solari/src/scene/blas.rs
+++ b/crates/bevy_solari/src/scene/blas.rs
@@ -105,20 +105,18 @@ pub fn compact_raytracing_blas(
     mut blas_manager: ResMut<BlasManager>,
     render_queue: Res<RenderQueue>,
 ) {
-    let mut first_mesh_processed = None;
-
+    let queue_size = blas_manager.compaction_queue.len();
+    let mut meshes_processed = 0;
     let mut vertices_compacted = 0;
-    while vertices_compacted < MAX_COMPACTION_VERTICES_PER_FRAME
-        && let Some((mesh, vertex_count, compaction_started)) =
-            blas_manager.compaction_queue.pop_front()
+
+    while !blas_manager.compaction_queue.is_empty()
+        && vertices_compacted < MAX_COMPACTION_VERTICES_PER_FRAME
+        && meshes_processed < queue_size
     {
-        // Stop iterating once we loop back around to the start of the list
-        if Some(mesh) == first_mesh_processed {
-            break;
-        }
-        if first_mesh_processed.is_none() {
-            first_mesh_processed = Some(mesh);
-        }
+        meshes_processed += 1;
+
+        let (mesh, vertex_count, compaction_started) =
+            blas_manager.compaction_queue.pop_front().unwrap();
 
         let Some(blas) = blas_manager.get(&mesh) else {
             continue;
diff --git a/release-content/release-notes/bevy_solari.md b/release-content/release-notes/bevy_solari.md
index 1ab28542a473c..c5d497ceba339 100644
--- a/release-content/release-notes/bevy_solari.md
+++ b/release-content/release-notes/bevy_solari.md
@@ -1,7 +1,7 @@
 ---
 title: Initial raytraced lighting progress (bevy_solari)
 authors: ["@JMS55", "@SparkyPotato"]
-pull_requests: [19058, 19620, 19790, 20020, 20113, 20156, 20213, 20242, 20259, 20406, 20457, 20580, 20596, 20622]
+pull_requests: [19058, 19620, 19790, 20020, 20113, 20156, 20213, 20242, 20259, 20406, 20457, 20580, 20596, 20622, 20659]
 ---
 
 ## Overview
