diff --git a/crates/bevy_core_pipeline/Cargo.toml b/crates/bevy_core_pipeline/Cargo.toml
index 3907fb499964a..e9b288498b0a8 100644
--- a/crates/bevy_core_pipeline/Cargo.toml
+++ b/crates/bevy_core_pipeline/Cargo.toml
@@ -2,10 +2,6 @@
 name = "bevy_core_pipeline"
 version = "0.17.0-dev"
 edition = "2024"
-authors = [
-  "Bevy Contributors <bevyengine@gmail.com>",
-  "Carter Anderson <mcanders1@gmail.com>",
-]
 description = "Provides a core render pipeline for Bevy Engine."
 homepage = "https://bevy.org"
 repository = "https://github.com/bevyengine/bevy"
