diff --git a/Cargo.toml b/Cargo.toml
index 28612f1c0a516..4f6f83a4ef92f 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -284,14 +284,7 @@ bevy_sprite_render = [
 bevy_text = ["bevy_internal/bevy_text", "bevy_asset"]
 
 # A custom ECS-driven UI framework
-bevy_ui = [
-  "bevy_internal/bevy_ui",
-  "bevy_core_pipeline",
-  "bevy_text",
-  "bevy_sprite",
-  "bevy_color",
-  "bevy_anti_aliasing",
-]
+bevy_ui = ["bevy_internal/bevy_ui", "bevy_text", "bevy_sprite", "bevy_color"]
 
 # Provides rendering functionality for bevy_ui
 bevy_ui_render = [
