diff --git a/tools/build-wasm-example/src/main.rs b/tools/build-wasm-example/src/main.rs
index 25c0b071528bb..44c3b15586ebe 100644
--- a/tools/build-wasm-example/src/main.rs
+++ b/tools/build-wasm-example/src/main.rs
@@ -80,14 +80,15 @@ fn main() {
 
         let profile = if cli.debug {
             "debug"
+        } else if cli.optimize_size {
+            "wasm-release"
         } else {
-            parameters.push("--release");
             "release"
         };
 
         let cmd = cmd!(
             sh,
-            "cargo build {parameters...} --target wasm32-unknown-unknown --example {example}"
+            "cargo build {parameters...} --profile {profile} --target wasm32-unknown-unknown --example {example}"
         );
         cmd.run().expect("Error building example");
 
