diff --git a/crates/bevy_mesh/src/lib.rs b/crates/bevy_mesh/src/lib.rs
index 6cb0fbe402b88..5741f874fdb4e 100644
--- a/crates/bevy_mesh/src/lib.rs
+++ b/crates/bevy_mesh/src/lib.rs
@@ -54,7 +54,7 @@ impl Plugin for MeshPlugin {
             .register_asset_reflect::<Mesh>()
             .add_systems(
                 PostUpdate,
-                mark_3d_meshes_as_changed_if_their_assets_changed.before(AssetEventSystems),
+                mark_3d_meshes_as_changed_if_their_assets_changed.after(AssetEventSystems),
             );
     }
 }
