diff --git a/examples/3d/render_to_texture.rs b/examples/3d/render_to_texture.rs
index ba17819ffbca7..f6b02ff98300d 100644
--- a/examples/3d/render_to_texture.rs
+++ b/examples/3d/render_to_texture.rs
@@ -64,6 +64,8 @@ fn setup(
     commands.spawn((
         Camera3d::default(),
         Camera {
+            // render before the "main pass" camera
+            order: -1,
             target: image_handle.clone().into(),
             clear_color: Color::WHITE.into(),
             ..default()
diff --git a/examples/ui/render_ui_to_texture.rs b/examples/ui/render_ui_to_texture.rs
index 7a8b211766dd5..f4d8d665f26a8 100644
--- a/examples/ui/render_ui_to_texture.rs
+++ b/examples/ui/render_ui_to_texture.rs
@@ -65,6 +65,8 @@ fn setup(
         .spawn((
             Camera2d,
             Camera {
+                // render before the "main pass" camera
+                order: -1,
                 target: RenderTarget::Image(image_handle.clone().into()),
                 ..default()
             },
