diff --git a/crates/bevy_ecs/src/hierarchy.rs b/crates/bevy_ecs/src/hierarchy.rs
index d325d5756cb1c..ad390db97bf72 100644
--- a/crates/bevy_ecs/src/hierarchy.rs
+++ b/crates/bevy_ecs/src/hierarchy.rs
@@ -808,6 +808,9 @@ mod tests {
     fn spawn_many_children() {
         let mut world = World::new();
 
+        // ensure an empty set can be mentioned
+        world.spawn(children![]);
+
         // 12 children should result in a flat tuple
         let id = world
             .spawn(children![(), (), (), (), (), (), (), (), (), (), (), ()])
diff --git a/crates/bevy_ecs/src/spawn.rs b/crates/bevy_ecs/src/spawn.rs
index e4b7340c9ccfc..b127c7f7dccb7 100644
--- a/crates/bevy_ecs/src/spawn.rs
+++ b/crates/bevy_ecs/src/spawn.rs
@@ -513,6 +513,7 @@ macro_rules! related {
 #[doc(hidden)]
 macro_rules! recursive_spawn {
     // direct expansion
+    () => { () };
     ($a:expr) => {
         $crate::spawn::Spawn($a)
     };
