diff --git a/crates/bevy_render/src/camera.rs b/crates/bevy_render/src/camera.rs
index 085e7733ad397..05df0b268c856 100644
--- a/crates/bevy_render/src/camera.rs
+++ b/crates/bevy_render/src/camera.rs
@@ -574,8 +574,8 @@ pub fn extract_cameras(
                 commands.remove::<RenderLayers>();
             }
 
-            if let Some(perspective) = projection {
-                commands.insert(perspective.clone());
+            if let Some(projection) = projection {
+                commands.insert(projection.clone());
             } else {
                 commands.remove::<Projection>();
             }
