diff --git a/examples/stress_tests/many_foxes.rs b/examples/stress_tests/many_foxes.rs
index 10d7ece5929c7..346a17a2d75d4 100644
--- a/examples/stress_tests/many_foxes.rs
+++ b/examples/stress_tests/many_foxes.rs
@@ -119,7 +119,7 @@ fn setup(
     ];
     let mut animation_graph = AnimationGraph::new();
     let node_indices = animation_graph
-        .add_clips(animation_clips.iter().cloned(), 1.0, animation_graph.root)
+        .add_clips(animation_clips, 1.0, animation_graph.root)
         .collect();
     commands.insert_resource(Animations {
         node_indices,
@@ -241,9 +241,10 @@ fn setup_scene_once_loaded(
             if !foxes.sync {
                 playing_animation.seek_to(scene_ready.entity.index_u32() as f32 / 10.0);
             }
-            commands
-                .entity(child)
-                .insert(AnimationGraphHandle(animations.graph.clone()));
+            commands.entity(child).insert((
+                AnimationGraphHandle(animations.graph.clone()),
+                AnimationTransitions::default(),
+            ));
         }
     }
 }
