diff --git a/.github/workflows/docs.yml b/.github/workflows/docs.yml
index 0b661e9986949..a4e319d710933 100644
--- a/.github/workflows/docs.yml
+++ b/.github/workflows/docs.yml
@@ -63,6 +63,8 @@ jobs:
           RUSTDOCFLAGS: -Zunstable-options --cfg=docsrs --generate-link-to-definition --html-after-content docs-rs/trait-tags.html
         run: |
           cargo doc \
+            -Zunstable-options \
+            -Zrustdoc-scrape-examples \
             --all-features \
             --workspace \
             --no-deps \
diff --git a/Cargo.toml b/Cargo.toml
index 7dda86c8f1418..46d7d7a7a6394 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -762,7 +762,8 @@ web-sys = { version = "0.3", features = ["Window"] }
 [[example]]
 name = "context_menu"
 path = "examples/usage/context_menu.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.context_menu]
 name = "Context Menu"
@@ -773,7 +774,8 @@ wasm = true
 [[example]]
 name = "hello_world"
 path = "examples/hello_world.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.hello_world]
 hidden = true
@@ -804,7 +806,8 @@ wasm = true
 [[example]]
 name = "2d_viewport_to_world"
 path = "examples/2d/2d_viewport_to_world.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.2d_viewport_to_world]
 name = "2D Viewport To World"
@@ -815,7 +818,8 @@ wasm = true
 [[example]]
 name = "rotation"
 path = "examples/2d/rotation.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.rotation]
 name = "2D Rotation"
@@ -826,7 +830,8 @@ wasm = true
 [[example]]
 name = "mesh2d"
 path = "examples/2d/mesh2d.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.mesh2d]
 name = "Mesh 2D"
@@ -837,7 +842,8 @@ wasm = true
 [[example]]
 name = "mesh2d_arcs"
 path = "examples/2d/mesh2d_arcs.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.mesh2d_arcs]
 name = "Arc 2D Meshes"
@@ -848,7 +854,8 @@ wasm = true
 [[example]]
 name = "mesh2d_manual"
 path = "examples/2d/mesh2d_manual.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.mesh2d_manual]
 name = "Manual Mesh 2D"
@@ -859,7 +866,8 @@ wasm = true
 [[example]]
 name = "mesh2d_vertex_color_texture"
 path = "examples/2d/mesh2d_vertex_color_texture.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.mesh2d_vertex_color_texture]
 name = "Mesh 2D With Vertex Colors"
@@ -870,7 +878,8 @@ wasm = true
 [[example]]
 name = "2d_shapes"
 path = "examples/2d/2d_shapes.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.2d_shapes]
 name = "2D Shapes"
@@ -881,7 +890,8 @@ wasm = true
 [[example]]
 name = "cpu_draw"
 path = "examples/2d/cpu_draw.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.cpu_draw]
 name = "CPU Drawing"
@@ -892,7 +902,8 @@ wasm = true
 [[example]]
 name = "sprite"
 path = "examples/2d/sprite.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.sprite]
 name = "Sprite"
@@ -903,7 +914,8 @@ wasm = true
 [[example]]
 name = "sprite_animation"
 path = "examples/2d/sprite_animation.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.sprite_animation]
 name = "Sprite Animation"
@@ -914,7 +926,8 @@ wasm = true
 [[example]]
 name = "sprite_scale"
 path = "examples/2d/sprite_scale.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.sprite_scale]
 name = "Sprite Scale"
@@ -925,7 +938,8 @@ wasm = true
 [[example]]
 name = "sprite_flipping"
 path = "examples/2d/sprite_flipping.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.sprite_flipping]
 name = "Sprite Flipping"
@@ -958,7 +972,8 @@ wasm = true
 [[example]]
 name = "sprite_slice"
 path = "examples/2d/sprite_slice.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.sprite_slice]
 name = "Sprite Slice"
@@ -969,7 +984,8 @@ wasm = true
 [[example]]
 name = "text2d"
 path = "examples/2d/text2d.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.text2d]
 name = "Text 2D"
@@ -980,7 +996,8 @@ wasm = true
 [[example]]
 name = "multi_window_text"
 path = "examples/window/multi_window_text.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.multi_window_text]
 name = "Multi-Window Text"
@@ -991,7 +1008,8 @@ wasm = true
 [[example]]
 name = "texture_atlas"
 path = "examples/2d/texture_atlas.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.texture_atlas]
 name = "Texture Atlas"
@@ -1003,7 +1021,8 @@ wasm = false
 [[example]]
 name = "tilemap_chunk"
 path = "examples/2d/tilemap_chunk.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.tilemap_chunk]
 name = "Tilemap Chunk"
@@ -1014,7 +1033,8 @@ wasm = true
 [[example]]
 name = "transparency_2d"
 path = "examples/2d/transparency_2d.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.transparency_2d]
 name = "Transparency in 2D"
@@ -1047,7 +1067,8 @@ wasm = true
 [[example]]
 name = "pixel_grid_snap"
 path = "examples/2d/pixel_grid_snap.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.pixel_grid_snap]
 name = "Pixel Grid Snapping"
@@ -1058,7 +1079,8 @@ wasm = true
 [[example]]
 name = "bounding_2d"
 path = "examples/math/bounding_2d.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.bounding_2d]
 name = "Bounding Volume Intersections (2D)"
@@ -1069,7 +1091,8 @@ wasm = true
 [[example]]
 name = "wireframe_2d"
 path = "examples/2d/wireframe_2d.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.wireframe_2d]
 name = "2D Wireframe"
@@ -1082,7 +1105,8 @@ wasm = false
 [[example]]
 name = "3d_scene"
 path = "examples/3d/3d_scene.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.3d_scene]
 name = "3D Scene"
@@ -1093,7 +1117,8 @@ wasm = true
 [[example]]
 name = "3d_shapes"
 path = "examples/3d/3d_shapes.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.3d_shapes]
 name = "3D Shapes"
@@ -1104,7 +1129,8 @@ wasm = true
 [[example]]
 name = "3d_viewport_to_world"
 path = "examples/3d/3d_viewport_to_world.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.3d_viewport_to_world]
 name = "3D Viewport To World"
@@ -1126,7 +1152,8 @@ wasm = true
 [[example]]
 name = "generate_custom_mesh"
 path = "examples/3d/generate_custom_mesh.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.generate_custom_mesh]
 name = "Generate Custom Mesh"
@@ -1137,7 +1164,8 @@ wasm = true
 [[example]]
 name = "anti_aliasing"
 path = "examples/3d/anti_aliasing.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.anti_aliasing]
 name = "Anti-aliasing"
@@ -1149,7 +1177,8 @@ wasm = false
 [[example]]
 name = "atmospheric_fog"
 path = "examples/3d/atmospheric_fog.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.atmospheric_fog]
 name = "Atmospheric Fog"
@@ -1160,7 +1189,8 @@ wasm = true
 [[example]]
 name = "atmosphere"
 path = "examples/3d/atmosphere.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.atmosphere]
 name = "Atmosphere"
@@ -1171,7 +1201,8 @@ wasm = true
 [[example]]
 name = "fog"
 path = "examples/3d/fog.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.fog]
 name = "Fog"
@@ -1182,7 +1213,8 @@ wasm = true
 [[example]]
 name = "auto_exposure"
 path = "examples/3d/auto_exposure.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.auto_exposure]
 name = "Auto Exposure"
@@ -1194,7 +1226,8 @@ wasm = false
 [[example]]
 name = "blend_modes"
 path = "examples/3d/blend_modes.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.blend_modes]
 name = "Blend Modes"
@@ -1205,7 +1238,8 @@ wasm = true
 [[example]]
 name = "lighting"
 path = "examples/3d/lighting.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.lighting]
 name = "Lighting"
@@ -1216,7 +1250,8 @@ wasm = true
 [[example]]
 name = "lines"
 path = "examples/3d/lines.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.lines]
 name = "Lines"
@@ -1228,7 +1263,8 @@ wasm = false
 [[example]]
 name = "ssao"
 path = "examples/3d/ssao.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.ssao]
 name = "Screen Space Ambient Occlusion"
@@ -1240,7 +1276,8 @@ wasm = false
 [[example]]
 name = "spotlight"
 path = "examples/3d/spotlight.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.spotlight]
 name = "Spotlight"
@@ -1251,7 +1288,8 @@ wasm = true
 [[example]]
 name = "bloom_3d"
 path = "examples/3d/bloom_3d.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.bloom_3d]
 name = "3D Bloom"
@@ -1262,7 +1300,8 @@ wasm = true
 [[example]]
 name = "decal"
 path = "examples/3d/decal.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["free_camera"]
 
 [package.metadata.example.decal]
@@ -1274,7 +1313,8 @@ wasm = true
 [[example]]
 name = "deferred_rendering"
 path = "examples/3d/deferred_rendering.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.deferred_rendering]
 name = "Deferred Rendering"
@@ -1285,7 +1325,8 @@ wasm = true
 [[example]]
 name = "motion_blur"
 path = "examples/3d/motion_blur.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.motion_blur]
 name = "Motion Blur"
@@ -1296,7 +1337,8 @@ wasm = true
 [[example]]
 name = "order_independent_transparency"
 path = "examples/3d/order_independent_transparency.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.order_independent_transparency]
 name = "Order Independent Transparency"
@@ -1308,7 +1350,8 @@ wasm = false
 [[example]]
 name = "tonemapping"
 path = "examples/3d/tonemapping.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.tonemapping]
 name = "Tonemapping"
@@ -1319,7 +1362,8 @@ wasm = true
 [[example]]
 name = "orthographic"
 path = "examples/3d/orthographic.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.orthographic]
 name = "Orthographic View"
@@ -1341,7 +1385,8 @@ wasm = true
 [[example]]
 name = "pbr"
 path = "examples/3d/pbr.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.pbr]
 name = "Physically Based Rendering"
@@ -1352,7 +1397,8 @@ wasm = true
 [[example]]
 name = "parallax_mapping"
 path = "examples/3d/parallax_mapping.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.parallax_mapping]
 name = "Parallax Mapping"
@@ -1363,7 +1409,8 @@ wasm = true
 [[example]]
 name = "render_to_texture"
 path = "examples/3d/render_to_texture.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.render_to_texture]
 name = "Render to Texture"
@@ -1374,7 +1421,8 @@ wasm = true
 [[example]]
 name = "shadow_biases"
 path = "examples/3d/shadow_biases.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["free_camera"]
 
 [package.metadata.example.shadow_biases]
@@ -1386,7 +1434,8 @@ wasm = true
 [[example]]
 name = "shadow_caster_receiver"
 path = "examples/3d/shadow_caster_receiver.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.shadow_caster_receiver]
 name = "Shadow Caster and Receiver"
@@ -1397,7 +1446,8 @@ wasm = true
 [[example]]
 name = "skybox"
 path = "examples/3d/skybox.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["free_camera"]
 
 [package.metadata.example.skybox]
@@ -1409,7 +1459,8 @@ wasm = true
 [[example]]
 name = "solari"
 path = "examples/3d/solari.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["bevy_solari", "https", "free_camera"]
 
 [package.metadata.example.solari]
@@ -1432,7 +1483,8 @@ wasm = true
 [[example]]
 name = "split_screen"
 path = "examples/3d/split_screen.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.split_screen]
 name = "Split Screen"
@@ -1443,7 +1495,8 @@ wasm = true
 [[example]]
 name = "texture"
 path = "examples/3d/texture.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.texture]
 name = "Texture"
@@ -1454,7 +1507,8 @@ wasm = true
 [[example]]
 name = "transparency_3d"
 path = "examples/3d/transparency_3d.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.transparency_3d]
 name = "Transparency in 3D"
@@ -1465,7 +1519,8 @@ wasm = true
 [[example]]
 name = "transmission"
 path = "examples/3d/transmission.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.transmission]
 name = "Transmission"
@@ -1476,7 +1531,8 @@ wasm = true
 [[example]]
 name = "two_passes"
 path = "examples/3d/two_passes.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.two_passes]
 name = "Two Passes"
@@ -1498,7 +1554,8 @@ wasm = true
 [[example]]
 name = "wireframe"
 path = "examples/3d/wireframe.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.wireframe]
 name = "Wireframe"
@@ -1510,7 +1567,8 @@ wasm = false
 [[example]]
 name = "irradiance_volumes"
 path = "examples/3d/irradiance_volumes.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.irradiance_volumes]
 name = "Irradiance Volumes"
@@ -1523,7 +1581,8 @@ wasm = false
 [[example]]
 name = "meshlet"
 path = "examples/3d/meshlet.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["meshlet", "https", "free_camera"]
 
 [package.metadata.example.meshlet]
@@ -1536,7 +1595,8 @@ wasm = false
 [[example]]
 name = "mesh_ray_cast"
 path = "examples/3d/mesh_ray_cast.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.mesh_ray_cast]
 name = "Mesh Ray Cast"
@@ -1547,7 +1607,8 @@ wasm = true
 [[example]]
 name = "lightmaps"
 path = "examples/3d/lightmaps.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.lightmaps]
 name = "Lightmaps"
@@ -1567,7 +1628,8 @@ hidden = true
 [[example]]
 name = "animation_events"
 path = "examples/animation/animation_events.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.animation_events]
 name = "Animation Events"
@@ -1578,7 +1640,8 @@ wasm = true
 [[example]]
 name = "animated_mesh"
 path = "examples/animation/animated_mesh.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.animated_mesh]
 name = "Animated Mesh"
@@ -1589,7 +1652,8 @@ wasm = true
 [[example]]
 name = "animated_mesh_control"
 path = "examples/animation/animated_mesh_control.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.animated_mesh_control]
 name = "Animated Mesh Control"
@@ -1600,7 +1664,8 @@ wasm = true
 [[example]]
 name = "animated_mesh_events"
 path = "examples/animation/animated_mesh_events.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.animated_mesh_events]
 name = "Animated Mesh Events"
@@ -1611,7 +1676,8 @@ wasm = true
 [[example]]
 name = "animation_graph"
 path = "examples/animation/animation_graph.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.animation_graph]
 name = "Animation Graph"
@@ -1622,7 +1688,8 @@ wasm = true
 [[example]]
 name = "morph_targets"
 path = "examples/animation/morph_targets.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.morph_targets]
 name = "Morph Targets"
@@ -1644,7 +1711,8 @@ wasm = true
 [[example]]
 name = "color_animation"
 path = "examples/animation/color_animation.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.color_animation]
 name = "Color animation"
@@ -1655,7 +1723,8 @@ wasm = true
 [[example]]
 name = "eased_motion"
 path = "examples/animation/eased_motion.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.eased_motion]
 name = "Eased Motion"
@@ -1666,7 +1735,8 @@ wasm = true
 [[example]]
 name = "easing_functions"
 path = "examples/animation/easing_functions.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.easing_functions]
 name = "Easing Functions"
@@ -1677,7 +1747,8 @@ wasm = true
 [[example]]
 name = "custom_skinned_mesh"
 path = "examples/animation/custom_skinned_mesh.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.custom_skinned_mesh]
 name = "Custom Skinned Mesh"
@@ -1689,7 +1760,8 @@ wasm = true
 [[example]]
 name = "custom_loop"
 path = "examples/app/custom_loop.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.custom_loop]
 name = "Custom Loop"
@@ -1773,7 +1845,8 @@ wasm = false
 [[example]]
 name = "log_layers_ecs"
 path = "examples/app/log_layers_ecs.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.log_layers_ecs]
 name = "Advanced log layers"
@@ -1785,7 +1858,8 @@ wasm = false
 [[example]]
 name = "plugin"
 path = "examples/app/plugin.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.plugin]
 name = "Plugin"
@@ -1840,7 +1914,8 @@ wasm = false
 [[example]]
 name = "headless_renderer"
 path = "examples/app/headless_renderer.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.headless_renderer]
 name = "Headless Renderer"
@@ -1851,7 +1926,8 @@ wasm = false
 [[example]]
 name = "without_winit"
 path = "examples/app/without_winit.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.without_winit]
 name = "Without Winit"
@@ -1863,7 +1939,8 @@ wasm = false
 [[example]]
 name = "alter_mesh"
 path = "examples/asset/alter_mesh.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.alter_mesh]
 name = "Alter Mesh"
@@ -1874,7 +1951,8 @@ wasm = false
 [[example]]
 name = "alter_sprite"
 path = "examples/asset/alter_sprite.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.alter_sprite]
 name = "Alter Sprite"
@@ -1885,7 +1963,8 @@ wasm = false
 [[example]]
 name = "asset_loading"
 path = "examples/asset/asset_loading.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.asset_loading]
 name = "Asset Loading"
@@ -1907,7 +1986,8 @@ wasm = false
 [[example]]
 name = "asset_decompression"
 path = "examples/asset/asset_decompression.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.asset_decompression]
 name = "Asset Decompression"
@@ -1918,7 +1998,8 @@ wasm = false
 [[example]]
 name = "custom_asset"
 path = "examples/asset/custom_asset.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.custom_asset]
 name = "Custom Asset"
@@ -1929,7 +2010,8 @@ wasm = true
 [[example]]
 name = "custom_asset_reader"
 path = "examples/asset/custom_asset_reader.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.custom_asset_reader]
 name = "Custom Asset IO"
@@ -1988,7 +2070,8 @@ wasm = false
 [[example]]
 name = "asset_processing"
 path = "examples/asset/processing/asset_processing.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["file_watcher", "asset_processor"]
 
 [package.metadata.example.asset_processing]
@@ -2000,7 +2083,8 @@ wasm = false
 [[example]]
 name = "repeated_texture"
 path = "examples/asset/repeated_texture.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.repeated_texture]
 name = "Repeated texture configuration"
@@ -2012,7 +2096,8 @@ wasm = true
 [[example]]
 name = "multi_asset_sync"
 path = "examples/asset/multi_asset_sync.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.multi_asset_sync]
 name = "Multi-asset synchronization"
@@ -2024,7 +2109,8 @@ wasm = true
 [[example]]
 name = "async_compute"
 path = "examples/async_tasks/async_compute.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.async_compute]
 name = "Async Compute"
@@ -2035,7 +2121,8 @@ wasm = false
 [[example]]
 name = "async_channel_pattern"
 path = "examples/async_tasks/async_channel_pattern.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.async_channel_pattern]
 name = "Async Channel Pattern"
@@ -2046,7 +2133,8 @@ wasm = true
 [[example]]
 name = "external_source_external_thread"
 path = "examples/async_tasks/external_source_external_thread.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.external_source_external_thread]
 name = "External Source of Data on an External Thread"
@@ -2080,7 +2168,8 @@ wasm = true
 [[example]]
 name = "decodable"
 path = "examples/audio/decodable.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.decodable]
 name = "Decodable"
@@ -2091,7 +2180,8 @@ wasm = true
 [[example]]
 name = "soundtrack"
 path = "examples/audio/soundtrack.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.soundtrack]
 name = "Soundtrack"
@@ -2102,7 +2192,8 @@ wasm = true
 [[example]]
 name = "spatial_audio_2d"
 path = "examples/audio/spatial_audio_2d.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.spatial_audio_2d]
 name = "Spatial Audio 2D"
@@ -2113,7 +2204,8 @@ wasm = true
 [[example]]
 name = "spatial_audio_3d"
 path = "examples/audio/spatial_audio_3d.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.spatial_audio_3d]
 name = "Spatial Audio 3D"
@@ -2124,7 +2216,8 @@ wasm = true
 [[example]]
 name = "pitch"
 path = "examples/audio/pitch.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.pitch]
 name = "Pitch"
@@ -2136,7 +2229,8 @@ wasm = true
 [[example]]
 name = "log_diagnostics"
 path = "examples/diagnostics/log_diagnostics.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.log_diagnostics]
 name = "Log Diagnostics"
@@ -2147,7 +2241,8 @@ wasm = true
 [[example]]
 name = "custom_diagnostic"
 path = "examples/diagnostics/custom_diagnostic.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.custom_diagnostic]
 name = "Custom Diagnostic"
@@ -2170,7 +2265,8 @@ wasm = true
 [[example]]
 name = "ecs_guide"
 path = "examples/ecs/ecs_guide.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.ecs_guide]
 name = "ECS Guide"
@@ -2181,7 +2277,8 @@ wasm = false
 [[example]]
 name = "change_detection"
 path = "examples/ecs/change_detection.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["track_location"]
 
 [package.metadata.example.change_detection]
@@ -2193,7 +2290,8 @@ wasm = false
 [[example]]
 name = "component_hooks"
 path = "examples/ecs/component_hooks.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.component_hooks]
 name = "Component Hooks"
@@ -2204,7 +2302,8 @@ wasm = false
 [[example]]
 name = "custom_schedule"
 path = "examples/ecs/custom_schedule.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.custom_schedule]
 name = "Custom Schedule"
@@ -2215,7 +2314,8 @@ wasm = false
 [[example]]
 name = "custom_query_param"
 path = "examples/ecs/custom_query_param.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.custom_query_param]
 name = "Custom Query Parameters"
@@ -2226,7 +2326,8 @@ wasm = false
 [[example]]
 name = "dynamic"
 path = "examples/ecs/dynamic.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["debug"]
 
 [package.metadata.example.dynamic]
@@ -2238,7 +2339,8 @@ wasm = false
 [[example]]
 name = "message"
 path = "examples/ecs/message.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.message]
 name = "Message"
@@ -2249,7 +2351,8 @@ wasm = false
 [[example]]
 name = "send_and_receive_messages"
 path = "examples/ecs/send_and_receive_messages.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.send_and_receive_messages]
 name = "Send and receive messages"
@@ -2260,7 +2363,8 @@ wasm = false
 [[example]]
 name = "entity_disabling"
 path = "examples/ecs/entity_disabling.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.entity_disabling]
 name = "Entity disabling"
@@ -2271,7 +2375,8 @@ wasm = true
 [[example]]
 name = "fixed_timestep"
 path = "examples/ecs/fixed_timestep.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.fixed_timestep]
 name = "Fixed Timestep"
@@ -2282,7 +2387,8 @@ wasm = false
 [[example]]
 name = "generic_system"
 path = "examples/ecs/generic_system.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.generic_system]
 name = "Generic System"
@@ -2293,7 +2399,8 @@ wasm = false
 [[example]]
 name = "hierarchy"
 path = "examples/ecs/hierarchy.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.hierarchy]
 name = "Hierarchy"
@@ -2304,7 +2411,8 @@ wasm = false
 [[example]]
 name = "immutable_components"
 path = "examples/ecs/immutable_components.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.immutable_components]
 name = "Immutable Components"
@@ -2315,7 +2423,8 @@ wasm = false
 [[example]]
 name = "iter_combinations"
 path = "examples/ecs/iter_combinations.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.iter_combinations]
 name = "Iter Combinations"
@@ -2326,7 +2435,8 @@ wasm = true
 [[example]]
 name = "one_shot_systems"
 path = "examples/ecs/one_shot_systems.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.one_shot_systems]
 name = "One Shot Systems"
@@ -2337,7 +2447,8 @@ wasm = false
 [[example]]
 name = "parallel_query"
 path = "examples/ecs/parallel_query.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.parallel_query]
 name = "Parallel Query"
@@ -2348,7 +2459,8 @@ wasm = false
 [[example]]
 name = "relationships"
 path = "examples/ecs/relationships.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.relationships]
 name = "Relationships"
@@ -2359,7 +2471,8 @@ wasm = false
 [[example]]
 name = "removal_detection"
 path = "examples/ecs/removal_detection.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.removal_detection]
 name = "Removal Detection"
@@ -2370,7 +2483,8 @@ wasm = false
 [[example]]
 name = "run_conditions"
 path = "examples/ecs/run_conditions.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.run_conditions]
 name = "Run Conditions"
@@ -2381,7 +2495,8 @@ wasm = false
 [[example]]
 name = "fallible_params"
 path = "examples/ecs/fallible_params.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.fallible_params]
 name = "Fallible System Parameters"
@@ -2392,7 +2507,8 @@ wasm = false
 [[example]]
 name = "error_handling"
 path = "examples/ecs/error_handling.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["mesh_picking"]
 
 [package.metadata.example.error_handling]
@@ -2404,7 +2520,8 @@ wasm = false
 [[example]]
 name = "startup_system"
 path = "examples/ecs/startup_system.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.startup_system]
 name = "Startup System"
@@ -2415,7 +2532,8 @@ wasm = false
 [[example]]
 name = "states"
 path = "examples/state/states.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.states]
 name = "States"
@@ -2426,7 +2544,8 @@ wasm = false
 [[example]]
 name = "sub_states"
 path = "examples/state/sub_states.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["bevy_dev_tools"]
 
 [package.metadata.example.sub_states]
@@ -2438,7 +2557,8 @@ wasm = false
 [[example]]
 name = "computed_states"
 path = "examples/state/computed_states.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["bevy_dev_tools"]
 
 [package.metadata.example.computed_states]
@@ -2450,7 +2570,8 @@ wasm = false
 [[example]]
 name = "custom_transitions"
 path = "examples/state/custom_transitions.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["bevy_dev_tools"]
 
 [package.metadata.example.custom_transitions]
@@ -2462,7 +2583,8 @@ wasm = false
 [[example]]
 name = "system_piping"
 path = "examples/ecs/system_piping.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.system_piping]
 name = "System Piping"
@@ -2473,7 +2595,8 @@ wasm = false
 [[example]]
 name = "state_scoped"
 path = "examples/ecs/state_scoped.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.state_scoped]
 name = "State Scoped"
@@ -2484,7 +2607,8 @@ wasm = false
 [[example]]
 name = "system_closure"
 path = "examples/ecs/system_closure.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.system_closure]
 name = "System Closure"
@@ -2519,7 +2643,8 @@ wasm = false
 [[example]]
 name = "time"
 path = "examples/time/time.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.time]
 name = "Time handling"
@@ -2530,7 +2655,8 @@ wasm = false
 [[example]]
 name = "virtual_time"
 path = "examples/time/virtual_time.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.virtual_time]
 name = "Virtual time"
@@ -2541,7 +2667,8 @@ wasm = false
 [[example]]
 name = "timers"
 path = "examples/time/timers.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.timers]
 name = "Timers"
@@ -2554,7 +2681,8 @@ wasm = false
 [[example]]
 name = "alien_cake_addict"
 path = "examples/games/alien_cake_addict.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.alien_cake_addict]
 name = "Alien Cake Addict"
@@ -2565,7 +2693,8 @@ wasm = true
 [[example]]
 name = "breakout"
 path = "examples/games/breakout.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.breakout]
 name = "Breakout"
@@ -2576,7 +2705,8 @@ wasm = true
 [[example]]
 name = "contributors"
 path = "examples/games/contributors.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.contributors]
 name = "Contributors"
@@ -2587,7 +2717,8 @@ wasm = true
 [[example]]
 name = "desk_toy"
 path = "examples/games/desk_toy.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.desk_toy]
 name = "Desk Toy"
@@ -2598,7 +2729,8 @@ wasm = false
 [[example]]
 name = "game_menu"
 path = "examples/games/game_menu.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.game_menu]
 name = "Game Menu"
@@ -2609,7 +2741,8 @@ wasm = true
 [[example]]
 name = "loading_screen"
 path = "examples/games/loading_screen.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.loading_screen]
 name = "Loading Screen"
@@ -2621,7 +2754,8 @@ wasm = true
 [[example]]
 name = "char_input_events"
 path = "examples/input/char_input_events.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.char_input_events]
 name = "Char Input Events"
@@ -2753,7 +2887,8 @@ wasm = false
 [[example]]
 name = "text_input"
 path = "examples/input/text_input.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.text_input]
 name = "Text Input"
@@ -2765,7 +2900,8 @@ wasm = false
 [[example]]
 name = "reflection"
 path = "examples/reflection/reflection.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.reflection]
 name = "Reflection"
@@ -2867,7 +3003,8 @@ wasm = false
 [[example]]
 name = "scene"
 path = "examples/scene/scene.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.scene]
 name = "Scene"
@@ -2889,7 +3026,8 @@ There are also compute shaders which are used for more general processing levera
 [[example]]
 name = "shader_defs"
 path = "examples/shader/shader_defs.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.shader_defs]
 name = "Shader Defs"
@@ -2922,7 +3060,8 @@ wasm = true
 [[example]]
 name = "extended_material"
 path = "examples/shader/extended_material.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.extended_material]
 name = "Extended Material"
@@ -2933,7 +3072,8 @@ wasm = true
 [[example]]
 name = "shader_prepass"
 path = "examples/shader/shader_prepass.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.shader_prepass]
 name = "Material Prepass"
@@ -2944,7 +3084,8 @@ wasm = false
 [[example]]
 name = "shader_material_screenspace_texture"
 path = "examples/shader/shader_material_screenspace_texture.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.shader_material_screenspace_texture]
 name = "Material - Screenspace Texture"
@@ -2979,7 +3120,8 @@ wasm = true
 [[example]]
 name = "automatic_instancing"
 path = "examples/shader/automatic_instancing.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.automatic_instancing]
 name = "Instancing"
@@ -2990,7 +3132,8 @@ wasm = true
 [[example]]
 name = "animate_shader"
 path = "examples/shader/animate_shader.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.animate_shader]
 name = "Animated"
@@ -3001,7 +3144,8 @@ wasm = true
 [[example]]
 name = "compute_shader_game_of_life"
 path = "examples/shader/compute_shader_game_of_life.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.compute_shader_game_of_life]
 name = "Compute - Game of Life"
@@ -3012,7 +3156,8 @@ wasm = false
 [[example]]
 name = "gpu_readback"
 path = "examples/shader/gpu_readback.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.gpu_readback]
 name = "GPU readback"
@@ -3023,7 +3168,8 @@ wasm = false
 [[example]]
 name = "array_texture"
 path = "examples/shader/array_texture.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.array_texture]
 name = "Array Texture"
@@ -3065,7 +3211,8 @@ wasm = true
 [[example]]
 name = "custom_post_processing"
 path = "examples/shader_advanced/custom_post_processing.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.custom_post_processing]
 name = "Post Processing - Custom Render Pass"
@@ -3076,7 +3223,8 @@ wasm = true
 [[example]]
 name = "custom_shader_instancing"
 path = "examples/shader_advanced/custom_shader_instancing.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.custom_shader_instancing]
 name = "Instancing"
@@ -3087,7 +3235,8 @@ wasm = true
 [[example]]
 name = "custom_render_phase"
 path = "examples/shader_advanced/custom_render_phase.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.custom_render_phase]
 name = "Custom Render Phase"
@@ -3098,7 +3247,8 @@ wasm = true
 [[example]]
 name = "manual_material"
 path = "examples/shader_advanced/manual_material.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.manual_material]
 name = "Manual Material Implementation"
@@ -3109,7 +3259,8 @@ wasm = true
 [[example]]
 name = "texture_binding_array"
 path = "examples/shader_advanced/texture_binding_array.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.texture_binding_array]
 name = "Texture Binding Array (Bindless Textures)"
@@ -3120,7 +3271,8 @@ wasm = false
 [[example]]
 name = "specialized_mesh_pipeline"
 path = "examples/shader_advanced/specialized_mesh_pipeline.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.specialized_mesh_pipeline]
 name = "Specialized Mesh Pipeline"
@@ -3131,7 +3283,8 @@ wasm = true
 [[example]]
 name = "fullscreen_material"
 path = "examples/shader_advanced/fullscreen_material.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.fullscreen_material]
 name = "Fullscreen Material"
@@ -3155,7 +3308,8 @@ cargo run --release --example <example name>
 [[example]]
 name = "bevymark"
 path = "examples/stress_tests/bevymark.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.bevymark]
 name = "Bevymark"
@@ -3166,7 +3320,8 @@ wasm = true
 [[example]]
 name = "many_animated_sprites"
 path = "examples/stress_tests/many_animated_sprites.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.many_animated_sprites]
 name = "Many Animated Sprites"
@@ -3177,7 +3332,8 @@ wasm = true
 [[example]]
 name = "many_buttons"
 path = "examples/stress_tests/many_buttons.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.many_buttons]
 name = "Many Buttons"
@@ -3188,7 +3344,8 @@ wasm = true
 [[example]]
 name = "many_gradients"
 path = "examples/stress_tests/many_gradients.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.many_gradients]
 name = "Many Gradients"
@@ -3199,7 +3356,8 @@ wasm = true
 [[example]]
 name = "many_cameras_lights"
 path = "examples/stress_tests/many_cameras_lights.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.many_cameras_lights]
 name = "Many Cameras & Lights"
@@ -3210,7 +3368,8 @@ wasm = true
 [[example]]
 name = "many_components"
 path = "examples/stress_tests/many_components.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.many_components]
 name = "Many Components (and Entities and Systems)"
@@ -3221,7 +3380,8 @@ wasm = false
 [[example]]
 name = "many_cubes"
 path = "examples/stress_tests/many_cubes.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.many_cubes]
 name = "Many Cubes"
@@ -3232,7 +3392,8 @@ wasm = true
 [[example]]
 name = "many_gizmos"
 path = "examples/stress_tests/many_gizmos.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.many_gizmos]
 name = "Many Gizmos"
@@ -3243,7 +3404,8 @@ wasm = true
 [[example]]
 name = "many_foxes"
 path = "examples/stress_tests/many_foxes.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.many_foxes]
 name = "Many Foxes"
@@ -3254,7 +3416,8 @@ wasm = true
 [[example]]
 name = "many_glyphs"
 path = "examples/stress_tests/many_glyphs.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.many_glyphs]
 name = "Many Glyphs"
@@ -3265,7 +3428,8 @@ wasm = true
 [[example]]
 name = "many_lights"
 path = "examples/stress_tests/many_lights.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.many_lights]
 name = "Many Lights"
@@ -3276,7 +3440,8 @@ wasm = true
 [[example]]
 name = "many_sprites"
 path = "examples/stress_tests/many_sprites.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.many_sprites]
 name = "Many Sprites"
@@ -3287,7 +3452,8 @@ wasm = true
 [[example]]
 name = "many_text2d"
 path = "examples/stress_tests/many_text2d.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.many_text2d]
 name = "Many Text2d"
@@ -3298,7 +3464,8 @@ wasm = true
 [[example]]
 name = "many_materials"
 path = "examples/stress_tests/many_materials.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.many_materials]
 name = "Many Animated Materials"
@@ -3309,7 +3476,8 @@ wasm = true
 [[example]]
 name = "transform_hierarchy"
 path = "examples/stress_tests/transform_hierarchy.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.transform_hierarchy]
 name = "Transform Hierarchy"
@@ -3320,7 +3488,8 @@ wasm = false
 [[example]]
 name = "text_pipeline"
 path = "examples/stress_tests/text_pipeline.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.text_pipeline]
 name = "Text Pipeline"
@@ -3332,7 +3501,8 @@ wasm = false
 [[example]]
 name = "scene_viewer"
 path = "examples/tools/scene_viewer/main.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["free_camera"]
 
 [package.metadata.example.scene_viewer]
@@ -3344,8 +3514,9 @@ wasm = true
 [[example]]
 name = "gamepad_viewer"
 path = "examples/tools/gamepad_viewer.rs"
-doc-scrape-examples = true
-
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
+
 [package.metadata.example.gamepad_viewer]
 name = "Gamepad Viewer"
 description = "Shows a visualization of gamepad buttons, sticks, and triggers"
@@ -3355,7 +3526,8 @@ wasm = true
 [[example]]
 name = "nondeterministic_system_order"
 path = "examples/ecs/nondeterministic_system_order.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.nondeterministic_system_order]
 name = "Nondeterministic System Order"
@@ -3366,7 +3538,8 @@ wasm = false
 [[example]]
 name = "observers"
 path = "examples/ecs/observers.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.observers]
 name = "Observers"
@@ -3377,7 +3550,8 @@ wasm = true
 [[example]]
 name = "observer_propagation"
 path = "examples/ecs/observer_propagation.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.observer_propagation]
 name = "Observer Propagation"
@@ -3388,7 +3562,8 @@ wasm = true
 [[example]]
 name = "3d_rotation"
 path = "examples/transforms/3d_rotation.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.3d_rotation]
 name = "3D Rotation"
@@ -3399,7 +3574,8 @@ wasm = true
 [[example]]
 name = "align"
 path = "examples/transforms/align.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.align]
 name = "Alignment"
@@ -3410,7 +3586,8 @@ wasm = true
 [[example]]
 name = "scale"
 path = "examples/transforms/scale.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.scale]
 name = "Scale"
@@ -3421,7 +3598,8 @@ wasm = true
 [[example]]
 name = "transform"
 path = "examples/transforms/transform.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.transform]
 name = "Transform"
@@ -3432,7 +3610,8 @@ wasm = true
 [[example]]
 name = "translation"
 path = "examples/transforms/translation.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.translation]
 name = "Translation"
@@ -3444,7 +3623,8 @@ wasm = true
 [[example]]
 name = "anchor_layout"
 path = "examples/ui/anchor_layout.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.anchor_layout]
 name = "Anchor Layout"
@@ -3455,7 +3635,8 @@ wasm = true
 [[example]]
 name = "borders"
 path = "examples/ui/borders.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.borders]
 name = "Borders"
@@ -3466,7 +3647,8 @@ wasm = true
 [[example]]
 name = "box_shadow"
 path = "examples/ui/box_shadow.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.box_shadow]
 name = "Box Shadow"
@@ -3477,7 +3659,8 @@ wasm = true
 [[example]]
 name = "button"
 path = "examples/ui/button.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.button]
 name = "Button"
@@ -3499,7 +3682,8 @@ wasm = true
 [[example]]
 name = "display_and_visibility"
 path = "examples/ui/display_and_visibility.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.display_and_visibility]
 name = "Display and Visibility"
@@ -3510,7 +3694,8 @@ wasm = true
 [[example]]
 name = "font_weights"
 path = "examples/ui/font_weights.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.font_weights]
 name = "Font Weights"
@@ -3532,7 +3717,8 @@ wasm = false
 [[example]]
 name = "vertical_slider"
 path = "examples/ui/vertical_slider.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["experimental_bevy_ui_widgets"]
 
 [package.metadata.example.vertical_slider]
@@ -3555,7 +3741,8 @@ wasm = true
 [[example]]
 name = "overflow"
 path = "examples/ui/overflow.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.overflow]
 name = "Overflow"
@@ -3566,7 +3753,8 @@ wasm = true
 [[example]]
 name = "overflow_clip_margin"
 path = "examples/ui/overflow_clip_margin.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.overflow_clip_margin]
 name = "Overflow Clip Margin"
@@ -3578,7 +3766,8 @@ wasm = true
 [[example]]
 name = "overflow_debug"
 path = "examples/ui/overflow_debug.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.overflow_debug]
 name = "Overflow and Clipping Debug"
@@ -3589,7 +3778,8 @@ wasm = true
 [[example]]
 name = "relative_cursor_position"
 path = "examples/ui/relative_cursor_position.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.relative_cursor_position]
 name = "Relative Cursor Position"
@@ -3600,7 +3790,8 @@ wasm = true
 [[example]]
 name = "render_ui_to_texture"
 path = "examples/ui/render_ui_to_texture.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.render_ui_to_texture]
 name = "Render UI to Texture"
@@ -3611,7 +3802,8 @@ wasm = true
 [[example]]
 name = "size_constraints"
 path = "examples/ui/size_constraints.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.size_constraints]
 name = "Size Constraints"
@@ -3622,7 +3814,8 @@ wasm = true
 [[example]]
 name = "strikethrough_and_underline"
 path = "examples/ui/strikethrough_and_underline.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.strikethrough_and_underline]
 name = "Strikethrough and Underline"
@@ -3633,7 +3826,8 @@ wasm = true
 [[example]]
 name = "text"
 path = "examples/ui/text.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.text]
 name = "Text"
@@ -3644,7 +3838,8 @@ wasm = true
 [[example]]
 name = "text_background_colors"
 path = "examples/ui/text_background_colors.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.text_background_colors]
 name = "Text Background Colors"
@@ -3655,7 +3850,8 @@ wasm = true
 [[example]]
 name = "text_debug"
 path = "examples/ui/text_debug.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.text_debug]
 name = "Text Debug"
@@ -3666,7 +3862,8 @@ wasm = true
 [[example]]
 name = "flex_layout"
 path = "examples/ui/flex_layout.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.flex_layout]
 name = "Flex Layout"
@@ -3677,7 +3874,8 @@ wasm = true
 [[example]]
 name = "text_wrap_debug"
 path = "examples/ui/text_wrap_debug.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.text_wrap_debug]
 name = "Text Wrap Debug"
@@ -3688,7 +3886,8 @@ wasm = true
 [[example]]
 name = "ghost_nodes"
 path = "examples/ui/ghost_nodes.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["ghost_nodes"]
 
 [package.metadata.example.ghost_nodes]
@@ -3700,7 +3899,8 @@ wasm = true
 [[example]]
 name = "grid"
 path = "examples/ui/grid.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.grid]
 name = "CSS Grid"
@@ -3711,7 +3911,8 @@ wasm = true
 [[example]]
 name = "gradients"
 path = "examples/ui/gradients.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.gradients]
 name = "Gradients"
@@ -3722,7 +3923,8 @@ wasm = true
 [[example]]
 name = "stacked_gradients"
 path = "examples/ui/stacked_gradients.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.stacked_gradients]
 name = "Stacked Gradients"
@@ -3733,7 +3935,8 @@ wasm = true
 [[example]]
 name = "scroll"
 path = "examples/ui/scroll.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.scroll]
 name = "Scroll"
@@ -3744,7 +3947,8 @@ wasm = true
 [[example]]
 name = "transparency_ui"
 path = "examples/ui/transparency_ui.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.transparency_ui]
 name = "Transparency UI"
@@ -3778,7 +3982,8 @@ wasm = true
 [[example]]
 name = "ui_scaling"
 path = "examples/ui/ui_scaling.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.ui_scaling]
 name = "UI Scaling"
@@ -3789,7 +3994,8 @@ wasm = true
 [[example]]
 name = "ui_texture_atlas"
 path = "examples/ui/ui_texture_atlas.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.ui_texture_atlas]
 name = "UI Texture Atlas"
@@ -3833,7 +4039,8 @@ wasm = true
 [[example]]
 name = "ui_transform"
 path = "examples/ui/ui_transform.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.ui_transform]
 name = "UI Transform"
@@ -3844,7 +4051,8 @@ wasm = true
 [[example]]
 name = "ui_target_camera"
 path = "examples/ui/ui_target_camera.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.ui_target_camera]
 name = "UI Target Camera"
@@ -3855,7 +4063,8 @@ wasm = true
 [[example]]
 name = "viewport_node"
 path = "examples/ui/viewport_node.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.viewport_node]
 name = "Viewport Node"
@@ -3867,7 +4076,8 @@ wasm = true
 [[example]]
 name = "clear_color"
 path = "examples/window/clear_color.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.clear_color]
 name = "Clear Color"
@@ -3878,7 +4088,8 @@ wasm = true
 [[example]]
 name = "custom_cursor_image"
 path = "examples/window/custom_cursor_image.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["custom_cursor"]
 
 [package.metadata.example.custom_cursor_image]
@@ -3890,7 +4101,8 @@ wasm = true
 [[example]]
 name = "low_power"
 path = "examples/window/low_power.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.low_power]
 name = "Low Power"
@@ -3901,7 +4113,8 @@ wasm = true
 [[example]]
 name = "multiple_windows"
 path = "examples/window/multiple_windows.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.multiple_windows]
 name = "Multiple Windows"
@@ -3945,7 +4158,8 @@ wasm = false
 [[example]]
 name = "window_settings"
 path = "examples/window/window_settings.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.window_settings]
 name = "Window Settings"
@@ -3956,7 +4170,8 @@ wasm = true
 [[example]]
 name = "window_drag_move"
 path = "examples/window/window_drag_move.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.window_drag_move]
 name = "Window Drag Move"
@@ -3967,7 +4182,8 @@ wasm = false
 [[example]]
 name = "ambiguity_detection"
 path = "tests/ecs/ambiguity_detection.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.ambiguity_detection]
 hidden = true
@@ -3975,7 +4191,8 @@ hidden = true
 [[example]]
 name = "resizing"
 path = "tests/window/resizing.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.resizing]
 hidden = true
@@ -3983,7 +4200,8 @@ hidden = true
 [[example]]
 name = "minimizing"
 path = "tests/window/minimizing.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.minimizing]
 hidden = true
@@ -3991,7 +4209,8 @@ hidden = true
 [[example]]
 name = "desktop_request_redraw"
 path = "tests/window/desktop_request_redraw.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["bevy_dev_tools"]
 
 [package.metadata.example.desktop_request_redraw]
@@ -4000,7 +4219,8 @@ hidden = true
 [[example]]
 name = "window_resizing"
 path = "examples/window/window_resizing.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [[example]]
 name = "fallback_image"
@@ -4013,7 +4233,8 @@ hidden = true
 [[example]]
 name = "reflection_probes"
 path = "examples/3d/reflection_probes.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.reflection_probes]
 name = "Reflection Probes"
@@ -4030,7 +4251,8 @@ wasm = true
 [[example]]
 name = "drag_to_scroll"
 path = "examples/ui/drag_to_scroll.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.drag_to_scroll]
 name = "Drag to Scroll"
@@ -4041,7 +4263,8 @@ wasm = true
 [[example]]
 name = "ui_material"
 path = "examples/ui/ui_material.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.ui_material]
 name = "UI Material"
@@ -4052,7 +4275,8 @@ wasm = true
 [[example]]
 name = "cubic_splines"
 path = "examples/math/cubic_splines.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.cubic_splines]
 name = "Cubic Splines"
@@ -4063,7 +4287,8 @@ wasm = true
 [[example]]
 name = "render_primitives"
 path = "examples/math/render_primitives.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.render_primitives]
 name = "Rendering Primitives"
@@ -4075,7 +4300,8 @@ wasm = true
 [[example]]
 name = "sampling_primitives"
 path = "examples/math/sampling_primitives.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.sampling_primitives]
 name = "Sampling Primitives"
@@ -4086,7 +4312,8 @@ wasm = true
 [[example]]
 name = "custom_primitives"
 path = "examples/math/custom_primitives.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.custom_primitives]
 name = "Custom Primitives"
@@ -4097,7 +4324,8 @@ wasm = true
 [[example]]
 name = "random_sampling"
 path = "examples/math/random_sampling.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.random_sampling]
 name = "Random Sampling"
@@ -4108,7 +4336,8 @@ wasm = true
 [[example]]
 name = "smooth_follow"
 path = "examples/movement/smooth_follow.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.smooth_follow]
 name = "Smooth Follow"
@@ -4120,7 +4349,8 @@ wasm = true
 [[example]]
 name = "2d_gizmos"
 path = "examples/gizmos/2d_gizmos.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.2d_gizmos]
 name = "2D Gizmos"
@@ -4131,7 +4361,8 @@ wasm = true
 [[example]]
 name = "3d_gizmos"
 path = "examples/gizmos/3d_gizmos.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["free_camera"]
 
 [package.metadata.example.3d_gizmos]
@@ -4143,7 +4374,8 @@ wasm = true
 [[example]]
 name = "axes"
 path = "examples/gizmos/axes.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.axes]
 name = "Axes"
@@ -4154,7 +4386,8 @@ wasm = true
 [[example]]
 name = "light_gizmos"
 path = "examples/gizmos/light_gizmos.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.light_gizmos]
 name = "Light Gizmos"
@@ -4165,7 +4398,8 @@ wasm = true
 [[example]]
 name = "custom_gltf_vertex_attribute"
 path = "examples/gltf/custom_gltf_vertex_attribute.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.custom_gltf_vertex_attribute]
 name = "Custom glTF vertex attribute 2D"
@@ -4176,7 +4410,8 @@ wasm = true
 [[example]]
 name = "edit_material_on_gltf"
 path = "examples/gltf/edit_material_on_gltf.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.edit_material_on_gltf]
 name = "Edit glTF Material"
@@ -4220,7 +4455,8 @@ wasm = true
 [[example]]
 name = "gltf_extension_animation_graph"
 path = "examples/gltf/gltf_extension_animation_graph.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.gltf_extension_animation_graph]
 name = "glTF extension AnimationGraph"
@@ -4231,7 +4467,8 @@ wasm = true
 [[example]]
 name = "gltf_extension_mesh_2d"
 path = "examples/gltf/gltf_extension_mesh_2d.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.gltf_extension_mesh_2d]
 name = "glTF extension processing to build Mesh2ds from glTF data"
@@ -4242,7 +4479,8 @@ wasm = true
 [[example]]
 name = "query_gltf_primitives"
 path = "examples/gltf/query_gltf_primitives.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.query_gltf_primitives]
 name = "Query glTF primitives"
@@ -4270,7 +4508,8 @@ required-features = ["bevy_dev_tools"]
 [[example]]
 name = "2d_top_down_camera"
 path = "examples/camera/2d_top_down_camera.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.2d_top_down_camera]
 name = "2D top-down camera"
@@ -4281,7 +4520,8 @@ wasm = true
 [[example]]
 name = "custom_projection"
 path = "examples/camera/custom_projection.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.custom_projection]
 name = "Custom Projection"
@@ -4292,7 +4532,8 @@ wasm = true
 [[example]]
 name = "first_person_view_model"
 path = "examples/camera/first_person_view_model.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.first_person_view_model]
 name = "First person view model"
@@ -4303,7 +4544,8 @@ wasm = true
 [[example]]
 name = "free_camera_controller"
 path = "examples/camera/free_camera_controller.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["free_camera"]
 
 [package.metadata.example.free_camera_controller]
@@ -4315,7 +4557,8 @@ wasm = true
 [[example]]
 name = "projection_zoom"
 path = "examples/camera/projection_zoom.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.projection_zoom]
 name = "Projection Zoom"
@@ -4326,7 +4569,8 @@ wasm = true
 [[example]]
 name = "camera_orbit"
 path = "examples/camera/camera_orbit.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.camera_orbit]
 name = "Camera Orbit"
@@ -4337,7 +4581,8 @@ wasm = true
 [[example]]
 name = "2d_screen_shake"
 path = "examples/camera/2d_screen_shake.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.2d_screen_shake]
 name = "Screen Shake"
@@ -4348,7 +4593,8 @@ wasm = true
 [[example]]
 name = "2d_on_ui"
 path = "examples/camera/2d_on_ui.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.2d_on_ui]
 name = "2D on Bevy UI"
@@ -4365,7 +4611,8 @@ wasm = true
 [[example]]
 name = "visibility_range"
 path = "examples/3d/visibility_range.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.visibility_range]
 name = "Visibility range"
@@ -4376,7 +4623,8 @@ wasm = true
 [[example]]
 name = "ssr"
 path = "examples/3d/ssr.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.ssr]
 name = "Screen Space Reflections"
@@ -4387,7 +4635,8 @@ wasm = false
 [[example]]
 name = "camera_sub_view"
 path = "examples/3d/camera_sub_view.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.camera_sub_view]
 name = "Camera sub view"
@@ -4398,7 +4647,8 @@ wasm = true
 [[example]]
 name = "color_grading"
 path = "examples/3d/color_grading.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.color_grading]
 name = "Color grading"
@@ -4409,7 +4659,8 @@ wasm = true
 [[example]]
 name = "clearcoat"
 path = "examples/3d/clearcoat.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["pbr_multi_layer_material_textures"]
 
 [package.metadata.example.clearcoat]
@@ -4421,7 +4672,8 @@ wasm = false
 [[example]]
 name = "depth_of_field"
 path = "examples/3d/depth_of_field.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.depth_of_field]
 name = "Depth of field"
@@ -4432,7 +4684,8 @@ wasm = false
 [[example]]
 name = "volumetric_fog"
 path = "examples/3d/volumetric_fog.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.volumetric_fog]
 name = "Volumetric fog"
@@ -4443,7 +4696,8 @@ wasm = true
 [[example]]
 name = "client"
 path = "examples/remote/client.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["bevy_remote"]
 
 [package.metadata.example.client]
@@ -4467,7 +4721,8 @@ wasm = false
 [[example]]
 name = "anisotropy"
 path = "examples/3d/anisotropy.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["jpeg", "pbr_anisotropy_texture"]
 
 [package.metadata.example.anisotropy]
@@ -4479,7 +4734,8 @@ wasm = false
 [[example]]
 name = "custom_phase_item"
 path = "examples/shader_advanced/custom_phase_item.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.custom_phase_item]
 name = "Custom phase item"
@@ -4490,7 +4746,8 @@ wasm = true
 [[example]]
 name = "fog_volumes"
 path = "examples/3d/fog_volumes.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.fog_volumes]
 name = "Fog volumes"
@@ -4501,7 +4758,8 @@ wasm = false
 [[example]]
 name = "scrolling_fog"
 path = "examples/3d/scrolling_fog.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.scrolling_fog]
 name = "Scrolling fog"
@@ -4512,7 +4770,8 @@ wasm = false
 [[example]]
 name = "physics_in_fixed_timestep"
 path = "examples/movement/physics_in_fixed_timestep.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.physics_in_fixed_timestep]
 name = "Run physics in a fixed timestep"
@@ -4523,7 +4782,8 @@ wasm = true
 [[example]]
 name = "post_processing"
 path = "examples/3d/post_processing.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.post_processing]
 name = "Built-in postprocessing"
@@ -4534,7 +4794,8 @@ wasm = true
 [[example]]
 name = "rotate_environment_map"
 path = "examples/3d/rotate_environment_map.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["pbr_multi_layer_material_textures"]
 
 [package.metadata.example.rotate_environment_map]
@@ -4546,7 +4807,8 @@ wasm = false
 [[example]]
 name = "mesh_picking"
 path = "examples/picking/mesh_picking.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["mesh_picking"]
 
 [package.metadata.example.mesh_picking]
@@ -4558,7 +4820,8 @@ wasm = true
 [[example]]
 name = "simple_picking"
 path = "examples/picking/simple_picking.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["bevy_picking"]
 
 [package.metadata.example.simple_picking]
@@ -4570,7 +4833,8 @@ wasm = true
 [[example]]
 name = "sprite_picking"
 path = "examples/picking/sprite_picking.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["sprite_picking"]
 
 [package.metadata.example.sprite_picking]
@@ -4582,7 +4846,8 @@ wasm = true
 [[example]]
 name = "debug_picking"
 path = "examples/picking/debug_picking.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["bevy_dev_tools"]
 
 [package.metadata.example.debug_picking]
@@ -4594,7 +4859,8 @@ wasm = true
 [[example]]
 name = "animation_masks"
 path = "examples/animation/animation_masks.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.animation_masks]
 name = "Animation Masks"
@@ -4605,7 +4871,8 @@ wasm = true
 [[example]]
 name = "pcss"
 path = "examples/3d/pcss.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["experimental_pbr_pcss"]
 
 [package.metadata.example.pcss]
@@ -4617,7 +4884,8 @@ wasm = false
 [[example]]
 name = "mixed_lighting"
 path = "examples/3d/mixed_lighting.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["jpeg"]
 
 [package.metadata.example.mixed_lighting]
@@ -4629,7 +4897,8 @@ wasm = true
 [[example]]
 name = "animated_ui"
 path = "examples/animation/animated_ui.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.animated_ui]
 name = "Animated UI"
@@ -4640,7 +4909,8 @@ wasm = true
 [[example]]
 name = "shader_material_bindless"
 path = "examples/shader/shader_material_bindless.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.shader_material_bindless]
 name = "Material - Bindless"
@@ -4651,7 +4921,8 @@ wasm = true
 [[example]]
 name = "specular_tint"
 path = "examples/3d/specular_tint.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["pbr_specular_textures"]
 
 [package.metadata.example.specular_tint]
@@ -4663,7 +4934,8 @@ wasm = true
 [[example]]
 name = "test_invalid_skinned_mesh"
 path = "tests/3d/test_invalid_skinned_mesh.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.test_invalid_skinned_mesh]
 hidden = true
@@ -4685,6 +4957,7 @@ panic = "abort"
 # for details on why this is needed. Since dependencies don't expect to be built
 # with `--cfg docsrs` (and thus fail to compile) we use a different cfg.
 rustc-args = ["--cfg", "docsrs_dep"]
+cargo-args = ["-Zunstable-options", "-Zrustdoc-scrape-examples"]
 rustdoc-args = [
   "-Zunstable-options",
   "--generate-link-to-definition",
@@ -4699,7 +4972,8 @@ all-features = true
 [[example]]
 name = "monitor_info"
 path = "examples/window/monitor_info.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.monitor_info]
 name = "Monitor info"
@@ -4711,7 +4985,8 @@ wasm = false
 [[example]]
 name = "testbed_2d"
 path = "examples/testbed/2d.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.testbed_2d]
 hidden = true
@@ -4719,7 +4994,8 @@ hidden = true
 [[example]]
 name = "testbed_3d"
 path = "examples/testbed/3d.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.testbed_3d]
 hidden = true
@@ -4727,7 +5003,8 @@ hidden = true
 [[example]]
 name = "testbed_ui"
 path = "examples/testbed/ui.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.testbed_ui]
 hidden = true
@@ -4735,7 +5012,8 @@ hidden = true
 [[example]]
 name = "testbed_full_ui"
 path = "examples/testbed/full_ui.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["experimental_bevy_ui_widgets"]
 
 [package.metadata.example.testbed_full_ui]
@@ -4755,7 +5033,8 @@ wasm = true
 [[example]]
 name = "directional_navigation"
 path = "examples/ui/directional_navigation.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.directional_navigation]
 name = "Directional Navigation"
@@ -4766,7 +5045,8 @@ wasm = true
 [[example]]
 name = "auto_directional_navigation"
 path = "examples/ui/auto_directional_navigation.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.auto_directional_navigation]
 name = "Automatic Directional Navigation"
@@ -4777,7 +5057,8 @@ wasm = true
 [[example]]
 name = "clustered_decals"
 path = "examples/3d/clustered_decals.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["pbr_clustered_decals"]
 
 [package.metadata.example.clustered_decals]
@@ -4789,7 +5070,8 @@ wasm = false
 [[example]]
 name = "light_textures"
 path = "examples/3d/light_textures.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["pbr_light_textures"]
 
 [package.metadata.example.light_textures]
@@ -4801,7 +5083,8 @@ wasm = false
 [[example]]
 name = "occlusion_culling"
 path = "examples/3d/occlusion_culling.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.occlusion_culling]
 name = "Occlusion Culling"
@@ -4812,7 +5095,8 @@ wasm = false
 [[example]]
 name = "widgets"
 path = "examples/helpers/widgets.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 crate-type = ["lib"]
 
 [package.metadata.example.widgets]
@@ -4824,7 +5108,8 @@ wasm = true
 [[example]]
 name = "no_std_library"
 path = "examples/no_std/library/src/lib.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 crate-type = ["lib"]
 
 [package.metadata.example.no_std_library]
@@ -4836,7 +5121,8 @@ wasm = true
 [[example]]
 name = "extended_material_bindless"
 path = "examples/shader/extended_material_bindless.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.extended_material_bindless]
 name = "Extended Bindless Material"
@@ -4847,7 +5133,8 @@ wasm = false
 [[example]]
 name = "cooldown"
 path = "examples/usage/cooldown.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.cooldown]
 name = "Cooldown"
@@ -4858,7 +5145,8 @@ wasm = true
 [[example]]
 name = "hotpatching_systems"
 path = "examples/ecs/hotpatching_systems.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["hotpatching"]
 
 [package.metadata.example.hotpatching_systems]
@@ -4870,7 +5158,8 @@ wasm = false
 [[example]]
 name = "standard_widgets"
 path = "examples/ui/standard_widgets.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["experimental_bevy_ui_widgets"]
 
 [package.metadata.example.standard_widgets]
@@ -4882,7 +5171,8 @@ wasm = true
 [[example]]
 name = "standard_widgets_observers"
 path = "examples/ui/standard_widgets_observers.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["experimental_bevy_ui_widgets"]
 
 [package.metadata.example.standard_widgets_observers]
@@ -4894,7 +5184,8 @@ wasm = true
 [[example]]
 name = "scrollbars"
 path = "examples/ui/scrollbars.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["experimental_bevy_ui_widgets"]
 
 [package.metadata.example.scrollbars]
@@ -4906,7 +5197,8 @@ wasm = true
 [[example]]
 name = "feathers"
 path = "examples/ui/feathers.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["experimental_bevy_feathers"]
 
 [package.metadata.example.feathers]
@@ -4918,7 +5210,8 @@ wasm = true
 [[example]]
 name = "render_depth_to_texture"
 path = "examples/shader_advanced/render_depth_to_texture.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.render_depth_to_texture]
 name = "Render Depth to Texture"
@@ -4929,7 +5222,8 @@ wasm = true
 [[example]]
 name = "pan_camera_controller"
 path = "examples/camera/pan_camera_controller.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["pan_camera"]
 
 [package.metadata.example.pan_camera_controller]
@@ -4941,7 +5235,8 @@ wasm = true
 [[example]]
 name = "clustered_decal_maps"
 path = "examples/3d/clustered_decal_maps.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 required-features = ["pbr_clustered_decals", "https"]
 
 [package.metadata.example.clustered_decal_maps]
@@ -4953,7 +5248,8 @@ wasm = false
 [[example]]
 name = "mirror"
 path = "examples/3d/mirror.rs"
-doc-scrape-examples = true
+# Causes an ICE on docs.rs
+doc-scrape-examples = false
 
 [package.metadata.example.mirror]
 name = "Mirror"
