diff --git a/examples/3d/clustered_decal_maps.rs b/examples/3d/clustered_decal_maps.rs
index f72aa1df40b99..0f88747fd1fd9 100644
--- a/examples/3d/clustered_decal_maps.rs
+++ b/examples/3d/clustered_decal_maps.rs
@@ -11,7 +11,8 @@ use bevy::{
     prelude::*,
     render::view::Hdr,
 };
-use rand::Rng;
+use rand::{Rng, SeedableRng};
+use rand_chacha::ChaCha8Rng;
 
 use crate::widgets::{RadioButton, RadioButtonText, WidgetClickEvent, WidgetClickSender};
 
@@ -159,9 +160,13 @@ fn main() {
             Update,
             handle_emission_type_change.after(widgets::handle_ui_interactions::<AppSetting>),
         )
+        .insert_resource(SeededRng(ChaCha8Rng::seed_from_u64(19878367467712)))
         .run();
 }
 
+#[derive(Resource)]
+struct SeededRng(ChaCha8Rng);
+
 /// Spawns all the objects in the scene.
 fn setup(
     mut commands: Commands,
@@ -279,6 +284,7 @@ fn spawn_decal(
     app_textures: Res<AppTextures>,
     time: Res<Time>,
     mut decal_spawn_timer: Local<Option<Timer>>,
+    mut seeded_rng: ResMut<SeededRng>,
 ) {
     // Tick the decal spawn timer. Check to see if we should spawn a new decal,
     // and bail out if it's not yet time to.
@@ -290,18 +296,17 @@ fn spawn_decal(
     }
 
     // Generate a random position along the plane.
-    let mut rng = rand::rng();
     let decal_position = vec3(
-        rng.random_range(-PLANE_HALF_SIZE..PLANE_HALF_SIZE),
-        rng.random_range(-PLANE_HALF_SIZE..PLANE_HALF_SIZE),
+        seeded_rng.0.random_range(-PLANE_HALF_SIZE..PLANE_HALF_SIZE),
+        seeded_rng.0.random_range(-PLANE_HALF_SIZE..PLANE_HALF_SIZE),
         0.0,
     );
 
     // Generate a random size for the decal.
-    let decal_size = rng.random_range(DECAL_MIN_SIZE..DECAL_MAX_SIZE);
+    let decal_size = seeded_rng.0.random_range(DECAL_MIN_SIZE..DECAL_MAX_SIZE);
 
     // Generate a random rotation for the decal.
-    let theta = rng.random_range(0.0f32..PI);
+    let theta = seeded_rng.0.random_range(0.0f32..PI);
 
     // Now spawn the decal.
     commands.spawn((
