diff --git a/crates/bevy_state/macros/src/lib.rs b/crates/bevy_state/macros/src/lib.rs
index f461f0ead2ce7..3c5a2d0674347 100644
--- a/crates/bevy_state/macros/src/lib.rs
+++ b/crates/bevy_state/macros/src/lib.rs
@@ -1,6 +1,7 @@
-#![expect(missing_docs, reason = "Not all docs are written yet, see #3492.")]
 #![cfg_attr(docsrs, feature(doc_auto_cfg))]
 
+//! Macros for deriving `States` and `SubStates` traits.
+
 extern crate proc_macro;
 
 mod states;
@@ -8,11 +9,15 @@ mod states;
 use bevy_macro_utils::BevyManifest;
 use proc_macro::TokenStream;
 
+/// Implements the `States` trait for a type - see the trait
+/// docs for an example usage.
 #[proc_macro_derive(States, attributes(states))]
 pub fn derive_states(input: TokenStream) -> TokenStream {
     states::derive_states(input)
 }
 
+/// Implements the `SubStates` trait for a type - see the trait
+/// docs for an example usage.
 #[proc_macro_derive(SubStates, attributes(states, source))]
 pub fn derive_substates(input: TokenStream) -> TokenStream {
     states::derive_substates(input)
