diff --git a/Cargo.toml b/Cargo.toml
index d2e5b92258d79..73c5534243c1d 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -608,7 +608,7 @@ web-sys = { version = "0.3", features = ["Window"] }
 
 [[example]]
 name = "context_menu"
-path = "examples/usages/context_menu.rs"
+path = "examples/usage/context_menu.rs"
 doc-scrape-examples = true
 
 [package.metadata.example.context_menu]
diff --git a/examples/README.md b/examples/README.md
index bccadd438eab9..7b21d15da3ba0 100644
--- a/examples/README.md
+++ b/examples/README.md
@@ -584,7 +584,7 @@ Example | Description
 
 Example | Description
 --- | ---
-[Context Menu](../examples/usages/context_menu.rs) | Example of a context menu
+[Context Menu](../examples/usage/context_menu.rs) | Example of a context menu
 [Cooldown](../examples/usage/cooldown.rs) | Example for cooldown on button clicks
 
 ## Window
diff --git a/examples/usages/context_menu.rs b/examples/usage/context_menu.rs
similarity index 100%
rename from examples/usages/context_menu.rs
rename to examples/usage/context_menu.rs
