diff --git a/.github/workflows/post-release.yml b/.github/workflows/post-release.yml
index 91a98f3ea7acc..485861ebdfa0f 100644
--- a/.github/workflows/post-release.yml
+++ b/.github/workflows/post-release.yml
@@ -49,7 +49,8 @@ jobs:
             --exclude ci \
             --exclude errors \
             --exclude bevy_mobile_example \
-            --exclude build-wasm-example
+            --exclude build-wasm-example \
+            --exclude no_std_library
 
       - name: Create PR
         uses: peter-evans/create-pull-request@v7
diff --git a/Cargo.toml b/Cargo.toml
index ca29624ef92a2..1e2e3d94fe719 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 categories = ["game-engines", "graphics", "gui", "rendering"]
 description = "A refreshingly simple data-driven game engine and app framework"
@@ -558,12 +558,12 @@ hotpatching = ["bevy_internal/hotpatching"]
 debug = ["bevy_internal/debug"]
 
 [dependencies]
-bevy_internal = { path = "crates/bevy_internal", version = "0.16.0-dev", default-features = false }
+bevy_internal = { path = "crates/bevy_internal", version = "0.17.0-dev", default-features = false }
 tracing = { version = "0.1", default-features = false, optional = true }
 
 # Wasm does not support dynamic linking.
 [target.'cfg(not(target_family = "wasm"))'.dependencies]
-bevy_dylib = { path = "crates/bevy_dylib", version = "0.16.0-dev", default-features = false, optional = true }
+bevy_dylib = { path = "crates/bevy_dylib", version = "0.17.0-dev", default-features = false, optional = true }
 
 [dev-dependencies]
 rand = "0.8.0"
@@ -573,14 +573,14 @@ flate2 = "1.0"
 serde = { version = "1", features = ["derive"] }
 serde_json = "1.0.140"
 bytemuck = "1.7"
-bevy_render = { path = "crates/bevy_render", version = "0.16.0-dev", default-features = false }
+bevy_render = { path = "crates/bevy_render", version = "0.17.0-dev", default-features = false }
 # The following explicit dependencies are needed for proc macros to work inside of examples as they are part of the bevy crate itself.
-bevy_ecs = { path = "crates/bevy_ecs", version = "0.16.0-dev", default-features = false }
-bevy_state = { path = "crates/bevy_state", version = "0.16.0-dev", default-features = false }
-bevy_asset = { path = "crates/bevy_asset", version = "0.16.0-dev", default-features = false }
-bevy_reflect = { path = "crates/bevy_reflect", version = "0.16.0-dev", default-features = false }
-bevy_image = { path = "crates/bevy_image", version = "0.16.0-dev", default-features = false }
-bevy_gizmos = { path = "crates/bevy_gizmos", version = "0.16.0-dev", default-features = false }
+bevy_ecs = { path = "crates/bevy_ecs", version = "0.17.0-dev", default-features = false }
+bevy_state = { path = "crates/bevy_state", version = "0.17.0-dev", default-features = false }
+bevy_asset = { path = "crates/bevy_asset", version = "0.17.0-dev", default-features = false }
+bevy_reflect = { path = "crates/bevy_reflect", version = "0.17.0-dev", default-features = false }
+bevy_image = { path = "crates/bevy_image", version = "0.17.0-dev", default-features = false }
+bevy_gizmos = { path = "crates/bevy_gizmos", version = "0.17.0-dev", default-features = false }
 # Needed to poll Task examples
 futures-lite = "2.0.1"
 async-std = "1.13"
diff --git a/crates/bevy_a11y/Cargo.toml b/crates/bevy_a11y/Cargo.toml
index 39628ec046117..70ee16cf77cfc 100644
--- a/crates/bevy_a11y/Cargo.toml
+++ b/crates/bevy_a11y/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_a11y"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Provides accessibility support for Bevy Engine"
 homepage = "https://bevy.org"
@@ -40,10 +40,10 @@ critical-section = [
 
 [dependencies]
 # bevy
-bevy_app = { path = "../bevy_app", version = "0.16.0-dev", default-features = false }
-bevy_derive = { path = "../bevy_derive", version = "0.16.0-dev" }
-bevy_ecs = { path = "../bevy_ecs", version = "0.16.0-dev", default-features = false }
-bevy_reflect = { path = "../bevy_reflect", version = "0.16.0-dev", default-features = false, optional = true }
+bevy_app = { path = "../bevy_app", version = "0.17.0-dev", default-features = false }
+bevy_derive = { path = "../bevy_derive", version = "0.17.0-dev" }
+bevy_ecs = { path = "../bevy_ecs", version = "0.17.0-dev", default-features = false }
+bevy_reflect = { path = "../bevy_reflect", version = "0.17.0-dev", default-features = false, optional = true }
 
 # other
 accesskit = { version = "0.19", default-features = false }
diff --git a/crates/bevy_animation/Cargo.toml b/crates/bevy_animation/Cargo.toml
index 9db4a97fd0867..adf17a858073b 100644
--- a/crates/bevy_animation/Cargo.toml
+++ b/crates/bevy_animation/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_animation"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Provides animation functionality for Bevy Engine"
 homepage = "https://bevy.org"
@@ -10,22 +10,22 @@ keywords = ["bevy"]
 
 [dependencies]
 # bevy
-bevy_app = { path = "../bevy_app", version = "0.16.0-dev" }
-bevy_asset = { path = "../bevy_asset", version = "0.16.0-dev" }
-bevy_color = { path = "../bevy_color", version = "0.16.0-dev" }
-bevy_derive = { path = "../bevy_derive", version = "0.16.0-dev" }
-bevy_log = { path = "../bevy_log", version = "0.16.0-dev" }
-bevy_math = { path = "../bevy_math", version = "0.16.0-dev" }
-bevy_mesh = { path = "../bevy_mesh", version = "0.16.0-dev" }
-bevy_reflect = { path = "../bevy_reflect", version = "0.16.0-dev", features = [
+bevy_app = { path = "../bevy_app", version = "0.17.0-dev" }
+bevy_asset = { path = "../bevy_asset", version = "0.17.0-dev" }
+bevy_color = { path = "../bevy_color", version = "0.17.0-dev" }
+bevy_derive = { path = "../bevy_derive", version = "0.17.0-dev" }
+bevy_log = { path = "../bevy_log", version = "0.17.0-dev" }
+bevy_math = { path = "../bevy_math", version = "0.17.0-dev" }
+bevy_mesh = { path = "../bevy_mesh", version = "0.17.0-dev" }
+bevy_reflect = { path = "../bevy_reflect", version = "0.17.0-dev", features = [
   "petgraph",
 ] }
-bevy_render = { path = "../bevy_render", version = "0.16.0-dev" }
-bevy_time = { path = "../bevy_time", version = "0.16.0-dev" }
-bevy_utils = { path = "../bevy_utils", version = "0.16.0-dev" }
-bevy_ecs = { path = "../bevy_ecs", version = "0.16.0-dev" }
-bevy_transform = { path = "../bevy_transform", version = "0.16.0-dev" }
-bevy_platform = { path = "../bevy_platform", version = "0.16.0-dev", default-features = false, features = [
+bevy_render = { path = "../bevy_render", version = "0.17.0-dev" }
+bevy_time = { path = "../bevy_time", version = "0.17.0-dev" }
+bevy_utils = { path = "../bevy_utils", version = "0.17.0-dev" }
+bevy_ecs = { path = "../bevy_ecs", version = "0.17.0-dev" }
+bevy_transform = { path = "../bevy_transform", version = "0.17.0-dev" }
+bevy_platform = { path = "../bevy_platform", version = "0.17.0-dev", default-features = false, features = [
   "std",
   "serialize",
 ] }
diff --git a/crates/bevy_anti_aliasing/Cargo.toml b/crates/bevy_anti_aliasing/Cargo.toml
index c54608a883bca..8c32d70fff0bb 100644
--- a/crates/bevy_anti_aliasing/Cargo.toml
+++ b/crates/bevy_anti_aliasing/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_anti_aliasing"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Provides various anti aliasing implementations for Bevy Engine"
 homepage = "https://bevy.org"
@@ -16,17 +16,17 @@ smaa_luts = ["bevy_render/ktx2", "bevy_image/ktx2", "bevy_image/zstd"]
 
 [dependencies]
 # bevy
-bevy_asset = { path = "../bevy_asset", version = "0.16.0-dev" }
-bevy_reflect = { path = "../bevy_reflect", version = "0.16.0-dev" }
-bevy_render = { path = "../bevy_render", version = "0.16.0-dev" }
-bevy_math = { path = "../bevy_math", version = "0.16.0-dev" }
-bevy_utils = { path = "../bevy_utils", version = "0.16.0-dev" }
-bevy_app = { path = "../bevy_app", version = "0.16.0-dev" }
-bevy_image = { path = "../bevy_image", version = "0.16.0-dev" }
-bevy_derive = { path = "../bevy_derive", version = "0.16.0-dev" }
-bevy_ecs = { path = "../bevy_ecs", version = "0.16.0-dev" }
-bevy_core_pipeline = { path = "../bevy_core_pipeline", version = "0.16.0-dev" }
-bevy_diagnostic = { path = "../bevy_diagnostic", version = "0.16.0-dev" }
+bevy_asset = { path = "../bevy_asset", version = "0.17.0-dev" }
+bevy_reflect = { path = "../bevy_reflect", version = "0.17.0-dev" }
+bevy_render = { path = "../bevy_render", version = "0.17.0-dev" }
+bevy_math = { path = "../bevy_math", version = "0.17.0-dev" }
+bevy_utils = { path = "../bevy_utils", version = "0.17.0-dev" }
+bevy_app = { path = "../bevy_app", version = "0.17.0-dev" }
+bevy_image = { path = "../bevy_image", version = "0.17.0-dev" }
+bevy_derive = { path = "../bevy_derive", version = "0.17.0-dev" }
+bevy_ecs = { path = "../bevy_ecs", version = "0.17.0-dev" }
+bevy_core_pipeline = { path = "../bevy_core_pipeline", version = "0.17.0-dev" }
+bevy_diagnostic = { path = "../bevy_diagnostic", version = "0.17.0-dev" }
 
 # other
 tracing = { version = "0.1", default-features = false, features = ["std"] }
diff --git a/crates/bevy_app/Cargo.toml b/crates/bevy_app/Cargo.toml
index 6b6120f182a86..a0c5222b0b1a3 100644
--- a/crates/bevy_app/Cargo.toml
+++ b/crates/bevy_app/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_app"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Provides core App functionality for Bevy Engine"
 homepage = "https://bevy.org"
@@ -79,12 +79,12 @@ hotpatching = [
 
 [dependencies]
 # bevy
-bevy_derive = { path = "../bevy_derive", version = "0.16.0-dev" }
-bevy_ecs = { path = "../bevy_ecs", version = "0.16.0-dev", default-features = false }
-bevy_reflect = { path = "../bevy_reflect", version = "0.16.0-dev", default-features = false, optional = true }
-bevy_utils = { path = "../bevy_utils", version = "0.16.0-dev", default-features = false }
-bevy_tasks = { path = "../bevy_tasks", version = "0.16.0-dev", default-features = false }
-bevy_platform = { path = "../bevy_platform", version = "0.16.0-dev", default-features = false }
+bevy_derive = { path = "../bevy_derive", version = "0.17.0-dev" }
+bevy_ecs = { path = "../bevy_ecs", version = "0.17.0-dev", default-features = false }
+bevy_reflect = { path = "../bevy_reflect", version = "0.17.0-dev", default-features = false, optional = true }
+bevy_utils = { path = "../bevy_utils", version = "0.17.0-dev", default-features = false }
+bevy_tasks = { path = "../bevy_tasks", version = "0.17.0-dev", default-features = false }
+bevy_platform = { path = "../bevy_platform", version = "0.17.0-dev", default-features = false }
 
 # other
 downcast-rs = { version = "2", default-features = false }
diff --git a/crates/bevy_asset/Cargo.toml b/crates/bevy_asset/Cargo.toml
index e91987f40af59..0835b3b49aaa5 100644
--- a/crates/bevy_asset/Cargo.toml
+++ b/crates/bevy_asset/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_asset"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Provides asset functionality for Bevy Engine"
 homepage = "https://bevy.org"
@@ -19,19 +19,19 @@ watch = []
 trace = []
 
 [dependencies]
-bevy_app = { path = "../bevy_app", version = "0.16.0-dev", default-features = false, features = [
+bevy_app = { path = "../bevy_app", version = "0.17.0-dev", default-features = false, features = [
   "bevy_reflect",
 ] }
-bevy_asset_macros = { path = "macros", version = "0.16.0-dev" }
-bevy_ecs = { path = "../bevy_ecs", version = "0.16.0-dev", default-features = false }
-bevy_reflect = { path = "../bevy_reflect", version = "0.16.0-dev", default-features = false, features = [
+bevy_asset_macros = { path = "macros", version = "0.17.0-dev" }
+bevy_ecs = { path = "../bevy_ecs", version = "0.17.0-dev", default-features = false }
+bevy_reflect = { path = "../bevy_reflect", version = "0.17.0-dev", default-features = false, features = [
   "uuid",
 ] }
-bevy_tasks = { path = "../bevy_tasks", version = "0.16.0-dev", default-features = false, features = [
+bevy_tasks = { path = "../bevy_tasks", version = "0.17.0-dev", default-features = false, features = [
   "async_executor",
 ] }
-bevy_utils = { path = "../bevy_utils", version = "0.16.0-dev", default-features = false }
-bevy_platform = { path = "../bevy_platform", version = "0.16.0-dev", default-features = false, features = [
+bevy_utils = { path = "../bevy_utils", version = "0.17.0-dev", default-features = false }
+bevy_platform = { path = "../bevy_platform", version = "0.17.0-dev", default-features = false, features = [
   "std",
 ] }
 
@@ -65,7 +65,7 @@ uuid = { version = "1.13.1", default-features = false, features = [
 tracing = { version = "0.1", default-features = false }
 
 [target.'cfg(target_os = "android")'.dependencies]
-bevy_window = { path = "../bevy_window", version = "0.16.0-dev" }
+bevy_window = { path = "../bevy_window", version = "0.17.0-dev" }
 
 [target.'cfg(target_arch = "wasm32")'.dependencies]
 # TODO: Assuming all wasm builds are for the browser. Require `no_std` support to break assumption.
@@ -78,13 +78,13 @@ web-sys = { version = "0.3", features = [
 wasm-bindgen-futures = "0.4"
 js-sys = "0.3"
 uuid = { version = "1.13.1", default-features = false, features = ["js"] }
-bevy_app = { path = "../bevy_app", version = "0.16.0-dev", default-features = false, features = [
+bevy_app = { path = "../bevy_app", version = "0.17.0-dev", default-features = false, features = [
   "web",
 ] }
-bevy_tasks = { path = "../bevy_tasks", version = "0.16.0-dev", default-features = false, features = [
+bevy_tasks = { path = "../bevy_tasks", version = "0.17.0-dev", default-features = false, features = [
   "web",
 ] }
-bevy_reflect = { path = "../bevy_reflect", version = "0.16.0-dev", default-features = false, features = [
+bevy_reflect = { path = "../bevy_reflect", version = "0.17.0-dev", default-features = false, features = [
   "web",
 ] }
 
diff --git a/crates/bevy_asset/macros/Cargo.toml b/crates/bevy_asset/macros/Cargo.toml
index 4d99d228d3dd2..0b525b3c1d146 100644
--- a/crates/bevy_asset/macros/Cargo.toml
+++ b/crates/bevy_asset/macros/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_asset_macros"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Derive implementations for bevy_asset"
 homepage = "https://bevy.org"
@@ -12,7 +12,7 @@ keywords = ["bevy"]
 proc-macro = true
 
 [dependencies]
-bevy_macro_utils = { path = "../../bevy_macro_utils", version = "0.16.0-dev" }
+bevy_macro_utils = { path = "../../bevy_macro_utils", version = "0.17.0-dev" }
 
 syn = "2.0"
 proc-macro2 = "1.0"
diff --git a/crates/bevy_audio/Cargo.toml b/crates/bevy_audio/Cargo.toml
index ae5385870d907..2ffa62db9da42 100644
--- a/crates/bevy_audio/Cargo.toml
+++ b/crates/bevy_audio/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_audio"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Provides audio functionality for Bevy Engine"
 homepage = "https://bevy.org"
@@ -10,13 +10,13 @@ keywords = ["bevy"]
 
 [dependencies]
 # bevy
-bevy_app = { path = "../bevy_app", version = "0.16.0-dev" }
-bevy_asset = { path = "../bevy_asset", version = "0.16.0-dev" }
-bevy_ecs = { path = "../bevy_ecs", version = "0.16.0-dev" }
-bevy_math = { path = "../bevy_math", version = "0.16.0-dev" }
-bevy_reflect = { path = "../bevy_reflect", version = "0.16.0-dev" }
-bevy_transform = { path = "../bevy_transform", version = "0.16.0-dev" }
-bevy_derive = { path = "../bevy_derive", version = "0.16.0-dev" }
+bevy_app = { path = "../bevy_app", version = "0.17.0-dev" }
+bevy_asset = { path = "../bevy_asset", version = "0.17.0-dev" }
+bevy_ecs = { path = "../bevy_ecs", version = "0.17.0-dev" }
+bevy_math = { path = "../bevy_math", version = "0.17.0-dev" }
+bevy_reflect = { path = "../bevy_reflect", version = "0.17.0-dev" }
+bevy_transform = { path = "../bevy_transform", version = "0.17.0-dev" }
+bevy_derive = { path = "../bevy_derive", version = "0.17.0-dev" }
 
 # other
 # TODO: Remove `coreaudio-sys` dep below when updating `cpal`.
@@ -36,10 +36,10 @@ coreaudio-sys = { version = "0.2.17", default-features = false }
 rodio = { version = "0.20", default-features = false, features = [
   "wasm-bindgen",
 ] }
-bevy_app = { path = "../bevy_app", version = "0.16.0-dev", default-features = false, features = [
+bevy_app = { path = "../bevy_app", version = "0.17.0-dev", default-features = false, features = [
   "web",
 ] }
-bevy_reflect = { path = "../bevy_reflect", version = "0.16.0-dev", default-features = false, features = [
+bevy_reflect = { path = "../bevy_reflect", version = "0.17.0-dev", default-features = false, features = [
   "web",
 ] }
 
diff --git a/crates/bevy_color/Cargo.toml b/crates/bevy_color/Cargo.toml
index ca7a7a74f5deb..298f80e54fc86 100644
--- a/crates/bevy_color/Cargo.toml
+++ b/crates/bevy_color/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_color"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Types for representing and manipulating color values"
 homepage = "https://bevy.org"
@@ -10,10 +10,10 @@ keywords = ["bevy", "color"]
 rust-version = "1.85.0"
 
 [dependencies]
-bevy_math = { path = "../bevy_math", version = "0.16.0-dev", default-features = false, features = [
+bevy_math = { path = "../bevy_math", version = "0.17.0-dev", default-features = false, features = [
   "curve",
 ] }
-bevy_reflect = { path = "../bevy_reflect", version = "0.16.0-dev", default-features = false, optional = true }
+bevy_reflect = { path = "../bevy_reflect", version = "0.17.0-dev", default-features = false, optional = true }
 bytemuck = { version = "1", features = ["derive"] }
 serde = { version = "1.0", features = [
   "derive",
diff --git a/crates/bevy_core_pipeline/Cargo.toml b/crates/bevy_core_pipeline/Cargo.toml
index 2d903d2cc47af..9b3f158af2044 100644
--- a/crates/bevy_core_pipeline/Cargo.toml
+++ b/crates/bevy_core_pipeline/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_core_pipeline"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 authors = [
   "Bevy Contributors <bevyengine@gmail.com>",
@@ -20,20 +20,20 @@ tonemapping_luts = ["bevy_render/ktx2", "bevy_image/ktx2", "bevy_image/zstd"]
 
 [dependencies]
 # bevy
-bevy_app = { path = "../bevy_app", version = "0.16.0-dev" }
-bevy_asset = { path = "../bevy_asset", version = "0.16.0-dev" }
-bevy_color = { path = "../bevy_color", version = "0.16.0-dev" }
-bevy_derive = { path = "../bevy_derive", version = "0.16.0-dev" }
-bevy_diagnostic = { path = "../bevy_diagnostic", version = "0.16.0-dev" }
-bevy_ecs = { path = "../bevy_ecs", version = "0.16.0-dev" }
-bevy_image = { path = "../bevy_image", version = "0.16.0-dev" }
-bevy_reflect = { path = "../bevy_reflect", version = "0.16.0-dev" }
-bevy_render = { path = "../bevy_render", version = "0.16.0-dev" }
-bevy_transform = { path = "../bevy_transform", version = "0.16.0-dev" }
-bevy_math = { path = "../bevy_math", version = "0.16.0-dev" }
-bevy_utils = { path = "../bevy_utils", version = "0.16.0-dev" }
-bevy_window = { path = "../bevy_window", version = "0.16.0-dev" }
-bevy_platform = { path = "../bevy_platform", version = "0.16.0-dev", default-features = false, features = [
+bevy_app = { path = "../bevy_app", version = "0.17.0-dev" }
+bevy_asset = { path = "../bevy_asset", version = "0.17.0-dev" }
+bevy_color = { path = "../bevy_color", version = "0.17.0-dev" }
+bevy_derive = { path = "../bevy_derive", version = "0.17.0-dev" }
+bevy_diagnostic = { path = "../bevy_diagnostic", version = "0.17.0-dev" }
+bevy_ecs = { path = "../bevy_ecs", version = "0.17.0-dev" }
+bevy_image = { path = "../bevy_image", version = "0.17.0-dev" }
+bevy_reflect = { path = "../bevy_reflect", version = "0.17.0-dev" }
+bevy_render = { path = "../bevy_render", version = "0.17.0-dev" }
+bevy_transform = { path = "../bevy_transform", version = "0.17.0-dev" }
+bevy_math = { path = "../bevy_math", version = "0.17.0-dev" }
+bevy_utils = { path = "../bevy_utils", version = "0.17.0-dev" }
+bevy_window = { path = "../bevy_window", version = "0.17.0-dev" }
+bevy_platform = { path = "../bevy_platform", version = "0.17.0-dev", default-features = false, features = [
   "std",
   "serialize",
 ] }
diff --git a/crates/bevy_core_widgets/Cargo.toml b/crates/bevy_core_widgets/Cargo.toml
index 83ecd67edeefa..e93891d8f7585 100644
--- a/crates/bevy_core_widgets/Cargo.toml
+++ b/crates/bevy_core_widgets/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_core_widgets"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Unstyled common widgets for Bevy Engine"
 homepage = "https://bevyengine.org"
@@ -10,17 +10,17 @@ keywords = ["bevy"]
 
 [dependencies]
 # bevy
-bevy_app = { path = "../bevy_app", version = "0.16.0-dev" }
-bevy_a11y = { path = "../bevy_a11y", version = "0.16.0-dev" }
-bevy_ecs = { path = "../bevy_ecs", version = "0.16.0-dev" }
-bevy_input = { path = "../bevy_input", version = "0.16.0-dev" }
-bevy_input_focus = { path = "../bevy_input_focus", version = "0.16.0-dev" }
-bevy_log = { path = "../bevy_log", version = "0.16.0-dev" }
-bevy_math = { path = "../bevy_math", version = "0.16.0-dev" }
-bevy_picking = { path = "../bevy_picking", version = "0.16.0-dev" }
-bevy_render = { path = "../bevy_render", version = "0.16.0-dev" }
-bevy_transform = { path = "../bevy_transform", version = "0.16.0-dev" }
-bevy_ui = { path = "../bevy_ui", version = "0.16.0-dev", features = [
+bevy_app = { path = "../bevy_app", version = "0.17.0-dev" }
+bevy_a11y = { path = "../bevy_a11y", version = "0.17.0-dev" }
+bevy_ecs = { path = "../bevy_ecs", version = "0.17.0-dev" }
+bevy_input = { path = "../bevy_input", version = "0.17.0-dev" }
+bevy_input_focus = { path = "../bevy_input_focus", version = "0.17.0-dev" }
+bevy_log = { path = "../bevy_log", version = "0.17.0-dev" }
+bevy_math = { path = "../bevy_math", version = "0.17.0-dev" }
+bevy_picking = { path = "../bevy_picking", version = "0.17.0-dev" }
+bevy_render = { path = "../bevy_render", version = "0.17.0-dev" }
+bevy_transform = { path = "../bevy_transform", version = "0.17.0-dev" }
+bevy_ui = { path = "../bevy_ui", version = "0.17.0-dev", features = [
   "bevy_ui_picking_backend",
 ] }
 
diff --git a/crates/bevy_derive/Cargo.toml b/crates/bevy_derive/Cargo.toml
index f127dbffb5833..f1a1cd44b3323 100644
--- a/crates/bevy_derive/Cargo.toml
+++ b/crates/bevy_derive/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_derive"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Provides derive implementations for Bevy Engine"
 homepage = "https://bevy.org"
@@ -12,7 +12,7 @@ keywords = ["bevy"]
 proc-macro = true
 
 [dependencies]
-bevy_macro_utils = { path = "../bevy_macro_utils", version = "0.16.0-dev" }
+bevy_macro_utils = { path = "../bevy_macro_utils", version = "0.17.0-dev" }
 
 quote = "1.0"
 syn = { version = "2.0", features = ["full"] }
diff --git a/crates/bevy_dev_tools/Cargo.toml b/crates/bevy_dev_tools/Cargo.toml
index ab5a04f2b2532..ef31767c0e3eb 100644
--- a/crates/bevy_dev_tools/Cargo.toml
+++ b/crates/bevy_dev_tools/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_dev_tools"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Collection of developer tools for the Bevy Engine"
 homepage = "https://bevy.org"
@@ -13,21 +13,21 @@ bevy_ci_testing = ["serde", "ron"]
 
 [dependencies]
 # bevy
-bevy_app = { path = "../bevy_app", version = "0.16.0-dev" }
-bevy_asset = { path = "../bevy_asset", version = "0.16.0-dev" }
-bevy_color = { path = "../bevy_color", version = "0.16.0-dev" }
-bevy_diagnostic = { path = "../bevy_diagnostic", version = "0.16.0-dev" }
-bevy_ecs = { path = "../bevy_ecs", version = "0.16.0-dev" }
-bevy_input = { path = "../bevy_input", version = "0.16.0-dev" }
-bevy_picking = { path = "../bevy_picking", version = "0.16.0-dev" }
-bevy_render = { path = "../bevy_render", version = "0.16.0-dev" }
-bevy_reflect = { path = "../bevy_reflect", version = "0.16.0-dev" }
-bevy_time = { path = "../bevy_time", version = "0.16.0-dev" }
-bevy_text = { path = "../bevy_text", version = "0.16.0-dev" }
-bevy_ui = { path = "../bevy_ui", version = "0.16.0-dev" }
-bevy_utils = { path = "../bevy_utils", version = "0.16.0-dev" }
-bevy_window = { path = "../bevy_window", version = "0.16.0-dev" }
-bevy_state = { path = "../bevy_state", version = "0.16.0-dev" }
+bevy_app = { path = "../bevy_app", version = "0.17.0-dev" }
+bevy_asset = { path = "../bevy_asset", version = "0.17.0-dev" }
+bevy_color = { path = "../bevy_color", version = "0.17.0-dev" }
+bevy_diagnostic = { path = "../bevy_diagnostic", version = "0.17.0-dev" }
+bevy_ecs = { path = "../bevy_ecs", version = "0.17.0-dev" }
+bevy_input = { path = "../bevy_input", version = "0.17.0-dev" }
+bevy_picking = { path = "../bevy_picking", version = "0.17.0-dev" }
+bevy_render = { path = "../bevy_render", version = "0.17.0-dev" }
+bevy_reflect = { path = "../bevy_reflect", version = "0.17.0-dev" }
+bevy_time = { path = "../bevy_time", version = "0.17.0-dev" }
+bevy_text = { path = "../bevy_text", version = "0.17.0-dev" }
+bevy_ui = { path = "../bevy_ui", version = "0.17.0-dev" }
+bevy_utils = { path = "../bevy_utils", version = "0.17.0-dev" }
+bevy_window = { path = "../bevy_window", version = "0.17.0-dev" }
+bevy_state = { path = "../bevy_state", version = "0.17.0-dev" }
 
 # other
 serde = { version = "1.0", features = ["derive"], optional = true }
diff --git a/crates/bevy_diagnostic/Cargo.toml b/crates/bevy_diagnostic/Cargo.toml
index 708f3b9ee17e8..e930da149a26a 100644
--- a/crates/bevy_diagnostic/Cargo.toml
+++ b/crates/bevy_diagnostic/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_diagnostic"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Provides diagnostic functionality for Bevy Engine"
 homepage = "https://bevy.org"
@@ -53,12 +53,12 @@ critical-section = [
 
 [dependencies]
 # bevy
-bevy_app = { path = "../bevy_app", version = "0.16.0-dev", default-features = false }
-bevy_ecs = { path = "../bevy_ecs", version = "0.16.0-dev", default-features = false }
-bevy_time = { path = "../bevy_time", version = "0.16.0-dev", default-features = false }
-bevy_utils = { path = "../bevy_utils", version = "0.16.0-dev", default-features = false }
-bevy_tasks = { path = "../bevy_tasks", version = "0.16.0-dev", default-features = false }
-bevy_platform = { path = "../bevy_platform", version = "0.16.0-dev", default-features = false, features = [
+bevy_app = { path = "../bevy_app", version = "0.17.0-dev", default-features = false }
+bevy_ecs = { path = "../bevy_ecs", version = "0.17.0-dev", default-features = false }
+bevy_time = { path = "../bevy_time", version = "0.17.0-dev", default-features = false }
+bevy_utils = { path = "../bevy_utils", version = "0.17.0-dev", default-features = false }
+bevy_tasks = { path = "../bevy_tasks", version = "0.17.0-dev", default-features = false }
+bevy_platform = { path = "../bevy_platform", version = "0.17.0-dev", default-features = false, features = [
   "alloc",
 ] }
 
diff --git a/crates/bevy_dylib/Cargo.toml b/crates/bevy_dylib/Cargo.toml
index a980a35b1074b..334ea10f1fb1e 100644
--- a/crates/bevy_dylib/Cargo.toml
+++ b/crates/bevy_dylib/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_dylib"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Force the Bevy Engine to be dynamically linked for faster linking"
 homepage = "https://bevy.org"
@@ -12,7 +12,7 @@ keywords = ["bevy"]
 crate-type = ["dylib"]
 
 [dependencies]
-bevy_internal = { path = "../bevy_internal", version = "0.16.0-dev", default-features = false }
+bevy_internal = { path = "../bevy_internal", version = "0.17.0-dev", default-features = false }
 
 [lints]
 workspace = true
diff --git a/crates/bevy_ecs/Cargo.toml b/crates/bevy_ecs/Cargo.toml
index c2c663b7d7fb5..5aa3dfe5ef0db 100644
--- a/crates/bevy_ecs/Cargo.toml
+++ b/crates/bevy_ecs/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_ecs"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Bevy Engine's entity component system"
 homepage = "https://bevy.org"
@@ -86,14 +86,14 @@ critical-section = [
 hotpatching = ["dep:subsecond"]
 
 [dependencies]
-bevy_ptr = { path = "../bevy_ptr", version = "0.16.0-dev" }
-bevy_reflect = { path = "../bevy_reflect", version = "0.16.0-dev", features = [
+bevy_ptr = { path = "../bevy_ptr", version = "0.17.0-dev" }
+bevy_reflect = { path = "../bevy_reflect", version = "0.17.0-dev", features = [
   "smallvec",
 ], default-features = false, optional = true }
-bevy_tasks = { path = "../bevy_tasks", version = "0.16.0-dev", default-features = false }
-bevy_utils = { path = "../bevy_utils", version = "0.16.0-dev", default-features = false }
-bevy_ecs_macros = { path = "macros", version = "0.16.0-dev" }
-bevy_platform = { path = "../bevy_platform", version = "0.16.0-dev", default-features = false, features = [
+bevy_tasks = { path = "../bevy_tasks", version = "0.17.0-dev", default-features = false }
+bevy_utils = { path = "../bevy_utils", version = "0.17.0-dev", default-features = false }
+bevy_ecs_macros = { path = "macros", version = "0.17.0-dev" }
+bevy_platform = { path = "../bevy_platform", version = "0.17.0-dev", default-features = false, features = [
   "alloc",
 ] }
 
diff --git a/crates/bevy_ecs/macros/Cargo.toml b/crates/bevy_ecs/macros/Cargo.toml
index 28605a5d67a5c..4f15bd59fd375 100644
--- a/crates/bevy_ecs/macros/Cargo.toml
+++ b/crates/bevy_ecs/macros/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_ecs_macros"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 description = "Bevy ECS Macros"
 edition = "2024"
 license = "MIT OR Apache-2.0"
@@ -9,7 +9,7 @@ license = "MIT OR Apache-2.0"
 proc-macro = true
 
 [dependencies]
-bevy_macro_utils = { path = "../../bevy_macro_utils", version = "0.16.0-dev" }
+bevy_macro_utils = { path = "../../bevy_macro_utils", version = "0.17.0-dev" }
 
 syn = { version = "2.0.99", features = ["full", "extra-traits"] }
 quote = "1.0"
diff --git a/crates/bevy_encase_derive/Cargo.toml b/crates/bevy_encase_derive/Cargo.toml
index 60a6515fd830f..9e05bc7a853fe 100644
--- a/crates/bevy_encase_derive/Cargo.toml
+++ b/crates/bevy_encase_derive/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_encase_derive"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Bevy derive macro for encase"
 homepage = "https://bevy.org"
@@ -12,7 +12,7 @@ keywords = ["bevy"]
 proc-macro = true
 
 [dependencies]
-bevy_macro_utils = { path = "../bevy_macro_utils", version = "0.16.0-dev" }
+bevy_macro_utils = { path = "../bevy_macro_utils", version = "0.17.0-dev" }
 encase_derive_impl = "0.10"
 
 [lints]
diff --git a/crates/bevy_gilrs/Cargo.toml b/crates/bevy_gilrs/Cargo.toml
index 0afb6babbf6cf..afb20318bc3d4 100644
--- a/crates/bevy_gilrs/Cargo.toml
+++ b/crates/bevy_gilrs/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_gilrs"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Gamepad system made using Gilrs for Bevy Engine"
 homepage = "https://bevy.org"
@@ -10,12 +10,12 @@ keywords = ["bevy"]
 
 [dependencies]
 # bevy
-bevy_app = { path = "../bevy_app", version = "0.16.0-dev" }
-bevy_ecs = { path = "../bevy_ecs", version = "0.16.0-dev" }
-bevy_input = { path = "../bevy_input", version = "0.16.0-dev" }
-bevy_utils = { path = "../bevy_utils", version = "0.16.0-dev" }
-bevy_time = { path = "../bevy_time", version = "0.16.0-dev" }
-bevy_platform = { path = "../bevy_platform", version = "0.16.0-dev", default-features = false, features = [
+bevy_app = { path = "../bevy_app", version = "0.17.0-dev" }
+bevy_ecs = { path = "../bevy_ecs", version = "0.17.0-dev" }
+bevy_input = { path = "../bevy_input", version = "0.17.0-dev" }
+bevy_utils = { path = "../bevy_utils", version = "0.17.0-dev" }
+bevy_time = { path = "../bevy_time", version = "0.17.0-dev" }
+bevy_platform = { path = "../bevy_platform", version = "0.17.0-dev", default-features = false, features = [
   "std",
 ] }
 
diff --git a/crates/bevy_gizmos/Cargo.toml b/crates/bevy_gizmos/Cargo.toml
index 97a41f15b6c67..c4833dbe7e052 100644
--- a/crates/bevy_gizmos/Cargo.toml
+++ b/crates/bevy_gizmos/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_gizmos"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Provides gizmos for Bevy Engine"
 homepage = "https://bevy.org"
@@ -15,21 +15,21 @@ bevy_render = ["dep:bevy_render", "bevy_core_pipeline"]
 
 [dependencies]
 # Bevy
-bevy_pbr = { path = "../bevy_pbr", version = "0.16.0-dev", optional = true }
-bevy_sprite = { path = "../bevy_sprite", version = "0.16.0-dev", optional = true }
-bevy_app = { path = "../bevy_app", version = "0.16.0-dev" }
-bevy_color = { path = "../bevy_color", version = "0.16.0-dev" }
-bevy_ecs = { path = "../bevy_ecs", version = "0.16.0-dev" }
-bevy_image = { path = "../bevy_image", version = "0.16.0-dev" }
-bevy_math = { path = "../bevy_math", version = "0.16.0-dev" }
-bevy_asset = { path = "../bevy_asset", version = "0.16.0-dev" }
-bevy_render = { path = "../bevy_render", version = "0.16.0-dev", optional = true }
-bevy_utils = { path = "../bevy_utils", version = "0.16.0-dev" }
-bevy_reflect = { path = "../bevy_reflect", version = "0.16.0-dev" }
-bevy_core_pipeline = { path = "../bevy_core_pipeline", version = "0.16.0-dev", optional = true }
-bevy_transform = { path = "../bevy_transform", version = "0.16.0-dev" }
-bevy_gizmos_macros = { path = "macros", version = "0.16.0-dev" }
-bevy_time = { path = "../bevy_time", version = "0.16.0-dev" }
+bevy_pbr = { path = "../bevy_pbr", version = "0.17.0-dev", optional = true }
+bevy_sprite = { path = "../bevy_sprite", version = "0.17.0-dev", optional = true }
+bevy_app = { path = "../bevy_app", version = "0.17.0-dev" }
+bevy_color = { path = "../bevy_color", version = "0.17.0-dev" }
+bevy_ecs = { path = "../bevy_ecs", version = "0.17.0-dev" }
+bevy_image = { path = "../bevy_image", version = "0.17.0-dev" }
+bevy_math = { path = "../bevy_math", version = "0.17.0-dev" }
+bevy_asset = { path = "../bevy_asset", version = "0.17.0-dev" }
+bevy_render = { path = "../bevy_render", version = "0.17.0-dev", optional = true }
+bevy_utils = { path = "../bevy_utils", version = "0.17.0-dev" }
+bevy_reflect = { path = "../bevy_reflect", version = "0.17.0-dev" }
+bevy_core_pipeline = { path = "../bevy_core_pipeline", version = "0.17.0-dev", optional = true }
+bevy_transform = { path = "../bevy_transform", version = "0.17.0-dev" }
+bevy_gizmos_macros = { path = "macros", version = "0.17.0-dev" }
+bevy_time = { path = "../bevy_time", version = "0.17.0-dev" }
 
 # other
 bytemuck = "1.0"
diff --git a/crates/bevy_gizmos/macros/Cargo.toml b/crates/bevy_gizmos/macros/Cargo.toml
index e15d0367b2f94..b7effe24b094a 100644
--- a/crates/bevy_gizmos/macros/Cargo.toml
+++ b/crates/bevy_gizmos/macros/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_gizmos_macros"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Derive implementations for bevy_gizmos"
 homepage = "https://bevy.org"
@@ -13,7 +13,7 @@ proc-macro = true
 
 
 [dependencies]
-bevy_macro_utils = { path = "../../bevy_macro_utils", version = "0.16.0-dev" }
+bevy_macro_utils = { path = "../../bevy_macro_utils", version = "0.17.0-dev" }
 
 syn = "2.0"
 proc-macro2 = "1.0"
diff --git a/crates/bevy_gltf/Cargo.toml b/crates/bevy_gltf/Cargo.toml
index 05b16d1fc90c1..c46b74b7ca270 100644
--- a/crates/bevy_gltf/Cargo.toml
+++ b/crates/bevy_gltf/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_gltf"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Bevy Engine GLTF loading"
 homepage = "https://bevy.org"
@@ -18,25 +18,25 @@ pbr_specular_textures = ["bevy_pbr/pbr_specular_textures"]
 
 [dependencies]
 # bevy
-bevy_animation = { path = "../bevy_animation", version = "0.16.0-dev", optional = true }
-bevy_app = { path = "../bevy_app", version = "0.16.0-dev" }
-bevy_asset = { path = "../bevy_asset", version = "0.16.0-dev" }
-bevy_color = { path = "../bevy_color", version = "0.16.0-dev" }
-bevy_core_pipeline = { path = "../bevy_core_pipeline", version = "0.16.0-dev" }
-bevy_ecs = { path = "../bevy_ecs", version = "0.16.0-dev" }
-bevy_image = { path = "../bevy_image", version = "0.16.0-dev" }
-bevy_math = { path = "../bevy_math", version = "0.16.0-dev" }
-bevy_mesh = { path = "../bevy_mesh", version = "0.16.0-dev" }
-bevy_pbr = { path = "../bevy_pbr", version = "0.16.0-dev" }
-bevy_reflect = { path = "../bevy_reflect", version = "0.16.0-dev" }
-bevy_render = { path = "../bevy_render", version = "0.16.0-dev" }
-bevy_scene = { path = "../bevy_scene", version = "0.16.0-dev", features = [
+bevy_animation = { path = "../bevy_animation", version = "0.17.0-dev", optional = true }
+bevy_app = { path = "../bevy_app", version = "0.17.0-dev" }
+bevy_asset = { path = "../bevy_asset", version = "0.17.0-dev" }
+bevy_color = { path = "../bevy_color", version = "0.17.0-dev" }
+bevy_core_pipeline = { path = "../bevy_core_pipeline", version = "0.17.0-dev" }
+bevy_ecs = { path = "../bevy_ecs", version = "0.17.0-dev" }
+bevy_image = { path = "../bevy_image", version = "0.17.0-dev" }
+bevy_math = { path = "../bevy_math", version = "0.17.0-dev" }
+bevy_mesh = { path = "../bevy_mesh", version = "0.17.0-dev" }
+bevy_pbr = { path = "../bevy_pbr", version = "0.17.0-dev" }
+bevy_reflect = { path = "../bevy_reflect", version = "0.17.0-dev" }
+bevy_render = { path = "../bevy_render", version = "0.17.0-dev" }
+bevy_scene = { path = "../bevy_scene", version = "0.17.0-dev", features = [
   "bevy_render",
 ] }
-bevy_transform = { path = "../bevy_transform", version = "0.16.0-dev" }
-bevy_tasks = { path = "../bevy_tasks", version = "0.16.0-dev" }
-bevy_utils = { path = "../bevy_utils", version = "0.16.0-dev" }
-bevy_platform = { path = "../bevy_platform", version = "0.16.0-dev", default-features = false, features = [
+bevy_transform = { path = "../bevy_transform", version = "0.17.0-dev" }
+bevy_tasks = { path = "../bevy_tasks", version = "0.17.0-dev" }
+bevy_utils = { path = "../bevy_utils", version = "0.17.0-dev" }
+bevy_platform = { path = "../bevy_platform", version = "0.17.0-dev", default-features = false, features = [
   "std",
   "serialize",
 ] }
@@ -66,7 +66,7 @@ smallvec = "1.11"
 tracing = { version = "0.1", default-features = false, features = ["std"] }
 
 [dev-dependencies]
-bevy_log = { path = "../bevy_log", version = "0.16.0-dev" }
+bevy_log = { path = "../bevy_log", version = "0.17.0-dev" }
 
 [lints]
 workspace = true
diff --git a/crates/bevy_image/Cargo.toml b/crates/bevy_image/Cargo.toml
index 7f8128e3656bc..c26a9da1bba7c 100644
--- a/crates/bevy_image/Cargo.toml
+++ b/crates/bevy_image/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_image"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Provides image types for Bevy Engine"
 homepage = "https://bevy.org"
@@ -40,17 +40,17 @@ zstd = ["ruzstd"]
 
 [dependencies]
 # bevy
-bevy_app = { path = "../bevy_app", version = "0.16.0-dev" }
-bevy_asset = { path = "../bevy_asset", version = "0.16.0-dev" }
-bevy_color = { path = "../bevy_color", version = "0.16.0-dev", features = [
+bevy_app = { path = "../bevy_app", version = "0.17.0-dev" }
+bevy_asset = { path = "../bevy_asset", version = "0.17.0-dev" }
+bevy_color = { path = "../bevy_color", version = "0.17.0-dev", features = [
   "serialize",
   "wgpu-types",
 ] }
-bevy_ecs = { path = "../bevy_ecs", version = "0.16.0-dev", default-features = false }
-bevy_math = { path = "../bevy_math", version = "0.16.0-dev" }
-bevy_reflect = { path = "../bevy_reflect", version = "0.16.0-dev" }
-bevy_utils = { path = "../bevy_utils", version = "0.16.0-dev" }
-bevy_platform = { path = "../bevy_platform", version = "0.16.0-dev", default-features = false, features = [
+bevy_ecs = { path = "../bevy_ecs", version = "0.17.0-dev", default-features = false }
+bevy_math = { path = "../bevy_math", version = "0.17.0-dev" }
+bevy_reflect = { path = "../bevy_reflect", version = "0.17.0-dev" }
+bevy_utils = { path = "../bevy_utils", version = "0.17.0-dev" }
+bevy_platform = { path = "../bevy_platform", version = "0.17.0-dev", default-features = false, features = [
   "std",
 ] }
 
@@ -77,7 +77,7 @@ tracing = { version = "0.1", default-features = false, features = ["std"] }
 half = { version = "2.4.1" }
 
 [dev-dependencies]
-bevy_ecs = { path = "../bevy_ecs", version = "0.16.0-dev" }
+bevy_ecs = { path = "../bevy_ecs", version = "0.17.0-dev" }
 
 [lints]
 workspace = true
diff --git a/crates/bevy_input/Cargo.toml b/crates/bevy_input/Cargo.toml
index 6b805b83bf387..2961c0d115fdb 100644
--- a/crates/bevy_input/Cargo.toml
+++ b/crates/bevy_input/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_input"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Provides input functionality for Bevy Engine"
 homepage = "https://bevy.org"
@@ -60,14 +60,14 @@ libm = ["bevy_math/libm"]
 
 [dependencies]
 # bevy
-bevy_app = { path = "../bevy_app", version = "0.16.0-dev", default-features = false }
-bevy_ecs = { path = "../bevy_ecs", version = "0.16.0-dev", default-features = false }
-bevy_math = { path = "../bevy_math", version = "0.16.0-dev", default-features = false }
-bevy_utils = { path = "../bevy_utils", version = "0.16.0-dev", default-features = false }
-bevy_reflect = { path = "../bevy_reflect", version = "0.16.0-dev", features = [
+bevy_app = { path = "../bevy_app", version = "0.17.0-dev", default-features = false }
+bevy_ecs = { path = "../bevy_ecs", version = "0.17.0-dev", default-features = false }
+bevy_math = { path = "../bevy_math", version = "0.17.0-dev", default-features = false }
+bevy_utils = { path = "../bevy_utils", version = "0.17.0-dev", default-features = false }
+bevy_reflect = { path = "../bevy_reflect", version = "0.17.0-dev", features = [
   "glam",
 ], default-features = false, optional = true }
-bevy_platform = { path = "../bevy_platform", version = "0.16.0-dev", default-features = false }
+bevy_platform = { path = "../bevy_platform", version = "0.17.0-dev", default-features = false }
 
 # other
 serde = { version = "1", features = [
diff --git a/crates/bevy_input_focus/Cargo.toml b/crates/bevy_input_focus/Cargo.toml
index f63d436af1f5a..60b824258df31 100644
--- a/crates/bevy_input_focus/Cargo.toml
+++ b/crates/bevy_input_focus/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_input_focus"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Keyboard focus management"
 homepage = "https://bevy.org"
@@ -60,13 +60,13 @@ libm = ["bevy_math/libm", "bevy_window/libm"]
 
 [dependencies]
 # bevy
-bevy_app = { path = "../bevy_app", version = "0.16.0-dev", default-features = false }
-bevy_ecs = { path = "../bevy_ecs", version = "0.16.0-dev", default-features = false }
-bevy_input = { path = "../bevy_input", version = "0.16.0-dev", default-features = false }
-bevy_math = { path = "../bevy_math", version = "0.16.0-dev", default-features = false }
-bevy_picking = { path = "../bevy_picking", version = "0.16.0-dev", default-features = false }
-bevy_window = { path = "../bevy_window", version = "0.16.0-dev", default-features = false }
-bevy_reflect = { path = "../bevy_reflect", version = "0.16.0-dev", features = [
+bevy_app = { path = "../bevy_app", version = "0.17.0-dev", default-features = false }
+bevy_ecs = { path = "../bevy_ecs", version = "0.17.0-dev", default-features = false }
+bevy_input = { path = "../bevy_input", version = "0.17.0-dev", default-features = false }
+bevy_math = { path = "../bevy_math", version = "0.17.0-dev", default-features = false }
+bevy_picking = { path = "../bevy_picking", version = "0.17.0-dev", default-features = false }
+bevy_window = { path = "../bevy_window", version = "0.17.0-dev", default-features = false }
+bevy_reflect = { path = "../bevy_reflect", version = "0.17.0-dev", features = [
   "glam",
 ], default-features = false, optional = true }
 
diff --git a/crates/bevy_internal/Cargo.toml b/crates/bevy_internal/Cargo.toml
index c9639f1950aec..4692fe9d15b23 100644
--- a/crates/bevy_internal/Cargo.toml
+++ b/crates/bevy_internal/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_internal"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "An internal Bevy crate used to facilitate optional dynamic linking via the 'dynamic_linking' feature"
 homepage = "https://bevy.org"
@@ -350,80 +350,80 @@ debug = ["bevy_utils/debug"]
 
 [dependencies]
 # bevy (no_std)
-bevy_app = { path = "../bevy_app", version = "0.16.0-dev", default-features = false, features = [
+bevy_app = { path = "../bevy_app", version = "0.17.0-dev", default-features = false, features = [
   "bevy_reflect",
 ] }
-bevy_derive = { path = "../bevy_derive", version = "0.16.0-dev", default-features = false }
-bevy_diagnostic = { path = "../bevy_diagnostic", version = "0.16.0-dev", default-features = false }
-bevy_ecs = { path = "../bevy_ecs", version = "0.16.0-dev", default-features = false, features = [
+bevy_derive = { path = "../bevy_derive", version = "0.17.0-dev", default-features = false }
+bevy_diagnostic = { path = "../bevy_diagnostic", version = "0.17.0-dev", default-features = false }
+bevy_ecs = { path = "../bevy_ecs", version = "0.17.0-dev", default-features = false, features = [
   "bevy_reflect",
 ] }
-bevy_input = { path = "../bevy_input", version = "0.16.0-dev", default-features = false, features = [
+bevy_input = { path = "../bevy_input", version = "0.17.0-dev", default-features = false, features = [
   "bevy_reflect",
 ] }
-bevy_math = { path = "../bevy_math", version = "0.16.0-dev", default-features = false, features = [
+bevy_math = { path = "../bevy_math", version = "0.17.0-dev", default-features = false, features = [
   "bevy_reflect",
   "nostd-libm",
 ] }
-bevy_platform = { path = "../bevy_platform", version = "0.16.0-dev", default-features = false, features = [
+bevy_platform = { path = "../bevy_platform", version = "0.17.0-dev", default-features = false, features = [
   "alloc",
 ] }
-bevy_ptr = { path = "../bevy_ptr", version = "0.16.0-dev", default-features = false }
-bevy_reflect = { path = "../bevy_reflect", version = "0.16.0-dev", default-features = false, features = [
+bevy_ptr = { path = "../bevy_ptr", version = "0.17.0-dev", default-features = false }
+bevy_reflect = { path = "../bevy_reflect", version = "0.17.0-dev", default-features = false, features = [
   "smallvec",
 ] }
-bevy_time = { path = "../bevy_time", version = "0.16.0-dev", default-features = false, features = [
+bevy_time = { path = "../bevy_time", version = "0.17.0-dev", default-features = false, features = [
   "bevy_reflect",
 ] }
-bevy_transform = { path = "../bevy_transform", version = "0.16.0-dev", default-features = false, features = [
+bevy_transform = { path = "../bevy_transform", version = "0.17.0-dev", default-features = false, features = [
   "bevy-support",
   "bevy_reflect",
 ] }
-bevy_utils = { path = "../bevy_utils", version = "0.16.0-dev", default-features = false }
-bevy_tasks = { path = "../bevy_tasks", version = "0.16.0-dev", default-features = false }
+bevy_utils = { path = "../bevy_utils", version = "0.17.0-dev", default-features = false }
+bevy_tasks = { path = "../bevy_tasks", version = "0.17.0-dev", default-features = false }
 
 # bevy (std required)
-bevy_log = { path = "../bevy_log", version = "0.16.0-dev", optional = true }
+bevy_log = { path = "../bevy_log", version = "0.17.0-dev", optional = true }
 
 # bevy (optional)
-bevy_a11y = { path = "../bevy_a11y", optional = true, version = "0.16.0-dev", features = [
+bevy_a11y = { path = "../bevy_a11y", optional = true, version = "0.17.0-dev", features = [
   "bevy_reflect",
 ] }
-bevy_animation = { path = "../bevy_animation", optional = true, version = "0.16.0-dev" }
-bevy_asset = { path = "../bevy_asset", optional = true, version = "0.16.0-dev" }
-bevy_audio = { path = "../bevy_audio", optional = true, version = "0.16.0-dev" }
-bevy_color = { path = "../bevy_color", optional = true, version = "0.16.0-dev", default-features = false, features = [
+bevy_animation = { path = "../bevy_animation", optional = true, version = "0.17.0-dev" }
+bevy_asset = { path = "../bevy_asset", optional = true, version = "0.17.0-dev" }
+bevy_audio = { path = "../bevy_audio", optional = true, version = "0.17.0-dev" }
+bevy_color = { path = "../bevy_color", optional = true, version = "0.17.0-dev", default-features = false, features = [
   "alloc",
   "bevy_reflect",
 ] }
-bevy_core_pipeline = { path = "../bevy_core_pipeline", optional = true, version = "0.16.0-dev" }
-bevy_core_widgets = { path = "../bevy_core_widgets", optional = true, version = "0.16.0-dev" }
-bevy_anti_aliasing = { path = "../bevy_anti_aliasing", optional = true, version = "0.16.0-dev" }
-bevy_dev_tools = { path = "../bevy_dev_tools", optional = true, version = "0.16.0-dev" }
-bevy_gilrs = { path = "../bevy_gilrs", optional = true, version = "0.16.0-dev" }
-bevy_gizmos = { path = "../bevy_gizmos", optional = true, version = "0.16.0-dev", default-features = false }
-bevy_gltf = { path = "../bevy_gltf", optional = true, version = "0.16.0-dev" }
-bevy_image = { path = "../bevy_image", optional = true, version = "0.16.0-dev" }
-bevy_input_focus = { path = "../bevy_input_focus", optional = true, version = "0.16.0-dev", default-features = false, features = [
+bevy_core_pipeline = { path = "../bevy_core_pipeline", optional = true, version = "0.17.0-dev" }
+bevy_core_widgets = { path = "../bevy_core_widgets", optional = true, version = "0.17.0-dev" }
+bevy_anti_aliasing = { path = "../bevy_anti_aliasing", optional = true, version = "0.17.0-dev" }
+bevy_dev_tools = { path = "../bevy_dev_tools", optional = true, version = "0.17.0-dev" }
+bevy_gilrs = { path = "../bevy_gilrs", optional = true, version = "0.17.0-dev" }
+bevy_gizmos = { path = "../bevy_gizmos", optional = true, version = "0.17.0-dev", default-features = false }
+bevy_gltf = { path = "../bevy_gltf", optional = true, version = "0.17.0-dev" }
+bevy_image = { path = "../bevy_image", optional = true, version = "0.17.0-dev" }
+bevy_input_focus = { path = "../bevy_input_focus", optional = true, version = "0.17.0-dev", default-features = false, features = [
   "bevy_reflect",
 ] }
-bevy_pbr = { path = "../bevy_pbr", optional = true, version = "0.16.0-dev" }
-bevy_picking = { path = "../bevy_picking", optional = true, version = "0.16.0-dev" }
-bevy_remote = { path = "../bevy_remote", optional = true, version = "0.16.0-dev" }
-bevy_render = { path = "../bevy_render", optional = true, version = "0.16.0-dev" }
-bevy_scene = { path = "../bevy_scene", optional = true, version = "0.16.0-dev" }
-bevy_solari = { path = "../bevy_solari", optional = true, version = "0.16.0-dev" }
-bevy_sprite = { path = "../bevy_sprite", optional = true, version = "0.16.0-dev" }
-bevy_state = { path = "../bevy_state", optional = true, version = "0.16.0-dev", default-features = false, features = [
+bevy_pbr = { path = "../bevy_pbr", optional = true, version = "0.17.0-dev" }
+bevy_picking = { path = "../bevy_picking", optional = true, version = "0.17.0-dev" }
+bevy_remote = { path = "../bevy_remote", optional = true, version = "0.17.0-dev" }
+bevy_render = { path = "../bevy_render", optional = true, version = "0.17.0-dev" }
+bevy_scene = { path = "../bevy_scene", optional = true, version = "0.17.0-dev" }
+bevy_solari = { path = "../bevy_solari", optional = true, version = "0.17.0-dev" }
+bevy_sprite = { path = "../bevy_sprite", optional = true, version = "0.17.0-dev" }
+bevy_state = { path = "../bevy_state", optional = true, version = "0.17.0-dev", default-features = false, features = [
   "bevy_app",
   "bevy_reflect",
 ] }
-bevy_text = { path = "../bevy_text", optional = true, version = "0.16.0-dev" }
-bevy_ui = { path = "../bevy_ui", optional = true, version = "0.16.0-dev" }
-bevy_window = { path = "../bevy_window", optional = true, version = "0.16.0-dev", default-features = false, features = [
+bevy_text = { path = "../bevy_text", optional = true, version = "0.17.0-dev" }
+bevy_ui = { path = "../bevy_ui", optional = true, version = "0.17.0-dev" }
+bevy_window = { path = "../bevy_window", optional = true, version = "0.17.0-dev", default-features = false, features = [
   "bevy_reflect",
 ] }
-bevy_winit = { path = "../bevy_winit", optional = true, version = "0.16.0-dev", default-features = false }
+bevy_winit = { path = "../bevy_winit", optional = true, version = "0.17.0-dev", default-features = false }
 
 [lints]
 workspace = true
diff --git a/crates/bevy_log/Cargo.toml b/crates/bevy_log/Cargo.toml
index fce602f7ad3e7..3dcfa277942f1 100644
--- a/crates/bevy_log/Cargo.toml
+++ b/crates/bevy_log/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_log"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Provides logging for Bevy Engine"
 homepage = "https://bevy.org"
@@ -14,10 +14,10 @@ trace_tracy_memory = ["dep:tracy-client"]
 
 [dependencies]
 # bevy
-bevy_app = { path = "../bevy_app", version = "0.16.0-dev" }
-bevy_utils = { path = "../bevy_utils", version = "0.16.0-dev" }
-bevy_platform = { path = "../bevy_platform", version = "0.16.0-dev" }
-bevy_ecs = { path = "../bevy_ecs", version = "0.16.0-dev" }
+bevy_app = { path = "../bevy_app", version = "0.17.0-dev" }
+bevy_utils = { path = "../bevy_utils", version = "0.17.0-dev" }
+bevy_platform = { path = "../bevy_platform", version = "0.17.0-dev" }
+bevy_ecs = { path = "../bevy_ecs", version = "0.17.0-dev" }
 
 # other
 tracing-subscriber = { version = "0.3.1", features = [
@@ -40,7 +40,7 @@ android_log-sys = "0.3.0"
 [target.'cfg(target_arch = "wasm32")'.dependencies]
 tracing-wasm = "0.2.1"
 # TODO: Assuming all wasm builds are for the browser. Require `no_std` support to break assumption.
-bevy_app = { path = "../bevy_app", version = "0.16.0-dev", default-features = false, features = [
+bevy_app = { path = "../bevy_app", version = "0.17.0-dev", default-features = false, features = [
   "web",
 ] }
 
diff --git a/crates/bevy_macro_utils/Cargo.toml b/crates/bevy_macro_utils/Cargo.toml
index bc5989f0f3d3e..b998ae4fd4553 100644
--- a/crates/bevy_macro_utils/Cargo.toml
+++ b/crates/bevy_macro_utils/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_macro_utils"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "A collection of utils for Bevy Engine"
 homepage = "https://bevy.org"
diff --git a/crates/bevy_math/Cargo.toml b/crates/bevy_math/Cargo.toml
index c420c3fe3ce2d..f28e9466ecea4 100644
--- a/crates/bevy_math/Cargo.toml
+++ b/crates/bevy_math/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_math"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Provides math functionality for Bevy Engine"
 homepage = "https://bevy.org"
@@ -25,7 +25,7 @@ approx = { version = "0.5", default-features = false, optional = true }
 rand = { version = "0.8", default-features = false, optional = true }
 rand_distr = { version = "0.4.3", optional = true }
 smallvec = { version = "1.11" }
-bevy_reflect = { path = "../bevy_reflect", version = "0.16.0-dev", default-features = false, features = [
+bevy_reflect = { path = "../bevy_reflect", version = "0.17.0-dev", default-features = false, features = [
   "glam",
 ], optional = true }
 variadics_please = "1.1"
diff --git a/crates/bevy_mesh/Cargo.toml b/crates/bevy_mesh/Cargo.toml
index a235fea5efc53..c65c648cfda03 100644
--- a/crates/bevy_mesh/Cargo.toml
+++ b/crates/bevy_mesh/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_mesh"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Provides mesh types for Bevy Engine"
 homepage = "https://bevy.org"
@@ -10,16 +10,16 @@ keywords = ["bevy"]
 
 [dependencies]
 # bevy
-bevy_asset = { path = "../bevy_asset", version = "0.16.0-dev" }
-bevy_image = { path = "../bevy_image", version = "0.16.0-dev" }
-bevy_math = { path = "../bevy_math", version = "0.16.0-dev" }
-bevy_reflect = { path = "../bevy_reflect", version = "0.16.0-dev" }
-bevy_ecs = { path = "../bevy_ecs", version = "0.16.0-dev" }
-bevy_transform = { path = "../bevy_transform", version = "0.16.0-dev" }
-bevy_mikktspace = { path = "../bevy_mikktspace", version = "0.16.0-dev" }
-bevy_derive = { path = "../bevy_derive", version = "0.16.0-dev" }
-bevy_utils = { path = "../bevy_utils", version = "0.16.0-dev" }
-bevy_platform = { path = "../bevy_platform", version = "0.16.0-dev", default-features = false, features = [
+bevy_asset = { path = "../bevy_asset", version = "0.17.0-dev" }
+bevy_image = { path = "../bevy_image", version = "0.17.0-dev" }
+bevy_math = { path = "../bevy_math", version = "0.17.0-dev" }
+bevy_reflect = { path = "../bevy_reflect", version = "0.17.0-dev" }
+bevy_ecs = { path = "../bevy_ecs", version = "0.17.0-dev" }
+bevy_transform = { path = "../bevy_transform", version = "0.17.0-dev" }
+bevy_mikktspace = { path = "../bevy_mikktspace", version = "0.17.0-dev" }
+bevy_derive = { path = "../bevy_derive", version = "0.17.0-dev" }
+bevy_utils = { path = "../bevy_utils", version = "0.17.0-dev" }
+bevy_platform = { path = "../bevy_platform", version = "0.17.0-dev", default-features = false, features = [
   "std",
   "serialize",
 ] }
diff --git a/crates/bevy_mikktspace/Cargo.toml b/crates/bevy_mikktspace/Cargo.toml
index 7428504adc785..82c2d86553c74 100644
--- a/crates/bevy_mikktspace/Cargo.toml
+++ b/crates/bevy_mikktspace/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_mikktspace"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 authors = [
   "Benjamin Wasty <benny.wasty@gmail.com>",
diff --git a/crates/bevy_pbr/Cargo.toml b/crates/bevy_pbr/Cargo.toml
index 41a058b52200f..4ecf53a7733ec 100644
--- a/crates/bevy_pbr/Cargo.toml
+++ b/crates/bevy_pbr/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_pbr"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Adds PBR rendering to Bevy Engine"
 homepage = "https://bevy.org"
@@ -31,22 +31,22 @@ meshlet_processor = [
 
 [dependencies]
 # bevy
-bevy_app = { path = "../bevy_app", version = "0.16.0-dev" }
-bevy_asset = { path = "../bevy_asset", version = "0.16.0-dev" }
-bevy_color = { path = "../bevy_color", version = "0.16.0-dev" }
-bevy_core_pipeline = { path = "../bevy_core_pipeline", version = "0.16.0-dev" }
-bevy_derive = { path = "../bevy_derive", version = "0.16.0-dev" }
-bevy_diagnostic = { path = "../bevy_diagnostic", version = "0.16.0-dev" }
-bevy_ecs = { path = "../bevy_ecs", version = "0.16.0-dev" }
-bevy_image = { path = "../bevy_image", version = "0.16.0-dev" }
-bevy_math = { path = "../bevy_math", version = "0.16.0-dev" }
-bevy_reflect = { path = "../bevy_reflect", version = "0.16.0-dev" }
-bevy_render = { path = "../bevy_render", version = "0.16.0-dev" }
-bevy_tasks = { path = "../bevy_tasks", version = "0.16.0-dev", optional = true }
-bevy_transform = { path = "../bevy_transform", version = "0.16.0-dev" }
-bevy_utils = { path = "../bevy_utils", version = "0.16.0-dev" }
-bevy_window = { path = "../bevy_window", version = "0.16.0-dev" }
-bevy_platform = { path = "../bevy_platform", version = "0.16.0-dev", default-features = false, features = [
+bevy_app = { path = "../bevy_app", version = "0.17.0-dev" }
+bevy_asset = { path = "../bevy_asset", version = "0.17.0-dev" }
+bevy_color = { path = "../bevy_color", version = "0.17.0-dev" }
+bevy_core_pipeline = { path = "../bevy_core_pipeline", version = "0.17.0-dev" }
+bevy_derive = { path = "../bevy_derive", version = "0.17.0-dev" }
+bevy_diagnostic = { path = "../bevy_diagnostic", version = "0.17.0-dev" }
+bevy_ecs = { path = "../bevy_ecs", version = "0.17.0-dev" }
+bevy_image = { path = "../bevy_image", version = "0.17.0-dev" }
+bevy_math = { path = "../bevy_math", version = "0.17.0-dev" }
+bevy_reflect = { path = "../bevy_reflect", version = "0.17.0-dev" }
+bevy_render = { path = "../bevy_render", version = "0.17.0-dev" }
+bevy_tasks = { path = "../bevy_tasks", version = "0.17.0-dev", optional = true }
+bevy_transform = { path = "../bevy_transform", version = "0.17.0-dev" }
+bevy_utils = { path = "../bevy_utils", version = "0.17.0-dev" }
+bevy_window = { path = "../bevy_window", version = "0.17.0-dev" }
+bevy_platform = { path = "../bevy_platform", version = "0.17.0-dev", default-features = false, features = [
   "std",
 ] }
 
diff --git a/crates/bevy_picking/Cargo.toml b/crates/bevy_picking/Cargo.toml
index b42f0db14cbd5..ac18b1dd897b9 100644
--- a/crates/bevy_picking/Cargo.toml
+++ b/crates/bevy_picking/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_picking"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Provides screen picking functionality for Bevy Engine"
 homepage = "https://bevy.org"
@@ -13,20 +13,20 @@ bevy_mesh_picking_backend = ["dep:bevy_mesh", "dep:crossbeam-channel"]
 
 [dependencies]
 # bevy
-bevy_app = { path = "../bevy_app", version = "0.16.0-dev" }
-bevy_asset = { path = "../bevy_asset", version = "0.16.0-dev" }
-bevy_derive = { path = "../bevy_derive", version = "0.16.0-dev" }
-bevy_ecs = { path = "../bevy_ecs", version = "0.16.0-dev" }
-bevy_input = { path = "../bevy_input", version = "0.16.0-dev" }
-bevy_math = { path = "../bevy_math", version = "0.16.0-dev" }
-bevy_mesh = { path = "../bevy_mesh", version = "0.16.0-dev", optional = true }
-bevy_reflect = { path = "../bevy_reflect", version = "0.16.0-dev" }
-bevy_render = { path = "../bevy_render", version = "0.16.0-dev" }
-bevy_time = { path = "../bevy_time", version = "0.16.0-dev" }
-bevy_transform = { path = "../bevy_transform", version = "0.16.0-dev" }
-bevy_utils = { path = "../bevy_utils", version = "0.16.0-dev" }
-bevy_window = { path = "../bevy_window", version = "0.16.0-dev" }
-bevy_platform = { path = "../bevy_platform", version = "0.16.0-dev", default-features = false, features = [
+bevy_app = { path = "../bevy_app", version = "0.17.0-dev" }
+bevy_asset = { path = "../bevy_asset", version = "0.17.0-dev" }
+bevy_derive = { path = "../bevy_derive", version = "0.17.0-dev" }
+bevy_ecs = { path = "../bevy_ecs", version = "0.17.0-dev" }
+bevy_input = { path = "../bevy_input", version = "0.17.0-dev" }
+bevy_math = { path = "../bevy_math", version = "0.17.0-dev" }
+bevy_mesh = { path = "../bevy_mesh", version = "0.17.0-dev", optional = true }
+bevy_reflect = { path = "../bevy_reflect", version = "0.17.0-dev" }
+bevy_render = { path = "../bevy_render", version = "0.17.0-dev" }
+bevy_time = { path = "../bevy_time", version = "0.17.0-dev" }
+bevy_transform = { path = "../bevy_transform", version = "0.17.0-dev" }
+bevy_utils = { path = "../bevy_utils", version = "0.17.0-dev" }
+bevy_window = { path = "../bevy_window", version = "0.17.0-dev" }
+bevy_platform = { path = "../bevy_platform", version = "0.17.0-dev", default-features = false, features = [
   "std",
 ] }
 
diff --git a/crates/bevy_platform/Cargo.toml b/crates/bevy_platform/Cargo.toml
index 7a4313af39fd0..614ff15fce0fb 100644
--- a/crates/bevy_platform/Cargo.toml
+++ b/crates/bevy_platform/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_platform"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Provides common platform agnostic APIs, as well as platform-specific features for Bevy Engine"
 homepage = "https://bevy.org"
diff --git a/crates/bevy_ptr/Cargo.toml b/crates/bevy_ptr/Cargo.toml
index b6e72e24f050a..07c6eeae68dd8 100644
--- a/crates/bevy_ptr/Cargo.toml
+++ b/crates/bevy_ptr/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_ptr"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Utilities for working with untyped pointers in a more safe way"
 homepage = "https://bevy.org"
diff --git a/crates/bevy_reflect/Cargo.toml b/crates/bevy_reflect/Cargo.toml
index 4d72ab1d6ab4b..087cdb44db96e 100644
--- a/crates/bevy_reflect/Cargo.toml
+++ b/crates/bevy_reflect/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_reflect"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Dynamically interact with rust types"
 homepage = "https://bevy.org"
@@ -74,10 +74,10 @@ web = ["bevy_platform/web", "uuid?/js"]
 
 [dependencies]
 # bevy
-bevy_reflect_derive = { path = "derive", version = "0.16.0-dev" }
-bevy_utils = { path = "../bevy_utils", version = "0.16.0-dev", default-features = false }
-bevy_ptr = { path = "../bevy_ptr", version = "0.16.0-dev" }
-bevy_platform = { path = "../bevy_platform", version = "0.16.0-dev", default-features = false, features = [
+bevy_reflect_derive = { path = "derive", version = "0.17.0-dev" }
+bevy_utils = { path = "../bevy_utils", version = "0.17.0-dev", default-features = false }
+bevy_ptr = { path = "../bevy_ptr", version = "0.17.0-dev" }
+bevy_platform = { path = "../bevy_platform", version = "0.17.0-dev", default-features = false, features = [
   "alloc",
   "serialize",
 ] }
diff --git a/crates/bevy_reflect/derive/Cargo.toml b/crates/bevy_reflect/derive/Cargo.toml
index a3685941cc322..032046ae2fdb3 100644
--- a/crates/bevy_reflect/derive/Cargo.toml
+++ b/crates/bevy_reflect/derive/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_reflect_derive"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Derive implementations for bevy_reflect"
 homepage = "https://bevy.org"
@@ -19,7 +19,7 @@ documentation = []
 functions = []
 
 [dependencies]
-bevy_macro_utils = { path = "../../bevy_macro_utils", version = "0.16.0-dev" }
+bevy_macro_utils = { path = "../../bevy_macro_utils", version = "0.17.0-dev" }
 proc-macro2 = "1.0"
 quote = "1.0"
 syn = { version = "2.0", features = ["full"] }
diff --git a/crates/bevy_remote/Cargo.toml b/crates/bevy_remote/Cargo.toml
index 28e0fc6f2b4b3..7fc8d2b4ded19 100644
--- a/crates/bevy_remote/Cargo.toml
+++ b/crates/bevy_remote/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_remote"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "The Bevy Remote Protocol"
 homepage = "https://bevy.org"
@@ -15,21 +15,21 @@ bevy_asset = ["dep:bevy_asset"]
 
 [dependencies]
 # bevy
-bevy_app = { path = "../bevy_app", version = "0.16.0-dev" }
-bevy_derive = { path = "../bevy_derive", version = "0.16.0-dev" }
-bevy_ecs = { path = "../bevy_ecs", version = "0.16.0-dev", features = [
+bevy_app = { path = "../bevy_app", version = "0.17.0-dev" }
+bevy_derive = { path = "../bevy_derive", version = "0.17.0-dev" }
+bevy_ecs = { path = "../bevy_ecs", version = "0.17.0-dev", features = [
   "serialize",
 ] }
-bevy_reflect = { path = "../bevy_reflect", version = "0.16.0-dev" }
-bevy_tasks = { path = "../bevy_tasks", version = "0.16.0-dev" }
-bevy_utils = { path = "../bevy_utils", version = "0.16.0-dev", features = [
+bevy_reflect = { path = "../bevy_reflect", version = "0.17.0-dev" }
+bevy_tasks = { path = "../bevy_tasks", version = "0.17.0-dev" }
+bevy_utils = { path = "../bevy_utils", version = "0.17.0-dev", features = [
   "debug",
 ] }
-bevy_platform = { path = "../bevy_platform", version = "0.16.0-dev", default-features = false, features = [
+bevy_platform = { path = "../bevy_platform", version = "0.17.0-dev", default-features = false, features = [
   "std",
   "serialize",
 ] }
-bevy_asset = { path = "../bevy_asset", version = "0.16.0-dev", optional = true }
+bevy_asset = { path = "../bevy_asset", version = "0.17.0-dev", optional = true }
 
 # other
 anyhow = "1"
diff --git a/crates/bevy_render/Cargo.toml b/crates/bevy_render/Cargo.toml
index d9775e9c8fea6..143a1f0270745 100644
--- a/crates/bevy_render/Cargo.toml
+++ b/crates/bevy_render/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_render"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Provides rendering functionality for Bevy Engine"
 homepage = "https://bevy.org"
@@ -49,29 +49,29 @@ detailed_trace = []
 
 [dependencies]
 # bevy
-bevy_app = { path = "../bevy_app", version = "0.16.0-dev" }
-bevy_asset = { path = "../bevy_asset", version = "0.16.0-dev" }
-bevy_color = { path = "../bevy_color", version = "0.16.0-dev", features = [
+bevy_app = { path = "../bevy_app", version = "0.17.0-dev" }
+bevy_asset = { path = "../bevy_asset", version = "0.17.0-dev" }
+bevy_color = { path = "../bevy_color", version = "0.17.0-dev", features = [
   "serialize",
   "wgpu-types",
 ] }
-bevy_derive = { path = "../bevy_derive", version = "0.16.0-dev" }
-bevy_diagnostic = { path = "../bevy_diagnostic", version = "0.16.0-dev" }
-bevy_ecs = { path = "../bevy_ecs", version = "0.16.0-dev" }
-bevy_encase_derive = { path = "../bevy_encase_derive", version = "0.16.0-dev" }
-bevy_math = { path = "../bevy_math", version = "0.16.0-dev" }
-bevy_reflect = { path = "../bevy_reflect", version = "0.16.0-dev" }
-bevy_render_macros = { path = "macros", version = "0.16.0-dev" }
-bevy_time = { path = "../bevy_time", version = "0.16.0-dev" }
-bevy_transform = { path = "../bevy_transform", version = "0.16.0-dev" }
-bevy_window = { path = "../bevy_window", version = "0.16.0-dev" }
-bevy_utils = { path = "../bevy_utils", version = "0.16.0-dev", features = [
+bevy_derive = { path = "../bevy_derive", version = "0.17.0-dev" }
+bevy_diagnostic = { path = "../bevy_diagnostic", version = "0.17.0-dev" }
+bevy_ecs = { path = "../bevy_ecs", version = "0.17.0-dev" }
+bevy_encase_derive = { path = "../bevy_encase_derive", version = "0.17.0-dev" }
+bevy_math = { path = "../bevy_math", version = "0.17.0-dev" }
+bevy_reflect = { path = "../bevy_reflect", version = "0.17.0-dev" }
+bevy_render_macros = { path = "macros", version = "0.17.0-dev" }
+bevy_time = { path = "../bevy_time", version = "0.17.0-dev" }
+bevy_transform = { path = "../bevy_transform", version = "0.17.0-dev" }
+bevy_window = { path = "../bevy_window", version = "0.17.0-dev" }
+bevy_utils = { path = "../bevy_utils", version = "0.17.0-dev", features = [
   "wgpu_wrapper",
 ] }
-bevy_tasks = { path = "../bevy_tasks", version = "0.16.0-dev" }
-bevy_image = { path = "../bevy_image", version = "0.16.0-dev" }
-bevy_mesh = { path = "../bevy_mesh", version = "0.16.0-dev" }
-bevy_platform = { path = "../bevy_platform", version = "0.16.0-dev", default-features = false, features = [
+bevy_tasks = { path = "../bevy_tasks", version = "0.17.0-dev" }
+bevy_image = { path = "../bevy_image", version = "0.17.0-dev" }
+bevy_mesh = { path = "../bevy_mesh", version = "0.17.0-dev" }
+bevy_platform = { path = "../bevy_platform", version = "0.17.0-dev", default-features = false, features = [
   "std",
   "serialize",
 ] }
@@ -140,16 +140,16 @@ web-sys = { version = "0.3.67", features = [
 ] }
 wasm-bindgen = "0.2"
 # TODO: Assuming all wasm builds are for the browser. Require `no_std` support to break assumption.
-bevy_app = { path = "../bevy_app", version = "0.16.0-dev", default-features = false, features = [
+bevy_app = { path = "../bevy_app", version = "0.17.0-dev", default-features = false, features = [
   "web",
 ] }
-bevy_tasks = { path = "../bevy_tasks", version = "0.16.0-dev", default-features = false, features = [
+bevy_tasks = { path = "../bevy_tasks", version = "0.17.0-dev", default-features = false, features = [
   "web",
 ] }
-bevy_platform = { path = "../bevy_platform", version = "0.16.0-dev", default-features = false, features = [
+bevy_platform = { path = "../bevy_platform", version = "0.17.0-dev", default-features = false, features = [
   "web",
 ] }
-bevy_reflect = { path = "../bevy_reflect", version = "0.16.0-dev", default-features = false, features = [
+bevy_reflect = { path = "../bevy_reflect", version = "0.17.0-dev", default-features = false, features = [
   "web",
 ] }
 
diff --git a/crates/bevy_render/macros/Cargo.toml b/crates/bevy_render/macros/Cargo.toml
index 74348598fe0d3..016fe887652e4 100644
--- a/crates/bevy_render/macros/Cargo.toml
+++ b/crates/bevy_render/macros/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_render_macros"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Derive implementations for bevy_render"
 homepage = "https://bevy.org"
@@ -12,7 +12,7 @@ keywords = ["bevy"]
 proc-macro = true
 
 [dependencies]
-bevy_macro_utils = { path = "../../bevy_macro_utils", version = "0.16.0-dev" }
+bevy_macro_utils = { path = "../../bevy_macro_utils", version = "0.17.0-dev" }
 
 syn = "2.0"
 proc-macro2 = "1.0"
diff --git a/crates/bevy_scene/Cargo.toml b/crates/bevy_scene/Cargo.toml
index 8a6fe517fde04..78de17e26f90b 100644
--- a/crates/bevy_scene/Cargo.toml
+++ b/crates/bevy_scene/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_scene"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Provides scene functionality for Bevy Engine"
 homepage = "https://bevy.org"
@@ -19,15 +19,15 @@ serialize = [
 
 [dependencies]
 # bevy
-bevy_app = { path = "../bevy_app", version = "0.16.0-dev" }
-bevy_asset = { path = "../bevy_asset", version = "0.16.0-dev" }
-bevy_derive = { path = "../bevy_derive", version = "0.16.0-dev" }
-bevy_ecs = { path = "../bevy_ecs", version = "0.16.0-dev" }
-bevy_reflect = { path = "../bevy_reflect", version = "0.16.0-dev" }
-bevy_transform = { path = "../bevy_transform", version = "0.16.0-dev" }
-bevy_utils = { path = "../bevy_utils", version = "0.16.0-dev" }
-bevy_render = { path = "../bevy_render", version = "0.16.0-dev", optional = true }
-bevy_platform = { path = "../bevy_platform", version = "0.16.0-dev", default-features = false, features = [
+bevy_app = { path = "../bevy_app", version = "0.17.0-dev" }
+bevy_asset = { path = "../bevy_asset", version = "0.17.0-dev" }
+bevy_derive = { path = "../bevy_derive", version = "0.17.0-dev" }
+bevy_ecs = { path = "../bevy_ecs", version = "0.17.0-dev" }
+bevy_reflect = { path = "../bevy_reflect", version = "0.17.0-dev" }
+bevy_transform = { path = "../bevy_transform", version = "0.17.0-dev" }
+bevy_utils = { path = "../bevy_utils", version = "0.17.0-dev" }
+bevy_render = { path = "../bevy_render", version = "0.17.0-dev", optional = true }
+bevy_platform = { path = "../bevy_platform", version = "0.17.0-dev", default-features = false, features = [
   "std",
 ] }
 
diff --git a/crates/bevy_solari/Cargo.toml b/crates/bevy_solari/Cargo.toml
index 85cf85bd49a63..76cb751fe6e94 100644
--- a/crates/bevy_solari/Cargo.toml
+++ b/crates/bevy_solari/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_solari"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Provides raytraced lighting for Bevy Engine"
 homepage = "https://bevyengine.org"
@@ -10,21 +10,21 @@ keywords = ["bevy"]
 
 [dependencies]
 # bevy
-bevy_app = { path = "../bevy_app", version = "0.16.0-dev" }
-bevy_asset = { path = "../bevy_asset", version = "0.16.0-dev" }
-bevy_color = { path = "../bevy_color", version = "0.16.0-dev" }
-bevy_core_pipeline = { path = "../bevy_core_pipeline", version = "0.16.0-dev" }
-bevy_derive = { path = "../bevy_derive", version = "0.16.0-dev" }
-bevy_ecs = { path = "../bevy_ecs", version = "0.16.0-dev" }
-bevy_math = { path = "../bevy_math", version = "0.16.0-dev" }
-bevy_mesh = { path = "../bevy_mesh", version = "0.16.0-dev" }
-bevy_pbr = { path = "../bevy_pbr", version = "0.16.0-dev" }
-bevy_platform = { path = "../bevy_platform", version = "0.16.0-dev", default-features = false, features = [
+bevy_app = { path = "../bevy_app", version = "0.17.0-dev" }
+bevy_asset = { path = "../bevy_asset", version = "0.17.0-dev" }
+bevy_color = { path = "../bevy_color", version = "0.17.0-dev" }
+bevy_core_pipeline = { path = "../bevy_core_pipeline", version = "0.17.0-dev" }
+bevy_derive = { path = "../bevy_derive", version = "0.17.0-dev" }
+bevy_ecs = { path = "../bevy_ecs", version = "0.17.0-dev" }
+bevy_math = { path = "../bevy_math", version = "0.17.0-dev" }
+bevy_mesh = { path = "../bevy_mesh", version = "0.17.0-dev" }
+bevy_pbr = { path = "../bevy_pbr", version = "0.17.0-dev" }
+bevy_platform = { path = "../bevy_platform", version = "0.17.0-dev", default-features = false, features = [
   "std",
 ] }
-bevy_reflect = { path = "../bevy_reflect", version = "0.16.0-dev" }
-bevy_render = { path = "../bevy_render", version = "0.16.0-dev" }
-bevy_transform = { path = "../bevy_transform", version = "0.16.0-dev" }
+bevy_reflect = { path = "../bevy_reflect", version = "0.17.0-dev" }
+bevy_render = { path = "../bevy_render", version = "0.17.0-dev" }
+bevy_transform = { path = "../bevy_transform", version = "0.17.0-dev" }
 
 # other
 tracing = { version = "0.1", default-features = false, features = ["std"] }
diff --git a/crates/bevy_sprite/Cargo.toml b/crates/bevy_sprite/Cargo.toml
index 99d526b1e8deb..1d356fdc40cc8 100644
--- a/crates/bevy_sprite/Cargo.toml
+++ b/crates/bevy_sprite/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_sprite"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Provides sprite functionality for Bevy Engine"
 homepage = "https://bevy.org"
@@ -15,21 +15,21 @@ webgpu = []
 
 [dependencies]
 # bevy
-bevy_app = { path = "../bevy_app", version = "0.16.0-dev" }
-bevy_asset = { path = "../bevy_asset", version = "0.16.0-dev" }
-bevy_color = { path = "../bevy_color", version = "0.16.0-dev" }
-bevy_core_pipeline = { path = "../bevy_core_pipeline", version = "0.16.0-dev" }
-bevy_ecs = { path = "../bevy_ecs", version = "0.16.0-dev" }
-bevy_image = { path = "../bevy_image", version = "0.16.0-dev" }
-bevy_math = { path = "../bevy_math", version = "0.16.0-dev" }
-bevy_picking = { path = "../bevy_picking", version = "0.16.0-dev", optional = true }
-bevy_reflect = { path = "../bevy_reflect", version = "0.16.0-dev" }
-bevy_render = { path = "../bevy_render", version = "0.16.0-dev" }
-bevy_transform = { path = "../bevy_transform", version = "0.16.0-dev" }
-bevy_utils = { path = "../bevy_utils", version = "0.16.0-dev" }
-bevy_window = { path = "../bevy_window", version = "0.16.0-dev", optional = true }
-bevy_derive = { path = "../bevy_derive", version = "0.16.0-dev" }
-bevy_platform = { path = "../bevy_platform", version = "0.16.0-dev", default-features = false, features = [
+bevy_app = { path = "../bevy_app", version = "0.17.0-dev" }
+bevy_asset = { path = "../bevy_asset", version = "0.17.0-dev" }
+bevy_color = { path = "../bevy_color", version = "0.17.0-dev" }
+bevy_core_pipeline = { path = "../bevy_core_pipeline", version = "0.17.0-dev" }
+bevy_ecs = { path = "../bevy_ecs", version = "0.17.0-dev" }
+bevy_image = { path = "../bevy_image", version = "0.17.0-dev" }
+bevy_math = { path = "../bevy_math", version = "0.17.0-dev" }
+bevy_picking = { path = "../bevy_picking", version = "0.17.0-dev", optional = true }
+bevy_reflect = { path = "../bevy_reflect", version = "0.17.0-dev" }
+bevy_render = { path = "../bevy_render", version = "0.17.0-dev" }
+bevy_transform = { path = "../bevy_transform", version = "0.17.0-dev" }
+bevy_utils = { path = "../bevy_utils", version = "0.17.0-dev" }
+bevy_window = { path = "../bevy_window", version = "0.17.0-dev", optional = true }
+bevy_derive = { path = "../bevy_derive", version = "0.17.0-dev" }
+bevy_platform = { path = "../bevy_platform", version = "0.17.0-dev", default-features = false, features = [
   "std",
 ] }
 
diff --git a/crates/bevy_state/Cargo.toml b/crates/bevy_state/Cargo.toml
index fd780fc486925..17f9d6c787cb1 100644
--- a/crates/bevy_state/Cargo.toml
+++ b/crates/bevy_state/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_state"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Finite state machines for Bevy"
 homepage = "https://bevy.org"
@@ -46,12 +46,12 @@ critical-section = [
 
 [dependencies]
 # bevy
-bevy_ecs = { path = "../bevy_ecs", version = "0.16.0-dev", default-features = false }
-bevy_state_macros = { path = "macros", version = "0.16.0-dev" }
-bevy_utils = { path = "../bevy_utils", version = "0.16.0-dev", default-features = false }
-bevy_reflect = { path = "../bevy_reflect", version = "0.16.0-dev", default-features = false, optional = true }
-bevy_app = { path = "../bevy_app", version = "0.16.0-dev", default-features = false, optional = true }
-bevy_platform = { path = "../bevy_platform", version = "0.16.0-dev", default-features = false }
+bevy_ecs = { path = "../bevy_ecs", version = "0.17.0-dev", default-features = false }
+bevy_state_macros = { path = "macros", version = "0.17.0-dev" }
+bevy_utils = { path = "../bevy_utils", version = "0.17.0-dev", default-features = false }
+bevy_reflect = { path = "../bevy_reflect", version = "0.17.0-dev", default-features = false, optional = true }
+bevy_app = { path = "../bevy_app", version = "0.17.0-dev", default-features = false, optional = true }
+bevy_platform = { path = "../bevy_platform", version = "0.17.0-dev", default-features = false }
 variadics_please = "1.1"
 
 # other
diff --git a/crates/bevy_state/macros/Cargo.toml b/crates/bevy_state/macros/Cargo.toml
index 2f569f395ef03..2ab531a9b2eea 100644
--- a/crates/bevy_state/macros/Cargo.toml
+++ b/crates/bevy_state/macros/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_state_macros"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 description = "Macros for bevy_state"
 edition = "2024"
 license = "MIT OR Apache-2.0"
@@ -9,7 +9,7 @@ license = "MIT OR Apache-2.0"
 proc-macro = true
 
 [dependencies]
-bevy_macro_utils = { path = "../../bevy_macro_utils", version = "0.16.0-dev" }
+bevy_macro_utils = { path = "../../bevy_macro_utils", version = "0.17.0-dev" }
 
 syn = { version = "2.0", features = ["full"] }
 quote = "1.0"
diff --git a/crates/bevy_tasks/Cargo.toml b/crates/bevy_tasks/Cargo.toml
index ad162a7ef74ff..61b07448ed202 100644
--- a/crates/bevy_tasks/Cargo.toml
+++ b/crates/bevy_tasks/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_tasks"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "A task executor for Bevy Engine"
 homepage = "https://bevy.org"
@@ -47,7 +47,7 @@ web = [
 ]
 
 [dependencies]
-bevy_platform = { path = "../bevy_platform", version = "0.16.0-dev", default-features = false, features = [
+bevy_platform = { path = "../bevy_platform", version = "0.17.0-dev", default-features = false, features = [
   "alloc",
 ] }
 
diff --git a/crates/bevy_text/Cargo.toml b/crates/bevy_text/Cargo.toml
index 9a8c8e8eeeb67..c71fb5ce0c1bd 100644
--- a/crates/bevy_text/Cargo.toml
+++ b/crates/bevy_text/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_text"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Provides text functionality for Bevy Engine"
 homepage = "https://bevy.org"
@@ -13,21 +13,21 @@ default_font = []
 
 [dependencies]
 # bevy
-bevy_app = { path = "../bevy_app", version = "0.16.0-dev" }
-bevy_asset = { path = "../bevy_asset", version = "0.16.0-dev" }
-bevy_color = { path = "../bevy_color", version = "0.16.0-dev" }
-bevy_derive = { path = "../bevy_derive", version = "0.16.0-dev" }
-bevy_ecs = { path = "../bevy_ecs", version = "0.16.0-dev" }
-bevy_image = { path = "../bevy_image", version = "0.16.0-dev" }
-bevy_log = { path = "../bevy_log", version = "0.16.0-dev" }
-bevy_math = { path = "../bevy_math", version = "0.16.0-dev" }
-bevy_reflect = { path = "../bevy_reflect", version = "0.16.0-dev" }
-bevy_render = { path = "../bevy_render", version = "0.16.0-dev" }
-bevy_sprite = { path = "../bevy_sprite", version = "0.16.0-dev" }
-bevy_transform = { path = "../bevy_transform", version = "0.16.0-dev" }
-bevy_window = { path = "../bevy_window", version = "0.16.0-dev" }
-bevy_utils = { path = "../bevy_utils", version = "0.16.0-dev" }
-bevy_platform = { path = "../bevy_platform", version = "0.16.0-dev", default-features = false, features = [
+bevy_app = { path = "../bevy_app", version = "0.17.0-dev" }
+bevy_asset = { path = "../bevy_asset", version = "0.17.0-dev" }
+bevy_color = { path = "../bevy_color", version = "0.17.0-dev" }
+bevy_derive = { path = "../bevy_derive", version = "0.17.0-dev" }
+bevy_ecs = { path = "../bevy_ecs", version = "0.17.0-dev" }
+bevy_image = { path = "../bevy_image", version = "0.17.0-dev" }
+bevy_log = { path = "../bevy_log", version = "0.17.0-dev" }
+bevy_math = { path = "../bevy_math", version = "0.17.0-dev" }
+bevy_reflect = { path = "../bevy_reflect", version = "0.17.0-dev" }
+bevy_render = { path = "../bevy_render", version = "0.17.0-dev" }
+bevy_sprite = { path = "../bevy_sprite", version = "0.17.0-dev" }
+bevy_transform = { path = "../bevy_transform", version = "0.17.0-dev" }
+bevy_window = { path = "../bevy_window", version = "0.17.0-dev" }
+bevy_utils = { path = "../bevy_utils", version = "0.17.0-dev" }
+bevy_platform = { path = "../bevy_platform", version = "0.17.0-dev", default-features = false, features = [
   "std",
   "serialize",
 ] }
diff --git a/crates/bevy_time/Cargo.toml b/crates/bevy_time/Cargo.toml
index 494725b08cc4d..7c8e840ace2ba 100644
--- a/crates/bevy_time/Cargo.toml
+++ b/crates/bevy_time/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_time"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Provides time functionality for Bevy Engine"
 homepage = "https://bevy.org"
@@ -48,10 +48,10 @@ critical-section = [
 
 [dependencies]
 # bevy
-bevy_app = { path = "../bevy_app", version = "0.16.0-dev", default-features = false }
-bevy_ecs = { path = "../bevy_ecs", version = "0.16.0-dev", default-features = false }
-bevy_reflect = { path = "../bevy_reflect", version = "0.16.0-dev", default-features = false, optional = true }
-bevy_platform = { path = "../bevy_platform", version = "0.16.0-dev", default-features = false }
+bevy_app = { path = "../bevy_app", version = "0.17.0-dev", default-features = false }
+bevy_ecs = { path = "../bevy_ecs", version = "0.17.0-dev", default-features = false }
+bevy_reflect = { path = "../bevy_reflect", version = "0.17.0-dev", default-features = false, optional = true }
+bevy_platform = { path = "../bevy_platform", version = "0.17.0-dev", default-features = false }
 
 # other
 crossbeam-channel = { version = "0.5.0", default-features = false, features = [
diff --git a/crates/bevy_transform/Cargo.toml b/crates/bevy_transform/Cargo.toml
index 6801f5734b6c3..6c86be0da654e 100644
--- a/crates/bevy_transform/Cargo.toml
+++ b/crates/bevy_transform/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_transform"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Provides transform functionality for Bevy Engine"
 homepage = "https://bevy.org"
@@ -10,13 +10,13 @@ keywords = ["bevy"]
 
 [dependencies]
 # bevy
-bevy_app = { path = "../bevy_app", version = "0.16.0-dev", default-features = false, optional = true }
-bevy_ecs = { path = "../bevy_ecs", version = "0.16.0-dev", default-features = false, optional = true }
-bevy_log = { path = "../bevy_log", version = "0.16.0-dev", default-features = false, optional = true }
-bevy_math = { path = "../bevy_math", version = "0.16.0-dev", default-features = false }
-bevy_reflect = { path = "../bevy_reflect", version = "0.16.0-dev", default-features = false, optional = true }
-bevy_tasks = { path = "../bevy_tasks", version = "0.16.0-dev", default-features = false }
-bevy_utils = { path = "../bevy_utils", version = "0.16.0-dev", default-features = false, optional = true }
+bevy_app = { path = "../bevy_app", version = "0.17.0-dev", default-features = false, optional = true }
+bevy_ecs = { path = "../bevy_ecs", version = "0.17.0-dev", default-features = false, optional = true }
+bevy_log = { path = "../bevy_log", version = "0.17.0-dev", default-features = false, optional = true }
+bevy_math = { path = "../bevy_math", version = "0.17.0-dev", default-features = false }
+bevy_reflect = { path = "../bevy_reflect", version = "0.17.0-dev", default-features = false, optional = true }
+bevy_tasks = { path = "../bevy_tasks", version = "0.17.0-dev", default-features = false }
+bevy_utils = { path = "../bevy_utils", version = "0.17.0-dev", default-features = false, optional = true }
 serde = { version = "1", default-features = false, features = [
   "derive",
 ], optional = true }
@@ -24,8 +24,8 @@ thiserror = { version = "2", default-features = false }
 derive_more = { version = "1", default-features = false, features = ["from"] }
 
 [dev-dependencies]
-bevy_tasks = { path = "../bevy_tasks", version = "0.16.0-dev" }
-bevy_math = { path = "../bevy_math", version = "0.16.0-dev", default-features = false, features = [
+bevy_tasks = { path = "../bevy_tasks", version = "0.17.0-dev" }
+bevy_math = { path = "../bevy_math", version = "0.17.0-dev", default-features = false, features = [
   "approx",
 ] }
 approx = "0.5.1"
diff --git a/crates/bevy_ui/Cargo.toml b/crates/bevy_ui/Cargo.toml
index 5224d902968b9..1f05d7a53b6f1 100644
--- a/crates/bevy_ui/Cargo.toml
+++ b/crates/bevy_ui/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_ui"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "A custom ECS-driven UI framework built specifically for Bevy Engine"
 homepage = "https://bevy.org"
@@ -10,25 +10,25 @@ keywords = ["bevy"]
 
 [dependencies]
 # bevy
-bevy_a11y = { path = "../bevy_a11y", version = "0.16.0-dev" }
-bevy_app = { path = "../bevy_app", version = "0.16.0-dev" }
-bevy_asset = { path = "../bevy_asset", version = "0.16.0-dev" }
-bevy_color = { path = "../bevy_color", version = "0.16.0-dev" }
-bevy_core_pipeline = { path = "../bevy_core_pipeline", version = "0.16.0-dev" }
-bevy_derive = { path = "../bevy_derive", version = "0.16.0-dev" }
-bevy_ecs = { path = "../bevy_ecs", version = "0.16.0-dev" }
-bevy_image = { path = "../bevy_image", version = "0.16.0-dev" }
-bevy_input = { path = "../bevy_input", version = "0.16.0-dev" }
-bevy_math = { path = "../bevy_math", version = "0.16.0-dev" }
-bevy_reflect = { path = "../bevy_reflect", version = "0.16.0-dev" }
-bevy_render = { path = "../bevy_render", version = "0.16.0-dev" }
-bevy_sprite = { path = "../bevy_sprite", version = "0.16.0-dev" }
-bevy_text = { path = "../bevy_text", version = "0.16.0-dev" }
-bevy_picking = { path = "../bevy_picking", version = "0.16.0-dev", optional = true }
-bevy_transform = { path = "../bevy_transform", version = "0.16.0-dev" }
-bevy_window = { path = "../bevy_window", version = "0.16.0-dev" }
-bevy_utils = { path = "../bevy_utils", version = "0.16.0-dev" }
-bevy_platform = { path = "../bevy_platform", version = "0.16.0-dev", default-features = false, features = [
+bevy_a11y = { path = "../bevy_a11y", version = "0.17.0-dev" }
+bevy_app = { path = "../bevy_app", version = "0.17.0-dev" }
+bevy_asset = { path = "../bevy_asset", version = "0.17.0-dev" }
+bevy_color = { path = "../bevy_color", version = "0.17.0-dev" }
+bevy_core_pipeline = { path = "../bevy_core_pipeline", version = "0.17.0-dev" }
+bevy_derive = { path = "../bevy_derive", version = "0.17.0-dev" }
+bevy_ecs = { path = "../bevy_ecs", version = "0.17.0-dev" }
+bevy_image = { path = "../bevy_image", version = "0.17.0-dev" }
+bevy_input = { path = "../bevy_input", version = "0.17.0-dev" }
+bevy_math = { path = "../bevy_math", version = "0.17.0-dev" }
+bevy_reflect = { path = "../bevy_reflect", version = "0.17.0-dev" }
+bevy_render = { path = "../bevy_render", version = "0.17.0-dev" }
+bevy_sprite = { path = "../bevy_sprite", version = "0.17.0-dev" }
+bevy_text = { path = "../bevy_text", version = "0.17.0-dev" }
+bevy_picking = { path = "../bevy_picking", version = "0.17.0-dev", optional = true }
+bevy_transform = { path = "../bevy_transform", version = "0.17.0-dev" }
+bevy_window = { path = "../bevy_window", version = "0.17.0-dev" }
+bevy_utils = { path = "../bevy_utils", version = "0.17.0-dev" }
+bevy_platform = { path = "../bevy_platform", version = "0.17.0-dev", default-features = false, features = [
   "std",
 ] }
 
diff --git a/crates/bevy_utils/Cargo.toml b/crates/bevy_utils/Cargo.toml
index aaf6a0834e4dc..4e74e6ea94b60 100644
--- a/crates/bevy_utils/Cargo.toml
+++ b/crates/bevy_utils/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_utils"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "A collection of utils for Bevy Engine"
 homepage = "https://bevy.org"
@@ -21,7 +21,7 @@ std = ["disqualified/alloc"]
 debug = []
 
 [dependencies]
-bevy_platform = { path = "../bevy_platform", version = "0.16.0-dev", default-features = false }
+bevy_platform = { path = "../bevy_platform", version = "0.17.0-dev", default-features = false }
 
 disqualified = { version = "1.0", default-features = false }
 thread_local = { version = "1.0", optional = true }
diff --git a/crates/bevy_window/Cargo.toml b/crates/bevy_window/Cargo.toml
index 2e3e7edf1882a..1a5bca6916fa6 100644
--- a/crates/bevy_window/Cargo.toml
+++ b/crates/bevy_window/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_window"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "Provides windowing functionality for Bevy Engine"
 homepage = "https://bevy.org"
@@ -50,16 +50,16 @@ libm = ["bevy_math/libm"]
 
 [dependencies]
 # bevy
-bevy_app = { path = "../bevy_app", version = "0.16.0-dev", default-features = false }
-bevy_ecs = { path = "../bevy_ecs", version = "0.16.0-dev", default-features = false }
-bevy_input = { path = "../bevy_input", version = "0.16.0-dev", default-features = false }
-bevy_math = { path = "../bevy_math", version = "0.16.0-dev", default-features = false }
-bevy_reflect = { path = "../bevy_reflect", version = "0.16.0-dev", default-features = false, features = [
+bevy_app = { path = "../bevy_app", version = "0.17.0-dev", default-features = false }
+bevy_ecs = { path = "../bevy_ecs", version = "0.17.0-dev", default-features = false }
+bevy_input = { path = "../bevy_input", version = "0.17.0-dev", default-features = false }
+bevy_math = { path = "../bevy_math", version = "0.17.0-dev", default-features = false }
+bevy_reflect = { path = "../bevy_reflect", version = "0.17.0-dev", default-features = false, features = [
   "glam",
   "smol_str",
 ], optional = true }
-bevy_utils = { path = "../bevy_utils", version = "0.16.0-dev", default-features = false }
-bevy_platform = { path = "../bevy_platform", version = "0.16.0-dev", default-features = false }
+bevy_utils = { path = "../bevy_utils", version = "0.17.0-dev", default-features = false }
+bevy_platform = { path = "../bevy_platform", version = "0.17.0-dev", default-features = false }
 
 # other
 serde = { version = "1.0", features = [
diff --git a/crates/bevy_winit/Cargo.toml b/crates/bevy_winit/Cargo.toml
index 43db87a1d2372..bb28b84fb041c 100644
--- a/crates/bevy_winit/Cargo.toml
+++ b/crates/bevy_winit/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_winit"
-version = "0.16.0-dev"
+version = "0.17.0-dev"
 edition = "2024"
 description = "A winit window and input backend for Bevy Engine"
 homepage = "https://bevy.org"
@@ -28,25 +28,25 @@ custom_cursor = ["bevy_image", "bevy_asset", "bytemuck", "wgpu-types"]
 
 [dependencies]
 # bevy
-bevy_a11y = { path = "../bevy_a11y", version = "0.16.0-dev" }
-bevy_app = { path = "../bevy_app", version = "0.16.0-dev" }
-bevy_derive = { path = "../bevy_derive", version = "0.16.0-dev" }
-bevy_ecs = { path = "../bevy_ecs", version = "0.16.0-dev" }
-bevy_input = { path = "../bevy_input", version = "0.16.0-dev" }
-bevy_input_focus = { path = "../bevy_input_focus", version = "0.16.0-dev" }
-bevy_log = { path = "../bevy_log", version = "0.16.0-dev" }
-bevy_math = { path = "../bevy_math", version = "0.16.0-dev" }
-bevy_reflect = { path = "../bevy_reflect", version = "0.16.0-dev" }
-bevy_window = { path = "../bevy_window", version = "0.16.0-dev" }
-bevy_utils = { path = "../bevy_utils", version = "0.16.0-dev" }
-bevy_tasks = { path = "../bevy_tasks", version = "0.16.0-dev" }
-bevy_platform = { path = "../bevy_platform", version = "0.16.0-dev", default-features = false, features = [
+bevy_a11y = { path = "../bevy_a11y", version = "0.17.0-dev" }
+bevy_app = { path = "../bevy_app", version = "0.17.0-dev" }
+bevy_derive = { path = "../bevy_derive", version = "0.17.0-dev" }
+bevy_ecs = { path = "../bevy_ecs", version = "0.17.0-dev" }
+bevy_input = { path = "../bevy_input", version = "0.17.0-dev" }
+bevy_input_focus = { path = "../bevy_input_focus", version = "0.17.0-dev" }
+bevy_log = { path = "../bevy_log", version = "0.17.0-dev" }
+bevy_math = { path = "../bevy_math", version = "0.17.0-dev" }
+bevy_reflect = { path = "../bevy_reflect", version = "0.17.0-dev" }
+bevy_window = { path = "../bevy_window", version = "0.17.0-dev" }
+bevy_utils = { path = "../bevy_utils", version = "0.17.0-dev" }
+bevy_tasks = { path = "../bevy_tasks", version = "0.17.0-dev" }
+bevy_platform = { path = "../bevy_platform", version = "0.17.0-dev", default-features = false, features = [
   "std",
 ] }
 
 # bevy optional
-bevy_asset = { path = "../bevy_asset", version = "0.16.0-dev", optional = true }
-bevy_image = { path = "../bevy_image", version = "0.16.0-dev", optional = true }
+bevy_asset = { path = "../bevy_asset", version = "0.17.0-dev", optional = true }
+bevy_image = { path = "../bevy_image", version = "0.17.0-dev", optional = true }
 
 # other
 # feature rwh_06 refers to window_raw_handle@v0.6
@@ -68,16 +68,16 @@ wasm-bindgen = { version = "0.2" }
 web-sys = "0.3"
 crossbeam-channel = "0.5"
 # TODO: Assuming all wasm builds are for the browser. Require `no_std` support to break assumption.
-bevy_app = { path = "../bevy_app", version = "0.16.0-dev", default-features = false, features = [
+bevy_app = { path = "../bevy_app", version = "0.17.0-dev", default-features = false, features = [
   "web",
 ] }
-bevy_tasks = { path = "../bevy_tasks", version = "0.16.0-dev", default-features = false, features = [
+bevy_tasks = { path = "../bevy_tasks", version = "0.17.0-dev", default-features = false, features = [
   "web",
 ] }
-bevy_platform = { path = "../bevy_platform", version = "0.16.0-dev", default-features = false, features = [
+bevy_platform = { path = "../bevy_platform", version = "0.17.0-dev", default-features = false, features = [
   "web",
 ] }
-bevy_reflect = { path = "../bevy_reflect", version = "0.16.0-dev", default-features = false, features = [
+bevy_reflect = { path = "../bevy_reflect", version = "0.17.0-dev", default-features = false, features = [
   "web",
 ] }
 
