diff --git a/crates/bevy_pbr/src/lib.rs b/crates/bevy_pbr/src/lib.rs
index ed8a993fee4e2..4ef2fb6c030d8 100644
--- a/crates/bevy_pbr/src/lib.rs
+++ b/crates/bevy_pbr/src/lib.rs
@@ -139,8 +139,8 @@ pub struct PbrPlugin {
     pub use_gpu_instance_buffer_builder: bool,
     /// Debugging flags that can optionally be set when constructing the renderer.
     pub debug_flags: RenderDebugFlags,
-    /// Renders GLTFs with PBR.
-    pub gltf_render_enabled: bool,
+    /// Builds and inserts `StandardMaterial` when loading glTF files
+    pub gltf_enable_standard_materials: bool,
 }
 
 impl Default for PbrPlugin {
@@ -150,7 +150,7 @@ impl Default for PbrPlugin {
             add_default_deferred_lighting_plugin: true,
             use_gpu_instance_buffer_builder: true,
             debug_flags: RenderDebugFlags::default(),
-            gltf_render_enabled: true,
+            gltf_enable_standard_materials: true,
         }
     }
 }
@@ -224,7 +224,7 @@ impl Plugin for PbrPlugin {
             ))
             .add_plugins((ScatteringMediumPlugin, AtmospherePlugin));
 
-        if self.gltf_render_enabled {
+        if self.gltf_enable_standard_materials {
             gltf::add_gltf(app);
         }
 
diff --git a/examples/gltf/gltf_extension_mesh_2d.rs b/examples/gltf/gltf_extension_mesh_2d.rs
index 11cdd792b6bf7..d1eb8ec578f65 100644
--- a/examples/gltf/gltf_extension_mesh_2d.rs
+++ b/examples/gltf/gltf_extension_mesh_2d.rs
@@ -44,7 +44,7 @@ fn main() {
                         .add_custom_vertex_attribute("_BARYCENTRIC", ATTRIBUTE_BARYCENTRIC),
                 )
                 .set(PbrPlugin {
-                    gltf_render_enabled: false,
+                    gltf_enable_standard_materials: false,
                     ..Default::default()
                 }),
             GltfToMesh2dPlugin,
diff --git a/release-content/migration-guides/gltf_pbr.md b/release-content/migration-guides/gltf_pbr.md
index a051a192fab85..0547aec98e90b 100644
--- a/release-content/migration-guides/gltf_pbr.md
+++ b/release-content/migration-guides/gltf_pbr.md
@@ -23,7 +23,7 @@ You can disable PBR rendering by initializing `PbrPlugin` as so:
 
 ```rs
 PbrPlugin {
-    gltf_render_enabled: false,
+    gltf_enable_standard_materials: false,
     ..Default::default()
 }
 ```
