diff --git a/examples/3d/post_processing.rs b/examples/3d/post_processing.rs
index 25873b6d1df96..e472bcdb97d2f 100644
--- a/examples/3d/post_processing.rs
+++ b/examples/3d/post_processing.rs
@@ -237,7 +237,6 @@ fn update_chromatic_aberration_settings(
 /// [`AppSettings`].
 fn update_help_text(mut text: Single<&mut Text>, app_settings: Res<AppSettings>) {
     text.clear();
-    //let vignette_mode_list = ["Cosine Fourth Law", "Higher-order Powers", "Smoothstep"];
     let text_list = [
         format!(
             "Chromatic aberration intensity: {:.2}\n",
