diff --git a/crates/bevy_pbr/src/prepass/mod.rs b/crates/bevy_pbr/src/prepass/mod.rs
index 581fb02878648..26b5f4d5ab0f4 100644
--- a/crates/bevy_pbr/src/prepass/mod.rs
+++ b/crates/bevy_pbr/src/prepass/mod.rs
@@ -226,19 +226,25 @@ pub fn update_mesh_previous_global_transforms(
         (Entity, &GlobalTransform),
         (PreviousMeshFilter, Without<PreviousGlobalTransform>),
     >,
-    mut meshes: Query<(&GlobalTransform, &mut PreviousGlobalTransform), PreviousMeshFilter>,
+    mut meshes: Query<(Ref<GlobalTransform>, &mut PreviousGlobalTransform), PreviousMeshFilter>,
+    system_change_tick: SystemChangeTick,
 ) {
-    let should_run = views.iter().any(|camera| camera.is_active);
+    if !views.iter().any(|camera| camera.is_active) {
+        return;
+    }
 
-    if should_run {
-        for (entity, transform) in &new_meshes {
-            let new_previous_transform = PreviousGlobalTransform(transform.affine());
-            commands.entity(entity).try_insert(new_previous_transform);
-        }
-        meshes.par_iter_mut().for_each(|(transform, mut previous)| {
-            previous.set_if_neq(PreviousGlobalTransform(transform.affine()));
-        });
+    for (entity, transform) in &new_meshes {
+        let new_previous_transform = PreviousGlobalTransform(transform.affine());
+        commands.entity(entity).try_insert(new_previous_transform);
     }
+    meshes.par_iter_mut().for_each(|(transform, mut previous)| {
+        if transform
+            .last_changed()
+            .is_newer_than(previous.last_changed(), system_change_tick.this_run())
+        {
+            *previous = PreviousGlobalTransform(transform.affine());
+        }
+    });
 }
 
 #[derive(Resource, Clone)]
