diff --git a/crates/bevy_core_pipeline/src/experimental/mod.rs b/crates/bevy_core_pipeline/src/experimental/mod.rs
deleted file mode 100644
index 9eb1b8a6b4491..0000000000000
--- a/crates/bevy_core_pipeline/src/experimental/mod.rs
+++ /dev/null
@@ -1,7 +0,0 @@
-//! Experimental rendering features.
-//!
-//! Experimental features are features with known problems, missing features,
-//! compatibility issues, low performance, and/or future breaking changes, but
-//! are included nonetheless for testing purposes.
-
-pub mod depth;
