diff --git a/examples/large_scenes/bistro/src/main.rs b/examples/large_scenes/bistro/src/main.rs
index 895c0d78ae1b7..9e15eb4934eb0 100644
--- a/examples/large_scenes/bistro/src/main.rs
+++ b/examples/large_scenes/bistro/src/main.rs
@@ -307,8 +307,8 @@ pub fn setup(mut commands: Commands, asset_server: Res<AssetServer>, args: Res<A
         commands
             .spawn((
                 Node {
-                    left: Val::Px(1.5),
-                    top: Val::Px(1.5),
+                    left: px(1.5),
+                    top: px(1.5),
                     ..default()
                 },
                 GlobalZIndex(-1),
diff --git a/examples/large_scenes/caldera_hotel/src/main.rs b/examples/large_scenes/caldera_hotel/src/main.rs
index 9a508fdc36085..787e970348b84 100644
--- a/examples/large_scenes/caldera_hotel/src/main.rs
+++ b/examples/large_scenes/caldera_hotel/src/main.rs
@@ -244,8 +244,8 @@ pub fn setup(
         commands
             .spawn((
                 Node {
-                    left: Val::Px(1.5),
-                    top: Val::Px(1.5),
+                    left: px(1.5),
+                    top: px(1.5),
                     ..default()
                 },
                 GlobalZIndex(-1),
diff --git a/examples/large_scenes/mipmap_generator/src/lib.rs b/examples/large_scenes/mipmap_generator/src/lib.rs
index f57bec192c36f..5c84aed86a3c0 100644
--- a/examples/large_scenes/mipmap_generator/src/lib.rs
+++ b/examples/large_scenes/mipmap_generator/src/lib.rs
@@ -155,8 +155,8 @@ fn init_loading_text(mut commands: Commands) {
     commands
         .spawn((
             Node {
-                left: Val::Px(1.5),
-                top: Val::Px(1.5),
+                left: px(1.5),
+                top: px(1.5),
                 ..default()
             },
             GlobalZIndex(-1),
diff --git a/examples/picking/dragdrop_picking.rs b/examples/picking/dragdrop_picking.rs
index e2616db41dc7f..b5772690dc022 100644
--- a/examples/picking/dragdrop_picking.rs
+++ b/examples/picking/dragdrop_picking.rs
@@ -49,9 +49,9 @@ fn setup(
                 .spawn((
                     DraggableButton,
                     Node {
-                        width: Val::Px(BUTTON_WIDTH),
-                        height: Val::Px(BUTTON_HEIGHT),
-                        margin: UiRect::all(Val::Px(10.0)),
+                        width: px(BUTTON_WIDTH),
+                        height: px(BUTTON_HEIGHT),
+                        margin: UiRect::all(px(10.0)),
                         justify_content: JustifyContent::Center,
                         align_items: AlignItems::Center,
                         ..default()
diff --git a/examples/shader_advanced/render_depth_to_texture.rs b/examples/shader_advanced/render_depth_to_texture.rs
index 4607c96b40aba..ac3c3365fd3fe 100644
--- a/examples/shader_advanced/render_depth_to_texture.rs
+++ b/examples/shader_advanced/render_depth_to_texture.rs
@@ -281,8 +281,8 @@ fn spawn_instructions(commands: &mut Commands) {
         Text::new("Use WASD to move the secondary camera"),
         Node {
             position_type: PositionType::Absolute,
-            top: Val::Px(12.0),
-            left: Val::Px(12.0),
+            top: px(12.0),
+            left: px(12.0),
             ..Node::default()
         },
     ));
diff --git a/examples/ui/images/image_node.rs b/examples/ui/images/image_node.rs
index 70e91d035d5f2..a374d4516a264 100644
--- a/examples/ui/images/image_node.rs
+++ b/examples/ui/images/image_node.rs
@@ -30,8 +30,8 @@ fn setup(mut commands: Commands, asset_server: Res<AssetServer>) {
             ImageNode::new(asset_server.load("branding/icon.png")),
             // Child Node control `ImageNode` size
             Node {
-                width: Val::Px(256.),
-                height: Val::Px(256.),
+                width: px(256.),
+                height: px(256.),
                 ..default()
             }
         )],
diff --git a/examples/ui/layout/anchor_layout.rs b/examples/ui/layout/anchor_layout.rs
index 5f0c6f9cc1050..8ad44661449c0 100644
--- a/examples/ui/layout/anchor_layout.rs
+++ b/examples/ui/layout/anchor_layout.rs
@@ -49,14 +49,14 @@ fn spawn_layout(mut commands: Commands, asset_server: Res<AssetServer>) {
             "left: 10px\ncenter: 10px",
             Node {
                 left: px(10),
-                margin: UiRect::vertical(Val::Auto),
+                margin: UiRect::vertical(auto()),
                 ..default()
             },
         ),
         (
             "center: 10px\ncenter: 10px",
             Node {
-                margin: UiRect::all(Val::Auto),
+                margin: UiRect::all(auto()),
                 ..default()
             },
         ),
@@ -64,7 +64,7 @@ fn spawn_layout(mut commands: Commands, asset_server: Res<AssetServer>) {
             "right: 10px\ncenter: 10px",
             Node {
                 right: px(10),
-                margin: UiRect::vertical(Val::Auto),
+                margin: UiRect::vertical(auto()),
                 ..default()
             },
         ),
@@ -79,7 +79,7 @@ fn spawn_layout(mut commands: Commands, asset_server: Res<AssetServer>) {
         (
             "center: 10px\nbottom: 10px",
             Node {
-                margin: UiRect::horizontal(Val::Auto),
+                margin: UiRect::horizontal(auto()),
                 bottom: px(10),
                 ..default()
             },
diff --git a/examples/ui/render_ui_to_texture.rs b/examples/ui/render_ui_to_texture.rs
index d2ff76bad34b4..6a9f3b582fe17 100644
--- a/examples/ui/render_ui_to_texture.rs
+++ b/examples/ui/render_ui_to_texture.rs
@@ -92,11 +92,11 @@ fn setup(
                 .spawn((
                     Node {
                         position_type: PositionType::Absolute,
-                        width: Val::Auto,
-                        height: Val::Auto,
+                        width: auto(),
+                        height: auto(),
                         align_items: AlignItems::Center,
-                        padding: UiRect::all(Val::Px(20.)),
-                        border_radius: BorderRadius::all(Val::Px(10.)),
+                        padding: UiRect::all(px(20.)),
+                        border_radius: BorderRadius::all(px(10.)),
                         ..default()
                     },
                     BackgroundColor(BLUE.into()),
@@ -104,9 +104,8 @@ fn setup(
                 .observe(
                     |drag: On<Pointer<Drag>>, mut nodes: Query<(&mut Node, &ComputedNode)>| {
                         let (mut node, computed) = nodes.get_mut(drag.entity).unwrap();
-                        node.left =
-                            Val::Px(drag.pointer_location.position.x - computed.size.x / 2.0);
-                        node.top = Val::Px(drag.pointer_location.position.y - 50.0);
+                        node.left = px(drag.pointer_location.position.x - computed.size.x / 2.0);
+                        node.top = px(drag.pointer_location.position.y - 50.0);
                     },
                 )
                 .observe(
diff --git a/examples/ui/text/text.rs b/examples/ui/text/text.rs
index e7c3f371ffafd..96b8f7a6bd8d2 100644
--- a/examples/ui/text/text.rs
+++ b/examples/ui/text/text.rs
@@ -86,10 +86,10 @@ fn setup(mut commands: Commands, asset_server: Res<AssetServer>) {
     commands
         .spawn((
             Node {
-                margin: UiRect::all(Val::Px(12.0)),
+                margin: UiRect::all(px(12.0)),
                 position_type: PositionType::Absolute,
-                top: Val::Px(5.0),
-                right: Val::Px(5.0),
+                top: px(5.0),
+                right: px(5.0),
                 ..default()
             },
             Text::new("Opentype features:\n"),
diff --git a/examples/window/multi_window_text.rs b/examples/window/multi_window_text.rs
index bd7af84625511..c88c22c185291 100644
--- a/examples/window/multi_window_text.rs
+++ b/examples/window/multi_window_text.rs
@@ -54,8 +54,8 @@ fn setup_scene(mut commands: Commands) {
 
     let node = Node {
         position_type: PositionType::Absolute,
-        top: Val::Px(12.0),
-        left: Val::Px(12.0),
+        top: px(12.0),
+        left: px(12.0),
         ..default()
     };
 
