diff --git a/crates/bevy_camera_controller/src/free_camera.rs b/crates/bevy_camera_controller/src/free_camera.rs
index b46be5f61223e..89b655bd02d7f 100644
--- a/crates/bevy_camera_controller/src/free_camera.rs
+++ b/crates/bevy_camera_controller/src/free_camera.rs
@@ -112,7 +112,7 @@ impl Default for FreeCamera {
             key_up: KeyCode::KeyE,
             key_down: KeyCode::KeyQ,
             key_run: KeyCode::ShiftLeft,
-            mouse_key_cursor_grab: MouseButton::Left,
+            mouse_key_cursor_grab: MouseButton::Right,
             keyboard_key_toggle_cursor_grab: KeyCode::KeyM,
             walk_speed: 5.0,
             run_speed: 15.0,
diff --git a/examples/camera/free_camera_controller.rs b/examples/camera/free_camera_controller.rs
index 12b15cad603ce..1de40b22b556f 100644
--- a/examples/camera/free_camera_controller.rs
+++ b/examples/camera/free_camera_controller.rs
@@ -14,7 +14,7 @@
 //! | Default Key Binding | Action                 |
 //! |:--------------------|:-----------------------|
 //! | Mouse               | Look around            |
-//! | Left click          | Capture mouse (hold)   |
+//! | Right click         | Capture mouse (hold)   |
 //! | M                   | Capture mouse (toggle) |
 //! | WASD                | Horizontal movement    |
 //! | QE                  | Vertical movement      |
