diff --git a/crates/bevy_anti_alias/src/contrast_adaptive_sharpening/mod.rs b/crates/bevy_anti_alias/src/contrast_adaptive_sharpening/mod.rs
index 9478167c0fdd0..6a6b854a5b951 100644
--- a/crates/bevy_anti_alias/src/contrast_adaptive_sharpening/mod.rs
+++ b/crates/bevy_anti_alias/src/contrast_adaptive_sharpening/mod.rs
@@ -16,6 +16,7 @@ use bevy_render::{
         *,
     },
     renderer::RenderDevice,
+    sync_component::SyncComponent,
     view::{ExtractedView, ViewTarget},
     Render, RenderApp, RenderStartup, RenderSystems,
 };
@@ -75,10 +76,13 @@ pub struct CasUniform {
     sharpness: f32,
 }
 
+impl SyncComponent for ContrastAdaptiveSharpening {
+    type Out = (DenoiseCas, CasUniform);
+}
+
 impl ExtractComponent for ContrastAdaptiveSharpening {
     type QueryData = &'static Self;
     type QueryFilter = With<Camera>;
-    type Out = (DenoiseCas, CasUniform);
 
     fn extract_component(item: QueryItem<Self::QueryData>) -> Option<Self::Out> {
         if !item.enabled || item.sharpening_strength == 0.0 {
diff --git a/crates/bevy_anti_alias/src/taa/mod.rs b/crates/bevy_anti_alias/src/taa/mod.rs
index d14473fb981cc..6601ba8661145 100644
--- a/crates/bevy_anti_alias/src/taa/mod.rs
+++ b/crates/bevy_anti_alias/src/taa/mod.rs
@@ -32,7 +32,7 @@ use bevy_render::{
         TextureFormat, TextureSampleType, TextureUsages, Variants,
     },
     renderer::{RenderContext, RenderDevice, ViewQuery},
-    sync_component::SyncComponentPlugin,
+    sync_component::{SyncComponent, SyncComponentPlugin},
     sync_world::RenderEntity,
     texture::{CachedTexture, TextureCache},
     view::{ExtractedView, Msaa, ViewTarget},
@@ -133,6 +133,10 @@ impl Default for TemporalAntiAliasing {
     }
 }
 
+impl SyncComponent for TemporalAntiAliasing {
+    type Out = Self;
+}
+
 fn temporal_anti_alias(
     view: ViewQuery<(
         &ExtractedCamera,
diff --git a/crates/bevy_pbr/src/atmosphere/mod.rs b/crates/bevy_pbr/src/atmosphere/mod.rs
index 44d5e111d5f82..a065daff40bc0 100644
--- a/crates/bevy_pbr/src/atmosphere/mod.rs
+++ b/crates/bevy_pbr/src/atmosphere/mod.rs
@@ -56,6 +56,7 @@ use bevy_reflect::{std_traits::ReflectDefault, Reflect};
 use bevy_render::{
     extract_component::UniformComponentPlugin,
     render_resource::{DownlevelFlags, ShaderType, SpecializedRenderPipelines},
+    sync_component::SyncComponent,
     sync_world::RenderEntity,
     Extract, ExtractSchedule, RenderStartup,
 };
@@ -361,13 +362,14 @@ impl From<AtmosphereSettings> for GpuAtmosphereSettings {
     }
 }
 
+impl SyncComponent for GpuAtmosphereSettings {
+    type Out = Self;
+}
+
 impl ExtractComponent for GpuAtmosphereSettings {
     type QueryData = Read<AtmosphereSettings>;
-
     type QueryFilter = (With<Camera3d>, With<Atmosphere>);
 
-    type Out = GpuAtmosphereSettings;
-
     fn extract_component(item: QueryItem<'_, '_, Self::QueryData>) -> Option<Self::Out> {
         Some(item.clone().into())
     }
diff --git a/crates/bevy_pbr/src/contact_shadows.rs b/crates/bevy_pbr/src/contact_shadows.rs
index f38e2c89fcf2a..2e6d1c585c495 100644
--- a/crates/bevy_pbr/src/contact_shadows.rs
+++ b/crates/bevy_pbr/src/contact_shadows.rs
@@ -16,6 +16,7 @@ use bevy_render::{
     extract_component::{ExtractComponent, ExtractComponentPlugin},
     render_resource::{DynamicUniformBuffer, ShaderType},
     renderer::{RenderDevice, RenderQueue},
+    sync_component::SyncComponent,
     view::ExtractedView,
     Render, RenderApp, RenderSystems,
 };
@@ -79,10 +80,13 @@ impl From<ContactShadows> for ContactShadowsUniform {
     }
 }
 
+impl SyncComponent for ContactShadows {
+    type Out = Self;
+}
+
 impl ExtractComponent for ContactShadows {
     type QueryData = &'static ContactShadows;
     type QueryFilter = ();
-    type Out = ContactShadows;
 
     fn extract_component(settings: QueryItem<'_, '_, Self::QueryData>) -> Option<Self::Out> {
         Some(*settings)
diff --git a/crates/bevy_pbr/src/decal/clustered.rs b/crates/bevy_pbr/src/decal/clustered.rs
index 47e95804a1614..832b25c8e2f70 100644
--- a/crates/bevy_pbr/src/decal/clustered.rs
+++ b/crates/bevy_pbr/src/decal/clustered.rs
@@ -43,7 +43,7 @@ use bevy_render::{
     },
     renderer::{RenderAdapter, RenderDevice, RenderQueue},
     settings::WgpuFeatures,
-    sync_component::SyncComponentPlugin,
+    sync_component::{SyncComponent, SyncComponentPlugin},
     sync_world::RenderEntity,
     texture::{FallbackImage, GpuImage},
     Extract, ExtractSchedule, Render, RenderApp, RenderSystems,
@@ -147,7 +147,7 @@ impl Plugin for ClusteredDecalPlugin {
     fn build(&self, app: &mut App) {
         load_shader_library!(app, "clustered.wgsl");
 
-        app.add_plugins(SyncComponentPlugin::<ClusteredDecal>::default());
+        app.add_plugins(SyncComponentPlugin::<ClusteredDecal, Self>::default());
 
         let Some(render_app) = app.get_sub_app_mut(RenderApp) else {
             return;
@@ -170,6 +170,10 @@ impl Plugin for ClusteredDecalPlugin {
     }
 }
 
+impl SyncComponent<ClusteredDecalPlugin> for ClusteredDecal {
+    type Out = Self;
+}
+
 // This is needed because of the orphan rule not allowing implementing
 // foreign trait ExtractComponent on foreign type ClusteredDecal
 fn extract_clustered_decal(
diff --git a/crates/bevy_pbr/src/lib.rs b/crates/bevy_pbr/src/lib.rs
index 4b792e25bb064..8e67727ccf990 100644
--- a/crates/bevy_pbr/src/lib.rs
+++ b/crates/bevy_pbr/src/lib.rs
@@ -29,6 +29,7 @@ mod cluster;
 mod components;
 pub mod contact_shadows;
 use bevy_gltf::extensions::{GltfExtensionHandler, GltfExtensionHandlers};
+use bevy_render::sync_component::SyncComponent;
 pub use contact_shadows::{
     ContactShadows, ContactShadowsBuffer, ContactShadowsPlugin, ContactShadowsUniform,
     ViewContactShadowsUniformOffset,
@@ -204,7 +205,7 @@ impl Plugin for PbrPlugin {
                 ScreenSpaceAmbientOcclusionPlugin,
                 FogPlugin,
                 ExtractResourcePlugin::<DefaultOpaqueRendererMethod>::default(),
-                SyncComponentPlugin::<ShadowFilteringMethod>::default(),
+                SyncComponentPlugin::<ShadowFilteringMethod, Self>::default(),
                 LightmapPlugin,
                 LightProbePlugin,
                 GpuMeshPreprocessPlugin {
@@ -218,10 +219,10 @@ impl Plugin for PbrPlugin {
             ))
             .add_plugins((
                 decal::ForwardDecalPlugin,
-                SyncComponentPlugin::<DirectionalLight>::default(),
-                SyncComponentPlugin::<PointLight>::default(),
-                SyncComponentPlugin::<SpotLight>::default(),
-                SyncComponentPlugin::<AmbientLight>::default(),
+                SyncComponentPlugin::<DirectionalLight, Self>::default(),
+                SyncComponentPlugin::<PointLight, Self>::default(),
+                SyncComponentPlugin::<SpotLight, Self>::default(),
+                SyncComponentPlugin::<AmbientLight, Self>::default(),
             ))
             .add_plugins((ScatteringMediumPlugin, AtmospherePlugin));
 
@@ -463,3 +464,19 @@ impl GltfExtensionHandler for GltfExtensionHandlerPbr {
         entity.insert(MeshMaterial3d(handle));
     }
 }
+
+impl SyncComponent<PbrPlugin> for DirectionalLight {
+    type Out = Self;
+}
+impl SyncComponent<PbrPlugin> for PointLight {
+    type Out = Self;
+}
+impl SyncComponent<PbrPlugin> for SpotLight {
+    type Out = Self;
+}
+impl SyncComponent<PbrPlugin> for AmbientLight {
+    type Out = Self;
+}
+impl SyncComponent<PbrPlugin> for ShadowFilteringMethod {
+    type Out = Self;
+}
diff --git a/crates/bevy_pbr/src/light_probe/generate.rs b/crates/bevy_pbr/src/light_probe/generate.rs
index b4c2609af3730..7d902d6baf199 100644
--- a/crates/bevy_pbr/src/light_probe/generate.rs
+++ b/crates/bevy_pbr/src/light_probe/generate.rs
@@ -38,7 +38,7 @@ use bevy_render::{
     },
     renderer::{RenderAdapter, RenderContext, RenderDevice, RenderQueue},
     settings::WgpuFeatures,
-    sync_component::SyncComponentPlugin,
+    sync_component::{SyncComponent, SyncComponentPlugin},
     sync_world::RenderEntity,
     texture::{CachedTexture, GpuImage, TextureCache},
     Extract, ExtractSchedule, Render, RenderApp, RenderStartup, RenderSystems,
@@ -128,7 +128,7 @@ impl Plugin for EnvironmentMapGenerationPlugin {
         embedded_asset!(app, "environment_filter.wgsl");
         embedded_asset!(app, "copy.wgsl");
 
-        app.add_plugins(SyncComponentPlugin::<GeneratedEnvironmentMapLight>::default())
+        app.add_plugins(SyncComponentPlugin::<EnvironmentMapLight, Self>::default())
             .add_systems(Update, generate_environment_map_light);
 
         let Some(render_app) = app.get_sub_app_mut(RenderApp) else {
@@ -1108,3 +1108,7 @@ pub fn generate_environment_map_light(
         });
     }
 }
+
+impl SyncComponent<EnvironmentMapGenerationPlugin> for EnvironmentMapLight {
+    type Out = GeneratedEnvironmentMapLight;
+}
diff --git a/crates/bevy_pbr/src/ssr/mod.rs b/crates/bevy_pbr/src/ssr/mod.rs
index 7c2c238e4e644..e2c70954c23fa 100644
--- a/crates/bevy_pbr/src/ssr/mod.rs
+++ b/crates/bevy_pbr/src/ssr/mod.rs
@@ -37,6 +37,7 @@ use bevy_render::{
         TextureViewDimension,
     },
     renderer::{RenderAdapter, RenderContext, RenderDevice, RenderQueue, ViewQuery},
+    sync_component::SyncComponent,
     texture::GpuImage,
     view::{ExtractedView, Msaa, ViewTarget, ViewUniformOffset},
     Render, RenderApp, RenderStartup, RenderSystems,
@@ -511,13 +512,14 @@ pub fn prepare_ssr_settings(
     }
 }
 
+impl SyncComponent for ScreenSpaceReflections {
+    type Out = ScreenSpaceReflectionsUniform;
+}
+
 impl ExtractComponent for ScreenSpaceReflections {
     type QueryData = Read<ScreenSpaceReflections>;
-
     type QueryFilter = ();
 
-    type Out = ScreenSpaceReflectionsUniform;
-
     fn extract_component(settings: QueryItem<'_, '_, Self::QueryData>) -> Option<Self::Out> {
         if !DEPTH_TEXTURE_SAMPLING_SUPPORTED {
             once!(info!(
diff --git a/crates/bevy_pbr/src/volumetric_fog/mod.rs b/crates/bevy_pbr/src/volumetric_fog/mod.rs
index e3b6c57c75824..d6d1bf353c5f0 100644
--- a/crates/bevy_pbr/src/volumetric_fog/mod.rs
+++ b/crates/bevy_pbr/src/volumetric_fog/mod.rs
@@ -43,7 +43,8 @@ use bevy_math::{
 };
 use bevy_mesh::{Mesh, Meshable};
 use bevy_render::{
-    render_resource::SpecializedRenderPipelines, sync_component::SyncComponentPlugin,
+    render_resource::SpecializedRenderPipelines,
+    sync_component::{SyncComponent, SyncComponentPlugin},
     ExtractSchedule, Render, RenderApp, RenderStartup, RenderSystems,
 };
 use render::{volumetric_fog, VolumetricFogPipeline, VolumetricFogUniformBuffer};
@@ -69,7 +70,7 @@ impl Plugin for VolumetricFogPlugin {
         let plane_mesh = meshes.add(Plane3d::new(Vec3::Z, Vec2::ONE).mesh());
         let cube_mesh = meshes.add(Cuboid::new(1.0, 1.0, 1.0).mesh());
 
-        app.add_plugins(SyncComponentPlugin::<FogVolume>::default());
+        app.add_plugins(SyncComponentPlugin::<FogVolume, Self>::default());
 
         let Some(render_app) = app.get_sub_app_mut(RenderApp) else {
             return;
@@ -102,3 +103,7 @@ impl Plugin for VolumetricFogPlugin {
             );
     }
 }
+
+impl SyncComponent<VolumetricFogPlugin> for FogVolume {
+    type Out = Self;
+}
diff --git a/crates/bevy_post_process/src/bloom/settings.rs b/crates/bevy_post_process/src/bloom/settings.rs
index 8a1080df4607e..272f853bd356a 100644
--- a/crates/bevy_post_process/src/bloom/settings.rs
+++ b/crates/bevy_post_process/src/bloom/settings.rs
@@ -7,7 +7,7 @@ use bevy_ecs::{
 };
 use bevy_math::{AspectRatio, URect, UVec4, Vec2, Vec4};
 use bevy_reflect::{std_traits::ReflectDefault, Reflect};
-use bevy_render::extract_component::ExtractComponent;
+use bevy_render::{extract_component::ExtractComponent, sync_component::SyncComponent};
 
 /// Applies a bloom effect to an HDR-enabled 2d or 3d camera.
 ///
@@ -222,11 +222,13 @@ pub enum BloomCompositeMode {
     Additive,
 }
 
+impl SyncComponent for Bloom {
+    type Out = (Self, BloomUniforms);
+}
+
 impl ExtractComponent for Bloom {
     type QueryData = (&'static Self, &'static Camera);
-
     type QueryFilter = With<Hdr>;
-    type Out = (Self, BloomUniforms);
 
     fn extract_component((bloom, camera): QueryItem<'_, '_, Self::QueryData>) -> Option<Self::Out> {
         match (
diff --git a/crates/bevy_post_process/src/dof/mod.rs b/crates/bevy_post_process/src/dof/mod.rs
index e9143f61e55c4..2adf62952e8cb 100644
--- a/crates/bevy_post_process/src/dof/mod.rs
+++ b/crates/bevy_post_process/src/dof/mod.rs
@@ -44,7 +44,7 @@ use bevy_render::{
         TextureDescriptor, TextureDimension, TextureFormat, TextureSampleType, TextureUsages,
     },
     renderer::{RenderContext, RenderDevice, ViewQuery},
-    sync_component::SyncComponentPlugin,
+    sync_component::{SyncComponent, SyncComponentPlugin},
     sync_world::RenderEntity,
     texture::{CachedTexture, TextureCache},
     view::{
@@ -653,6 +653,16 @@ impl SpecializedRenderPipeline for DepthOfFieldPipeline {
     }
 }
 
+impl SyncComponent for DepthOfField {
+    type Out = (
+        DepthOfField,
+        DepthOfFieldUniform,
+        DepthOfFieldPipelines,
+        AuxiliaryDepthOfFieldTexture,
+        ViewDepthOfFieldBindGroupLayouts,
+    );
+}
+
 /// Extracts all [`DepthOfField`] components into the render world.
 fn extract_depth_of_field_settings(
     mut commands: Commands,
@@ -672,15 +682,8 @@ fn extract_depth_of_field_settings(
 
         // Depth of field is nonsensical without a perspective projection.
         let Projection::Perspective(ref perspective_projection) = *projection else {
-            // TODO: needs better strategy for cleaning up
-            entity_commands.remove::<(
-                DepthOfField,
-                DepthOfFieldUniform,
-                // components added in prepare systems (because `DepthOfFieldNode` does not query extracted components)
-                DepthOfFieldPipelines,
-                AuxiliaryDepthOfFieldTexture,
-                ViewDepthOfFieldBindGroupLayouts,
-            )>();
+            entity_commands.remove::<<DepthOfField as SyncComponent>::Out>();
+
             continue;
         };
 
diff --git a/crates/bevy_post_process/src/effect_stack/chromatic_aberration.rs b/crates/bevy_post_process/src/effect_stack/chromatic_aberration.rs
index d0c8f91f21549..4dc06716db368 100644
--- a/crates/bevy_post_process/src/effect_stack/chromatic_aberration.rs
+++ b/crates/bevy_post_process/src/effect_stack/chromatic_aberration.rs
@@ -9,7 +9,9 @@ use bevy_ecs::{
 };
 use bevy_image::Image;
 use bevy_reflect::{std_traits::ReflectDefault, Reflect};
-use bevy_render::{extract_component::ExtractComponent, render_resource::ShaderType};
+use bevy_render::{
+    extract_component::ExtractComponent, render_resource::ShaderType, sync_component::SyncComponent,
+};
 
 /// The raw RGBA data for the default chromatic aberration gradient.
 ///
@@ -83,13 +85,14 @@ impl Default for ChromaticAberration {
     }
 }
 
+impl SyncComponent for ChromaticAberration {
+    type Out = Self;
+}
+
 impl ExtractComponent for ChromaticAberration {
     type QueryData = Read<ChromaticAberration>;
-
     type QueryFilter = With<Camera>;
 
-    type Out = ChromaticAberration;
-
     fn extract_component(
         chromatic_aberration: QueryItem<'_, '_, Self::QueryData>,
     ) -> Option<Self::Out> {
diff --git a/crates/bevy_post_process/src/effect_stack/vignette.rs b/crates/bevy_post_process/src/effect_stack/vignette.rs
index cf96708b6c0df..45740fbd4bd68 100644
--- a/crates/bevy_post_process/src/effect_stack/vignette.rs
+++ b/crates/bevy_post_process/src/effect_stack/vignette.rs
@@ -8,7 +8,9 @@ use bevy_ecs::{
 };
 use bevy_math::{Vec2, Vec4};
 use bevy_reflect::{std_traits::ReflectDefault, Reflect};
-use bevy_render::{extract_component::ExtractComponent, render_resource::ShaderType};
+use bevy_render::{
+    extract_component::ExtractComponent, render_resource::ShaderType, sync_component::SyncComponent,
+};
 
 /// The default vignette intensity amount.
 const DEFAULT_VIGNETTE_INTENSITY: f32 = 1.0;
@@ -100,13 +102,14 @@ impl Default for Vignette {
     }
 }
 
+impl SyncComponent for Vignette {
+    type Out = Self;
+}
+
 impl ExtractComponent for Vignette {
     type QueryData = Read<Vignette>;
-
     type QueryFilter = With<Camera>;
 
-    type Out = Vignette;
-
     fn extract_component(vignette: QueryItem<'_, '_, Self::QueryData>) -> Option<Self::Out> {
         // Skip the postprocessing phase entirely if the intensity is zero.
         if vignette.intensity > 0.0 {
diff --git a/crates/bevy_post_process/src/motion_blur/mod.rs b/crates/bevy_post_process/src/motion_blur/mod.rs
index 167cf6298e3a0..b6c34815cb533 100644
--- a/crates/bevy_post_process/src/motion_blur/mod.rs
+++ b/crates/bevy_post_process/src/motion_blur/mod.rs
@@ -32,6 +32,7 @@ use bevy_render::{
         RenderPassDescriptor, ShaderType, SpecializedRenderPipelines,
     },
     renderer::{RenderContext, ViewQuery},
+    sync_component::SyncComponent,
     view::{Msaa, ViewTarget},
     Render, RenderApp, RenderStartup, RenderSystems,
 };
@@ -114,10 +115,13 @@ impl Default for MotionBlur {
     }
 }
 
+impl SyncComponent for MotionBlur {
+    type Out = MotionBlurUniform;
+}
+
 impl ExtractComponent for MotionBlur {
     type QueryData = &'static Self;
     type QueryFilter = With<Camera>;
-    type Out = MotionBlurUniform;
 
     fn extract_component(item: QueryItem<Self::QueryData>) -> Option<Self::Out> {
         Some(MotionBlurUniform {
diff --git a/crates/bevy_render/macros/src/extract_component.rs b/crates/bevy_render/macros/src/extract_component.rs
index ce06fcebd2132..8b57d0a47d646 100644
--- a/crates/bevy_render/macros/src/extract_component.rs
+++ b/crates/bevy_render/macros/src/extract_component.rs
@@ -39,11 +39,14 @@ pub fn derive_extract_component(input: TokenStream) -> TokenStream {
     };
 
     TokenStream::from(quote! {
+        impl #impl_generics #bevy_render_path::sync_component::SyncComponent for #struct_name #type_generics #where_clause {
+            type Out = Self;
+        }
+
         impl #impl_generics #bevy_render_path::extract_component::ExtractComponent for #struct_name #type_generics #where_clause {
             type QueryData = &'static Self;
 
             type QueryFilter = #filter;
-            type Out = Self;
 
             fn extract_component(item: #bevy_ecs_path::query::QueryItem<'_, '_, Self::QueryData>) -> Option<Self::Out> {
                 Some(item.clone())
diff --git a/crates/bevy_render/src/camera.rs b/crates/bevy_render/src/camera.rs
index 6b19e8daa85e4..26ed7a2492c27 100644
--- a/crates/bevy_render/src/camera.rs
+++ b/crates/bevy_render/src/camera.rs
@@ -4,6 +4,7 @@ use crate::{
     extract_resource::{ExtractResource, ExtractResourcePlugin},
     render_asset::RenderAssets,
     render_resource::TextureView,
+    sync_component::SyncComponent,
     sync_world::{RenderEntity, SyncToRenderWorld},
     texture::{GpuImage, ManualTextureViews},
     view::{
@@ -96,28 +97,40 @@ impl ExtractResource for ClearColor {
         source.clone()
     }
 }
+
+impl SyncComponent for CameraMainTextureUsages {
+    type Out = Self;
+}
+
 impl ExtractComponent for CameraMainTextureUsages {
     type QueryData = &'static Self;
     type QueryFilter = ();
-    type Out = Self;
 
     fn extract_component(item: QueryItem<Self::QueryData>) -> Option<Self::Out> {
         Some(*item)
     }
 }
+
+impl SyncComponent for Camera2d {
+    type Out = Self;
+}
+
 impl ExtractComponent for Camera2d {
     type QueryData = &'static Self;
     type QueryFilter = With<Camera>;
-    type Out = Self;
 
     fn extract_component(item: QueryItem<Self::QueryData>) -> Option<Self::Out> {
         Some(item.clone())
     }
 }
+
+impl SyncComponent for Camera3d {
+    type Out = Self;
+}
+
 impl ExtractComponent for Camera3d {
     type QueryData = &'static Self;
     type QueryFilter = With<Camera>;
-    type Out = Self;
 
     fn extract_component(item: QueryItem<Self::QueryData>) -> Option<Self::Out> {
         Some(item.clone())
diff --git a/crates/bevy_render/src/extract_component.rs b/crates/bevy_render/src/extract_component.rs
index ce656c357924e..57a307ba7df88 100644
--- a/crates/bevy_render/src/extract_component.rs
+++ b/crates/bevy_render/src/extract_component.rs
@@ -1,14 +1,13 @@
 use crate::{
     render_resource::{encase::internal::WriteInto, DynamicUniformBuffer, ShaderType},
     renderer::{RenderDevice, RenderQueue},
-    sync_component::SyncComponentPlugin,
+    sync_component::{SyncComponent, SyncComponentPlugin},
     sync_world::RenderEntity,
     Extract, ExtractSchedule, Render, RenderApp, RenderSystems,
 };
 use bevy_app::{App, Plugin};
 use bevy_camera::visibility::ViewVisibility;
 use bevy_ecs::{
-    bundle::NoBundleEffect,
     component::Component,
     prelude::*,
     query::{QueryFilter, QueryItem, ReadOnlyQueryData},
@@ -33,32 +32,21 @@ impl<C: Component> DynamicUniformIndex<C> {
 
 /// Describes how a component gets extracted for rendering.
 ///
-/// Therefore the component is transferred from the "app world" into the "render world"
-/// in the [`ExtractSchedule`] step.
-pub trait ExtractComponent: Component {
+/// Therefore the component is transferred from the "app world" into the "render
+/// world" in the [`ExtractSchedule`] step. This functionality is enabled by
+/// adding [`ExtractComponentPlugin`] with the component type.
+///
+/// The Out type is defined in [`SyncComponent`].
+///
+/// The marker type is only used as a way to bypass the orphan rules. To
+/// implement the trait for a foreign type you can use a local type as the
+/// marker, e.g. the type of the plugin that calls [`ExtractComponentPlugin`].
+pub trait ExtractComponent<Marker = ()>: SyncComponent {
     /// ECS [`ReadOnlyQueryData`] to fetch the components to extract.
     type QueryData: ReadOnlyQueryData;
     /// Filters the entities with additional constraints.
     type QueryFilter: QueryFilter;
 
-    /// The output from extraction.
-    ///
-    /// Returning `None` based on the queried item will remove the component from the entity in
-    /// the render world. This can be used, for example, to conditionally extract camera settings
-    /// in order to disable a rendering feature on the basis of those settings, without removing
-    /// the component from the entity in the main world.
-    ///
-    /// The output may be different from the queried component.
-    /// This can be useful for example if only a subset of the fields are useful
-    /// in the render world.
-    ///
-    /// `Out` has a [`Bundle`] trait bound instead of a [`Component`] trait bound in order to allow use cases
-    /// such as tuples of components as output.
-    type Out: Bundle<Effect: NoBundleEffect>;
-
-    // TODO: https://github.com/rust-lang/rust/issues/29661
-    // type Out: Component = Self;
-
     /// Defines how the component is transferred into the "render world".
     fn extract_component(item: QueryItem<'_, '_, Self::QueryData>) -> Option<Self::Out>;
 }
@@ -157,9 +145,12 @@ fn prepare_uniform_components<C>(
     commands.try_insert_batch(entities);
 }
 
-/// This plugin extracts the components into the render world for synced entities.
+/// This plugin extracts the components into the render world for synced
+/// entities. To do so, it sets up the [`ExtractSchedule`] step for the
+/// specified [`ExtractComponent`].
 ///
-/// To do so, it sets up the [`ExtractSchedule`] step for the specified [`ExtractComponent`].
+/// It also registers [`SyncComponentPlugin`] to ensure the extracted components
+/// are deleted if the main world components are removed.
 pub struct ExtractComponentPlugin<C, F = ()> {
     only_extract_visible: bool,
     marker: PhantomData<fn() -> (C, F)>,
@@ -183,22 +174,22 @@ impl<C, F> ExtractComponentPlugin<C, F> {
     }
 }
 
-impl<C: ExtractComponent> Plugin for ExtractComponentPlugin<C> {
+impl<C: ExtractComponent<Marker>, Marker: 'static> Plugin for ExtractComponentPlugin<C, Marker> {
     fn build(&self, app: &mut App) {
         app.add_plugins(SyncComponentPlugin::<C>::default());
 
         if let Some(render_app) = app.get_sub_app_mut(RenderApp) {
             if self.only_extract_visible {
-                render_app.add_systems(ExtractSchedule, extract_visible_components::<C>);
+                render_app.add_systems(ExtractSchedule, extract_visible_components::<C, Marker>);
             } else {
-                render_app.add_systems(ExtractSchedule, extract_components::<C>);
+                render_app.add_systems(ExtractSchedule, extract_components::<C, Marker>);
             }
         }
     }
 }
 
 /// This system extracts all components of the corresponding [`ExtractComponent`], for entities that are synced via [`crate::sync_world::SyncToRenderWorld`].
-fn extract_components<C: ExtractComponent>(
+fn extract_components<C: ExtractComponent<Marker>, Marker>(
     mut commands: Commands,
     mut previous_len: Local<usize>,
     query: Extract<Query<(RenderEntity, C::QueryData), C::QueryFilter>>,
@@ -216,7 +207,7 @@ fn extract_components<C: ExtractComponent>(
 }
 
 /// This system extracts all components of the corresponding [`ExtractComponent`], for entities that are visible and synced via [`crate::sync_world::SyncToRenderWorld`].
-fn extract_visible_components<C: ExtractComponent>(
+fn extract_visible_components<C: ExtractComponent<Marker>, Marker>(
     mut commands: Commands,
     mut previous_len: Local<usize>,
     query: Extract<Query<(RenderEntity, &ViewVisibility, C::QueryData), C::QueryFilter>>,
diff --git a/crates/bevy_render/src/extract_plugin.rs b/crates/bevy_render/src/extract_plugin.rs
index 55380c0eaf2ab..79db16580c45b 100644
--- a/crates/bevy_render/src/extract_plugin.rs
+++ b/crates/bevy_render/src/extract_plugin.rs
@@ -134,6 +134,7 @@ mod test {
     use crate::{
         extract_component::{ExtractComponent, ExtractComponentPlugin},
         extract_plugin::ExtractPlugin,
+        sync_component::SyncComponent,
         sync_world::MainEntity,
         Render, RenderApp,
     };
@@ -150,13 +151,15 @@ mod test {
     #[derive(Component, Clone, Debug)]
     struct RenderComponentNoExtract;
 
+    impl SyncComponent for RenderComponent {
+        type Out = (RenderComponent, RenderComponentExtra);
+    }
+
     impl ExtractComponent for RenderComponent {
         type QueryData = &'static Self;
 
         type QueryFilter = ();
 
-        type Out = (RenderComponent, RenderComponentExtra);
-
         fn extract_component(
             _item: bevy_ecs::query::QueryItem<'_, '_, Self::QueryData>,
         ) -> Option<Self::Out> {
@@ -243,8 +246,7 @@ mod test {
                         assert!(entity.1.is_none());
                         assert!(entity.2.is_none());
                         assert!(entity.3.is_some());
-                        // TODO: this is a bug
-                        // assert!(entity.4.is_some());
+                        assert!(entity.4.is_some());
                     },
                 )
                 .unwrap();
diff --git a/crates/bevy_render/src/sync_component.rs b/crates/bevy_render/src/sync_component.rs
index dd7eca1a2840c..0b5ddf68e87cd 100644
--- a/crates/bevy_render/src/sync_component.rs
+++ b/crates/bevy_render/src/sync_component.rs
@@ -1,7 +1,10 @@
 use core::marker::PhantomData;
 
 use bevy_app::{App, Plugin};
-use bevy_ecs::component::Component;
+use bevy_ecs::{
+    bundle::{Bundle, NoBundleEffect},
+    component::Component,
+};
 
 use crate::sync_world::{EntityRecord, PendingSyncEntity, SyncToRenderWorld};
 
@@ -14,21 +17,43 @@ use crate::sync_world::{EntityRecord, PendingSyncEntity, SyncToRenderWorld};
 /// It adds [`SyncToRenderWorld`] as a required component to make the [`SyncWorldPlugin`] aware of the component, and
 /// handles cleanup of the component in the render world when it is removed from an entity.
 ///
-/// # Warning
-/// When the component is removed from the main world entity, all components are removed from the entity in the render world.
-/// This is done in order to handle components with custom extraction logic and derived state.
-///
 /// [`ExtractComponentPlugin`]: crate::extract_component::ExtractComponentPlugin
 /// [`SyncWorldPlugin`]: crate::sync_world::SyncWorldPlugin
-pub struct SyncComponentPlugin<C: Component>(PhantomData<C>);
+pub struct SyncComponentPlugin<C, Marker = ()>(PhantomData<(C, Marker)>);
 
-impl<C: Component> Default for SyncComponentPlugin<C> {
+impl<C: SyncComponent<Marker>, Marker> Default for SyncComponentPlugin<C, Marker> {
     fn default() -> Self {
         Self(PhantomData)
     }
 }
 
-impl<C: Component> Plugin for SyncComponentPlugin<C> {
+/// Trait that links components from the main world with output components in
+/// the render world. It is used by [`SyncComponentPlugin`].
+///
+/// This trait is a subtrait of [`ExtractComponent`], which uses it to determine
+/// which components to extract.
+///
+/// The marker type is only used as a way to bypass the orphan rules. To
+/// implement the trait for a foreign type you can use a local type as the
+/// marker, e.g. the type of the plugin that calls [`SyncComponentPlugin`].
+///
+/// [`ExtractComponent`]: crate::extract_component::ExtractComponent
+pub trait SyncComponent<Marker = ()>: Component {
+    /// Describes what components should be removed from the render world if the
+    /// implementing component is removed.
+    ///
+    /// It is also used by the [`ExtractComponent`] trait to determine which
+    /// components are generated during extraction.
+    ///
+    /// [`ExtractComponent`]: crate::extract_component::ExtractComponent
+    type Out: Bundle<Effect: NoBundleEffect>;
+    // TODO: https://github.com/rust-lang/rust/issues/29661
+    // type Out: Component = Self;
+}
+
+impl<C: SyncComponent<Marker>, Marker: Send + Sync + 'static> Plugin
+    for SyncComponentPlugin<C, Marker>
+{
     fn build(&self, app: &mut App) {
         app.register_required_components::<C, SyncToRenderWorld>();
 
@@ -36,7 +61,12 @@ impl<C: Component> Plugin for SyncComponentPlugin<C> {
             .register_component_hooks::<C>()
             .on_remove(|mut world, context| {
                 let mut pending = world.resource_mut::<PendingSyncEntity>();
-                pending.push(EntityRecord::ComponentRemoved(context.entity));
+                pending.push(EntityRecord::ComponentRemoved(
+                    context.entity,
+                    |mut entity| {
+                        entity.remove::<C::Out>();
+                    },
+                ));
             });
     }
 }
diff --git a/crates/bevy_render/src/sync_world.rs b/crates/bevy_render/src/sync_world.rs
index c59cde47bf71f..1a89da5d34e37 100644
--- a/crates/bevy_render/src/sync_world.rs
+++ b/crates/bevy_render/src/sync_world.rs
@@ -9,7 +9,7 @@ use bevy_ecs::{
     reflect::ReflectComponent,
     resource::Resource,
     system::{Local, Query, ResMut, SystemState},
-    world::{Mut, World},
+    world::{EntityWorldMut, Mut, World},
 };
 use bevy_platform::collections::{HashMap, HashSet};
 use bevy_reflect::{std_traits::ReflectDefault, Reflect};
@@ -202,7 +202,7 @@ pub(crate) enum EntityRecord {
     Removed(RenderEntity),
     /// When a component is removed from an entity, notify the render world so that the corresponding component can be
     /// removed. This contains the main world entity.
-    ComponentRemoved(Entity),
+    ComponentRemoved(Entity, fn(EntityWorldMut<'_>)),
 }
 
 // Entity Record in MainWorld pending to Sync
@@ -235,17 +235,12 @@ pub(crate) fn entity_sync_system(main_world: &mut World, render_world: &mut Worl
                         ec.despawn();
                     };
                 }
-                EntityRecord::ComponentRemoved(main_entity) => {
-                    let Some(mut render_entity) = world.get_mut::<RenderEntity>(main_entity) else {
+                EntityRecord::ComponentRemoved(main_entity, removal_function) => {
+                    let Some(render_entity) = world.get::<RenderEntity>(main_entity) else {
                         continue;
                     };
                     if let Ok(render_world_entity) = render_world.get_entity_mut(render_entity.id()) {
-                        // In order to handle components that extract to derived components, we clear the entity
-                        // and let the extraction system re-add the components.
-                        render_world_entity.despawn();
-
-                        let id = render_world.spawn(MainEntity(main_entity)).id();
-                        render_entity.0 = id;
+                        removal_function(render_world_entity);
                     }
                 },
             }
diff --git a/examples/2d/mesh2d_manual.rs b/examples/2d/mesh2d_manual.rs
index c080d77c7deac..484e67cbad518 100644
--- a/examples/2d/mesh2d_manual.rs
+++ b/examples/2d/mesh2d_manual.rs
@@ -26,7 +26,7 @@ use bevy::{
             SpecializedRenderPipelines, StencilFaceState, StencilState, TextureFormat,
             VertexFormat, VertexState, VertexStepMode,
         },
-        sync_component::SyncComponentPlugin,
+        sync_component::{SyncComponent, SyncComponentPlugin},
         sync_world::{MainEntityHashMap, RenderEntity},
         view::{ExtractedView, RenderVisibleEntities, ViewTarget},
         Extract, Render, RenderApp, RenderStartup, RenderSystems,
@@ -125,6 +125,10 @@ fn star(
 #[derive(Component, Default)]
 pub struct ColoredMesh2d;
 
+impl SyncComponent for ColoredMesh2d {
+    type Out = Self;
+}
+
 /// Custom pipeline for 2d meshes with vertex colors
 #[derive(Resource)]
 pub struct ColoredMesh2dPipeline {
diff --git a/examples/shader_advanced/custom_shader_instancing.rs b/examples/shader_advanced/custom_shader_instancing.rs
index 4f5943bf2f188..cee6947bfc034 100644
--- a/examples/shader_advanced/custom_shader_instancing.rs
+++ b/examples/shader_advanced/custom_shader_instancing.rs
@@ -30,6 +30,7 @@ use bevy::{
         },
         render_resource::*,
         renderer::RenderDevice,
+        sync_component::SyncComponent,
         sync_world::MainEntity,
         view::{ExtractedView, NoIndirectDrawing},
         Render, RenderApp, RenderStartup, RenderSystems,
@@ -84,10 +85,13 @@ fn setup(mut commands: Commands, mut meshes: ResMut<Assets<Mesh>>) {
 #[derive(Component, Deref)]
 struct InstanceMaterialData(Vec<InstanceData>);
 
+impl SyncComponent for InstanceMaterialData {
+    type Out = Self;
+}
+
 impl ExtractComponent for InstanceMaterialData {
     type QueryData = &'static InstanceMaterialData;
     type QueryFilter = ();
-    type Out = Self;
 
     fn extract_component(item: QueryItem<'_, '_, Self::QueryData>) -> Option<Self> {
         Some(InstanceMaterialData(item.0.clone()))
diff --git a/release-content/migration-guides/extract_refactor.md b/release-content/migration-guides/extract_refactor.md
new file mode 100644
index 0000000000000..bc33d5bb9bf96
--- /dev/null
+++ b/release-content/migration-guides/extract_refactor.md
@@ -0,0 +1,41 @@
+---
+title: "`ExtractComponent` refactor"
+pull_requests: [22766]
+---
+
+The `Out` type from `ExtractComponent` has been split into a separate `SyncComponent` trait.
+
+Both traits have also gotten an optional marker type that can be used to bypass orphan rules, see the docs for details.
+
+```rust,ignore
+impl ExtractComponent for MyComponent {
+    type QueryData = ();
+    type QueryFilter = ();
+    type Out = Self;
+
+    fn extract_component(
+        item: QueryItem<'_, '_, Self::QueryData>,
+    ) -> Option<Self::Out> {
+        Some(*item)
+    }
+}
+```
+
+After:
+
+```rust,ignore
+impl SyncComponent for MyComponent {
+    type Out = Self;
+}
+
+impl ExtractComponent for MyComponent {
+    type QueryData = ();
+    type QueryFilter = ();
+
+    fn extract_component(
+        item: QueryItem<'_, '_, Self::QueryData>,
+    ) -> Option<Self::Out> {
+        Some(*item)
+    }
+}
+```
