diff --git a/crates/bevy_ecs/src/world/deferred_world.rs b/crates/bevy_ecs/src/world/deferred_world.rs
index e6fd07d0fad93..54ab26e6057d6 100644
--- a/crates/bevy_ecs/src/world/deferred_world.rs
+++ b/crates/bevy_ecs/src/world/deferred_world.rs
@@ -4,7 +4,7 @@ use bevy_utils::prelude::DebugName;
 
 use crate::{
     archetype::Archetype,
-    change_detection::{MaybeLocation, MutUntyped},
+    change_detection::{MaybeLocation, MutUntyped, Tick},
     component::{ComponentId, Mutable},
     entity::Entity,
     event::{EntityComponentsTrigger, Event, EventKey, Trigger},
@@ -823,4 +823,10 @@ impl<'w> DeferredWorld<'w> {
     pub fn as_unsafe_world_cell(&mut self) -> UnsafeWorldCell<'_> {
         self.world
     }
+
+    /// Gets the current change tick of [`DeferredWorld`].
+    #[inline]
+    pub fn change_tick(&mut self) -> Tick {
+        self.world.change_tick()
+    }
 }
