diff --git a/crates/bevy_ecs/src/component/register.rs b/crates/bevy_ecs/src/component/register.rs
index d38f749255c74..76c10810b8061 100644
--- a/crates/bevy_ecs/src/component/register.rs
+++ b/crates/bevy_ecs/src/component/register.rs
@@ -5,6 +5,7 @@ use core::any::Any;
 use core::{any::TypeId, fmt::Debug, ops::Deref};
 
 use crate::component::{enforce_no_required_components_recursion, RequiredComponentsRegistrator};
+use crate::lifecycle::ComponentHooks;
 use crate::{
     component::{
         Component, ComponentDescriptor, ComponentId, Components, RequiredComponents, StorageType,
@@ -161,13 +162,22 @@ impl<'w> ComponentsRegistrator<'w> {
     /// * [`ComponentsRegistrator::register_component_with_descriptor()`]
     #[inline]
     pub fn register_component<T: Component>(&mut self) -> ComponentId {
-        self.register_component_checked::<T>()
+        self.register_component_checked(
+            TypeId::of::<T>(),
+            ComponentDescriptor::new::<T>,
+            T::register_required_components,
+            ComponentHooks::update_from_component::<T>,
+        )
     }
 
-    /// Same as [`Self::register_component_unchecked`] but keeps a checks for safety.
-    #[inline]
-    pub(super) fn register_component_checked<T: Component>(&mut self) -> ComponentId {
-        let type_id = TypeId::of::<T>();
+    // This exists to cut down on monomorphized code in register_component, which reduces compile times and binary sizes.
+    fn register_component_checked(
+        &mut self,
+        type_id: TypeId,
+        descriptor: fn() -> ComponentDescriptor,
+        register_required_components: fn(ComponentId, &mut RequiredComponentsRegistrator),
+        update_from_component: fn(&mut ComponentHooks) -> &mut ComponentHooks,
+    ) -> ComponentId {
         if let Some(&id) = self.indices.get(&type_id) {
             enforce_no_required_components_recursion(self, &self.recursion_check_stack, id);
             return id;
@@ -189,7 +199,13 @@ impl<'w> ComponentsRegistrator<'w> {
         let id = self.ids.next_mut();
         // SAFETY: The component is not currently registered, and the id is fresh.
         unsafe {
-            self.register_component_unchecked::<T>(id);
+            self.register_component_unchecked(
+                type_id,
+                id,
+                descriptor(),
+                register_required_components,
+                update_from_component,
+            );
         }
         id
     }
@@ -197,14 +213,19 @@ impl<'w> ComponentsRegistrator<'w> {
     /// # Safety
     ///
     /// Neither this component, nor its id may be registered or queued. This must be a new registration.
-    #[inline]
-    unsafe fn register_component_unchecked<T: Component>(&mut self, id: ComponentId) {
+    // This was written in a type-erased way to cut down on monomorphized code in register_component, which reduces compile times and binary sizes.
+    unsafe fn register_component_unchecked(
+        &mut self,
+        type_id: TypeId,
+        id: ComponentId,
+        descriptor: ComponentDescriptor,
+        register_required_components: fn(ComponentId, &mut RequiredComponentsRegistrator),
+        update_from_component: fn(&mut ComponentHooks) -> &mut ComponentHooks,
+    ) {
         // SAFETY: ensured by caller.
         unsafe {
-            self.components
-                .register_component_inner(id, ComponentDescriptor::new::<T>());
+            self.components.register_component_inner(id, descriptor);
         }
-        let type_id = TypeId::of::<T>();
         let prev = self.components.indices.insert(type_id, id);
         debug_assert!(prev.is_none());
 
@@ -213,7 +234,7 @@ impl<'w> ComponentsRegistrator<'w> {
         // SAFETY: `required_components` is empty
         let mut required_components_registrator =
             unsafe { RequiredComponentsRegistrator::new(self, &mut required_components) };
-        T::register_required_components(id, &mut required_components_registrator);
+        register_required_components(id, &mut required_components_registrator);
         // SAFETY:
         // - `id` was just registered in `self`
         // - RequiredComponentsRegistrator guarantees that only components from `self` are included in `required_components`;
@@ -235,7 +256,7 @@ impl<'w> ComponentsRegistrator<'w> {
                 .debug_checked_unwrap()
         };
 
-        info.hooks.update_from_component::<T>();
+        update_from_component(&mut info.hooks);
 
         info.required_components = required_components;
     }
@@ -535,7 +556,13 @@ impl<'w> ComponentsQueuedRegistrator<'w> {
                         // SAFETY: We just checked that this is not currently registered or queued, and if it was registered since, this would have been dropped from the queue.
                         #[expect(unused_unsafe, reason = "More precise to specify.")]
                         unsafe {
-                            registrator.register_component_unchecked::<T>(id);
+                            registrator.register_component_unchecked(
+                                TypeId::of::<T>(),
+                                id,
+                                ComponentDescriptor::new::<T>(),
+                                T::register_required_components,
+                                ComponentHooks::update_from_component::<T>,
+                            );
                         }
                     },
                 )
