diff --git a/crates/bevy_internal/Cargo.toml b/crates/bevy_internal/Cargo.toml
index a4f558de5091d..aeaa24ca9e8f0 100644
--- a/crates/bevy_internal/Cargo.toml
+++ b/crates/bevy_internal/Cargo.toml
@@ -20,6 +20,7 @@ trace = [
   "bevy_pbr?/trace",
   "bevy_render?/trace",
   "bevy_winit?/trace",
+  "bevy_post_process?/trace",
 ]
 trace_chrome = ["bevy_log/tracing-chrome"]
 trace_tracy = ["bevy_render?/tracing-tracy", "bevy_log/tracing-tracy"]
@@ -191,6 +192,7 @@ webgl = [
   "bevy_gizmos_render?/webgl",
   "bevy_sprite_render?/webgl",
   "bevy_dev_tools?/webgl",
+  "bevy_post_process?/webgl",
   "bevy_feathers?/webgl",
 ]
 
@@ -202,6 +204,7 @@ webgpu = [
   "bevy_gizmos_render?/webgpu",
   "bevy_sprite_render?/webgpu",
   "bevy_dev_tools?/webgpu",
+  "bevy_post_process?/webgpu",
   "bevy_feathers?/webgpu",
 ]
 
diff --git a/crates/bevy_post_process/src/bloom/mod.rs b/crates/bevy_post_process/src/bloom/mod.rs
index 345b6b5d683dc..42d6eacbce29a 100644
--- a/crates/bevy_post_process/src/bloom/mod.rs
+++ b/crates/bevy_post_process/src/bloom/mod.rs
@@ -378,7 +378,14 @@ fn prepare_bloom_textures(
 
 #[derive(Component)]
 pub struct BloomBindGroups {
+    #[cfg(any(
+        not(feature = "webgl"),
+        not(target_arch = "wasm32"),
+        feature = "webgpu"
+    ))]
     cache_key: (TextureId, BufferId),
+    #[cfg(all(feature = "webgl", target_arch = "wasm32", not(feature = "webgpu")))]
+    cache_key: (Vec<TextureId>, BufferId),
     downsampling_bind_groups: Box<[BindGroup]>,
     upsampling_bind_groups: Box<[BindGroup]>,
     sampler: Sampler,
@@ -396,12 +403,27 @@ fn prepare_bloom_bind_groups(
     let sampler = &downsampling_pipeline.sampler;
 
     for (entity, bloom_texture, bloom_bind_groups) in &views {
+        #[cfg(any(
+            not(feature = "webgl"),
+            not(target_arch = "wasm32"),
+            feature = "webgpu"
+        ))]
+        let cache_key = (
+            bloom_texture.texture.texture.id(),
+            uniforms.buffer().unwrap().id(),
+        );
+        #[cfg(all(feature = "webgl", target_arch = "wasm32", not(feature = "webgpu")))]
+        let cache_key = (
+            bloom_texture
+                .texture
+                .iter()
+                .map(|tex| tex.texture.id())
+                .collect(),
+            uniforms.buffer().unwrap().id(),
+        );
+
         if let Some(b) = bloom_bind_groups
-            && b.cache_key
-                == (
-                    bloom_texture.texture.texture.id(),
-                    uniforms.buffer().unwrap().id(),
-                )
+            && b.cache_key == cache_key
         {
             continue;
         }
@@ -435,10 +457,7 @@ fn prepare_bloom_bind_groups(
         }
 
         commands.entity(entity).insert(BloomBindGroups {
-            cache_key: (
-                bloom_texture.texture.texture.id(),
-                uniforms.buffer().unwrap().id(),
-            ),
+            cache_key,
             downsampling_bind_groups: downsampling_bind_groups.into_boxed_slice(),
             upsampling_bind_groups: upsampling_bind_groups.into_boxed_slice(),
             sampler: sampler.clone(),
