diff --git a/examples/3d/solari.rs b/examples/3d/solari.rs
index de1bcb67ae4a7..5139f9dcd3445 100644
--- a/examples/3d/solari.rs
+++ b/examples/3d/solari.rs
@@ -211,7 +211,7 @@ fn setup_many_lights(
 
     let mut plane_mesh = Plane3d::default()
         .mesh()
-        .size(40.0, 40.0)
+        .size(400.0, 400.0)
         .build()
         .with_generated_tangents()
         .unwrap();
@@ -230,7 +230,7 @@ fn setup_many_lights(
             .unwrap(),
     );
     let sphere_mesh = meshes.add(
-        Sphere::default()
+        Sphere::new(1.0)
             .mesh()
             .build()
             .with_generated_tangents()
@@ -258,7 +258,7 @@ fn setup_many_lights(
         ))
         .insert_if(Mesh3d(plane_mesh), || args.pathtracer != Some(true));
 
-    for _ in 0..200 {
+    for _ in 0..8000 {
         commands
             .spawn((
                 RaytracingMesh3d(cube_mesh.clone()),
@@ -274,38 +274,40 @@ fn setup_many_lights(
                         z: rng.random_range(0.2..=2.0),
                     })
                     .with_translation(Vec3::new(
-                        rng.random_range(-18.0..=18.0),
+                        rng.random_range(-180.0..=180.0),
                         0.2,
-                        rng.random_range(-18.0..=18.0),
+                        rng.random_range(-180.0..=180.0),
                     )),
             ))
             .insert_if(Mesh3d(cube_mesh.clone()), || args.pathtracer != Some(true));
     }
 
-    for _ in 0..100 {
-        commands
-            .spawn((
-                RaytracingMesh3d(sphere_mesh.clone()),
-                MeshMaterial3d(
-                    materials.add(StandardMaterial {
-                        emissive: Color::linear_rgb(
-                            rng.random::<f32>() * 20000.0,
-                            rng.random::<f32>() * 20000.0,
-                            rng.random::<f32>() * 20000.0,
-                        )
-                        .into(),
-                        ..default()
-                    }),
-                ),
-                Transform::default().with_translation(Vec3::new(
-                    rng.random_range(-18.0..=18.0),
-                    rng.random_range(6.0..=9.0),
-                    rng.random_range(-18.0..=18.0),
-                )),
-            ))
-            .insert_if(Mesh3d(sphere_mesh.clone()), || {
-                args.pathtracer != Some(true)
-            });
+    for x in -10..=10 {
+        for y in -10..=10 {
+            commands
+                .spawn((
+                    RaytracingMesh3d(sphere_mesh.clone()),
+                    MeshMaterial3d(
+                        materials.add(StandardMaterial {
+                            emissive: Color::linear_rgb(
+                                rng.random::<f32>() * 60000.0,
+                                rng.random::<f32>() * 60000.0,
+                                rng.random::<f32>() * 60000.0,
+                            )
+                            .into(),
+                            ..default()
+                        }),
+                    ),
+                    Transform::default().with_translation(Vec3::new(
+                        (x * 20) as f32,
+                        7.0,
+                        (y * 20) as f32,
+                    )),
+                ))
+                .insert_if(Mesh3d(sphere_mesh.clone()), || {
+                    args.pathtracer != Some(true)
+                });
+        }
     }
 
     let mut camera = commands.spawn((
@@ -319,8 +321,8 @@ fn setup_many_lights(
             run_speed: 10.0,
             ..Default::default()
         },
-        Transform::from_translation(Vec3::new(0.0919233, 7.5015035, 28.449198)).with_rotation(
-            Quat::from_xyzw(-0.18394549, 0.0019948867, 0.0003733214, 0.98293436),
+        Transform::from_translation(Vec3::new(6.11329, 166.74896, 451.8226)).with_rotation(
+            Quat::from_xyzw(-0.183938, 0.009093744, 0.0017017953, 0.9828943),
         ),
         // Msaa::Off and CameraMainTextureUsages with STORAGE_BINDING are required for Solari
         CameraMainTextureUsages::default().with(TextureUsages::STORAGE_BINDING),
