diff --git a/examples/testbed/2d.rs b/examples/testbed/2d.rs
index 573c42427bd2b..9f1ca518b26b1 100644
--- a/examples/testbed/2d.rs
+++ b/examples/testbed/2d.rs
@@ -395,6 +395,14 @@ mod gizmos {
             )
             .resolution(64);
 
+        gizmos.text_2d(
+            Isometry2d::from_translation(Vec2::new(-200.0, 0.0)),
+            "text_2d gizmo",
+            15.,
+            Vec2 { x: 0., y: 0. },
+            Color::WHITE,
+        );
+
         // 2d grids with all variations of outer edges on or off
         for i in 0..4 {
             let x = 200.0 * (1.0 + (i % 2) as f32);
diff --git a/examples/testbed/3d.rs b/examples/testbed/3d.rs
index c1f1152057af6..de8b0c6e28a09 100644
--- a/examples/testbed/3d.rs
+++ b/examples/testbed/3d.rs
@@ -358,6 +358,14 @@ mod gizmos {
             .sphere(Isometry3d::from_translation(Vec3::X * -3.5), 0.75, GREEN)
             .resolution(30_000 / 3);
 
+        gizmos.text(
+            Isometry3d::from_translation(Vec3::Y * 1.5),
+            "text gizmo",
+            0.3,
+            Vec2 { x: 0., y: 0. },
+            Color::WHITE,
+        );
+
         // 3d grids with all variations of outer edges on or off
         for i in 0..8 {
             let x = 1.5 * (i % 4) as f32;
