diff --git a/crates/bevy_text/src/text.rs b/crates/bevy_text/src/text.rs
index 17e9a2f86c47e..3b7f172fb007b 100644
--- a/crates/bevy_text/src/text.rs
+++ b/crates/bevy_text/src/text.rs
@@ -410,6 +410,11 @@ impl TextFont {
         Self::default().with_font_size(font_size)
     }
 
+    /// Returns a new [`TextFont`] with the specified font weight
+    pub fn from_font_weight(weight: impl Into<FontWeight>) -> Self {
+        Self::default().with_font_weight(weight)
+    }
+
     /// Returns this [`TextFont`] with the specified font face handle.
     pub fn with_font(mut self, font: Handle<Font>) -> Self {
         self.font = FontSource::Handle(font);
@@ -433,6 +438,12 @@ impl TextFont {
         self.font_smoothing = font_smoothing;
         self
     }
+
+    /// Returns this [`TextFont`] with the specified [`FontWeight`].
+    pub fn with_font_weight(mut self, weight: impl Into<FontWeight>) -> Self {
+        self.weight = weight.into();
+        self
+    }
 }
 
 impl<T: Into<FontSource>> From<T> for TextFont {
