diff --git a/crates/bevy_anti_alias/Cargo.toml b/crates/bevy_anti_alias/Cargo.toml
index 25a3e35b43ac6..4277b4b1e2d51 100644
--- a/crates/bevy_anti_alias/Cargo.toml
+++ b/crates/bevy_anti_alias/Cargo.toml
@@ -35,7 +35,7 @@ bevy_post_process = { path = "../bevy_post_process", version = "0.19.0-dev" }
 
 # other
 tracing = { version = "0.1", default-features = false, features = ["std"] }
-dlss_wgpu = { version = "3", optional = true }
+dlss_wgpu = { version = "4.0.0-dev", optional = true }
 uuid = { version = "1", optional = true }
 
 [lints]
diff --git a/crates/bevy_anti_alias/src/dlss/mod.rs b/crates/bevy_anti_alias/src/dlss/mod.rs
index 6d504795182aa..3723ee752009f 100644
--- a/crates/bevy_anti_alias/src/dlss/mod.rs
+++ b/crates/bevy_anti_alias/src/dlss/mod.rs
@@ -265,11 +265,11 @@ impl DlssFeature for DlssSuperResolutionFeature {
     type Context = DlssSuperResolution;
 
     fn upscaled_resolution(context: &Self::Context) -> UVec2 {
-        context.upscaled_resolution()
+        context.upscaled_resolution().into()
     }
 
     fn render_resolution(context: &Self::Context) -> UVec2 {
-        context.render_resolution()
+        context.render_resolution().into()
     }
 
     fn suggested_jitter(
@@ -277,11 +277,13 @@ impl DlssFeature for DlssSuperResolutionFeature {
         frame_number: u32,
         render_resolution: UVec2,
     ) -> Vec2 {
-        context.suggested_jitter(frame_number, render_resolution)
+        context
+            .suggested_jitter(frame_number, render_resolution.to_array())
+            .into()
     }
 
     fn suggested_mip_bias(context: &Self::Context, render_resolution: UVec2) -> f32 {
-        context.suggested_mip_bias(render_resolution)
+        context.suggested_mip_bias(render_resolution.to_array())
     }
 
     fn new_context(
@@ -293,7 +295,7 @@ impl DlssFeature for DlssSuperResolutionFeature {
         queue: &RenderQueue,
     ) -> Result<Self::Context, dlss_wgpu::DlssError> {
         DlssSuperResolution::new(
-            upscaled_resolution,
+            upscaled_resolution.to_array(),
             perf_quality_mode,
             feature_flags,
             sdk,
@@ -313,11 +315,11 @@ impl DlssFeature for DlssRayReconstructionFeature {
     type Context = DlssRayReconstruction;
 
     fn upscaled_resolution(context: &Self::Context) -> UVec2 {
-        context.upscaled_resolution()
+        context.upscaled_resolution().into()
     }
 
     fn render_resolution(context: &Self::Context) -> UVec2 {
-        context.render_resolution()
+        context.render_resolution().into()
     }
 
     fn suggested_jitter(
@@ -325,11 +327,13 @@ impl DlssFeature for DlssRayReconstructionFeature {
         frame_number: u32,
         render_resolution: UVec2,
     ) -> Vec2 {
-        context.suggested_jitter(frame_number, render_resolution)
+        context
+            .suggested_jitter(frame_number, render_resolution.to_array())
+            .into()
     }
 
     fn suggested_mip_bias(context: &Self::Context, render_resolution: UVec2) -> f32 {
-        context.suggested_mip_bias(render_resolution)
+        context.suggested_mip_bias(render_resolution.to_array())
     }
 
     fn new_context(
@@ -341,7 +345,7 @@ impl DlssFeature for DlssRayReconstructionFeature {
         queue: &RenderQueue,
     ) -> Result<Self::Context, dlss_wgpu::DlssError> {
         DlssRayReconstruction::new(
-            upscaled_resolution,
+            upscaled_resolution.to_array(),
             perf_quality_mode,
             feature_flags,
             DlssRayReconstructionRoughnessMode::Packed,
diff --git a/crates/bevy_anti_alias/src/dlss/node.rs b/crates/bevy_anti_alias/src/dlss/node.rs
index aa356e37b3a51..5078e06e3743a 100644
--- a/crates/bevy_anti_alias/src/dlss/node.rs
+++ b/crates/bevy_anti_alias/src/dlss/node.rs
@@ -50,9 +50,9 @@ pub fn dlss_super_resolution(
         bias: None,                                       // TODO
         dlss_output: &view_target.destination,
         reset: dlss.reset,
-        jitter_offset: -temporal_jitter.offset,
-        partial_texture_size: Some(render_resolution),
-        motion_vector_scale: Some(-render_resolution.as_vec2()),
+        jitter_offset: (-temporal_jitter.offset).to_array(),
+        partial_texture_size: Some(render_resolution.to_array()),
+        motion_vector_scale: Some((-render_resolution.as_vec2()).to_array()),
     };
 
     let diagnostics = ctx.diagnostic_recorder();
@@ -117,9 +117,9 @@ pub fn dlss_ray_reconstruction(
         bias: None,                                     // TODO
         dlss_output: &view_target.destination,
         reset: dlss.reset,
-        jitter_offset: -temporal_jitter.offset,
-        partial_texture_size: Some(render_resolution),
-        motion_vector_scale: Some(-render_resolution.as_vec2()),
+        jitter_offset: (-temporal_jitter.offset).to_array(),
+        partial_texture_size: Some(render_resolution.to_array()),
+        motion_vector_scale: Some((-render_resolution.as_vec2()).to_array()),
     };
 
     let diagnostics = ctx.diagnostic_recorder();
