diff --git a/examples/stress_tests/many_cubes.rs b/examples/stress_tests/many_cubes.rs
index 48694ccb02365..a1d95acf19532 100644
--- a/examples/stress_tests/many_cubes.rs
+++ b/examples/stress_tests/many_cubes.rs
@@ -547,9 +547,9 @@ fn rotate_cubes(
     mut query: Query<&mut Transform, (With<Mesh3d>, Without<NotShadowCaster>)>,
     time: Res<Time>,
 ) {
-    for mut transform in query.iter_mut() {
+    query.par_iter_mut().for_each(|mut transform| {
         transform.rotate_y(10.0 * time.delta_secs());
-    }
+    });
 }
 
 #[inline]
