diff --git a/crates/bevy_ecs/src/world/mod.rs b/crates/bevy_ecs/src/world/mod.rs
index ff5081cddd1ae..a4f49c7460a35 100644
--- a/crates/bevy_ecs/src/world/mod.rs
+++ b/crates/bevy_ecs/src/world/mod.rs
@@ -75,7 +75,7 @@ use unsafe_world_cell::UnsafeWorldCell;
 /// and their associated metadata.
 ///
 /// Each [`Entity`] has a set of unique components, based on their type.
-/// Entity components can be created, updated, removed, and queried using a given
+/// Entity components can be created, updated, removed, and queried using a given [`World`].
 ///
 /// For complex access patterns involving [`SystemParam`](crate::system::SystemParam),
 /// consider using [`SystemState`](crate::system::SystemState).
