diff --git a/crates/bevy_pbr/src/prepass/mod.rs b/crates/bevy_pbr/src/prepass/mod.rs
index dcffe3d295377..9b5c9e99b3583 100644
--- a/crates/bevy_pbr/src/prepass/mod.rs
+++ b/crates/bevy_pbr/src/prepass/mod.rs
@@ -897,7 +897,7 @@ pub fn specialize_prepass_material_meshes(
             let Some(material) = render_materials.get(material_instance.asset_id) else {
                 continue;
             };
-            if !material.properties.prepass_enabled && !material.properties.shadows_enabled {
+            if !material.properties.prepass_enabled {
                 // If the material was previously specialized for prepass, remove it
                 view_specialized_material_pipeline_cache.remove(visible_entity);
                 continue;
