diff --git a/Cargo.toml b/Cargo.toml
index 9e1af044b7401..0c023fb96b200 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1186,17 +1186,6 @@ description = "Showcases different blend modes"
 category = "3D Rendering"
 wasm = true
 
-[[example]]
-name = "manual_material"
-path = "examples/3d/manual_material.rs"
-doc-scrape-examples = true
-
-[package.metadata.example.manual_material]
-name = "Manual Material Implementation"
-description = "Demonstrates how to implement a material manually using the mid-level render APIs"
-category = "3D Rendering"
-wasm = true
-
 [[example]]
 name = "lighting"
 path = "examples/3d/lighting.rs"
@@ -3091,6 +3080,17 @@ description = "Shows how to make a complete render phase"
 category = "Shaders"
 wasm = true
 
+[[example]]
+name = "manual_material"
+path = "examples/shader_advanced/manual_material.rs"
+doc-scrape-examples = true
+
+[package.metadata.example.manual_material]
+name = "Manual Material Implementation"
+description = "Demonstrates how to implement a material manually using the mid-level render APIs"
+category = "Shader Advanced"
+wasm = true
+
 [[example]]
 name = "texture_binding_array"
 path = "examples/shader_advanced/texture_binding_array.rs"
@@ -3102,7 +3102,6 @@ description = "A shader that shows how to bind and sample multiple textures as a
 category = "Shaders"
 wasm = false
 
-
 [[example]]
 name = "specialized_mesh_pipeline"
 path = "examples/shader_advanced/specialized_mesh_pipeline.rs"
diff --git a/examples/README.md b/examples/README.md
index c83abdb3dd1de..23e5e8ae19829 100644
--- a/examples/README.md
+++ b/examples/README.md
@@ -60,6 +60,7 @@ git checkout v0.4.0
   - [Reflection](#reflection)
   - [Remote Protocol](#remote-protocol)
   - [Scene](#scene)
+  - [Shader Advanced](#shader-advanced)
   - [Shaders](#shaders)
   - [State](#state)
   - [Stress Tests](#stress-tests)
@@ -164,7 +165,6 @@ Example | Description
 [Lighting](../examples/3d/lighting.rs) | Illustrates various lighting options in a simple scene
 [Lightmaps](../examples/3d/lightmaps.rs) | Rendering a scene with baked lightmaps
 [Lines](../examples/3d/lines.rs) | Create a custom material to draw 3d lines
-[Manual Material Implementation](../examples/3d/manual_material.rs) | Demonstrates how to implement a material manually using the mid-level render APIs
 [Mesh Ray Cast](../examples/3d/mesh_ray_cast.rs) | Demonstrates ray casting with the `MeshRayCast` system parameter
 [Meshlet](../examples/3d/meshlet.rs) | Meshlet rendering for dense high-poly scenes (experimental)
 [Mixed lighting](../examples/3d/mixed_lighting.rs) | Demonstrates how to combine baked and dynamic lighting
@@ -445,6 +445,12 @@ Example | Description
 --- | ---
 [Scene](../examples/scene/scene.rs) | Demonstrates loading from and saving scenes to files
 
+### Shader Advanced
+
+Example | Description
+--- | ---
+[Manual Material Implementation](../examples/shader_advanced/manual_material.rs) | Demonstrates how to implement a material manually using the mid-level render APIs
+
 ### Shaders
 
 These examples demonstrate how to implement different shaders in user code.
diff --git a/examples/3d/manual_material.rs b/examples/shader_advanced/manual_material.rs
similarity index 100%
rename from examples/3d/manual_material.rs
rename to examples/shader_advanced/manual_material.rs
