diff --git a/examples/gltf/edit_material_on_gltf.rs b/examples/gltf/edit_material_on_gltf.rs
index 9379a004b83e0..893a20e5813ad 100644
--- a/examples/gltf/edit_material_on_gltf.rs
+++ b/examples/gltf/edit_material_on_gltf.rs
@@ -14,7 +14,7 @@ fn main() {
 }
 
 /// This is added to a [`SceneRoot`] and will cause the [`StandardMaterial::base_color`]
-/// of all materials to be overwritten
+/// of materials with [`GltfMaterialName`] equal to `LeatherPartsMat`.
 #[derive(Component)]
 struct ColorOverride(Color);
 
@@ -48,6 +48,10 @@ fn setup_scene(mut commands: Commands, asset_server: Res<AssetServer>) {
     ));
 }
 
+/// On [`SceneInstanceReady`], iterates over all descendants of the scene
+/// and modifies the tint of the material for the materials named `LeatherPartsMat`.
+///
+/// If the [`SceneRoot`] does not have a [`ColorOverride`], it is skipped.
 fn change_material(
     scene_ready: On<SceneInstanceReady>,
     mut commands: Commands,
@@ -57,6 +61,13 @@ fn change_material(
     mut asset_materials: ResMut<Assets<StandardMaterial>>,
 ) {
     info!("processing Scene Entity: {}", scene_ready.entity);
+
+    // Get the `ColorOverride` of the entity, if it does not have a color override, return
+    let Ok(color_override) = color_override.get(scene_ready.entity) else {
+        info!("{} does not have a color override", scene_ready.entity);
+        return;
+    };
+
     // Iterate over all children recursively
     for descendant in children.iter_descendants(scene_ready.entity) {
         // Get the material id and name which were created from the glTF file information
@@ -72,10 +83,6 @@ fn change_material(
         match material_name.0.as_str() {
             "LeatherPartsMat" => {
                 info!("editing LeatherPartsMat to use ColorOverride tint");
-                // Get the `ColorOverride` of the entity, if it does not have a color override, skip
-                let Ok(color_override) = color_override.get(scene_ready.entity) else {
-                    continue;
-                };
                 // Create a copy of the material and override base color
                 // If you intend on creating multiple models with the same tint, it
                 // is best to cache the handle somewhere, as having multiple materials
