diff --git a/crates/bevy_ecs/src/entity/mod.rs b/crates/bevy_ecs/src/entity/mod.rs
index 82f8c7a64419e..62c1c3de94eb7 100644
--- a/crates/bevy_ecs/src/entity/mod.rs
+++ b/crates/bevy_ecs/src/entity/mod.rs
@@ -671,9 +671,13 @@ impl fmt::Debug for Entity {
 impl fmt::Display for Entity {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         if self == &Self::PLACEHOLDER {
-            write!(f, "PLACEHOLDER")
+            f.pad("PLACEHOLDER")
         } else {
-            write!(f, "{}v{}", self.index(), self.generation())
+            f.pad(&alloc::fmt::format(format_args!(
+                "{}v{}",
+                self.index(),
+                self.generation()
+            )))
         }
     }
 }
@@ -1514,6 +1518,12 @@ mod tests {
         let string = format!("{entity}");
         assert_eq!(string, "42v0");
 
+        let padded_left = format!("{entity:<5}");
+        assert_eq!(padded_left, "42v0 ");
+
+        let padded_right = format!("{entity:>6}");
+        assert_eq!(padded_right, "  42v0");
+
         let entity = Entity::PLACEHOLDER;
         let string = format!("{entity}");
         assert_eq!(string, "PLACEHOLDER");
