diff --git a/crates/bevy_platform/src/hash.rs b/crates/bevy_platform/src/hash.rs
index 66814bc583623..812b88cd10572 100644
--- a/crates/bevy_platform/src/hash.rs
+++ b/crates/bevy_platform/src/hash.rs
@@ -66,6 +66,12 @@ impl<V, H> Hash for Hashed<V, H> {
     }
 }
 
+impl<V: Hash, H: BuildHasher + Default> From<V> for Hashed<V, H> {
+    fn from(value: V) -> Self {
+        Self::new(value)
+    }
+}
+
 impl<V, H> Deref for Hashed<V, H> {
     type Target = V;
 
