diff --git a/crates/bevy_text/Cargo.toml b/crates/bevy_text/Cargo.toml
index 12a36d5826b7d..bc9b134f280a2 100644
--- a/crates/bevy_text/Cargo.toml
+++ b/crates/bevy_text/Cargo.toml
@@ -30,7 +30,7 @@ bevy_platform = { path = "../bevy_platform", version = "0.18.0-dev", default-fea
 
 # other
 wgpu-types = { version = "26", default-features = false }
-cosmic-text = { version = "0.14", features = ["shape-run-cache"] }
+cosmic-text = { version = "0.15", features = ["shape-run-cache"] }
 thiserror = { version = "2", default-features = false }
 serde = { version = "1", features = ["derive"] }
 smallvec = { version = "1", default-features = false }
diff --git a/crates/bevy_text/src/font.rs b/crates/bevy_text/src/font.rs
index b748f4a111fdd..8830f7158baef 100644
--- a/crates/bevy_text/src/font.rs
+++ b/crates/bevy_text/src/font.rs
@@ -2,6 +2,8 @@ use alloc::sync::Arc;
 
 use bevy_asset::Asset;
 use bevy_reflect::TypePath;
+use cosmic_text::skrifa::raw::ReadError;
+use cosmic_text::skrifa::FontRef;
 
 /// An [`Asset`] that contains the data for a loaded font, if loaded as an asset.
 ///
@@ -23,11 +25,8 @@ pub struct Font {
 
 impl Font {
     /// Creates a [`Font`] from bytes
-    pub fn try_from_bytes(
-        font_data: Vec<u8>,
-    ) -> Result<Self, cosmic_text::ttf_parser::FaceParsingError> {
-        use cosmic_text::ttf_parser;
-        ttf_parser::Face::parse(&font_data, 0)?;
+    pub fn try_from_bytes(font_data: Vec<u8>) -> Result<Self, ReadError> {
+        let _ = FontRef::from_index(&font_data, 0)?;
         Ok(Self {
             data: Arc::new(font_data),
         })
diff --git a/crates/bevy_text/src/font_loader.rs b/crates/bevy_text/src/font_loader.rs
index 9e0f2185a234e..77b38082f2dfc 100644
--- a/crates/bevy_text/src/font_loader.rs
+++ b/crates/bevy_text/src/font_loader.rs
@@ -1,5 +1,6 @@
 use crate::Font;
 use bevy_asset::{io::Reader, AssetLoader, LoadContext};
+use cosmic_text::skrifa::raw::ReadError;
 use thiserror::Error;
 
 #[derive(Default)]
@@ -12,7 +13,7 @@ pub struct FontLoader;
 pub enum FontLoaderError {
     /// The contents that could not be parsed
     #[error(transparent)]
-    Content(#[from] cosmic_text::ttf_parser::FaceParsingError),
+    Content(#[from] ReadError),
     /// An [IO](std::io) Error
     #[error(transparent)]
     Io(#[from] std::io::Error),
diff --git a/crates/bevy_text/src/pipeline.rs b/crates/bevy_text/src/pipeline.rs
index ea957f5bf3247..cd40cdab2a011 100644
--- a/crates/bevy_text/src/pipeline.rs
+++ b/crates/bevy_text/src/pipeline.rs
@@ -310,7 +310,12 @@ impl TextPipeline {
             let Some((id, _)) = self.map_handle_to_font_id.get(font) else {
                 continue;
             };
-            if let Some(font) = font_system.get_font(*id) {
+            let weight = font_system
+                .db()
+                .face(*id)
+                .map(|f| f.weight)
+                .unwrap_or(cosmic_text::Weight::NORMAL);
+            if let Some(font) = font_system.get_font(*id, weight) {
                 let swash = font.as_swash();
                 let metrics = swash.metrics(&[]);
                 let upem = metrics.units_per_em as f32;
