diff --git a/crates/bevy_tasks/Cargo.toml b/crates/bevy_tasks/Cargo.toml
index abe5a1485a6ad..80d77fa2dcd5b 100644
--- a/crates/bevy_tasks/Cargo.toml
+++ b/crates/bevy_tasks/Cargo.toml
@@ -61,7 +61,7 @@ async-channel = { version = "2.3.0", default-features = false }
 async-task = { version = "4.4.0", default-features = false, features = [
   "portable-atomic",
 ] }
-heapless = { version = "0.8", default-features = false, features = [
+heapless = { version = "0.9", default-features = false, features = [
   "portable-atomic",
 ] }
 atomic-waker = { version = "1", default-features = false, features = [
diff --git a/crates/bevy_tasks/src/edge_executor.rs b/crates/bevy_tasks/src/edge_executor.rs
index a8c80725cafe9..2bf719029d621 100644
--- a/crates/bevy_tasks/src/edge_executor.rs
+++ b/crates/bevy_tasks/src/edge_executor.rs
@@ -449,7 +449,7 @@ struct State<const C: usize> {
         target_has_atomic = "64",
         target_has_atomic = "ptr"
     )))]
-    queue: heapless::mpmc::MpMcQueue<Runnable, C>,
+    queue: heapless::mpmc::Queue<Runnable, C>,
     waker: AtomicWaker,
 }
 
@@ -471,7 +471,7 @@ impl<const C: usize> State<C> {
                 target_has_atomic = "64",
                 target_has_atomic = "ptr"
             )))]
-            queue: heapless::mpmc::MpMcQueue::new(),
+            queue: heapless::mpmc::Queue::new(),
             waker: AtomicWaker::new(),
         }
     }
