diff --git a/crates/bevy_internal/Cargo.toml b/crates/bevy_internal/Cargo.toml
index c4afedcf3fb2d..0c1eeb4c6b28d 100644
--- a/crates/bevy_internal/Cargo.toml
+++ b/crates/bevy_internal/Cargo.toml
@@ -430,7 +430,7 @@ bevy_ui = { path = "../bevy_ui", optional = true, version = "0.16.0-dev" }
 bevy_window = { path = "../bevy_window", optional = true, version = "0.16.0-dev", default-features = false, features = [
   "bevy_reflect",
 ] }
-bevy_winit = { path = "../bevy_winit", optional = true, version = "0.16.0-dev" }
+bevy_winit = { path = "../bevy_winit", optional = true, version = "0.16.0-dev", default-features = false }
 
 [lints]
 workspace = true
diff --git a/crates/bevy_winit/Cargo.toml b/crates/bevy_winit/Cargo.toml
index ff888332ab765..31dbd63545572 100644
--- a/crates/bevy_winit/Cargo.toml
+++ b/crates/bevy_winit/Cargo.toml
@@ -9,6 +9,7 @@ license = "MIT OR Apache-2.0"
 keywords = ["bevy"]
 
 [features]
+default = ["x11"]
 trace = []
 wayland = ["winit/wayland", "winit/wayland-csd-adwaita"]
 x11 = ["winit/x11"]
