diff --git a/crates/bevy_diagnostic/Cargo.toml b/crates/bevy_diagnostic/Cargo.toml
index 0fc784db750ce..f3f73e76a5155 100644
--- a/crates/bevy_diagnostic/Cargo.toml
+++ b/crates/bevy_diagnostic/Cargo.toml
@@ -77,14 +77,14 @@ log = { version = "0.4", default-features = false }
 # macOS
 [target.'cfg(all(target_os="macos"))'.dependencies]
 # Some features of sysinfo are not supported by apple. This will disable those features on apple devices
-sysinfo = { version = "0.33.0", optional = true, default-features = false, features = [
+sysinfo = { version = "0.34.0", optional = true, default-features = false, features = [
   "apple-app-store",
   "system",
 ] }
 
 # Only include when on linux/windows/android/freebsd
 [target.'cfg(any(target_os = "linux", target_os = "windows", target_os = "android", target_os = "freebsd"))'.dependencies]
-sysinfo = { version = "0.33.0", optional = true, default-features = false, features = [
+sysinfo = { version = "0.34.0", optional = true, default-features = false, features = [
   "system",
 ] }
 
diff --git a/crates/bevy_diagnostic/src/system_information_diagnostics_plugin.rs b/crates/bevy_diagnostic/src/system_information_diagnostics_plugin.rs
index 5fef2a4f25f75..bae6c086a5900 100644
--- a/crates/bevy_diagnostic/src/system_information_diagnostics_plugin.rs
+++ b/crates/bevy_diagnostic/src/system_information_diagnostics_plugin.rs
@@ -224,8 +224,7 @@ pub mod internal {
                     .first()
                     .map(|cpu| cpu.brand().trim().to_string())
                     .unwrap_or_else(|| String::from("not available")),
-                core_count: sys
-                    .physical_core_count()
+                core_count: System::physical_core_count()
                     .map(|x| x.to_string())
                     .unwrap_or_else(|| String::from("not available")),
                 // Convert from Bytes to GibiBytes since it's probably what people expect most of the time
